// circuit-json-to-simple-3d/dist/index.js
// lib/index.ts
import { cju } from "@tscircuit/circuit-json-util";
import { renderScene } from "@tscircuit/simple-3d-svg";

// lib/getDefaultCameraForPcbBoard.ts
function getDefaultCameraForPcbBoard(pcbBoard, anglePreset = "angle1", defaultZoomMultiplier) {
  const w = pcbBoard.width;
  const h = pcbBoard.height;
  const cx = pcbBoard.center?.x || 0;
  const cz = pcbBoard.center?.y || 0;
  const boardSize = Math.max(w, h);
  let baseDist = boardSize * 1.5;
  let effectiveZoomLevel;
  if (defaultZoomMultiplier !== void 0) {
    if (defaultZoomMultiplier > 0) {
      effectiveZoomLevel = defaultZoomMultiplier;
      baseDist = baseDist / effectiveZoomLevel;
    }
  }
  const dist = baseDist;
  let position;
  if (anglePreset === "angle1") {
    position = { x: cx - dist, y: dist, z: cz - dist };
  } else if (anglePreset === "angle2") {
    position = { x: cx + dist, y: dist, z: cz - dist };
  } else if (anglePreset === "left") {
    position = { x: cx - dist, y: 0, z: cz };
  } else if (anglePreset === "right") {
    position = { x: cx + dist, y: 0, z: cz };
  } else if (anglePreset === "left-raised") {
    position = { x: cx - dist, y: dist, z: cz };
  } else if (anglePreset === "right-raised") {
    position = { x: cx + dist, y: dist, z: cz };
  } else {
    throw new Error(`Unknown angle preset: ${anglePreset}`);
  }
  const focalLength = effectiveZoomLevel ? Math.max(1, 2 * effectiveZoomLevel) : 2;
  return {
    position,
    lookAt: { x: cx, y: 0, z: cz },
    focalLength
  };
}

// lib/getDefaultCameraForComponents.ts
function getDefaultCameraForComponents(components, anglePreset = "angle1", defaultZoomMultiplier) {
  if (components.length === 0) {
    return {
      position: { x: 10, y: 10, z: 10 },
      lookAt: { x: 0, y: 0, z: 0 },
      focalLength: 2
    };
  }
  let minX = Number.POSITIVE_INFINITY;
  let maxX = Number.NEGATIVE_INFINITY;
  let minZ = Number.POSITIVE_INFINITY;
  let maxZ = Number.NEGATIVE_INFINITY;
  for (const comp of components) {
    const halfWidth = comp.width / 2;
    const halfHeight = comp.height / 2;
    minX = Math.min(minX, comp.center.x - halfWidth);
    maxX = Math.max(maxX, comp.center.x + halfWidth);
    minZ = Math.min(minZ, comp.center.y - halfHeight);
    maxZ = Math.max(maxZ, comp.center.y + halfHeight);
  }
  const cx = (minX + maxX) / 2;
  const cz = (minZ + maxZ) / 2;
  const w = maxX - minX;
  const h = maxZ - minZ;
  const componentSize = Math.max(w, h, 5);
  let baseDist = componentSize * 1.5;
  let effectiveZoomLevel;
  if (defaultZoomMultiplier !== void 0) {
    if (defaultZoomMultiplier > 0) {
      effectiveZoomLevel = defaultZoomMultiplier;
      baseDist = baseDist / effectiveZoomLevel;
    }
  }
  const dist = baseDist;
  let position;
  if (anglePreset === "angle1") {
    position = { x: cx - dist, y: dist, z: cz - dist };
  } else if (anglePreset === "angle2") {
    position = { x: cx + dist, y: dist, z: cz - dist };
  } else if (anglePreset === "left") {
    position = { x: cx - dist, y: 0, z: cz };
  } else if (anglePreset === "right") {
    position = { x: cx + dist, y: 0, z: cz };
  } else if (anglePreset === "left-raised") {
    position = { x: cx - dist, y: dist, z: cz };
  } else if (anglePreset === "right-raised") {
    position = { x: cx + dist, y: dist, z: cz };
  } else {
    throw new Error(`Unknown angle preset: ${anglePreset}`);
  }
  const focalLength = effectiveZoomLevel ? Math.max(1, 2 * effectiveZoomLevel) : 2;
  return {
    position,
    lookAt: { x: cx, y: 0, z: cz },
    focalLength
  };
}

// lib/index.ts
import { convertCircuitJsonToPcbSvg } from "circuit-to-svg";

// lib/utils/get-background-color-from-options.ts
import Color2 from "color";

// lib/utils/is-valid-color.ts
import Color from "color";
var isValidColor = (color) => {
  try {
    Color(color);
    return true;
  } catch {
    return false;
  }
};

// lib/utils/get-background-color-from-options.ts
function getColorFromBackgroundOptions(background) {
  if (!background) return "white";
  if (background.color) {
    if (background.opacity !== void 0) {
      try {
        const color = Color2(background.color);
        const opacity = Math.max(0, Math.min(1, background.opacity));
        return color.alpha(opacity).string();
      } catch {
        return "white";
      }
    }
    if (isValidColor(background.color)) {
      return background.color;
    }
    return "white";
  }
  return "white";
}

// lib/index.ts
var convertRotationFromCadRotation = (rot) => {
  return {
    x: rot.x * Math.PI / 180 - Math.PI / 2,
    y: rot.y * Math.PI / 180,
    z: rot.z * Math.PI / 180
  };
};
async function convertCircuitJsonToSimple3dScene(circuitJson, opts = {}) {
  const db = cju(circuitJson);
  const boxes = [];
  const pcbBoard = db.pcb_board.list()[0];
  const DEFAULT_COMP_HEIGHT = 2;
  const DEFAULT_BOARD_THICKNESS = 1.4;
  const cadComponents = db.cad_component?.list?.() ?? [];
  const pcbComponentIdsWith3d = /* @__PURE__ */ new Set();
  for (const cad of cadComponents) {
    const { model_stl_url, model_obj_url } = cad;
    if (!model_stl_url && !model_obj_url) continue;
    pcbComponentIdsWith3d.add(cad.pcb_component_id);
    const pcbComp = db.pcb_component.get(cad.pcb_component_id);
    const DEFAULT_SIZE = 2;
    const size = cad.size ?? (pcbComp ? { x: pcbComp.width, y: DEFAULT_COMP_HEIGHT, z: pcbComp.height } : { x: DEFAULT_SIZE, y: DEFAULT_SIZE, z: DEFAULT_SIZE });
    const boardThickness = pcbBoard?.thickness ?? DEFAULT_BOARD_THICKNESS;
    const center = cad.position ? { x: cad.position.x, y: cad.position.z, z: cad.position.y } : {
      x: pcbComp?.center.x ?? 0,
      y: boardThickness / 2 + size.y / 2,
      z: pcbComp?.center.y ?? 0
    };
    boxes.push({
      center,
      size,
      color: !(model_obj_url || model_stl_url) ? "rgba(128,128,128,0.5)" : void 0,
      // used only when model fails to load
      // STL / OBJ support
      ...model_stl_url ? {
        stlUrl: model_stl_url,
        scaleStlToBox: false,
        centerModel: true,
        drawBoundingBox: opts.showBoundingBoxes ?? false
      } : {},
      ...model_obj_url ? {
        objUrl: model_obj_url,
        scaleObjToBox: false,
        centerModel: true,
        drawBoundingBox: opts.showBoundingBoxes ?? false
      } : {},
      // Forward any explicit rotation/position overrides from the CAD record
      ...cad.rotation ? {
        // stlRotation: convertRotation(cad.rotation),
        objRotation: convertRotationFromCadRotation(cad.rotation)
      } : {}
    });
  }
  let camera = opts.camera;
  if (!camera) {
    if (pcbBoard) {
      camera = getDefaultCameraForPcbBoard(
        pcbBoard,
        opts.anglePreset ?? "angle1",
        opts.defaultZoomMultiplier
      );
    } else {
      const pcbComponents = db.pcb_component.list();
      camera = getDefaultCameraForComponents(
        pcbComponents,
        opts.anglePreset ?? "angle1",
        opts.defaultZoomMultiplier
      );
    }
  }
  if (!camera.focalLength) {
    camera.focalLength = 1;
  }
  if (pcbBoard) {
    const pcbTopSvg = convertCircuitJsonToPcbSvg(circuitJson, {
      layer: "top",
      matchBoardAspectRatio: true,
      backgroundColor: "transparent",
      drawPaddingOutsideBoard: false,
      colorOverrides: {
        copper: {
          top: "#ffe066",
          bottom: "#ffe066"
        },
        drill: "rgba(0,0,0,0.5)"
      }
    }).replace("<svg", "<svg transform='scale(1, -1)'");
    boxes.push({
      center: {
        x: pcbBoard.center.x,
        y: 0,
        z: pcbBoard.center.y
      },
      size: {
        x: pcbBoard.width,
        y: pcbBoard.thickness,
        z: pcbBoard.height
      },
      faceImages: {
        top: `data:image/svg+xml;base64,${btoa(pcbTopSvg)}`
      },
      projectionSubdivision: 10,
      color: "rgba(0,140,0,0.8)"
    });
  }
  for (const comp of db.pcb_component.list()) {
    if (pcbComponentIdsWith3d.has(comp.pcb_component_id)) continue;
    const sourceComponent = db.source_component.get(comp.source_component_id);
    const compHeight = Math.min(
      Math.min(comp.width, comp.height),
      DEFAULT_COMP_HEIGHT
    );
    const boardThickness = pcbBoard?.thickness ?? DEFAULT_BOARD_THICKNESS;
    boxes.push({
      center: {
        x: comp.center.x,
        y: boardThickness / 2 + compHeight / 2,
        z: comp.center.y
      },
      size: {
        x: comp.width,
        y: compHeight,
        z: comp.height
      },
      color: "rgba(128,128,128,0.5)",
      topLabel: sourceComponent?.name ?? "?",
      topLabelColor: "white"
    });
  }
  return { boxes, camera };
}
async function convertCircuitJsonToSimple3dSvg(circuitJson, opts = {}) {
  const backgroundColor = getColorFromBackgroundOptions(opts.background);
  const renderOptions = {
    backgroundColor,
    showGrid: opts.showGrid ?? false,
    showOrigin: opts.showOrigin ?? false,
    showAxes: opts.showAxes ?? false
  };
  if (opts.width) renderOptions.width = opts.width;
  if (opts.height) renderOptions.height = opts.height;
  const scene = await convertCircuitJsonToSimple3dScene(circuitJson, opts);
  return await renderScene(scene, {
    ...renderOptions
  });
}
export {
  convertCircuitJsonToSimple3dScene,
  convertCircuitJsonToSimple3dSvg
};
