// circuit-json-to-spice/dist/index.js
// lib/spice-utils/convertSpiceNetlistToString.ts
var convertSpiceNetlistToString = (netlist) => {
  const lines = [];
  lines.push(netlist.title);
  if (netlist.models.size > 0) {
    lines.push(...Array.from(netlist.models.values()));
  }
  for (const component of netlist.components) {
    lines.push(component.toSpiceString());
  }
  for (const subcircuit of netlist.subcircuits) {
    lines.push(subcircuit.toSpiceString());
  }
  if (netlist.printStatements.length > 0) {
    lines.push(...netlist.printStatements);
  }
  if (netlist.controls.length > 0) {
    lines.push(".control");
    lines.push(...netlist.controls);
    lines.push(".endc");
  }
  if (netlist.tranCommand && !lines.some((l) => l.trim().toLowerCase().startsWith(".tran"))) {
    lines.push(netlist.tranCommand);
  }
  lines.push(".END");
  return lines.join("\n");
};

// lib/spice-classes/SpiceNetlist.ts
var SpiceNetlist = class {
  title;
  components;
  nodes;
  controls;
  subcircuits;
  models;
  tranCommand;
  printStatements;
  constructor(title = "Circuit Netlist") {
    this.title = title;
    this.components = [];
    this.nodes = /* @__PURE__ */ new Set();
    this.controls = [];
    this.subcircuits = [];
    this.models = /* @__PURE__ */ new Map();
    this.tranCommand = null;
    this.printStatements = [];
  }
  addComponent(component) {
    this.components.push(component);
    for (const node of component.nodes) {
      this.nodes.add(node);
    }
  }
  addSubcircuit(subcircuit) {
    if (this.subcircuits.find((s) => s.name === subcircuit.name)) return;
    this.subcircuits.push(subcircuit);
  }
  toSpiceString() {
    return convertSpiceNetlistToString(this);
  }
};

// lib/spice-classes/SpiceComponent.ts
var SpiceComponent = class {
  name;
  command;
  nodes;
  constructor(name, command, nodes) {
    this.name = name;
    this.command = command;
    this.nodes = nodes;
  }
  toSpiceString() {
    return this.command.toSpiceString();
  }
};

// lib/spice-commands/ResistorCommand.ts
var ResistorCommand = class {
  commandName = "resistor";
  props;
  constructor(props) {
    this.props = props;
  }
  toSpiceString() {
    const { name, positiveNode, negativeNode, model, value } = this.props;
    let spiceString = `R${name} ${positiveNode} ${negativeNode}`;
    if (model) {
      spiceString += ` ${model}`;
    }
    spiceString += ` ${value}`;
    return spiceString;
  }
};

// lib/spice-commands/CapacitorCommand.ts
var CapacitorCommand = class {
  commandName = "capacitor";
  props;
  constructor(props) {
    this.props = props;
  }
  toSpiceString() {
    const {
      name,
      positiveNode,
      negativeNode,
      modelName,
      value,
      initialCondition
    } = this.props;
    let spiceString = `C${name} ${positiveNode} ${negativeNode}`;
    if (modelName) {
      spiceString += ` ${modelName}`;
    }
    spiceString += ` ${value}`;
    if (initialCondition) {
      spiceString += ` IC=${initialCondition}`;
    }
    return spiceString;
  }
};

// lib/spice-commands/VoltageSourceCommand.ts
var VoltageSourceCommand = class {
  commandName = "voltage_source";
  props;
  constructor(props) {
    this.props = props;
  }
  toSpiceString() {
    const { name, positiveNode, negativeNode, value, acMagnitude, acPhase } = this.props;
    let spiceString = `V${name} ${positiveNode} ${negativeNode}`;
    if (value) {
      spiceString += ` ${value}`;
    }
    if (acMagnitude) {
      spiceString += ` AC ${acMagnitude}`;
      if (acPhase) {
        spiceString += ` ${acPhase}`;
      }
    }
    return spiceString;
  }
};

// lib/spice-commands/DiodeCommand.ts
var DiodeCommand = class {
  commandName = "diode";
  props;
  constructor(props) {
    this.props = props;
  }
  toSpiceString() {
    const { name, positiveNode, negativeNode, model, area } = this.props;
    let spiceString = `D${name} ${positiveNode} ${negativeNode} ${model}`;
    if (area) {
      spiceString += ` ${area}`;
    }
    return spiceString;
  }
};

// lib/spice-commands/InductorCommand.ts
var InductorCommand = class {
  commandName = "inductor";
  props;
  constructor(props) {
    this.props = props;
  }
  toSpiceString() {
    const { name, positiveNode, negativeNode, model, value, initialCondition } = this.props;
    let spiceString = `L${name} ${positiveNode} ${negativeNode}`;
    if (model) {
      spiceString += ` ${model}`;
    }
    spiceString += ` ${value}`;
    if (initialCondition) {
      spiceString += ` IC=${initialCondition}`;
    }
    return spiceString;
  }
};

// lib/spice-commands/VoltageControlledSwitchCommand.ts
var VoltageControlledSwitchCommand = class {
  commandName = "voltage_controlled_switch";
  props;
  constructor(props) {
    this.props = props;
  }
  toSpiceString() {
    const {
      name,
      positiveNode,
      negativeNode,
      positiveControl,
      negativeControl,
      model
    } = this.props;
    return `S${name} ${positiveNode} ${negativeNode} ${positiveControl} ${negativeControl} ${model}`;
  }
};

// lib/circuitJsonToSpice.ts
import { getSourcePortConnectivityMapFromCircuitJson } from "circuit-json-to-connectivity-map";
import { su } from "@tscircuit/circuit-json-util";
function circuitJsonToSpice(circuitJson) {
  const netlist = new SpiceNetlist("* Circuit JSON to SPICE Netlist");
  const sourceComponents = su(circuitJson).source_component.list();
  const sourcePorts = su(circuitJson).source_port.list();
  const sourceTraces = su(circuitJson).source_trace.list();
  const simulationProbes = circuitJson.filter(
    (elm) => elm.type === "simulation_voltage_probe"
  );
  const simulationSwitches = circuitJson.filter(
    (element) => element.type === "simulation_switch"
  ).map((element) => element);
  const simulationSwitchMap = /* @__PURE__ */ new Map();
  for (const simSwitch of simulationSwitches) {
    if (simSwitch.source_component_id) {
      simulationSwitchMap.set(simSwitch.source_component_id, simSwitch);
    }
  }
  const connMap = getSourcePortConnectivityMapFromCircuitJson(circuitJson);
  const nodeMap = /* @__PURE__ */ new Map();
  const netToNodeName = /* @__PURE__ */ new Map();
  let nodeCounter = 1;
  const probeNames = /* @__PURE__ */ new Set();
  if (simulationProbes.length > 0) {
    for (const probe of simulationProbes) {
      if (probe.name) {
        probeNames.add(probe.name);
      }
    }
  }
  const numericProbeNames = [...probeNames].map((name) => /^N(\d+)$/i.exec(name)).filter((m) => m !== null).map((m) => parseInt(m[1], 10));
  if (numericProbeNames.length > 0) {
    nodeCounter = Math.max(...numericProbeNames) + 1;
  }
  const groundNets = /* @__PURE__ */ new Set();
  const gndSourceNetIds = new Set(
    su(circuitJson).source_net.list().filter((sn) => sn.name?.toLowerCase().includes("gnd")).map((sn) => sn.source_net_id)
  );
  if (gndSourceNetIds.size > 0) {
    for (const trace of su(circuitJson).source_trace.list()) {
      if (trace.connected_source_port_ids.length > 0) {
        const isOnGndNet = trace.connected_source_net_ids.some(
          (netId) => gndSourceNetIds.has(netId)
        );
        if (isOnGndNet) {
          const aPortOnGnd = trace.connected_source_port_ids[0];
          const gndNet = connMap.getNetConnectedToId(aPortOnGnd);
          if (gndNet) {
            groundNets.add(gndNet);
          }
        }
      }
    }
  }
  const groundPorts = sourcePorts.filter((p) => p.name?.toLowerCase() === "gnd");
  for (const groundPort of groundPorts) {
    const groundNet = connMap.getNetConnectedToId(groundPort.source_port_id);
    if (groundNet) {
      groundNets.add(groundNet);
    }
  }
  for (const simSource of su(circuitJson).simulation_voltage_source.list()) {
    const neg_port_id = simSource.negative_source_port_id ?? simSource.terminal2_source_port_id;
    if (neg_port_id) {
      const gnd_net = connMap.getNetConnectedToId(neg_port_id);
      if (gnd_net) {
        groundNets.add(gnd_net);
      }
    }
  }
  for (const groundNet of groundNets) {
    netToNodeName.set(groundNet, "0");
  }
  if (simulationProbes.length > 0) {
    for (const probe of simulationProbes) {
      if (!probe.name) continue;
      let net;
      if (probe.source_port_id) {
        net = connMap.getNetConnectedToId(probe.source_port_id);
      } else if (probe.source_net_id) {
        const trace = sourceTraces.find(
          (t) => t.connected_source_net_ids.includes(probe.source_net_id)
        );
        if (trace && trace.connected_source_port_ids.length > 0) {
          const portId = trace.connected_source_port_ids[0];
          net = connMap.getNetConnectedToId(portId);
        }
      }
      if (net) {
        if (!netToNodeName.has(net)) {
          netToNodeName.set(net, probe.name);
        }
      } else if (probe.source_port_id && probe.name) {
        nodeMap.set(probe.source_port_id, probe.name);
      }
    }
  }
  for (const port of sourcePorts) {
    const portId = port.source_port_id;
    const net = connMap.getNetConnectedToId(portId);
    if (net) {
      if (!netToNodeName.has(net)) {
        netToNodeName.set(net, `N${nodeCounter++}`);
      }
      nodeMap.set(portId, netToNodeName.get(net));
    }
  }
  for (const port of sourcePorts) {
    const portId = port.source_port_id;
    if (!nodeMap.has(portId)) {
      nodeMap.set(portId, `N${nodeCounter++}`);
    }
  }
  for (const component of sourceComponents) {
    if (component.type !== "source_component") continue;
    const componentPorts = su(circuitJson).source_port.list({
      source_component_id: component.source_component_id
    }).sort((a, b) => (a.pin_number ?? 0) - (b.pin_number ?? 0));
    const nodes = componentPorts.map((port) => {
      return nodeMap.get(port.source_port_id) || "0";
    });
    if ("ftype" in component) {
      let spiceComponent = null;
      switch (component.ftype) {
        case "simple_resistor": {
          if ("resistance" in component && "name" in component) {
            const resistorCmd = new ResistorCommand({
              name: component.name,
              positiveNode: nodes[0] || "0",
              negativeNode: nodes[1] || "0",
              value: formatResistance(component.resistance)
            });
            spiceComponent = new SpiceComponent(
              component.name,
              resistorCmd,
              nodes
            );
          }
          break;
        }
        case "simple_switch": {
          const sanitizedBase = sanitizeIdentifier(
            component.name ?? component.source_component_id,
            "SW"
          );
          const positiveNode = nodes[0] || "0";
          const negativeNode = nodes[1] || "0";
          const controlNode = `NCTRL_${sanitizedBase}`;
          const modelName = `SW_${sanitizedBase}`;
          const associatedSimulationSwitch = simulationSwitchMap.get(
            component.source_component_id
          );
          const controlValue = buildSimulationSwitchControlValue(
            associatedSimulationSwitch
          );
          const switchCmd = new VoltageControlledSwitchCommand({
            name: sanitizedBase,
            positiveNode,
            negativeNode,
            positiveControl: controlNode,
            negativeControl: "0",
            model: modelName
          });
          spiceComponent = new SpiceComponent(sanitizedBase, switchCmd, [
            positiveNode,
            negativeNode,
            controlNode,
            "0"
          ]);
          if (!netlist.models.has(modelName)) {
            netlist.models.set(
              modelName,
              `.MODEL ${modelName} SW(Ron=0.1 Roff=1e9 Vt=2.5 Vh=0.1)`
            );
          }
          const controlSourceName = `CTRL_${sanitizedBase}`;
          const controlSourceCmd = new VoltageSourceCommand({
            name: controlSourceName,
            positiveNode: controlNode,
            negativeNode: "0",
            value: controlValue
          });
          const controlComponent = new SpiceComponent(
            controlSourceName,
            controlSourceCmd,
            [controlNode, "0"]
          );
          netlist.addComponent(controlComponent);
          break;
        }
        case "simple_capacitor": {
          if ("capacitance" in component && "name" in component) {
            const capacitorCmd = new CapacitorCommand({
              name: component.name,
              positiveNode: nodes[0] || "0",
              negativeNode: nodes[1] || "0",
              value: formatCapacitance(component.capacitance)
            });
            spiceComponent = new SpiceComponent(
              component.name,
              capacitorCmd,
              nodes
            );
          }
          break;
        }
        case "simple_diode": {
          if ("name" in component) {
            const anodePort = componentPorts.find(
              (p) => p.name?.toLowerCase() === "anode" || p.port_hints?.includes("anode")
            );
            const cathodePort = componentPorts.find(
              (p) => p.name?.toLowerCase() === "cathode" || p.port_hints?.includes("cathode")
            );
            const positiveNode = nodeMap.get(anodePort?.source_port_id ?? "") || "0";
            const negativeNode = nodeMap.get(cathodePort?.source_port_id ?? "") || "0";
            const modelName = "D";
            const diodeCmd = new DiodeCommand({
              name: component.name,
              positiveNode,
              negativeNode,
              model: modelName
              // generic model
            });
            netlist.models.set(modelName, `.MODEL ${modelName} D`);
            spiceComponent = new SpiceComponent(component.name, diodeCmd, [
              positiveNode,
              negativeNode
            ]);
          }
          break;
        }
        case "simple_inductor": {
          if ("inductance" in component && "name" in component) {
            const inductorCmd = new InductorCommand({
              name: component.name,
              positiveNode: nodes[0] || "0",
              negativeNode: nodes[1] || "0",
              value: formatInductance(component.inductance)
            });
            spiceComponent = new SpiceComponent(
              component.name,
              inductorCmd,
              nodes
            );
          }
          break;
        }
        case "simple_mosfet": {
          if ("name" in component) {
            const drainPort = componentPorts.find(
              (p) => p.name?.toLowerCase() === "drain" || p.port_hints?.includes("drain")
            );
            const gatePort = componentPorts.find(
              (p) => p.name?.toLowerCase() === "gate" || p.port_hints?.includes("gate")
            );
            const sourcePort = componentPorts.find(
              (p) => p.name?.toLowerCase() === "source" || p.port_hints?.includes("source")
            );
            const drainNode = nodeMap.get(drainPort?.source_port_id ?? "") || "0";
            const gateNode = nodeMap.get(gatePort?.source_port_id ?? "") || "0";
            const sourceNode = nodeMap.get(sourcePort?.source_port_id ?? "") || "0";
            const modelName = "SWMOD";
            const switchCmd = new VoltageControlledSwitchCommand({
              name: component.name,
              positiveNode: drainNode,
              negativeNode: sourceNode,
              positiveControl: gateNode,
              negativeControl: sourceNode,
              model: modelName
            });
            netlist.models.set(modelName, `.MODEL ${modelName} SW`);
            spiceComponent = new SpiceComponent(component.name, switchCmd, [
              drainNode,
              gateNode,
              sourceNode
            ]);
          }
          break;
        }
      }
      if (spiceComponent) {
        netlist.addComponent(spiceComponent);
      }
    }
  }
  const simulationVoltageSources = su(circuitJson).simulation_voltage_source.list();
  for (const simSource of simulationVoltageSources) {
    if (simSource.type !== "simulation_voltage_source") continue;
    if (simSource.is_dc_source === false) {
      if ("terminal1_source_port_id" in simSource && "terminal2_source_port_id" in simSource && simSource.terminal1_source_port_id && simSource.terminal2_source_port_id) {
        const positiveNode = nodeMap.get(simSource.terminal1_source_port_id) || "0";
        const negativeNode = nodeMap.get(simSource.terminal2_source_port_id) || "0";
        let value = "";
        const wave_shape = simSource.wave_shape;
        if (wave_shape === "sinewave") {
          const v_offset = 0;
          const v_peak = simSource.voltage ?? 0;
          const freq = simSource.frequency ?? 0;
          const delay = 0;
          const damping_factor = 0;
          const phase = simSource.phase ?? 0;
          if (freq > 0) {
            value = `SIN(${v_offset} ${v_peak} ${freq} ${delay} ${damping_factor} ${phase})`;
          } else {
            value = `DC ${simSource.voltage ?? 0}`;
          }
        } else if (wave_shape === "square") {
          const v_initial = 0;
          const v_pulsed = simSource.voltage ?? 0;
          const freq = simSource.frequency ?? 0;
          const period_from_freq = freq === 0 ? Infinity : 1 / freq;
          const period = simSource.period ?? period_from_freq;
          const duty_cycle = simSource.duty_cycle ?? 0.5;
          const pulse_width = period * duty_cycle;
          const delay = 0;
          const rise_time = "1n";
          const fall_time = "1n";
          value = `PULSE(${v_initial} ${v_pulsed} ${delay} ${rise_time} ${fall_time} ${pulse_width} ${period})`;
        } else if (simSource.voltage !== void 0) {
          value = `DC ${simSource.voltage}`;
        }
        if (value) {
          const voltageSourceCmd = new VoltageSourceCommand({
            name: simSource.simulation_voltage_source_id,
            positiveNode,
            negativeNode,
            value
          });
          const spiceComponent = new SpiceComponent(
            simSource.simulation_voltage_source_id,
            voltageSourceCmd,
            [positiveNode, negativeNode]
          );
          netlist.addComponent(spiceComponent);
        }
      }
    } else {
      const positivePortId = simSource.positive_source_port_id ?? simSource.terminal1_source_port_id;
      const negativePortId = simSource.negative_source_port_id ?? simSource.terminal2_source_port_id;
      if (positivePortId && negativePortId && "voltage" in simSource && simSource.voltage !== void 0) {
        const positiveNode = nodeMap.get(positivePortId) || "0";
        const negativeNode = nodeMap.get(negativePortId) || "0";
        const voltageSourceCmd = new VoltageSourceCommand({
          name: simSource.simulation_voltage_source_id,
          positiveNode,
          negativeNode,
          value: `DC ${simSource.voltage}`
        });
        const spiceComponent = new SpiceComponent(
          simSource.simulation_voltage_source_id,
          voltageSourceCmd,
          [positiveNode, negativeNode]
        );
        netlist.addComponent(spiceComponent);
      }
    }
  }
  const simExperiment = circuitJson.find(
    (elm) => elm.type === "simulation_experiment"
  );
  if (simExperiment) {
    if (simulationProbes.length > 0) {
      const nodesToProbe = /* @__PURE__ */ new Set();
      for (const probe of simulationProbes) {
        let nodeName;
        if (probe.source_port_id) {
          nodeName = nodeMap.get(probe.source_port_id);
        } else if (probe.source_net_id) {
          const trace = sourceTraces.find(
            (t) => t.connected_source_net_ids.includes(probe.source_net_id)
          );
          if (trace && trace.connected_source_port_ids.length > 0) {
            const portId = trace.connected_source_port_ids[0];
            nodeName = nodeMap.get(portId);
          }
        }
        if (nodeName && nodeName !== "0") {
          nodesToProbe.add(`V(${nodeName})`);
        }
      }
      if (nodesToProbe.size > 0 && simExperiment.experiment_type?.includes("transient")) {
        netlist.printStatements.push(
          `.PRINT TRAN ${[...nodesToProbe].join(" ")}`
        );
      }
    }
    const timePerStep = simExperiment.time_per_step;
    const endTime = simExperiment.end_time_ms;
    const startTimeMs = simExperiment.start_time_ms;
    if (timePerStep && endTime) {
      const startTime = (startTimeMs ?? 0) / 1e3;
      let tranCmd = `.tran ${formatNumberForSpice(
        timePerStep / 1e3
      )} ${formatNumberForSpice(endTime / 1e3)}`;
      if (startTime > 0) {
        tranCmd += ` ${formatNumberForSpice(startTime)}`;
      }
      tranCmd += " UIC";
      netlist.tranCommand = tranCmd;
    }
  }
  return netlist;
}
function formatResistance(resistance) {
  if (resistance >= 1e6) return `${resistance / 1e6}MEG`;
  if (resistance >= 1e3) return `${resistance / 1e3}K`;
  return resistance.toString();
}
function formatCapacitance(capacitance) {
  if (capacitance >= 1e-3) return `${capacitance * 1e3}M`;
  if (capacitance >= 1e-6) return `${capacitance * 1e6}U`;
  if (capacitance >= 1e-9) return `${capacitance * 1e9}N`;
  if (capacitance >= 1e-12) return `${capacitance * 1e12}P`;
  return capacitance.toString();
}
function formatInductance(inductance) {
  if (inductance >= 1) return inductance.toString();
  if (inductance >= 1e-3) return `${inductance * 1e3}m`;
  if (inductance >= 1e-6) return `${inductance * 1e6}u`;
  if (inductance >= 1e-9) return `${inductance * 1e9}n`;
  if (inductance >= 1e-12) return `${inductance * 1e12}p`;
  return inductance.toString();
}
function sanitizeIdentifier(value, prefix) {
  if (!value) return prefix;
  const sanitized = value.replace(/[^A-Za-z0-9_]/g, "_");
  if (!sanitized) return prefix;
  if (/^[0-9]/.test(sanitized)) {
    return `${prefix}_${sanitized}`;
  }
  return sanitized;
}
function buildSimulationSwitchControlValue(simulationSwitch) {
  const highVoltage = 5;
  const lowVoltage = 0;
  const riseTime = "1n";
  const fallTime = "1n";
  if (!simulationSwitch) {
    return `DC ${lowVoltage}`;
  }
  const startsClosed = simulationSwitch.starts_closed ?? false;
  const closesAt = simulationSwitch.closes_at ?? 0;
  const opensAt = simulationSwitch.opens_at;
  const switchingFrequency = simulationSwitch.switching_frequency;
  const [initialVoltage, pulsedVoltage] = startsClosed ? [highVoltage, lowVoltage] : [lowVoltage, highVoltage];
  if (switchingFrequency && switchingFrequency > 0) {
    const period = 1 / switchingFrequency;
    const widthFromOpenClose = opensAt && opensAt > closesAt ? Math.min(opensAt - closesAt, period) : 0;
    const pulseWidth = widthFromOpenClose > 0 ? widthFromOpenClose : Math.max(period / 2, 1e-9);
    return `PULSE(${formatNumberForSpice(initialVoltage)} ${formatNumberForSpice(pulsedVoltage)} ${formatNumberForSpice(closesAt)} ${riseTime} ${fallTime} ${formatNumberForSpice(pulseWidth)} ${formatNumberForSpice(period)})`;
  }
  if (opensAt !== void 0 && opensAt > closesAt) {
    const pulseWidth = Math.max(opensAt - closesAt, 1e-9);
    const period = closesAt + pulseWidth * 2;
    return `PULSE(${formatNumberForSpice(initialVoltage)} ${formatNumberForSpice(pulsedVoltage)} ${formatNumberForSpice(closesAt)} ${riseTime} ${fallTime} ${formatNumberForSpice(pulseWidth)} ${formatNumberForSpice(period)})`;
  }
  if (closesAt > 0) {
    const period = closesAt * 2;
    const pulseWidth = Math.max(period / 2, 1e-9);
    return `PULSE(${formatNumberForSpice(initialVoltage)} ${formatNumberForSpice(pulsedVoltage)} ${formatNumberForSpice(closesAt)} ${riseTime} ${fallTime} ${formatNumberForSpice(pulseWidth)} ${formatNumberForSpice(period)})`;
  }
  return `DC ${startsClosed ? highVoltage : lowVoltage}`;
}
function formatNumberForSpice(value) {
  if (!Number.isFinite(value)) return `${value}`;
  if (value === 0) return "0";
  const absValue = Math.abs(value);
  if (absValue >= 1e3 || absValue <= 1e-3) {
    return Number(value.toExponential(6)).toString();
  }
  return Number(value.toPrecision(6)).toString();
}

// lib/spice-classes/SpiceSubcircuit.ts
var SpiceSubcircuit = class {
  name;
  pins;
  constructor(name, pins) {
    this.name = name;
    this.pins = pins;
  }
  toSpiceString() {
    const pinString = this.pins.join(" ");
    const header = `.SUBCKT ${this.name} ${pinString}`;
    const footer = `.ENDS ${this.name}`;
    const body = `* Placeholder for ${this.name}. No definition found in circuit JSON.`;
    return ["", header, body, footer].join("\n");
  }
};

// lib/spice-commands/BJTCommand.ts
var BJTCommand = class {
  commandName = "bjt";
  props;
  constructor(props) {
    this.props = props;
  }
  toSpiceString() {
    const { name, collector, base, emitter, substrate, model, area } = this.props;
    let spiceString = `Q${name} ${collector} ${base} ${emitter}`;
    if (substrate) {
      spiceString += ` ${substrate}`;
    }
    spiceString += ` ${model}`;
    if (area) {
      spiceString += ` ${area}`;
    }
    return spiceString;
  }
};

// lib/spice-commands/CurrentSourceCommand.ts
var CurrentSourceCommand = class {
  commandName = "current_source";
  props;
  constructor(props) {
    this.props = props;
  }
  toSpiceString() {
    const { name, positiveNode, negativeNode, value, acMagnitude, acPhase } = this.props;
    let spiceString = `I${name} ${positiveNode} ${negativeNode}`;
    if (value) {
      spiceString += ` ${value}`;
    }
    if (acMagnitude) {
      spiceString += ` AC ${acMagnitude}`;
      if (acPhase) {
        spiceString += ` ${acPhase}`;
      }
    }
    return spiceString;
  }
};

// lib/spice-commands/InductorCouplingCommand.ts
var InductorCouplingCommand = class {
  commandName = "inductor_coupling";
  props;
  constructor(props) {
    this.props = props;
  }
  toSpiceString() {
    const { name, inductors, coupling } = this.props;
    return `K${name} ${inductors.join(" ")} ${coupling}`;
  }
};

// lib/spice-commands/JFETCommand.ts
var JFETCommand = class {
  commandName = "jfet";
  props;
  constructor(props) {
    this.props = props;
  }
  toSpiceString() {
    const { name, drain, gate, source, model, area } = this.props;
    let spiceString = `J${name} ${drain} ${gate} ${source} ${model}`;
    if (area) {
      spiceString += ` ${area}`;
    }
    return spiceString;
  }
};

// lib/spice-commands/MOSFETCommand.ts
var MOSFETCommand = class {
  commandName = "mosfet";
  props;
  constructor(props) {
    this.props = props;
  }
  toSpiceString() {
    const {
      name,
      drain,
      gate,
      source,
      substrate,
      model,
      length,
      width,
      drainArea,
      sourceArea,
      drainPerimeter,
      sourcePerimeter,
      drainResistance,
      sourceResistance
    } = this.props;
    let spiceString = `M${name} ${drain} ${gate} ${source} ${substrate} ${model}`;
    const params = {
      L: length,
      W: width,
      AD: drainArea,
      AS: sourceArea,
      PD: drainPerimeter,
      PS: sourcePerimeter,
      NRD: drainResistance,
      NRS: sourceResistance
    };
    Object.entries(params).forEach(([key, value]) => {
      if (value) {
        spiceString += ` ${key}=${value}`;
      }
    });
    return spiceString;
  }
};

// lib/spice-commands/SubcircuitCallCommand.ts
var SubcircuitCallCommand = class {
  commandName = "subcircuit_call";
  props;
  constructor(props) {
    this.props = props;
  }
  toSpiceString() {
    const { name, nodes, subcircuitName } = this.props;
    return `X${name} ${nodes.join(" ")} ${subcircuitName}`;
  }
};

// lib/spice-commands/TransmissionLineCommand.ts
var TransmissionLineCommand = class {
  commandName = "transmission_line";
  props;
  constructor(props) {
    this.props = props;
  }
  toSpiceString() {
    const {
      name,
      aPositive,
      aNegative,
      bPositive,
      bNegative,
      impedance,
      delay,
      frequency,
      normalizedLength
    } = this.props;
    let spiceString = `T${name} ${aPositive} ${aNegative} ${bPositive} ${bNegative} Z0=${impedance}`;
    if (delay) {
      spiceString += ` TD=${delay}`;
    } else if (frequency) {
      spiceString += ` F=${frequency}`;
      if (normalizedLength) {
        spiceString += ` NL=${normalizedLength}`;
      }
    }
    return spiceString;
  }
};
export {
  BJTCommand,
  CapacitorCommand,
  CurrentSourceCommand,
  DiodeCommand,
  InductorCommand,
  InductorCouplingCommand,
  JFETCommand,
  MOSFETCommand,
  ResistorCommand,
  SpiceComponent,
  SpiceNetlist,
  SpiceSubcircuit,
  SubcircuitCallCommand,
  TransmissionLineCommand,
  VoltageControlledSwitchCommand,
  VoltageSourceCommand,
  circuitJsonToSpice,
  convertSpiceNetlistToString
};
