// @tscircuit/props/dist/index.js
// lib/typecheck.ts
var expectTypesMatch = (shouldBe) => {
};
expectTypesMatch("extra props b");
expectTypesMatch("missing props b");
expectTypesMatch(true);
expectTypesMatch("property a has mismatched types");

// lib/common/direction.ts
import { z } from "zod";
var direction = z.enum(["up", "down", "left", "right"]);
var directionAlongEdge = z.enum([
  "top-to-bottom",
  "left-to-right",
  "bottom-to-top",
  "right-to-left"
]);
expectTypesMatch(true);
expectTypesMatch(true);

// lib/common/portHints.ts
import { z as z2 } from "zod";
var portHints = z2.array(z2.string().or(z2.number()));
expectTypesMatch(true);

// lib/common/layout.ts
import {
  distance as distance4,
  layer_ref,
  rotation,
  supplier_name
} from "circuit-json";
import { z as z9 } from "zod";

// lib/common/cadModel.ts
import { z as z5 } from "zod";

// lib/common/distance.ts
import "zod";
import { distance, length } from "circuit-json";

// lib/common/point3.ts
import { distance as distance2 } from "circuit-json";
import { z as z4 } from "zod";
var point3 = z4.object({
  x: distance2,
  y: distance2,
  z: distance2
});

// lib/common/cadModel.ts
var rotationPoint3 = z5.object({
  x: z5.union([z5.number(), z5.string()]),
  y: z5.union([z5.number(), z5.string()]),
  z: z5.union([z5.number(), z5.string()])
});
var cadModelBase = z5.object({
  rotationOffset: z5.number().or(rotationPoint3).optional(),
  positionOffset: point3.optional(),
  size: point3.optional(),
  modelUnitToMmScale: distance.optional(),
  zOffsetFromSurface: distance.optional()
});
expectTypesMatch(true);
var cadModelStl = cadModelBase.extend({
  stlUrl: z5.string()
});
var cadModelObj = cadModelBase.extend({
  objUrl: z5.string(),
  mtlUrl: z5.string().optional()
});
var cadModelGltf = cadModelBase.extend({
  gltfUrl: z5.string()
});
var cadModelGlb = cadModelBase.extend({
  glbUrl: z5.string()
});
var cadModelStep = cadModelBase.extend({
  stepUrl: z5.string()
});
var cadModelWrl = cadModelBase.extend({
  wrlUrl: z5.string()
});
var cadModelJscad = cadModelBase.extend({
  jscad: z5.record(z5.any())
});
var cadModelProp = z5.union([
  z5.null(),
  z5.string(),
  z5.custom((v) => {
    return v && typeof v === "object" && "type" in v && "props" in v;
  }),
  cadModelStl,
  cadModelObj,
  cadModelGltf,
  cadModelGlb,
  cadModelStep,
  cadModelWrl,
  cadModelJscad
]);
expectTypesMatch(true);

// lib/common/footprintProp.ts
import { z as z6 } from "zod";
var footprintProp = z6.custom((v) => true);

// lib/common/pcbStyle.ts
import { distance as distance3 } from "circuit-json";
import { z as z7 } from "zod";
var pcbStyle = z7.object({
  silkscreenFontSize: distance3.optional(),
  silkscreenTextPosition: z7.union([
    z7.enum(["centered", "outside", "none"]),
    z7.object({
      offsetX: z7.number(),
      offsetY: z7.number()
    })
  ]).optional(),
  silkscreenTextVisibility: z7.enum(["hidden", "visible", "inherit"]).optional()
});
expectTypesMatch(true);

// lib/common/symbolProp.ts
import { z as z8 } from "zod";
var symbolProp = z8.custom((v) => true);

// lib/common/layout.ts
var pcbLayoutProps = z9.object({
  pcbX: distance4.optional(),
  pcbY: distance4.optional(),
  pcbOffsetX: distance4.optional(),
  pcbOffsetY: distance4.optional(),
  pcbRotation: rotation.optional(),
  pcbPositionAnchor: z9.string().optional(),
  pcbPositionMode: z9.enum([
    "relative_to_group_anchor",
    "auto",
    "relative_to_board_anchor",
    "relative_to_component_anchor"
  ]).optional(),
  layer: layer_ref.optional(),
  pcbMarginTop: distance4.optional(),
  pcbMarginRight: distance4.optional(),
  pcbMarginBottom: distance4.optional(),
  pcbMarginLeft: distance4.optional(),
  pcbMarginX: distance4.optional(),
  pcbMarginY: distance4.optional(),
  pcbStyle: pcbStyle.optional(),
  pcbRelative: z9.boolean().optional(),
  relative: z9.boolean().optional()
});
expectTypesMatch(true);
var commonLayoutProps = z9.object({
  pcbX: distance4.optional(),
  pcbY: distance4.optional(),
  pcbOffsetX: distance4.optional(),
  pcbOffsetY: distance4.optional(),
  pcbRotation: rotation.optional(),
  pcbPositionAnchor: z9.string().optional(),
  pcbPositionMode: z9.enum([
    "relative_to_group_anchor",
    "auto",
    "relative_to_board_anchor",
    "relative_to_component_anchor"
  ]).optional(),
  pcbMarginTop: distance4.optional(),
  pcbMarginRight: distance4.optional(),
  pcbMarginBottom: distance4.optional(),
  pcbMarginLeft: distance4.optional(),
  pcbMarginX: distance4.optional(),
  pcbMarginY: distance4.optional(),
  pcbStyle: pcbStyle.optional(),
  schMarginTop: distance4.optional(),
  schMarginRight: distance4.optional(),
  schMarginBottom: distance4.optional(),
  schMarginLeft: distance4.optional(),
  schMarginX: distance4.optional(),
  schMarginY: distance4.optional(),
  schX: distance4.optional(),
  schY: distance4.optional(),
  schRotation: rotation.optional(),
  layer: layer_ref.optional(),
  footprint: footprintProp.optional(),
  symbol: symbolProp.optional(),
  relative: z9.boolean().optional(),
  schRelative: z9.boolean().optional(),
  pcbRelative: z9.boolean().optional()
});
expectTypesMatch(true);
var supplierProps = z9.object({
  supplierPartNumbers: z9.record(supplier_name, z9.array(z9.string())).optional()
});
expectTypesMatch(true);
var pinAttributeMap = z9.object({
  providesPower: z9.boolean().optional(),
  requiresPower: z9.boolean().optional(),
  providesGround: z9.boolean().optional(),
  requiresGround: z9.boolean().optional(),
  providesVoltage: z9.union([z9.string(), z9.number()]).optional(),
  requiresVoltage: z9.union([z9.string(), z9.number()]).optional(),
  doNotConnect: z9.boolean().optional(),
  includeInBoardPinout: z9.boolean().optional(),
  highlightColor: z9.string().optional()
});
expectTypesMatch(true);
var commonComponentProps = commonLayoutProps.merge(supplierProps).extend({
  key: z9.any().optional(),
  name: z9.string(),
  cadModel: cadModelProp.optional(),
  children: z9.any().optional(),
  symbolName: z9.string().optional(),
  doNotPlace: z9.boolean().optional(),
  obstructsWithinBounds: z9.boolean().optional().describe(
    "Does this component take up all the space within its bounds on a layer. This is generally true except for when separated pin headers are being represented by a single component (in which case, chips can be placed between the pin headers) or for tall modules where chips fit underneath"
  ),
  pinAttributes: z9.record(z9.string(), pinAttributeMap).optional()
});
expectTypesMatch(true);
var componentProps = commonComponentProps;
var lrPins = ["pin1", "left", "pin2", "right"];
var lrPolarPins = [
  "pin1",
  "left",
  "anode",
  "pos",
  "pin2",
  "right",
  "cathode",
  "neg"
];
var distanceOrMultiplier = distance4.or(z9.enum(["2x", "3x", "4x"]));

// lib/common/schematicOrientation.ts
import { z as z10 } from "zod";
var schematicOrientation = z10.enum([
  "vertical",
  "horizontal",
  "pos_top",
  "pos_bottom",
  "pos_left",
  "pos_right",
  "neg_top",
  "neg_bottom",
  "neg_left",
  "neg_right"
]).describe(
  "horizontal means pins go 1->2 rightward and vertical means pins go 1->2 downward (generally, positive on top)"
);
expectTypesMatch(
  true
);

// lib/common/schematicPinDefinitions.ts
import { z as z11 } from "zod";
var explicitPinSideDefinition = z11.object({
  pins: z11.array(z11.union([z11.number(), z11.string()])),
  direction: z11.union([
    z11.literal("top-to-bottom"),
    z11.literal("left-to-right"),
    z11.literal("bottom-to-top"),
    z11.literal("right-to-left")
  ])
});
var schematicPortArrangement = z11.object({
  leftSize: z11.number().optional().describe("@deprecated, use leftPinCount"),
  topSize: z11.number().optional().describe("@deprecated, use topPinCount"),
  rightSize: z11.number().optional().describe("@deprecated, use rightPinCount"),
  bottomSize: z11.number().optional().describe("@deprecated, use bottomPinCount"),
  leftPinCount: z11.number().optional(),
  rightPinCount: z11.number().optional(),
  topPinCount: z11.number().optional(),
  bottomPinCount: z11.number().optional(),
  leftSide: explicitPinSideDefinition.optional(),
  rightSide: explicitPinSideDefinition.optional(),
  topSide: explicitPinSideDefinition.optional(),
  bottomSide: explicitPinSideDefinition.optional()
});
var schematicPinArrangement = schematicPortArrangement;
expectTypesMatch(true);

// lib/common/schematicPinStyle.ts
import { distance as distance5 } from "circuit-json";
import { z as z12 } from "zod";
var schematicPinStyle = z12.record(
  z12.object({
    marginLeft: distance5.optional(),
    marginRight: distance5.optional(),
    marginTop: distance5.optional(),
    marginBottom: distance5.optional(),
    leftMargin: distance5.optional(),
    rightMargin: distance5.optional(),
    topMargin: distance5.optional(),
    bottomMargin: distance5.optional()
  })
);
expectTypesMatch(true);

// lib/common/schematicPinLabel.ts
import { z as z13 } from "zod";
var schematicPinLabel = z13.string().regex(/^[A-Za-z0-9_]+$/);

// lib/common/ninePointAnchor.ts
import { z as z14 } from "zod";
var ninePointAnchor = z14.enum([
  "top_left",
  "top_center",
  "top_right",
  "center_left",
  "center",
  "center_right",
  "bottom_left",
  "bottom_center",
  "bottom_right"
]);

// lib/common/point.ts
import { distance as distance6 } from "circuit-json";
import { z as z15 } from "zod";
var point = z15.object({
  x: distance6,
  y: distance6
});

// lib/components/board.ts
import { z as z29 } from "zod";

// lib/components/group.ts
import { length as length2, distance as distance7 } from "circuit-json";
import { z as z28 } from "zod";

// lib/manual-edits/manual-edit-events/base_manual_edit_event.ts
import { z as z16 } from "zod";
var base_manual_edit_event = z16.object({
  edit_event_id: z16.string(),
  in_progress: z16.boolean().optional(),
  created_at: z16.number()
});
expectTypesMatch(
  true
);

// lib/manual-edits/manual-edit-events/edit_pcb_component_location_event.ts
import { z as z17 } from "zod";
var edit_pcb_component_location_event = base_manual_edit_event.extend({
  pcb_edit_event_type: z17.literal("edit_component_location").describe("deprecated"),
  edit_event_type: z17.literal("edit_pcb_component_location"),
  pcb_component_id: z17.string(),
  original_center: z17.object({ x: z17.number(), y: z17.number() }),
  new_center: z17.object({ x: z17.number(), y: z17.number() })
});
var edit_component_location_event = edit_pcb_component_location_event;
expectTypesMatch(true);

// lib/manual-edits/manual-edit-events/edit_trace_hint_event.ts
import { z as z18 } from "zod";
var edit_trace_hint_event = base_manual_edit_event.extend({
  pcb_edit_event_type: z18.literal("edit_trace_hint").describe("deprecated"),
  edit_event_type: z18.literal("edit_pcb_trace_hint").optional(),
  pcb_port_id: z18.string(),
  pcb_trace_hint_id: z18.string().optional(),
  route: z18.array(
    z18.object({ x: z18.number(), y: z18.number(), via: z18.boolean().optional() })
  )
});
expectTypesMatch(
  true
);

// lib/manual-edits/manual-edit-events/edit_schematic_component_location_event.ts
import { z as z19 } from "zod";
var edit_schematic_component_location_event = base_manual_edit_event.extend({
  edit_event_type: z19.literal("edit_schematic_component_location"),
  schematic_component_id: z19.string(),
  original_center: z19.object({ x: z19.number(), y: z19.number() }),
  new_center: z19.object({ x: z19.number(), y: z19.number() })
});
expectTypesMatch(true);

// lib/manual-edits/manual-edit-events/edit_pcb_group_location_event.ts
import { z as z20 } from "zod";
var edit_pcb_group_location_event = base_manual_edit_event.extend({
  edit_event_type: z20.literal("edit_pcb_group_location"),
  pcb_group_id: z20.string(),
  original_center: z20.object({ x: z20.number(), y: z20.number() }),
  new_center: z20.object({ x: z20.number(), y: z20.number() })
});
expectTypesMatch(true);

// lib/manual-edits/manual-edit-events/edit_schematic_group_location_event.ts
import { z as z21 } from "zod";
var edit_schematic_group_location_event = base_manual_edit_event.extend({
  edit_event_type: z21.literal("edit_schematic_group_location"),
  schematic_group_id: z21.string(),
  original_center: z21.object({ x: z21.number(), y: z21.number() }),
  new_center: z21.object({ x: z21.number(), y: z21.number() })
});
expectTypesMatch(true);

// lib/manual-edits/manual_edit_event.ts
import { z as z22 } from "zod";
var manual_edit_event = z22.union([
  edit_pcb_component_location_event,
  edit_trace_hint_event,
  edit_schematic_component_location_event
]);
expectTypesMatch(true);

// lib/manual-edits/manual_edits_file.ts
import { z as z26 } from "zod";

// lib/manual-edits/manual_pcb_placement.ts
import { z as z23 } from "zod";
import { point as point2 } from "circuit-json";
var manual_pcb_placement = z23.object({
  selector: z23.string(),
  relative_to: z23.string().optional().default("group_center").describe("Can be a selector or 'group_center'"),
  center: point2
});
expectTypesMatch(true);

// lib/manual-edits/manual_trace_hint.ts
import { z as z24 } from "zod";
import { route_hint_point } from "circuit-json";
var manual_trace_hint = z24.object({
  pcb_port_selector: z24.string(),
  offsets: z24.array(route_hint_point)
});
expectTypesMatch(true);

// lib/manual-edits/manual_schematic_placement.ts
import { z as z25 } from "zod";
import { point as point4 } from "circuit-json";
var manual_schematic_placement = z25.object({
  selector: z25.string(),
  relative_to: z25.string().optional().default("group_center").describe("Can be a selector or 'group_center'"),
  center: point4
});
expectTypesMatch(
  true
);

// lib/manual-edits/manual_edits_file.ts
var manual_edits_file = z26.object({
  pcb_placements: z26.array(manual_pcb_placement).optional(),
  manual_trace_hints: z26.array(manual_trace_hint).optional(),
  schematic_placements: z26.array(manual_schematic_placement).optional()
});
expectTypesMatch(true);

// lib/common/connectionsProp.ts
import { z as z27 } from "zod";
var connectionTarget = z27.string().or(z27.array(z27.string()).readonly()).or(z27.array(z27.string()));
var createConnectionsProp = (labels) => {
  return z27.record(z27.enum(labels), connectionTarget);
};

// lib/components/group.ts
var layoutConfig = z28.object({
  layoutMode: z28.enum(["grid", "flex", "match-adapt", "relative", "none"]).optional(),
  position: z28.enum(["absolute", "relative"]).optional(),
  grid: z28.boolean().optional(),
  gridCols: z28.number().or(z28.string()).optional(),
  gridRows: z28.number().or(z28.string()).optional(),
  gridTemplateRows: z28.string().optional(),
  gridTemplateColumns: z28.string().optional(),
  gridTemplate: z28.string().optional(),
  gridGap: z28.number().or(z28.string()).optional(),
  gridRowGap: z28.number().or(z28.string()).optional(),
  gridColumnGap: z28.number().or(z28.string()).optional(),
  flex: z28.boolean().or(z28.string()).optional(),
  flexDirection: z28.enum(["row", "column"]).optional(),
  alignItems: z28.enum(["start", "center", "end", "stretch"]).optional(),
  justifyContent: z28.enum([
    "start",
    "center",
    "end",
    "stretch",
    "space-between",
    "space-around",
    "space-evenly"
  ]).optional(),
  flexRow: z28.boolean().optional(),
  flexColumn: z28.boolean().optional(),
  gap: z28.number().or(z28.string()).optional(),
  pack: z28.boolean().optional().describe("Pack the contents of this group using a packing strategy"),
  packOrderStrategy: z28.enum([
    "largest_to_smallest",
    "first_to_last",
    "highest_to_lowest_pin_count"
  ]).optional(),
  packPlacementStrategy: z28.enum(["shortest_connection_along_outline"]).optional(),
  padding: length2.optional(),
  paddingLeft: length2.optional(),
  paddingRight: length2.optional(),
  paddingTop: length2.optional(),
  paddingBottom: length2.optional(),
  paddingX: length2.optional(),
  paddingY: length2.optional(),
  width: length2.optional(),
  height: length2.optional(),
  matchAdapt: z28.boolean().optional(),
  matchAdaptTemplate: z28.any().optional()
});
expectTypesMatch(true);
var border = z28.object({
  strokeWidth: length2.optional(),
  dashed: z28.boolean().optional(),
  solid: z28.boolean().optional()
});
var pcbPositionAnchorAutocomplete = z28.custom((value) => typeof value === "string");
var autorouterConfig = z28.object({
  serverUrl: z28.string().optional(),
  inputFormat: z28.enum(["simplified", "circuit-json"]).optional(),
  serverMode: z28.enum(["job", "solve-endpoint"]).optional(),
  serverCacheEnabled: z28.boolean().optional(),
  cache: z28.custom((v) => true).optional(),
  traceClearance: length2.optional(),
  groupMode: z28.enum(["sequential_trace", "subcircuit", "sequential-trace"]).optional(),
  algorithmFn: z28.custom(
    (v) => typeof v === "function" || v === void 0
  ).optional(),
  preset: z28.enum([
    "sequential_trace",
    "subcircuit",
    "auto",
    "auto_local",
    "auto_cloud",
    "freerouting",
    "laser_prefab",
    "sequential-trace",
    "auto-local",
    "auto-cloud"
  ]).optional(),
  local: z28.boolean().optional()
});
var autorouterPreset = z28.union([
  z28.literal("sequential_trace"),
  z28.literal("subcircuit"),
  z28.literal("auto"),
  z28.literal("auto_local"),
  z28.literal("auto_cloud"),
  z28.literal("freerouting"),
  z28.literal("laser_prefab"),
  // Prefabricated PCB with laser copper ablation
  z28.literal("sequential-trace"),
  z28.literal("auto-local"),
  z28.literal("auto-cloud")
]);
var autorouterString = z28.string();
var autorouterProp = z28.union([
  autorouterConfig,
  autorouterPreset,
  autorouterString
]);
var baseGroupProps = commonLayoutProps.extend({
  name: z28.string().optional(),
  children: z28.any().optional(),
  schTitle: z28.string().optional(),
  key: z28.any().optional(),
  showAsSchematicBox: z28.boolean().optional(),
  connections: z28.record(z28.string(), connectionTarget.optional()).optional(),
  schPinArrangement: schematicPinArrangement.optional(),
  schPinSpacing: length2.optional(),
  schPinStyle: schematicPinStyle.optional(),
  ...layoutConfig.shape,
  grid: layoutConfig.shape.grid.describe("@deprecated use pcbGrid"),
  flex: layoutConfig.shape.flex.describe("@deprecated use pcbFlex"),
  pcbGrid: z28.boolean().optional(),
  pcbGridCols: z28.number().or(z28.string()).optional(),
  pcbGridRows: z28.number().or(z28.string()).optional(),
  pcbGridTemplateRows: z28.string().optional(),
  pcbGridTemplateColumns: z28.string().optional(),
  pcbGridTemplate: z28.string().optional(),
  pcbGridGap: z28.number().or(z28.string()).optional(),
  pcbGridRowGap: z28.number().or(z28.string()).optional(),
  pcbGridColumnGap: z28.number().or(z28.string()).optional(),
  pcbFlex: z28.boolean().or(z28.string()).optional(),
  pcbFlexGap: z28.number().or(z28.string()).optional(),
  pcbFlexDirection: z28.enum(["row", "column"]).optional(),
  pcbAlignItems: z28.enum(["start", "center", "end", "stretch"]).optional(),
  pcbJustifyContent: z28.enum([
    "start",
    "center",
    "end",
    "stretch",
    "space-between",
    "space-around",
    "space-evenly"
  ]).optional(),
  pcbFlexRow: z28.boolean().optional(),
  pcbFlexColumn: z28.boolean().optional(),
  pcbGap: z28.number().or(z28.string()).optional(),
  pcbPack: z28.boolean().optional(),
  pcbPackGap: z28.number().or(z28.string()).optional(),
  schGrid: z28.boolean().optional(),
  schGridCols: z28.number().or(z28.string()).optional(),
  schGridRows: z28.number().or(z28.string()).optional(),
  schGridTemplateRows: z28.string().optional(),
  schGridTemplateColumns: z28.string().optional(),
  schGridTemplate: z28.string().optional(),
  schGridGap: z28.number().or(z28.string()).optional(),
  schGridRowGap: z28.number().or(z28.string()).optional(),
  schGridColumnGap: z28.number().or(z28.string()).optional(),
  schFlex: z28.boolean().or(z28.string()).optional(),
  schFlexGap: z28.number().or(z28.string()).optional(),
  schFlexDirection: z28.enum(["row", "column"]).optional(),
  schAlignItems: z28.enum(["start", "center", "end", "stretch"]).optional(),
  schJustifyContent: z28.enum([
    "start",
    "center",
    "end",
    "stretch",
    "space-between",
    "space-around",
    "space-evenly"
  ]).optional(),
  schFlexRow: z28.boolean().optional(),
  schFlexColumn: z28.boolean().optional(),
  schGap: z28.number().or(z28.string()).optional(),
  schPack: z28.boolean().optional(),
  schMatchAdapt: z28.boolean().optional(),
  pcbWidth: length2.optional(),
  pcbHeight: length2.optional(),
  minTraceWidth: length2.optional(),
  schWidth: length2.optional(),
  schHeight: length2.optional(),
  pcbLayout: layoutConfig.optional(),
  schLayout: layoutConfig.optional(),
  cellBorder: border.nullable().optional(),
  border: border.nullable().optional(),
  schPadding: length2.optional(),
  schPaddingLeft: length2.optional(),
  schPaddingRight: length2.optional(),
  schPaddingTop: length2.optional(),
  schPaddingBottom: length2.optional(),
  pcbPadding: length2.optional(),
  pcbPaddingLeft: length2.optional(),
  pcbPaddingRight: length2.optional(),
  pcbPaddingTop: length2.optional(),
  pcbPaddingBottom: length2.optional(),
  pcbPositionAnchor: pcbPositionAnchorAutocomplete.optional()
});
var partsEngine = z28.custom((v) => "findPart" in v);
var subcircuitGroupProps = baseGroupProps.extend({
  manualEdits: manual_edits_file.optional(),
  schAutoLayoutEnabled: z28.boolean().optional(),
  schTraceAutoLabelEnabled: z28.boolean().optional(),
  schMaxTraceDistance: distance7.optional(),
  routingDisabled: z28.boolean().optional(),
  defaultTraceWidth: length2.optional(),
  minTraceWidth: length2.optional(),
  partsEngine: partsEngine.optional(),
  pcbRouteCache: z28.custom((v) => true).optional(),
  autorouter: autorouterProp.optional(),
  square: z28.boolean().optional(),
  emptyArea: z28.string().optional(),
  filledArea: z28.string().optional(),
  width: distance7.optional(),
  height: distance7.optional(),
  outline: z28.array(point).optional(),
  outlineOffsetX: distance7.optional(),
  outlineOffsetY: distance7.optional(),
  circuitJson: z28.array(z28.any()).optional()
});
var subcircuitGroupPropsWithBool = subcircuitGroupProps.extend({
  subcircuit: z28.literal(true)
});
var groupProps = z28.discriminatedUnion("subcircuit", [
  baseGroupProps.extend({ subcircuit: z28.literal(false).optional() }),
  subcircuitGroupPropsWithBool
]);
expectTypesMatch(true);
expectTypesMatch(true);
expectTypesMatch(true);
expectTypesMatch(true);

// lib/components/board.ts
var boardColor = z29.custom((value) => typeof value === "string");
var boardProps = subcircuitGroupProps.omit({ connections: true }).extend({
  material: z29.enum(["fr4", "fr1"]).default("fr4"),
  layers: z29.union([z29.literal(2), z29.literal(4)]).default(2),
  borderRadius: distance.optional(),
  thickness: distance.optional(),
  boardAnchorPosition: point.optional(),
  boardAnchorAlignment: ninePointAnchor.optional(),
  title: z29.string().optional(),
  solderMaskColor: boardColor.optional(),
  topSolderMaskColor: boardColor.optional(),
  bottomSolderMaskColor: boardColor.optional(),
  silkscreenColor: boardColor.optional(),
  topSilkscreenColor: boardColor.optional(),
  bottomSilkscreenColor: boardColor.optional(),
  doubleSidedAssembly: z29.boolean().optional().default(false),
  schematicDisabled: z29.boolean().optional()
});
expectTypesMatch(true);

// lib/components/panel.ts
import { z as z30 } from "zod";
var panelProps = baseGroupProps.omit({
  width: true,
  height: true,
  children: true
}).extend({
  width: distance,
  height: distance,
  children: z30.any().optional(),
  noSolderMask: z30.boolean().optional()
});
expectTypesMatch(true);

// lib/components/breakout.ts
import { distance as distance8 } from "circuit-json";
import "zod";
var breakoutProps = subcircuitGroupProps.extend({
  padding: distance8.optional(),
  paddingLeft: distance8.optional(),
  paddingRight: distance8.optional(),
  paddingTop: distance8.optional(),
  paddingBottom: distance8.optional()
});
expectTypesMatch(true);

// lib/components/chip.ts
import { distance as distance9, supplier_name as supplier_name2 } from "circuit-json";
import { z as z32 } from "zod";
var connectionTarget2 = z32.string().or(z32.array(z32.string()).readonly()).or(z32.array(z32.string()));
var connectionsProp = z32.custom().pipe(z32.record(z32.string(), connectionTarget2));
var pinLabelsProp = z32.record(
  schematicPinLabel,
  schematicPinLabel.or(z32.array(schematicPinLabel).readonly()).or(z32.array(schematicPinLabel))
);
expectTypesMatch(true);
var pinCompatibleVariant = z32.object({
  manufacturerPartNumber: z32.string().optional(),
  supplierPartNumber: z32.record(supplier_name2, z32.array(z32.string())).optional()
});
var chipProps = commonComponentProps.extend({
  manufacturerPartNumber: z32.string().optional(),
  pinLabels: pinLabelsProp.optional(),
  showPinAliases: z32.boolean().optional(),
  pcbPinLabels: z32.record(z32.string(), z32.string()).optional(),
  internallyConnectedPins: z32.array(z32.array(z32.union([z32.string(), z32.number()]))).optional(),
  externallyConnectedPins: z32.array(z32.array(z32.string())).optional(),
  schPinArrangement: schematicPortArrangement.optional(),
  schPortArrangement: schematicPortArrangement.optional(),
  pinCompatibleVariants: z32.array(pinCompatibleVariant).optional(),
  schPinStyle: schematicPinStyle.optional(),
  schPinSpacing: distance9.optional(),
  schWidth: distance9.optional(),
  schHeight: distance9.optional(),
  noSchematicRepresentation: z32.boolean().optional(),
  connections: connectionsProp.optional()
});
var bugProps = chipProps;
expectTypesMatch(true);

// lib/components/pinout.ts
var pinoutProps = chipProps;
expectTypesMatch(true);

// lib/components/jumper.ts
import { distance as distance10 } from "circuit-json";
import { z as z33 } from "zod";
var jumperProps = commonComponentProps.extend({
  manufacturerPartNumber: z33.string().optional(),
  pinLabels: z33.record(
    z33.number().or(schematicPinLabel),
    schematicPinLabel.or(z33.array(schematicPinLabel))
  ).optional(),
  schPinStyle: schematicPinStyle.optional(),
  schPinSpacing: distance10.optional(),
  schWidth: distance10.optional(),
  schHeight: distance10.optional(),
  schDirection: z33.enum(["left", "right"]).optional(),
  schPinArrangement: schematicPinArrangement.optional(),
  schPortArrangement: schematicPortArrangement.optional(),
  pcbPinLabels: z33.record(z33.string(), z33.string()).optional(),
  pinCount: z33.union([z33.literal(2), z33.literal(3)]).optional(),
  internallyConnectedPins: z33.array(z33.array(z33.union([z33.string(), z33.number()]))).optional(),
  connections: z33.custom().pipe(z33.record(z33.string(), connectionTarget)).optional()
});
expectTypesMatch(true);

// lib/components/solderjumper.ts
import { z as z34 } from "zod";
var solderjumperProps = jumperProps.extend({
  bridgedPins: z34.array(z34.array(z34.string())).optional(),
  bridged: z34.boolean().optional()
});
expectTypesMatch(true);

// lib/components/connector.ts
import { distance as distance11 } from "circuit-json";
import { z as z35 } from "zod";
var connectorProps = commonComponentProps.extend({
  manufacturerPartNumber: z35.string().optional(),
  pinLabels: z35.record(
    z35.number().or(schematicPinLabel),
    schematicPinLabel.or(z35.array(schematicPinLabel))
  ).optional(),
  schPinStyle: schematicPinStyle.optional(),
  schPinSpacing: distance11.optional(),
  schWidth: distance11.optional(),
  schHeight: distance11.optional(),
  schDirection: z35.enum(["left", "right"]).optional(),
  schPortArrangement: schematicPortArrangement.optional(),
  internallyConnectedPins: z35.array(z35.array(z35.union([z35.string(), z35.number()]))).optional(),
  standard: z35.enum(["usb_c", "m2"]).optional()
});
expectTypesMatch(true);

// lib/components/fuse.ts
import { z as z36 } from "zod";
var fusePinLabels = ["pin1", "pin2"];
var fuseProps = commonComponentProps.extend({
  currentRating: z36.union([z36.number(), z36.string()]),
  voltageRating: z36.union([z36.number(), z36.string()]).optional(),
  schShowRatings: z36.boolean().optional(),
  schOrientation: schematicOrientation.optional(),
  connections: z36.record(
    z36.string(),
    z36.union([
      z36.string(),
      z36.array(z36.string()).readonly(),
      z36.array(z36.string())
    ])
  ).optional()
});

// lib/components/platedhole.ts
import { distance as distance12 } from "circuit-json";
import { z as z37 } from "zod";
var distanceHiddenUndefined = z37.custom().transform((a) => {
  if (a === void 0) return void 0;
  return distance12.parse(a);
});
var platedHoleProps = z37.discriminatedUnion("shape", [
  pcbLayoutProps.omit({ pcbRotation: true, layer: true }).extend({
    name: z37.string().optional(),
    connectsTo: z37.string().or(z37.array(z37.string())).optional(),
    shape: z37.literal("circle"),
    holeDiameter: distance12,
    outerDiameter: distance12,
    portHints: portHints.optional()
  }),
  pcbLayoutProps.omit({ layer: true }).extend({
    name: z37.string().optional(),
    connectsTo: z37.string().or(z37.array(z37.string())).optional(),
    shape: z37.literal("oval"),
    outerWidth: distance12,
    outerHeight: distance12,
    holeWidth: distanceHiddenUndefined,
    holeHeight: distanceHiddenUndefined,
    innerWidth: distance12.optional().describe("DEPRECATED use holeWidth"),
    innerHeight: distance12.optional().describe("DEPRECATED use holeHeight"),
    portHints: portHints.optional()
  }),
  pcbLayoutProps.omit({ layer: true }).extend({
    name: z37.string().optional(),
    connectsTo: z37.string().or(z37.array(z37.string())).optional(),
    shape: z37.literal("pill"),
    rectPad: z37.boolean().optional(),
    outerWidth: distance12,
    outerHeight: distance12,
    holeWidth: distanceHiddenUndefined,
    holeHeight: distanceHiddenUndefined,
    innerWidth: distance12.optional().describe("DEPRECATED use holeWidth"),
    innerHeight: distance12.optional().describe("DEPRECATED use holeHeight"),
    portHints: portHints.optional(),
    holeOffsetX: distance12.optional(),
    holeOffsetY: distance12.optional()
  }),
  pcbLayoutProps.omit({ pcbRotation: true, layer: true }).extend({
    name: z37.string().optional(),
    connectsTo: z37.string().or(z37.array(z37.string())).optional(),
    shape: z37.literal("circular_hole_with_rect_pad"),
    holeDiameter: distance12,
    rectPadWidth: distance12,
    rectPadHeight: distance12,
    rectBorderRadius: distance12.optional(),
    holeShape: z37.literal("circle").optional(),
    padShape: z37.literal("rect").optional(),
    portHints: portHints.optional(),
    holeOffsetX: distance12.optional(),
    holeOffsetY: distance12.optional()
  }),
  pcbLayoutProps.omit({ pcbRotation: true, layer: true }).extend({
    name: z37.string().optional(),
    connectsTo: z37.string().or(z37.array(z37.string())).optional(),
    shape: z37.literal("pill_hole_with_rect_pad"),
    holeShape: z37.literal("pill"),
    padShape: z37.literal("rect"),
    holeWidth: distance12,
    holeHeight: distance12,
    rectPadWidth: distance12,
    rectPadHeight: distance12,
    portHints: portHints.optional(),
    holeOffsetX: distance12.optional(),
    holeOffsetY: distance12.optional()
  }),
  pcbLayoutProps.omit({ pcbRotation: true, layer: true }).extend({
    name: z37.string().optional(),
    connectsTo: z37.string().or(z37.array(z37.string())).optional(),
    shape: z37.literal("hole_with_polygon_pad"),
    holeShape: z37.enum(["circle", "oval", "pill", "rotated_pill"]),
    holeDiameter: distance12.optional(),
    holeWidth: distance12.optional(),
    holeHeight: distance12.optional(),
    padOutline: z37.array(point),
    holeOffsetX: distance12,
    holeOffsetY: distance12,
    portHints: portHints.optional()
  })
]).refine((a) => {
  if ("innerWidth" in a && a.innerWidth !== void 0) {
    a.holeWidth ??= a.innerWidth;
  }
  if ("innerHeight" in a && a.innerHeight !== void 0) {
    a.holeHeight ??= a.innerHeight;
  }
  return a;
});
expectTypesMatch(true);

// lib/components/resistor.ts
import { resistance } from "circuit-json";
import { z as z38 } from "zod";
var resistorPinLabels = ["pin1", "pin2", "pos", "neg"];
var resistorProps = commonComponentProps.extend({
  resistance,
  pullupFor: z38.string().optional(),
  pullupTo: z38.string().optional(),
  pulldownFor: z38.string().optional(),
  pulldownTo: z38.string().optional(),
  schOrientation: schematicOrientation.optional(),
  connections: createConnectionsProp(resistorPinLabels).optional()
});
var resistorPins = lrPins;
expectTypesMatch(true);

// lib/components/potentiometer.ts
import { resistance as resistance2 } from "circuit-json";
import { z as z39 } from "zod";
var potentiometerProps = commonComponentProps.extend({
  maxResistance: resistance2,
  pinVariant: z39.enum(["two_pin", "three_pin"]).optional()
});
expectTypesMatch(true);

// lib/components/crystal.ts
import { frequency, capacitance } from "circuit-json";
import { z as z40 } from "zod";
var crystalPins = lrPins;
var crystalProps = commonComponentProps.extend({
  frequency,
  loadCapacitance: capacitance,
  manufacturerPartNumber: z40.string().optional(),
  mpn: z40.string().optional(),
  pinVariant: z40.enum(["two_pin", "four_pin"]).optional(),
  schOrientation: schematicOrientation.optional(),
  connections: createConnectionsProp(crystalPins).optional()
});
expectTypesMatch(true);

// lib/components/resonator.ts
import { frequency as frequency2, capacitance as capacitance2 } from "circuit-json";
import { z as z41 } from "zod";
var resonatorProps = commonComponentProps.extend({
  frequency: frequency2,
  loadCapacitance: capacitance2,
  pinVariant: z41.enum(["no_ground", "ground_pin", "two_ground_pins"]).optional()
});
expectTypesMatch(true);

// lib/components/stampboard.ts
import { distance as distance13 } from "circuit-json";
import { z as z42 } from "zod";
var stampboardProps = boardProps.extend({
  leftPinCount: z42.number().optional(),
  rightPinCount: z42.number().optional(),
  topPinCount: z42.number().optional(),
  bottomPinCount: z42.number().optional(),
  leftPins: z42.array(z42.string()).optional(),
  rightPins: z42.array(z42.string()).optional(),
  topPins: z42.array(z42.string()).optional(),
  bottomPins: z42.array(z42.string()).optional(),
  pinPitch: distance13.optional(),
  innerHoles: z42.boolean().optional()
});
expectTypesMatch(true);

// lib/components/capacitor.ts
import { capacitance as capacitance3, voltage } from "circuit-json";
import { z as z43 } from "zod";
var capacitorPinLabels = [
  "pin1",
  "pin2",
  "pos",
  "neg",
  "anode",
  "cathode"
];
var capacitorProps = commonComponentProps.extend({
  capacitance: capacitance3,
  maxVoltageRating: voltage.optional(),
  schShowRatings: z43.boolean().optional().default(false),
  polarized: z43.boolean().optional().default(false),
  decouplingFor: z43.string().optional(),
  decouplingTo: z43.string().optional(),
  bypassFor: z43.string().optional(),
  bypassTo: z43.string().optional(),
  maxDecouplingTraceLength: z43.number().optional(),
  schOrientation: schematicOrientation.optional(),
  connections: createConnectionsProp(capacitorPinLabels).optional()
});
var capacitorPins = lrPolarPins;
expectTypesMatch(true);

// lib/components/net.ts
import { z as z44 } from "zod";
var netProps = z44.object({
  name: z44.string(),
  connectsTo: z44.string().or(z44.array(z44.string())).optional(),
  highlightColor: z44.string().optional(),
  isPowerNet: z44.boolean().optional(),
  isGroundNet: z44.boolean().optional()
});
expectTypesMatch(true);

// lib/components/constrainedlayout.ts
import { z as z45 } from "zod";
var constrainedLayoutProps = z45.object({
  name: z45.string().optional(),
  pcbOnly: z45.boolean().optional(),
  schOnly: z45.boolean().optional()
});
expectTypesMatch(true);

// lib/components/constraint.ts
import { z as z46 } from "zod";
var pcbXDistConstraintProps = z46.object({
  pcb: z46.literal(true).optional(),
  xDist: distance,
  left: z46.string(),
  right: z46.string(),
  edgeToEdge: z46.literal(true).optional(),
  centerToCenter: z46.literal(true).optional()
});
expectTypesMatch(
  true
);
var pcbYDistConstraintProps = z46.object({
  pcb: z46.literal(true).optional(),
  yDist: distance,
  top: z46.string(),
  bottom: z46.string(),
  edgeToEdge: z46.literal(true).optional(),
  centerToCenter: z46.literal(true).optional()
});
expectTypesMatch(
  true
);
var pcbSameYConstraintProps = z46.object({
  pcb: z46.literal(true).optional(),
  sameY: z46.literal(true).optional(),
  for: z46.array(z46.string())
});
expectTypesMatch(
  true
);
var pcbSameXConstraintProps = z46.object({
  pcb: z46.literal(true).optional(),
  sameX: z46.literal(true).optional(),
  for: z46.array(z46.string())
});
expectTypesMatch(
  true
);
var constraintProps = z46.union([
  pcbXDistConstraintProps,
  pcbYDistConstraintProps,
  pcbSameYConstraintProps,
  pcbSameXConstraintProps
]);
expectTypesMatch(true);

// lib/components/cutout.ts
import { z as z47 } from "zod";
var rectCutoutProps = pcbLayoutProps.omit({
  layer: true,
  pcbRotation: true
}).extend({
  name: z47.string().optional(),
  shape: z47.literal("rect"),
  width: distance,
  height: distance
});
expectTypesMatch(true);
var circleCutoutProps = pcbLayoutProps.omit({
  layer: true,
  pcbRotation: true
}).extend({
  name: z47.string().optional(),
  shape: z47.literal("circle"),
  radius: distance
});
expectTypesMatch(true);
var polygonCutoutProps = pcbLayoutProps.omit({
  layer: true,
  pcbRotation: true
}).extend({
  name: z47.string().optional(),
  shape: z47.literal("polygon"),
  points: z47.array(point)
});
expectTypesMatch(true);
var cutoutProps = z47.discriminatedUnion("shape", [
  rectCutoutProps,
  circleCutoutProps,
  polygonCutoutProps
]);

// lib/components/smtpad.ts
import { z as z48 } from "zod";
var rectSmtPadProps = pcbLayoutProps.omit({ pcbRotation: true }).extend({
  name: z48.string().optional(),
  shape: z48.literal("rect"),
  width: distance,
  height: distance,
  rectBorderRadius: distance.optional(),
  cornerRadius: distance.optional(),
  portHints: portHints.optional(),
  coveredWithSolderMask: z48.boolean().optional()
});
expectTypesMatch(true);
var rotatedRectSmtPadProps = pcbLayoutProps.omit({ pcbRotation: true }).extend({
  name: z48.string().optional(),
  shape: z48.literal("rotated_rect"),
  width: distance,
  height: distance,
  ccwRotation: z48.number(),
  cornerRadius: distance.optional(),
  portHints: portHints.optional(),
  coveredWithSolderMask: z48.boolean().optional()
});
expectTypesMatch(true);
var circleSmtPadProps = pcbLayoutProps.omit({ pcbRotation: true }).extend({
  name: z48.string().optional(),
  shape: z48.literal("circle"),
  radius: distance,
  portHints: portHints.optional(),
  coveredWithSolderMask: z48.boolean().optional()
});
expectTypesMatch(true);
var pillSmtPadProps = pcbLayoutProps.omit({ pcbRotation: true }).extend({
  name: z48.string().optional(),
  shape: z48.literal("pill"),
  width: distance,
  height: distance,
  radius: distance,
  portHints: portHints.optional(),
  coveredWithSolderMask: z48.boolean().optional()
});
expectTypesMatch(true);
var polygonSmtPadProps = pcbLayoutProps.omit({ pcbRotation: true }).extend({
  name: z48.string().optional(),
  shape: z48.literal("polygon"),
  points: z48.array(point),
  portHints: portHints.optional(),
  coveredWithSolderMask: z48.boolean().optional()
});
expectTypesMatch(true);
var smtPadProps = z48.discriminatedUnion("shape", [
  circleSmtPadProps,
  rectSmtPadProps,
  rotatedRectSmtPadProps,
  pillSmtPadProps,
  polygonSmtPadProps
]);
expectTypesMatch(true);

// lib/components/solderpaste.ts
import { z as z49 } from "zod";
var rectSolderPasteProps = pcbLayoutProps.omit({ pcbRotation: true }).extend({
  shape: z49.literal("rect"),
  width: distance,
  height: distance
});
expectTypesMatch(true);
var circleSolderPasteProps = pcbLayoutProps.omit({ pcbRotation: true }).extend({
  shape: z49.literal("circle"),
  radius: distance
});
expectTypesMatch(true);
var solderPasteProps = z49.union([
  circleSolderPasteProps,
  rectSolderPasteProps
]);
expectTypesMatch(true);

// lib/components/hole.ts
import { z as z50 } from "zod";
var circleHoleProps = pcbLayoutProps.extend({
  name: z50.string().optional(),
  shape: z50.literal("circle").optional(),
  diameter: distance.optional(),
  radius: distance.optional()
}).transform((d) => ({
  ...d,
  diameter: d.diameter ?? 2 * d.radius,
  radius: d.radius ?? d.diameter / 2
}));
var pillHoleProps = pcbLayoutProps.extend({
  name: z50.string().optional(),
  shape: z50.literal("pill"),
  width: distance,
  height: distance
});
var rectHoleProps = pcbLayoutProps.extend({
  name: z50.string().optional(),
  shape: z50.literal("rect"),
  width: distance,
  height: distance
});
var holeProps = z50.union([
  circleHoleProps,
  pillHoleProps,
  rectHoleProps
]);
expectTypesMatch(true);

// lib/components/trace.ts
import { distance as distance14, route_hint_point as route_hint_point2 } from "circuit-json";
import { z as z51 } from "zod";
var portRef = z51.union([
  z51.string(),
  z51.custom(
    (v) => Boolean(v.getPortSelector)
  )
]);
var baseTraceProps = z51.object({
  key: z51.string().optional(),
  thickness: distance14.optional(),
  width: distance14.optional().describe("Alias for trace thickness"),
  schematicRouteHints: z51.array(point).optional(),
  pcbRouteHints: z51.array(route_hint_point2).optional(),
  pcbPathRelativeTo: z51.string().optional(),
  pcbPath: z51.array(z51.union([point, z51.string()])).optional(),
  schDisplayLabel: z51.string().optional(),
  schStroke: z51.string().optional(),
  highlightColor: z51.string().optional(),
  maxLength: distance14.optional()
});
var traceProps = z51.union([
  baseTraceProps.extend({
    path: z51.array(portRef)
  }),
  baseTraceProps.extend({
    from: portRef,
    to: portRef
  })
]);

// lib/components/footprint.ts
import { layer_ref as layer_ref4 } from "circuit-json";
import { z as z52 } from "zod";
var footprintProps = z52.object({
  children: z52.any().optional(),
  originalLayer: layer_ref4.default("top").optional(),
  circuitJson: z52.array(z52.any()).optional()
});
expectTypesMatch(true);

// lib/components/symbol.ts
import { z as z53 } from "zod";
var symbolProps = z53.object({
  originalFacingDirection: z53.enum(["up", "down", "left", "right"]).default("right").optional()
});
expectTypesMatch(true);

// lib/components/battery.ts
import { voltage as voltage2 } from "circuit-json";
import { z as z54 } from "zod";
var capacity = z54.number().or(z54.string().endsWith("mAh")).transform((v) => {
  if (typeof v === "string") {
    const valString = v.replace("mAh", "");
    const num = Number.parseFloat(valString);
    if (Number.isNaN(num)) {
      throw new Error("Invalid capacity");
    }
    return num;
  }
  return v;
}).describe("Battery capacity in mAh");
var batteryProps = commonComponentProps.extend({
  capacity: capacity.optional(),
  voltage: voltage2.optional(),
  standard: z54.enum(["AA", "AAA", "9V", "CR2032", "18650", "C"]).optional(),
  schOrientation: schematicOrientation.optional()
});
var batteryPins = lrPolarPins;
expectTypesMatch(true);

// lib/components/pin-header.ts
import { distance as distance15 } from "circuit-json";

// lib/common/pcbOrientation.ts
import { z as z55 } from "zod";
var pcbOrientation = z55.enum(["vertical", "horizontal"]).describe(
  "vertical means pins go 1->2 downward and horizontal means pins go 1->2 rightward"
);
expectTypesMatch(true);

// lib/components/pin-header.ts
import { z as z56 } from "zod";
var pinHeaderProps = commonComponentProps.extend({
  pinCount: z56.number(),
  pitch: distance15.optional(),
  schFacingDirection: z56.enum(["up", "down", "left", "right"]).optional(),
  gender: z56.enum(["male", "female", "unpopulated"]).optional().default("male"),
  showSilkscreenPinLabels: z56.boolean().optional(),
  pcbPinLabels: z56.record(z56.string(), z56.string()).optional(),
  doubleRow: z56.boolean().optional(),
  rightAngle: z56.boolean().optional(),
  pcbOrientation: pcbOrientation.optional(),
  holeDiameter: distance15.optional(),
  platedDiameter: distance15.optional(),
  pinLabels: z56.record(z56.string(), schematicPinLabel).or(z56.array(schematicPinLabel)).optional(),
  connections: z56.custom().pipe(z56.record(z56.string(), connectionTarget)).optional(),
  facingDirection: z56.enum(["left", "right"]).optional(),
  schPinArrangement: schematicPinArrangement.optional(),
  schPinStyle: schematicPinStyle.optional(),
  schPinSpacing: distance15.optional(),
  schWidth: distance15.optional(),
  schHeight: distance15.optional()
});
expectTypesMatch(true);

// lib/components/netalias.ts
import { z as z57 } from "zod";
import { rotation as rotation2 } from "circuit-json";
var netAliasProps = z57.object({
  net: z57.string().optional(),
  connection: z57.string().optional(),
  schX: distance.optional(),
  schY: distance.optional(),
  schRotation: rotation2.optional(),
  anchorSide: z57.enum(["left", "top", "right", "bottom"]).optional()
});
expectTypesMatch(true);

// lib/components/netlabel.ts
import { z as z58 } from "zod";
import { rotation as rotation3 } from "circuit-json";
var netLabelProps = z58.object({
  net: z58.string().optional(),
  connection: z58.string().optional(),
  connectsTo: z58.string().or(z58.array(z58.string())).optional(),
  schX: distance.optional(),
  schY: distance.optional(),
  schRotation: rotation3.optional(),
  anchorSide: z58.enum(["left", "top", "right", "bottom"]).optional()
});
expectTypesMatch(true);

// lib/components/push-button.ts
import "zod";
var pushButtonProps = chipProps.extend({});
expectTypesMatch(true);

// lib/components/subcircuit.ts
var subcircuitProps = subcircuitGroupProps;
expectTypesMatch(true);

// lib/components/analogsimulation.ts
import { ms } from "circuit-json";
import { z as z60 } from "zod";
var spiceEngine = z60.custom(
  (value) => typeof value === "string"
);
var analogSimulationProps = z60.object({
  simulationType: z60.literal("spice_transient_analysis").default("spice_transient_analysis"),
  duration: ms.optional(),
  timePerStep: ms.optional(),
  spiceEngine: spiceEngine.optional()
});
expectTypesMatch(
  true
);

// lib/components/transistor.ts
import { z as z61 } from "zod";
var transistorPinsLabels = [
  "pin1",
  "pin2",
  "pin3",
  "emitter",
  "collector",
  "base",
  "gate",
  "source",
  "drain"
];
var transistorProps = commonComponentProps.extend({
  type: z61.enum(["npn", "pnp", "bjt", "jfet", "mosfet", "igbt"]),
  connections: createConnectionsProp(transistorPinsLabels).optional()
});
var transistorPins = [
  "pin1",
  "emitter",
  "pin2",
  "collector",
  "pin3",
  "base"
];
expectTypesMatch(true);

// lib/components/mosfet.ts
import { z as z62 } from "zod";
var mosfetProps = commonComponentProps.extend({
  channelType: z62.enum(["n", "p"]),
  mosfetMode: z62.enum(["enhancement", "depletion"])
});
var mosfetPins = [
  "pin1",
  "drain",
  "pin2",
  "source",
  "pin3",
  "gate"
];
expectTypesMatch(true);

// lib/components/inductor.ts
import { inductance } from "circuit-json";
import { z as z63 } from "zod";
var inductorPins = lrPins;
var inductorProps = commonComponentProps.extend({
  inductance,
  maxCurrentRating: z63.union([z63.string(), z63.number()]).optional(),
  schOrientation: schematicOrientation.optional(),
  connections: createConnectionsProp(inductorPins).optional()
});
expectTypesMatch(true);

// lib/components/diode.ts
import { z as z64 } from "zod";
var diodeConnectionKeys = z64.enum([
  "anode",
  "cathode",
  "pin1",
  "pin2",
  "pos",
  "neg"
]);
var connectionTarget3 = z64.string().or(z64.array(z64.string()).readonly()).or(z64.array(z64.string()));
var connectionsProp2 = z64.record(diodeConnectionKeys, connectionTarget3);
var diodeVariant = z64.enum([
  "standard",
  "schottky",
  "zener",
  "avalanche",
  "photo",
  "tvs"
]);
var diodeProps = commonComponentProps.extend({
  connections: connectionsProp2.optional(),
  variant: diodeVariant.optional().default("standard"),
  standard: z64.boolean().optional(),
  schottky: z64.boolean().optional(),
  zener: z64.boolean().optional(),
  avalanche: z64.boolean().optional(),
  photo: z64.boolean().optional(),
  tvs: z64.boolean().optional(),
  schOrientation: schematicOrientation.optional()
}).superRefine((data, ctx) => {
  const enabledFlags = [
    data.standard,
    data.schottky,
    data.zener,
    data.avalanche,
    data.photo,
    data.tvs
  ].filter(Boolean).length;
  if (enabledFlags > 1) {
    ctx.addIssue({
      code: z64.ZodIssueCode.custom,
      message: "Exactly one diode variant must be enabled",
      path: []
    });
    return z64.INVALID;
  }
}).transform((data) => {
  const result = {
    ...data,
    standard: false,
    schottky: false,
    zener: false,
    avalanche: false,
    photo: false,
    tvs: false
  };
  if (data.standard) result.standard = true;
  else if (data.schottky) result.schottky = true;
  else if (data.zener) result.zener = true;
  else if (data.photo) result.photo = true;
  else if (data.tvs) result.tvs = true;
  else {
    switch (data.variant) {
      case "standard":
        result.standard = true;
        break;
      case "schottky":
        result.schottky = true;
        break;
      case "zener":
        result.zener = true;
        break;
      case "avalanche":
        result.avalanche = true;
        break;
      case "photo":
        result.photo = true;
        break;
      case "tvs":
        result.tvs = true;
        break;
      default:
        result.standard = true;
    }
  }
  return result;
});
var diodePins = lrPolarPins;
expectTypesMatch(true);

// lib/components/led.ts
import { z as z65 } from "zod";
var ledProps = commonComponentProps.extend({
  color: z65.string().optional(),
  wavelength: z65.string().optional(),
  schDisplayValue: z65.string().optional(),
  schOrientation: schematicOrientation.optional(),
  connections: createConnectionsProp(lrPolarPins).optional(),
  laser: z65.boolean().optional()
});
var ledPins = lrPolarPins;

// lib/components/switch.ts
import { ms as ms2, frequency as frequency3 } from "circuit-json";
import { z as z66 } from "zod";
var switchProps = commonComponentProps.extend({
  type: z66.enum(["spst", "spdt", "dpst", "dpdt"]).optional(),
  isNormallyClosed: z66.boolean().optional().default(false),
  spst: z66.boolean().optional(),
  spdt: z66.boolean().optional(),
  dpst: z66.boolean().optional(),
  dpdt: z66.boolean().optional(),
  simSwitchFrequency: frequency3.optional(),
  simCloseAt: ms2.optional(),
  simOpenAt: ms2.optional(),
  simStartClosed: z66.boolean().optional(),
  simStartOpen: z66.boolean().optional(),
  connections: z66.custom().pipe(z66.record(z66.string(), connectionTarget)).optional()
}).transform((props) => {
  const updatedProps = { ...props };
  if (updatedProps.dpdt) {
    updatedProps.type = "dpdt";
  } else if (updatedProps.spst) {
    updatedProps.type = "spst";
  } else if (updatedProps.spdt) {
    updatedProps.type = "spdt";
  } else if (updatedProps.dpst) {
    updatedProps.type = "dpst";
  }
  switch (updatedProps.type) {
    case "spdt":
      updatedProps.spdt = true;
      break;
    case "spst":
      updatedProps.spst = true;
      break;
    case "dpst":
      updatedProps.dpst = true;
      break;
    case "dpdt":
      updatedProps.dpdt = true;
      break;
  }
  return updatedProps;
});
expectTypesMatch(true);

// lib/components/fabrication-note-text.ts
import { length as length3 } from "circuit-json";
import { z as z67 } from "zod";
var fabricationNoteTextProps = pcbLayoutProps.extend({
  text: z67.string(),
  anchorAlignment: z67.enum(["center", "top_left", "top_right", "bottom_left", "bottom_right"]).default("center"),
  font: z67.enum(["tscircuit2024"]).optional(),
  fontSize: length3.optional(),
  color: z67.string().optional()
});
expectTypesMatch(true);

// lib/components/fabrication-note-rect.ts
import { distance as distance16 } from "circuit-json";
import { z as z68 } from "zod";
var fabricationNoteRectProps = pcbLayoutProps.omit({ pcbRotation: true }).extend({
  width: distance16,
  height: distance16,
  strokeWidth: distance16.optional(),
  isFilled: z68.boolean().optional(),
  hasStroke: z68.boolean().optional(),
  isStrokeDashed: z68.boolean().optional(),
  color: z68.string().optional(),
  cornerRadius: distance16.optional()
});

// lib/components/fabrication-note-path.ts
import { length as length4, route_hint_point as route_hint_point3 } from "circuit-json";
import { z as z69 } from "zod";
var fabricationNotePathProps = pcbLayoutProps.omit({
  pcbX: true,
  pcbY: true,
  pcbOffsetX: true,
  pcbOffsetY: true,
  pcbRotation: true
}).extend({
  route: z69.array(route_hint_point3),
  strokeWidth: length4.optional(),
  color: z69.string().optional()
});

// lib/components/fabrication-note-dimension.ts
import { distance as distance17, length as length5 } from "circuit-json";
import { z as z70 } from "zod";
var dimensionTarget = z70.union([z70.string(), point]);
var fabricationNoteDimensionProps = pcbLayoutProps.omit({
  pcbX: true,
  pcbY: true,
  pcbOffsetX: true,
  pcbOffsetY: true,
  pcbRotation: true
}).extend({
  from: dimensionTarget,
  to: dimensionTarget,
  text: z70.string().optional(),
  offset: distance17.optional(),
  font: z70.enum(["tscircuit2024"]).optional(),
  fontSize: length5.optional(),
  color: z70.string().optional(),
  arrowSize: distance17.optional(),
  units: z70.enum(["in", "mm"]).optional(),
  outerEdgeToEdge: z70.literal(true).optional(),
  centerToCenter: z70.literal(true).optional(),
  innerEdgeToEdge: z70.literal(true).optional()
});
expectTypesMatch(true);

// lib/components/pcb-trace.ts
import { distance as distance18, route_hint_point as route_hint_point4 } from "circuit-json";
import { z as z71 } from "zod";
var pcbTraceProps = z71.object({
  layer: z71.string().optional(),
  thickness: distance18.optional(),
  route: z71.array(route_hint_point4)
});

// lib/components/via.ts
import { distance as distance19, layer_ref as layer_ref5 } from "circuit-json";
import { z as z72 } from "zod";
var viaProps = commonLayoutProps.extend({
  name: z72.string().optional(),
  fromLayer: layer_ref5,
  toLayer: layer_ref5,
  holeDiameter: distance19,
  outerDiameter: distance19,
  connectsTo: z72.string().or(z72.array(z72.string())).optional(),
  netIsAssignable: z72.boolean().optional()
});
expectTypesMatch(true);

// lib/components/testpoint.ts
import { distance as distance20 } from "circuit-json";
import { z as z73 } from "zod";
var testpointPins = ["pin1"];
var testpointConnectionsProp = z73.object({
  pin1: connectionTarget
}).strict();
var testpointProps = commonComponentProps.extend({
  connections: testpointConnectionsProp.optional(),
  footprintVariant: z73.enum(["pad", "through_hole"]).optional(),
  padShape: z73.enum(["rect", "circle"]).optional().default("circle"),
  padDiameter: distance20.optional(),
  holeDiameter: distance20.optional(),
  width: distance20.optional(),
  height: distance20.optional()
}).refine(
  (props) => props.footprintVariant !== "through_hole" || props.holeDiameter !== void 0,
  { message: "holeDiameter is required for through_hole testpoints" }
);
expectTypesMatch(true);

// lib/components/breakoutpoint.ts
import { z as z74 } from "zod";
var breakoutPointProps = pcbLayoutProps.omit({ pcbRotation: true, layer: true }).extend({
  connection: z74.string()
});
expectTypesMatch(true);

// lib/components/pcb-keepout.ts
import { distance as distance21 } from "circuit-json";
import { z as z75 } from "zod";
var pcbKeepoutProps = z75.union([
  pcbLayoutProps.omit({ pcbRotation: true }).extend({
    shape: z75.literal("circle"),
    radius: distance21
  }),
  pcbLayoutProps.extend({
    shape: z75.literal("rect"),
    width: distance21,
    height: distance21
  })
]);

// lib/components/courtyard-rect.ts
import { distance as distance22 } from "circuit-json";
import { z as z76 } from "zod";
var courtyardRectProps = pcbLayoutProps.omit({ pcbRotation: true }).extend({
  width: distance22,
  height: distance22,
  strokeWidth: distance22.optional(),
  isFilled: z76.boolean().optional(),
  hasStroke: z76.boolean().optional(),
  isStrokeDashed: z76.boolean().optional(),
  color: z76.string().optional()
});

// lib/components/courtyard-outline.ts
import { length as length6 } from "circuit-json";
import { z as z77 } from "zod";
var courtyardOutlineProps = pcbLayoutProps.omit({
  pcbX: true,
  pcbY: true,
  pcbOffsetX: true,
  pcbOffsetY: true,
  pcbRotation: true
}).extend({
  outline: z77.array(point),
  strokeWidth: length6.optional(),
  isClosed: z77.boolean().optional(),
  isStrokeDashed: z77.boolean().optional(),
  color: z77.string().optional()
});

// lib/components/copper-pour.ts
import { z as z78 } from "zod";
import { layer_ref as layer_ref6 } from "circuit-json";
var copperPourProps = z78.object({
  name: z78.string().optional(),
  layer: layer_ref6,
  connectsTo: z78.string(),
  padMargin: distance.optional(),
  traceMargin: distance.optional(),
  clearance: distance.optional(),
  boardEdgeMargin: distance.optional(),
  cutoutMargin: distance.optional(),
  coveredWithSolderMask: z78.boolean().optional().default(true)
});
expectTypesMatch(true);

// lib/components/cadassembly.ts
import { layer_ref as layer_ref7 } from "circuit-json";
import { z as z79 } from "zod";
var cadassemblyProps = z79.object({
  originalLayer: layer_ref7.default("top").optional(),
  children: z79.any().optional()
});
expectTypesMatch(true);

// lib/components/cadmodel.ts
import { z as z80 } from "zod";
var pcbPosition = z80.object({
  pcbX: distance.optional(),
  pcbY: distance.optional(),
  pcbOffsetX: distance.optional(),
  pcbOffsetY: distance.optional(),
  pcbZ: distance.optional()
});
var cadModelBaseWithUrl = cadModelBase.extend({
  modelUrl: z80.string(),
  stepUrl: z80.string().optional()
});
var cadModelObject = cadModelBaseWithUrl.merge(pcbPosition);
expectTypesMatch(true);
var cadmodelProps = z80.union([z80.null(), z80.string(), cadModelObject]);

// lib/components/power-source.ts
import { voltage as voltage3 } from "circuit-json";
var powerSourceProps = commonComponentProps.extend({
  voltage: voltage3
});

// lib/components/voltagesource.ts
import { frequency as frequency4, rotation as rotation4, voltage as voltage4 } from "circuit-json";
import { z as z81 } from "zod";
var voltageSourcePinLabels = ["pin1", "pin2", "pos", "neg"];
var percentage = z81.union([z81.string(), z81.number()]).transform((val) => {
  if (typeof val === "string") {
    if (val.endsWith("%")) {
      return parseFloat(val.slice(0, -1)) / 100;
    }
    return parseFloat(val);
  }
  return val;
}).pipe(
  z81.number().min(0, "Duty cycle must be non-negative").max(1, "Duty cycle cannot be greater than 100%")
);
var voltageSourceProps = commonComponentProps.extend({
  voltage: voltage4.optional(),
  frequency: frequency4.optional(),
  peakToPeakVoltage: voltage4.optional(),
  waveShape: z81.enum(["sinewave", "square", "triangle", "sawtooth"]).optional(),
  phase: rotation4.optional(),
  dutyCycle: percentage.optional(),
  connections: createConnectionsProp(voltageSourcePinLabels).optional()
});
var voltageSourcePins = lrPolarPins;
expectTypesMatch(true);

// lib/components/voltageprobe.ts
import { z as z82 } from "zod";
var voltageProbeProps = commonComponentProps.omit({ name: true }).extend({
  name: z82.string().optional(),
  connectsTo: z82.string().or(z82.array(z82.string()))
});
expectTypesMatch(true);

// lib/components/schematic-arc.ts
import { distance as distance23, point as point5, rotation as rotation5 } from "circuit-json";
import { z as z83 } from "zod";
var schematicArcProps = z83.object({
  center: point5,
  radius: distance23,
  startAngleDegrees: rotation5,
  endAngleDegrees: rotation5,
  direction: z83.enum(["clockwise", "counterclockwise"]).default("counterclockwise"),
  strokeWidth: distance23.optional(),
  color: z83.string().optional(),
  isDashed: z83.boolean().optional().default(false)
});

// lib/components/schematic-box.ts
import { distance as distance24 } from "circuit-json";
import { z as z84 } from "zod";
var schematicBoxProps = z84.object({
  schX: distance24.optional(),
  schY: distance24.optional(),
  width: distance24.optional(),
  height: distance24.optional(),
  overlay: z84.array(z84.string()).optional(),
  padding: distance24.optional(),
  paddingLeft: distance24.optional(),
  paddingRight: distance24.optional(),
  paddingTop: distance24.optional(),
  paddingBottom: distance24.optional(),
  title: z84.string().optional(),
  titleAlignment: ninePointAnchor.default("top_left"),
  titleColor: z84.string().optional(),
  titleFontSize: distance24.optional(),
  titleInside: z84.boolean().default(false),
  strokeStyle: z84.enum(["solid", "dashed"]).default("solid")
}).refine(
  (elm) => elm.width !== void 0 && elm.height !== void 0 || Array.isArray(elm.overlay) && elm.overlay.length > 0,
  {
    message: "Must provide either both `width` and `height`, or a non-empty `overlay` array."
  }
).refine(
  (elm) => !(elm.width !== void 0 && elm.height !== void 0 && Array.isArray(elm.overlay) && elm.overlay.length > 0),
  {
    message: "Cannot provide both `width`/`height` and `overlay` at the same time."
  }
);

// lib/components/schematic-circle.ts
import { distance as distance25, point as point6 } from "circuit-json";
import { z as z85 } from "zod";
var schematicCircleProps = z85.object({
  center: point6,
  radius: distance25,
  strokeWidth: distance25.optional(),
  color: z85.string().optional(),
  isFilled: z85.boolean().optional().default(false),
  fillColor: z85.string().optional(),
  isDashed: z85.boolean().optional().default(false)
});

// lib/components/schematic-rect.ts
import { distance as distance26, rotation as rotation6 } from "circuit-json";
import { z as z86 } from "zod";
var schematicRectProps = z86.object({
  schX: distance26.optional(),
  schY: distance26.optional(),
  width: distance26,
  height: distance26,
  rotation: rotation6.default(0),
  strokeWidth: distance26.optional(),
  color: z86.string().optional(),
  isFilled: z86.boolean().optional().default(false),
  fillColor: z86.string().optional(),
  isDashed: z86.boolean().optional().default(false),
  cornerRadius: distance26.optional()
});

// lib/components/schematic-line.ts
import { distance as distance27 } from "circuit-json";
import { z as z87 } from "zod";
var schematicLineProps = z87.object({
  x1: distance27,
  y1: distance27,
  x2: distance27,
  y2: distance27,
  strokeWidth: distance27.optional(),
  color: z87.string().optional(),
  isDashed: z87.boolean().optional().default(false)
});

// lib/components/schematic-text.ts
import { distance as distance28, rotation as rotation7 } from "circuit-json";
import { z as z89 } from "zod";

// lib/common/fivePointAnchor.ts
import { z as z88 } from "zod";
var fivePointAnchor = z88.enum([
  "center",
  "left",
  "right",
  "top",
  "bottom"
]);

// lib/components/schematic-text.ts
var schematicTextProps = z89.object({
  schX: distance28.optional(),
  schY: distance28.optional(),
  text: z89.string(),
  fontSize: z89.number().default(1),
  anchor: z89.union([fivePointAnchor.describe("legacy"), ninePointAnchor]).default("center"),
  color: z89.string().default("#000000"),
  schRotation: rotation7.default(0)
});

// lib/components/schematic-path.ts
import { point as point8 } from "circuit-json";
import { z as z90 } from "zod";
var schematicPathProps = z90.object({
  points: z90.array(point8),
  isFilled: z90.boolean().optional().default(false),
  fillColor: z90.enum(["red", "blue"]).optional()
});

// lib/components/schematic-table.ts
import { distance as distance29 } from "circuit-json";
import { z as z91 } from "zod";
var schematicTableProps = z91.object({
  schX: distance29.optional(),
  schY: distance29.optional(),
  children: z91.any().optional(),
  cellPadding: distance29.optional(),
  borderWidth: distance29.optional(),
  anchor: ninePointAnchor.optional(),
  fontSize: distance29.optional()
});
expectTypesMatch(true);

// lib/components/schematic-row.ts
import { distance as distance30 } from "circuit-json";
import { z as z92 } from "zod";
var schematicRowProps = z92.object({
  children: z92.any().optional(),
  height: distance30.optional()
});
expectTypesMatch(true);

// lib/components/schematic-cell.ts
import { distance as distance31 } from "circuit-json";
import { z as z93 } from "zod";
var schematicCellProps = z93.object({
  children: z93.string().optional(),
  horizontalAlign: z93.enum(["left", "center", "right"]).optional(),
  verticalAlign: z93.enum(["top", "middle", "bottom"]).optional(),
  fontSize: distance31.optional(),
  rowSpan: z93.number().optional(),
  colSpan: z93.number().optional(),
  width: distance31.optional(),
  text: z93.string().optional()
});
expectTypesMatch(true);

// lib/components/copper-text.ts
import { layer_ref as layer_ref8, length as length7 } from "circuit-json";
import { z as z94 } from "zod";
var copperTextProps = pcbLayoutProps.extend({
  text: z94.string(),
  anchorAlignment: ninePointAnchor.default("center"),
  font: z94.enum(["tscircuit2024"]).optional(),
  fontSize: length7.optional(),
  layers: z94.array(layer_ref8).optional()
});

// lib/components/silkscreen-text.ts
import { layer_ref as layer_ref9, length as length8 } from "circuit-json";
import { z as z95 } from "zod";
var silkscreenTextProps = pcbLayoutProps.extend({
  text: z95.string(),
  anchorAlignment: ninePointAnchor.default("center"),
  font: z95.enum(["tscircuit2024"]).optional(),
  fontSize: length8.optional(),
  /**
   * If true, text will knock out underlying silkscreen
   */
  isKnockout: z95.boolean().optional(),
  knockoutPadding: length8.optional(),
  knockoutPaddingLeft: length8.optional(),
  knockoutPaddingRight: length8.optional(),
  knockoutPaddingTop: length8.optional(),
  knockoutPaddingBottom: length8.optional(),
  layers: z95.array(layer_ref9).optional()
});

// lib/components/silkscreen-path.ts
import { length as length9, route_hint_point as route_hint_point5 } from "circuit-json";
import { z as z96 } from "zod";
var silkscreenPathProps = pcbLayoutProps.omit({
  pcbX: true,
  pcbY: true,
  pcbOffsetX: true,
  pcbOffsetY: true,
  pcbRotation: true
}).extend({
  route: z96.array(route_hint_point5),
  strokeWidth: length9.optional()
});

// lib/components/silkscreen-line.ts
import { distance as distance32 } from "circuit-json";
var silkscreenLineProps = pcbLayoutProps.omit({
  pcbX: true,
  pcbY: true,
  pcbOffsetX: true,
  pcbOffsetY: true,
  pcbRotation: true
}).extend({
  strokeWidth: distance32,
  x1: distance32,
  y1: distance32,
  x2: distance32,
  y2: distance32
});

// lib/components/silkscreen-rect.ts
import { distance as distance33 } from "circuit-json";
import { z as z97 } from "zod";
var silkscreenRectProps = pcbLayoutProps.omit({ pcbRotation: true }).extend({
  filled: z97.boolean().default(true).optional(),
  stroke: z97.enum(["dashed", "solid", "none"]).optional(),
  strokeWidth: distance33.optional(),
  width: distance33,
  height: distance33,
  cornerRadius: distance33.optional()
});

// lib/components/silkscreen-circle.ts
import { distance as distance34 } from "circuit-json";
import { z as z98 } from "zod";
var silkscreenCircleProps = pcbLayoutProps.omit({ pcbRotation: true }).extend({
  isFilled: z98.boolean().optional(),
  isOutline: z98.boolean().optional(),
  strokeWidth: distance34.optional(),
  radius: distance34
});

// lib/components/trace-hint.ts
import { distance as distance35, layer_ref as layer_ref10, route_hint_point as route_hint_point6 } from "circuit-json";
import { z as z99 } from "zod";
var routeHintPointProps = z99.object({
  x: distance35,
  y: distance35,
  via: z99.boolean().optional(),
  toLayer: layer_ref10.optional()
});
var traceHintProps = z99.object({
  for: z99.string().optional().describe(
    "Selector for the port you're targeting, not required if you're inside a trace"
  ),
  order: z99.number().optional(),
  offset: route_hint_point6.or(routeHintPointProps).optional(),
  offsets: z99.array(route_hint_point6).or(z99.array(routeHintPointProps)).optional(),
  traceWidth: z99.number().optional()
});

// lib/components/port.ts
import { z as z100 } from "zod";
var portProps = commonLayoutProps.extend({
  name: z100.string(),
  pinNumber: z100.number().optional(),
  aliases: z100.array(z100.string()).optional(),
  direction,
  connectsTo: z100.string().or(z100.array(z100.string())).optional()
});

// lib/components/pcb-note-text.ts
import { length as length10 } from "circuit-json";
import { z as z101 } from "zod";
var pcbNoteTextProps = pcbLayoutProps.extend({
  text: z101.string(),
  anchorAlignment: z101.enum(["center", "top_left", "top_right", "bottom_left", "bottom_right"]).default("center"),
  font: z101.enum(["tscircuit2024"]).optional(),
  fontSize: length10.optional(),
  color: z101.string().optional()
});
expectTypesMatch(true);

// lib/components/pcb-note-rect.ts
import { distance as distance36 } from "circuit-json";
import { z as z102 } from "zod";
var pcbNoteRectProps = pcbLayoutProps.omit({ pcbRotation: true }).extend({
  width: distance36,
  height: distance36,
  strokeWidth: distance36.optional(),
  isFilled: z102.boolean().optional(),
  hasStroke: z102.boolean().optional(),
  isStrokeDashed: z102.boolean().optional(),
  color: z102.string().optional(),
  cornerRadius: distance36.optional()
});
expectTypesMatch(true);

// lib/components/pcb-note-path.ts
import {
  length as length11,
  route_hint_point as route_hint_point7
} from "circuit-json";
import { z as z103 } from "zod";
var pcbNotePathProps = pcbLayoutProps.omit({
  pcbX: true,
  pcbY: true,
  pcbOffsetX: true,
  pcbOffsetY: true,
  pcbRotation: true
}).extend({
  route: z103.array(route_hint_point7),
  strokeWidth: length11.optional(),
  color: z103.string().optional()
});
expectTypesMatch(true);

// lib/components/pcb-note-line.ts
import { distance as distance37 } from "circuit-json";
import { z as z104 } from "zod";
var pcbNoteLineProps = pcbLayoutProps.omit({
  pcbX: true,
  pcbY: true,
  pcbOffsetX: true,
  pcbOffsetY: true,
  pcbRotation: true
}).extend({
  x1: distance37,
  y1: distance37,
  x2: distance37,
  y2: distance37,
  strokeWidth: distance37.optional(),
  color: z104.string().optional(),
  isDashed: z104.boolean().optional()
});
expectTypesMatch(true);

// lib/components/pcb-note-dimension.ts
import { distance as distance38, length as length12 } from "circuit-json";
import { z as z105 } from "zod";
var dimensionTarget2 = z105.union([z105.string(), point]);
var pcbNoteDimensionProps = pcbLayoutProps.omit({
  pcbX: true,
  pcbY: true,
  pcbOffsetX: true,
  pcbOffsetY: true,
  pcbRotation: true
}).extend({
  from: dimensionTarget2,
  to: dimensionTarget2,
  text: z105.string().optional(),
  offset: distance38.optional(),
  font: z105.enum(["tscircuit2024"]).optional(),
  fontSize: length12.optional(),
  color: z105.string().optional(),
  arrowSize: distance38.optional(),
  units: z105.enum(["in", "mm"]).optional(),
  outerEdgeToEdge: z105.literal(true).optional(),
  centerToCenter: z105.literal(true).optional(),
  innerEdgeToEdge: z105.literal(true).optional()
});
expectTypesMatch(
  true
);

// lib/platformConfig.ts
import { z as z106 } from "zod";
var unvalidatedCircuitJson = z106.array(z106.any()).describe("Circuit JSON");
var footprintLibraryResult = z106.object({
  footprintCircuitJson: z106.array(z106.any()),
  cadModel: cadModelProp.optional()
});
var pathToCircuitJsonFn = z106.function().args(z106.string()).returns(z106.promise(footprintLibraryResult)).or(
  z106.function().args(
    z106.string(),
    z106.object({ resolvedPcbStyle: pcbStyle.optional() }).optional()
  ).returns(z106.promise(footprintLibraryResult))
).describe("A function that takes a path and returns Circuit JSON");
var footprintFileParserEntry = z106.object({
  loadFromUrl: z106.function().args(z106.string()).returns(z106.promise(footprintLibraryResult)).describe(
    "A function that takes a footprint file URL and returns Circuit JSON"
  )
});
var spiceEngineSimulationResult = z106.object({
  engineVersionString: z106.string().optional(),
  simulationResultCircuitJson: unvalidatedCircuitJson
});
var spiceEngineZod = z106.object({
  simulate: z106.function().args(z106.string()).returns(z106.promise(spiceEngineSimulationResult)).describe(
    "A function that takes a SPICE string and returns a simulation result"
  )
});
var defaultSpiceEngine = z106.custom(
  (value) => typeof value === "string"
);
var autorouterInstance = z106.object({
  run: z106.function().args().returns(z106.promise(z106.unknown())).describe("Run the autorouter"),
  getOutputSimpleRouteJson: z106.function().args().returns(z106.promise(z106.any())).describe("Get the resulting SimpleRouteJson")
});
var autorouterDefinition = z106.object({
  createAutorouter: z106.function().args(z106.any(), z106.any().optional()).returns(z106.union([autorouterInstance, z106.promise(autorouterInstance)])).describe("Create an autorouter instance")
});
var platformConfig = z106.object({
  partsEngine: partsEngine.optional(),
  autorouter: autorouterProp.optional(),
  autorouterMap: z106.record(z106.string(), autorouterDefinition).optional(),
  registryApiUrl: z106.string().optional(),
  cloudAutorouterUrl: z106.string().optional(),
  projectName: z106.string().optional(),
  projectBaseUrl: z106.string().optional(),
  version: z106.string().optional(),
  url: z106.string().optional(),
  printBoardInformationToSilkscreen: z106.boolean().optional(),
  includeBoardFiles: z106.array(z106.string()).describe(
    'The board files to automatically build with "tsci build", defaults to ["**/*.circuit.tsx"]. Can be an array of files or globs'
  ).optional(),
  snapshotsDir: z106.string().describe(
    'The directory where snapshots are stored for "tsci snapshot", defaults to "tests/__snapshots__"'
  ).optional(),
  defaultSpiceEngine: defaultSpiceEngine.optional(),
  localCacheEngine: z106.any().optional(),
  pcbDisabled: z106.boolean().optional(),
  schematicDisabled: z106.boolean().optional(),
  partsEngineDisabled: z106.boolean().optional(),
  spiceEngineMap: z106.record(z106.string(), spiceEngineZod).optional(),
  footprintLibraryMap: z106.record(
    z106.string(),
    z106.union([
      pathToCircuitJsonFn,
      z106.record(
        z106.string(),
        z106.union([unvalidatedCircuitJson, pathToCircuitJsonFn])
      )
    ])
  ).optional(),
  footprintFileParserMap: z106.record(z106.string(), footprintFileParserEntry).optional(),
  resolveProjectStaticFileImportUrl: z106.function().args(z106.string()).returns(z106.promise(z106.string())).describe(
    "A function that returns a string URL for static files for the project"
  ).optional()
});
expectTypesMatch(true);

// lib/projectConfig.ts
import "zod";
var platformConfigObject = platformConfig;
var projectConfig = platformConfigObject.pick({
  projectName: true,
  projectBaseUrl: true,
  version: true,
  url: true,
  printBoardInformationToSilkscreen: true,
  includeBoardFiles: true,
  snapshotsDir: true,
  defaultSpiceEngine: true
});
expectTypesMatch(true);
export {
  analogSimulationProps,
  autorouterConfig,
  autorouterPreset,
  autorouterProp,
  baseGroupProps,
  base_manual_edit_event,
  batteryPins,
  batteryProps,
  boardProps,
  border,
  breakoutPointProps,
  breakoutProps,
  bugProps,
  cadModelBase,
  cadModelGlb,
  cadModelGltf,
  cadModelJscad,
  cadModelObj,
  cadModelProp,
  cadModelStep,
  cadModelStl,
  cadModelWrl,
  cadassemblyProps,
  cadmodelProps,
  capacitorPinLabels,
  capacitorPins,
  capacitorProps,
  chipProps,
  circleCutoutProps,
  circleSmtPadProps,
  circleSolderPasteProps,
  commonComponentProps,
  commonLayoutProps,
  componentProps,
  connectorProps,
  constrainedLayoutProps,
  constraintProps,
  copperPourProps,
  copperTextProps,
  courtyardOutlineProps,
  courtyardRectProps,
  crystalPins,
  crystalProps,
  cutoutProps,
  diodePins,
  diodeProps,
  direction,
  directionAlongEdge,
  distanceOrMultiplier,
  edit_component_location_event,
  edit_pcb_component_location_event,
  edit_pcb_group_location_event,
  edit_schematic_component_location_event,
  edit_schematic_group_location_event,
  edit_trace_hint_event,
  explicitPinSideDefinition,
  fabricationNoteDimensionProps,
  fabricationNotePathProps,
  fabricationNoteRectProps,
  fabricationNoteTextProps,
  footprintProp,
  footprintProps,
  fusePinLabels,
  fuseProps,
  groupProps,
  holeProps,
  inductorPins,
  inductorProps,
  jumperProps,
  layoutConfig,
  ledPins,
  ledProps,
  lrPins,
  lrPolarPins,
  manual_edit_event,
  manual_edits_file,
  manual_pcb_placement,
  manual_schematic_placement,
  manual_trace_hint,
  mosfetPins,
  mosfetProps,
  netAliasProps,
  netLabelProps,
  netProps,
  ninePointAnchor,
  panelProps,
  partsEngine,
  pcbKeepoutProps,
  pcbLayoutProps,
  pcbNoteDimensionProps,
  pcbNoteLineProps,
  pcbNotePathProps,
  pcbNoteRectProps,
  pcbNoteTextProps,
  pcbSameXConstraintProps,
  pcbSameYConstraintProps,
  pcbStyle,
  pcbTraceProps,
  pcbXDistConstraintProps,
  pcbYDistConstraintProps,
  pillSmtPadProps,
  pinAttributeMap,
  pinCompatibleVariant,
  pinHeaderProps,
  pinLabelsProp,
  pinoutProps,
  platedHoleProps,
  platformConfig,
  point3,
  polygonCutoutProps,
  polygonSmtPadProps,
  portHints,
  portProps,
  portRef,
  potentiometerProps,
  powerSourceProps,
  projectConfig,
  pushButtonProps,
  rectCutoutProps,
  rectSmtPadProps,
  rectSolderPasteProps,
  resistorPinLabels,
  resistorPins,
  resistorProps,
  resonatorProps,
  rotatedRectSmtPadProps,
  rotationPoint3,
  routeHintPointProps,
  schematicArcProps,
  schematicBoxProps,
  schematicCellProps,
  schematicCircleProps,
  schematicLineProps,
  schematicOrientation,
  schematicPathProps,
  schematicPinArrangement,
  schematicPinLabel,
  schematicPinStyle,
  schematicPortArrangement,
  schematicRectProps,
  schematicRowProps,
  schematicTableProps,
  schematicTextProps,
  silkscreenCircleProps,
  silkscreenLineProps,
  silkscreenPathProps,
  silkscreenRectProps,
  silkscreenTextProps,
  smtPadProps,
  solderPasteProps,
  solderjumperProps,
  stampboardProps,
  subcircuitGroupProps,
  subcircuitGroupPropsWithBool,
  subcircuitProps,
  supplierProps,
  switchProps,
  symbolProp,
  symbolProps,
  testpointPins,
  testpointProps,
  traceHintProps,
  traceProps,
  transistorPins,
  transistorPinsLabels,
  transistorProps,
  viaProps,
  voltageProbeProps,
  voltageSourcePinLabels,
  voltageSourcePins,
  voltageSourceProps
};
