// @tsci/seveibar.xiao-esp32c6/index.cjs
'use strict';

var core = require('@tscircuit/core');

function _extends() { return _extends = Object.assign ? Object.assign.bind() : function (n) { for (var e = 1; e < arguments.length; e++) { var t = arguments[e]; for (var r in t) ({}).hasOwnProperty.call(t, r) && (n[r] = t[r]); } return n; }, _extends.apply(null, arguments); }
const pinLabels = {
  // Left side, top to bottom
  pin1: ["LP_GPIO0_A0_D0"],
  pin2: ["LP_GPIO1_A1_D1"],
  pin3: ["LP_GPIO2_A2_D2"],
  pin4: ["SDIO_DATA1_GPIO21_D3"],
  pin5: ["SDIO_DATA2_GPIO22_SDA_D4"],
  pin6: ["SDIO_DATA3_GPIO23_SCL_D5"],
  pin7: ["GPIO16_TX_D6"],
  // Now wrap around to the bottom-right pin and go up
  pin8: ["D7_RX_GPIO17"],
  pin9: ["D8_SCK_GPIO19_SDIO_CLK"],
  pin10: ["D9_MISO_GPIO20_SDIO_DATA0"],
  pin11: ["D10_MOSI_GPIO18_SDIO_CMD"],
  pin12: ["V3_3"],
  pin13: ["GND"],
  pin14: ["V5"]
};
const XIAO_ESP32C6 = props => {
  return /*#__PURE__*/React.createElement("chip", _extends({}, props, {
    pinLabels: pinLabels,
    schWidth: 5,
    schPortArrangement: {
      leftSide: {
        direction: "top-to-bottom",
        pins: [1, 2, 3, 4, 5, 6, 7]
      },
      rightSide: {
        direction: "bottom-to-top",
        pins: [8, 9, 10, 11, 12, 13, 14]
      }
    },
    footprint: /*#__PURE__*/React.createElement("footprint", null, /*#__PURE__*/React.createElement("smtpad", {
      portHints: ["1"],
      pcbX: "-8.2816mm",
      pcbY: "7.5985mm",
      width: "2.75mm",
      height: "2mm",
      shape: "rect"
    }), /*#__PURE__*/React.createElement("smtpad", {
      portHints: ["2"],
      pcbX: "-8.2816mm",
      pcbY: "5.0585mm",
      width: "2.75mm",
      height: "2mm",
      shape: "rect"
    }), /*#__PURE__*/React.createElement("smtpad", {
      portHints: ["3"],
      pcbX: "-8.2816mm",
      pcbY: "2.5185mm",
      width: "2.75mm",
      height: "2mm",
      shape: "rect"
    }), /*#__PURE__*/React.createElement("smtpad", {
      portHints: ["4"],
      pcbX: "-8.2816mm",
      pcbY: "-0.0215mm",
      width: "2.75mm",
      height: "2mm",
      shape: "rect"
    }), /*#__PURE__*/React.createElement("smtpad", {
      portHints: ["5"],
      pcbX: "-8.2816mm",
      pcbY: "-2.5615mm",
      width: "2.75mm",
      height: "2mm",
      shape: "rect"
    }), /*#__PURE__*/React.createElement("smtpad", {
      portHints: ["6"],
      pcbX: "-8.2816mm",
      pcbY: "-5.1015mm",
      width: "2.75mm",
      height: "2mm",
      shape: "rect"
    }), /*#__PURE__*/React.createElement("smtpad", {
      portHints: ["7"],
      pcbX: "-8.2816mm",
      pcbY: "-7.6415mm",
      width: "2.75mm",
      height: "2mm",
      shape: "rect"
    }), /*#__PURE__*/React.createElement("smtpad", {
      portHints: ["8"],
      pcbX: "7.8834mm",
      pcbY: "-7.6415mm",
      width: "2.75mm",
      height: "2mm",
      shape: "rect"
    }), /*#__PURE__*/React.createElement("smtpad", {
      portHints: ["9"],
      pcbX: "7.8834mm",
      pcbY: "-5.1015mm",
      width: "2.75mm",
      height: "2mm",
      shape: "rect"
    }), /*#__PURE__*/React.createElement("smtpad", {
      portHints: ["10"],
      pcbX: "7.8834mm",
      pcbY: "-2.5615mm",
      width: "2.75mm",
      height: "2mm",
      shape: "rect"
    }), /*#__PURE__*/React.createElement("smtpad", {
      portHints: ["11"],
      pcbX: "7.8834mm",
      pcbY: "-0.0215mm",
      width: "2.75mm",
      height: "2mm",
      shape: "rect"
    }), /*#__PURE__*/React.createElement("smtpad", {
      portHints: ["12"],
      pcbX: "7.8834mm",
      pcbY: "2.5185mm",
      width: "2.75mm",
      height: "2mm",
      shape: "rect"
    }), /*#__PURE__*/React.createElement("smtpad", {
      portHints: ["13"],
      pcbX: "7.8834mm",
      pcbY: "5.0585mm",
      width: "2.75mm",
      height: "2mm",
      shape: "rect"
    }), /*#__PURE__*/React.createElement("smtpad", {
      portHints: ["14"],
      pcbX: "7.8834mm",
      pcbY: "7.5985mm",
      width: "2.75mm",
      height: "2mm",
      shape: "rect"
    }), /*#__PURE__*/React.createElement("smtpad", {
      portHints: ["15"],
      pcbX: "-2.008mm",
      pcbY: "-5.5532mm",
      width: "2.5mm",
      height: "1.1mm",
      shape: "rect"
    }), /*#__PURE__*/React.createElement("smtpad", {
      portHints: ["16"],
      pcbX: "0.532mm",
      pcbY: "-5.5532mm",
      width: "2.5mm",
      height: "1.1mm",
      shape: "rect"
    }), /*#__PURE__*/React.createElement("smtpad", {
      portHints: ["17"],
      pcbX: "-1.5mm",
      pcbY: "8.62mm",
      width: "1.7mm",
      height: "1.7mm",
      shape: "rect"
    }), /*#__PURE__*/React.createElement("smtpad", {
      portHints: ["18"],
      pcbX: "1.04mm",
      pcbY: "8.62mm",
      width: "1.7mm",
      height: "1.7mm",
      shape: "rect"
    }), /*#__PURE__*/React.createElement("smtpad", {
      portHints: ["19"],
      pcbX: "-1.5mm",
      pcbY: "6.08mm",
      width: "1.7mm",
      height: "1.7mm",
      shape: "rect"
    }), /*#__PURE__*/React.createElement("smtpad", {
      portHints: ["20"],
      pcbX: "1.04mm",
      pcbY: "6.08mm",
      width: "1.7mm",
      height: "1.7mm",
      shape: "rect"
    }), /*#__PURE__*/React.createElement("smtpad", {
      portHints: ["21"],
      pcbX: "-1.5mm",
      pcbY: "3.54mm",
      width: "1.7mm",
      height: "1.7mm",
      shape: "rect"
    }), /*#__PURE__*/React.createElement("smtpad", {
      portHints: ["22"],
      pcbX: "1.04mm",
      pcbY: "3.54mm",
      width: "1.7mm",
      height: "1.7mm",
      shape: "rect"
    }), /*#__PURE__*/React.createElement("smtpad", {
      portHints: ["23"],
      pcbX: "-1.5mm",
      pcbY: "1mm",
      width: "1.7mm",
      height: "1.7mm",
      shape: "rect"
    }), /*#__PURE__*/React.createElement("smtpad", {
      portHints: ["24"],
      pcbX: "0.9474mm",
      pcbY: "1.0568mm",
      width: "1.7mm",
      height: "1.7mm",
      shape: "rect"
    }), /*#__PURE__*/React.createElement("silkscreenpath", {
      route: [{
        x: -9.1166,
        y: -8.5345
      }, {
        x: -9.1166,
        y: 8.6105
      }]
    }), /*#__PURE__*/React.createElement("silkscreenpath", {
      route: [{
        x: -7.2116,
        y: -10.4395
      }, {
        x: 6.7584,
        y: -10.4395
      }]
    }), /*#__PURE__*/React.createElement("silkscreenpath", {
      route: [{
        x: -4.7216,
        y: 10.5155
      }, {
        x: -4.717872,
        y: 11.525772
      }]
    }), /*#__PURE__*/React.createElement("silkscreenpath", {
      route: [{
        x: -4.217872,
        y: 12.0255
      }, {
        x: 3.7774,
        y: 12.0255
      }]
    }), /*#__PURE__*/React.createElement("silkscreenpath", {
      route: [{
        x: 4.2774,
        y: 11.5255
      }, {
        x: 4.2774,
        y: 10.5155
      }]
    }), /*#__PURE__*/React.createElement("silkscreenpath", {
      route: [{
        x: 6.7584,
        y: 10.5155
      }, {
        x: -7.2116,
        y: 10.5155
      }]
    }), /*#__PURE__*/React.createElement("silkscreenpath", {
      route: [{
        x: 8.6634,
        y: -8.5345
      }, {
        x: 8.6634,
        y: 8.6105
      }]
    }), /*#__PURE__*/React.createElement("silkscreenpath", {
      route: [{
        x: -9.1166,
        y: 8.6145
      }, {
        x: -8.86137796498721,
        y: 9.566999698293817
      }, {
        x: -8.164099698293818,
        y: 10.26427796498721
      }, {
        x: -7.2116,
        y: 10.5195
      }]
    }), /*#__PURE__*/React.createElement("silkscreenpath", {
      route: [{
        x: -7.2116,
        y: -10.4395
      }, {
        x: -8.164099698293818,
        y: -10.18427796498721
      }, {
        x: -8.861377964987211,
        y: -9.486999698293818
      }, {
        x: -9.116600000000002,
        y: -8.5345
      }]
    }), /*#__PURE__*/React.createElement("silkscreenpath", {
      route: [{
        x: -4.717872,
        y: 11.525772
      }, {
        x: -4.217872,
        y: 12.025499999999997
      }]
    }), /*#__PURE__*/React.createElement("silkscreenpath", {
      route: [{
        x: 3.7774,
        y: 12.0255
      }, {
        x: 4.277400000000001,
        y: 11.5255
      }]
    }), /*#__PURE__*/React.createElement("silkscreenpath", {
      route: [{
        x: 6.7584,
        y: 10.5155
      }, {
        x: 7.710899698293817,
        y: 10.26027796498721
      }, {
        x: 8.408177964987209,
        y: 9.562999698293817
      }, {
        x: 8.663399999999998,
        y: 8.6105
      }]
    }), /*#__PURE__*/React.createElement("silkscreenpath", {
      route: [{
        x: 8.6634,
        y: -8.5345
      }, {
        x: 8.40817796498721,
        y: -9.486999698293818
      }, {
        x: 7.710899698293818,
        y: -10.184277964987212
      }, {
        x: 6.7584,
        y: -10.4395
      }]
    }))
  }));
};
const useXIAO_ESP32C6 = core.createUseComponent(XIAO_ESP32C6, pinLabels);

exports.XIAO_ESP32C6 = XIAO_ESP32C6;
exports.useXIAO_ESP32C6 = useXIAO_ESP32C6;
