// @tscircuit/eval/dist/lib/index.js
// webworker/execution-context.ts
import { RootCircuit } from "@tscircuit/core";
import * as tscircuitCore from "@tscircuit/core";
import * as React from "react";
import * as tscircuitMathUtils from "@tscircuit/math-utils";

// node_modules/@tscircuit/parts-engine/lib/footprint-translators/get-footprinter-string-from-kicad.ts
var getFootprinterStringFromKicad = (kicadFootprint) => {
  let match = kicadFootprint.match(/:[RC]_(\d{4})_/);
  if (match) return match[1];
  match = kicadFootprint.match(
    /:(SOIC-\d+|SOT-\d+|SOD-\d+|SSOP-\d+|TSSOP-\d+|QFP-\d+|QFN-\d+)/
  );
  if (match) return match[1];
  return void 0;
};

// node_modules/@tscircuit/parts-engine/lib/footprint-translators/get-jlc-package-from-footprinter-string.ts
var getJlcPackageFromFootprinterString = (footprinterString) => {
  if (footprinterString.includes("cap")) {
    return footprinterString.replace(/cap/g, "");
  }
  return footprinterString;
};

// node_modules/@tscircuit/parts-engine/lib/footprint-translators/index.ts
var getJlcpcbPackageName = (footprint) => {
  if (!footprint) return void 0;
  if (footprint.startsWith("kicad:")) {
    const footprinterString = getFootprinterStringFromKicad(footprint);
    if (footprinterString) {
      return getJlcPackageFromFootprinterString(footprinterString);
    }
    return footprint;
  }
  return getJlcPackageFromFootprinterString(footprint);
};

// node_modules/@tscircuit/parts-engine/lib/jlc-parts-engine.ts
var cache = /* @__PURE__ */ new Map();
var getJlcPartsCached = async (name, params) => {
  const paramString = new URLSearchParams({
    ...params,
    json: "true"
  }).toString();
  if (cache.has(paramString)) {
    return cache.get(paramString);
  }
  const response = await fetch(
    `https://jlcsearch.tscircuit.com/${name}/list?${paramString}`
  );
  const responseJson = await response.json();
  cache.set(paramString, responseJson);
  return responseJson;
};
var withBasicPartPreference = (parts) => {
  if (!parts) return [];
  return [...parts].sort(
    (a, b) => Number(b.is_basic ?? false) - Number(a.is_basic ?? false)
  );
};
var jlcPartsEngine = {
  findPart: async ({
    sourceComponent,
    footprinterString
  }) => {
    const jlcpcbPackage = getJlcpcbPackageName(footprinterString);
    if (sourceComponent.type === "source_component" && sourceComponent.ftype === "simple_resistor") {
      const { resistors } = await getJlcPartsCached("resistors", {
        resistance: sourceComponent.resistance,
        package: jlcpcbPackage
      });
      return {
        jlcpcb: withBasicPartPreference(resistors).map((r) => `C${r.lcsc}`).slice(0, 3)
      };
    } else if (sourceComponent.type === "source_component" && sourceComponent.ftype === "simple_capacitor") {
      const { capacitors } = await getJlcPartsCached("capacitors", {
        capacitance: sourceComponent.capacitance,
        package: jlcpcbPackage
      });
      return {
        jlcpcb: withBasicPartPreference(capacitors).map((c) => `C${c.lcsc}`).slice(0, 3)
      };
    } else if (sourceComponent.type === "source_component" && sourceComponent.ftype === "simple_pin_header") {
      let pitch;
      if (footprinterString?.includes("_p")) {
        pitch = Number(footprinterString.split("_p")[1]);
      }
      const { headers } = await getJlcPartsCached(
        "headers",
        pitch ? {
          pitch,
          num_pins: sourceComponent.pin_count,
          gender: sourceComponent.gender
        } : {
          num_pins: sourceComponent.pin_count,
          gender: sourceComponent.gender
        }
      );
      return {
        jlcpcb: withBasicPartPreference(headers).map((h) => `C${h.lcsc}`).slice(0, 3)
      };
    } else if (sourceComponent.type === "source_component" && sourceComponent.ftype === "simple_potentiometer") {
      const { potentiometers } = await getJlcPartsCached("potentiometers", {
        resistance: sourceComponent.max_resistance,
        package: jlcpcbPackage
      });
      return {
        jlcpcb: withBasicPartPreference(potentiometers).map((p) => `C${p.lcsc}`).slice(0, 3)
      };
    } else if (sourceComponent.type === "source_component" && sourceComponent.ftype === "simple_diode") {
      const { diodes } = await getJlcPartsCached("diodes", {
        package: jlcpcbPackage
      });
      return {
        jlcpcb: withBasicPartPreference(diodes).map((d) => `C${d.lcsc}`).slice(0, 3)
      };
    } else if (sourceComponent.type === "source_component" && sourceComponent.ftype === "simple_chip") {
      const { chips } = await getJlcPartsCached("chips", {
        package: jlcpcbPackage
      });
      return {
        jlcpcb: withBasicPartPreference(chips).map((c) => `C${c.lcsc}`).slice(0, 3)
      };
    } else if (sourceComponent.type === "source_component" && sourceComponent.ftype === "simple_transistor") {
      const { transistors } = await getJlcPartsCached("transistors", {
        package: jlcpcbPackage,
        transistor_type: sourceComponent.transistor_type
      });
      return {
        jlcpcb: withBasicPartPreference(transistors).map((t) => `C${t.lcsc}`).slice(0, 3)
      };
    } else if (sourceComponent.type === "source_component" && sourceComponent.ftype === "simple_power_source") {
      const { power_sources } = await getJlcPartsCached("power_sources", {
        voltage: sourceComponent.voltage,
        package: jlcpcbPackage
      });
      return {
        jlcpcb: withBasicPartPreference(power_sources).map((p) => `C${p.lcsc}`).slice(0, 3)
      };
    } else if (sourceComponent.type === "source_component" && sourceComponent.ftype === "simple_inductor") {
      const { inductors } = await getJlcPartsCached("inductors", {
        inductance: sourceComponent.inductance,
        package: jlcpcbPackage
      });
      return {
        jlcpcb: withBasicPartPreference(inductors).map((i) => `C${i.lcsc}`).slice(0, 3)
      };
    } else if (sourceComponent.type === "source_component" && sourceComponent.ftype === "simple_crystal") {
      const { crystals } = await getJlcPartsCached("crystals", {
        frequency: sourceComponent.frequency,
        load_capacitance: sourceComponent.load_capacitance,
        package: jlcpcbPackage
      });
      return {
        jlcpcb: withBasicPartPreference(crystals).map((c) => `C${c.lcsc}`).slice(0, 3)
      };
    } else if (sourceComponent.type === "source_component" && sourceComponent.ftype === "simple_mosfet") {
      const { mosfets } = await getJlcPartsCached("mosfets", {
        package: jlcpcbPackage,
        mosfet_mode: sourceComponent.mosfet_mode,
        channel_type: sourceComponent.channel_type
      });
      return {
        jlcpcb: withBasicPartPreference(mosfets).map((m) => `C${m.lcsc}`).slice(0, 3)
      };
    } else if (sourceComponent.type === "source_component" && sourceComponent.ftype === "simple_resonator") {
      const { resonators } = await getJlcPartsCached("resonators", {
        frequency: sourceComponent.frequency,
        package: jlcpcbPackage
      });
      return {
        jlcpcb: withBasicPartPreference(resonators).map((r) => `C${r.lcsc}`).slice(0, 3)
      };
    } else if (sourceComponent.type === "source_component" && sourceComponent.ftype === "simple_switch") {
      const { switches } = await getJlcPartsCached("switches", {
        switch_type: sourceComponent.type,
        package: jlcpcbPackage
      });
      return {
        jlcpcb: withBasicPartPreference(switches).map((s) => `C${s.lcsc}`).slice(0, 3)
      };
    } else if (sourceComponent.type === "source_component" && sourceComponent.ftype === "simple_led") {
      const { leds } = await getJlcPartsCached("leds", {
        package: jlcpcbPackage
      });
      return {
        jlcpcb: withBasicPartPreference(leds).map((l) => `C${l.lcsc}`).slice(0, 3)
      };
    } else if (sourceComponent.type === "source_component" && sourceComponent.ftype === "simple_fuse") {
      const { fuses } = await getJlcPartsCached("fuses", {
        package: jlcpcbPackage
      });
      return {
        jlcpcb: withBasicPartPreference(fuses).map((l) => `C${l.lcsc}`).slice(0, 3)
      };
    }
    return {};
  }
};

// lib/getPlatformConfig.ts
import { parseKicadModToCircuitJson } from "kicad-component-converter";

// lib/utils/dynamically-load-dependency-with-cdn-backup.ts
var dynamicallyLoadDependencyWithCdnBackup = async (packageName) => {
  try {
    const module = await import(packageName);
    return module.default;
  } catch (e) {
    console.log(`Failed to load ${packageName} locally, trying CDN fallback...`);
    try {
      const res = await fetch(
        `https://cdn.jsdelivr.net/npm/${packageName}/+esm`
      );
      if (!res.ok) {
        throw new Error(
          `Failed to fetch ${packageName} from CDN: ${res.statusText}`
        );
      }
      const code = await res.text();
      const blob = new Blob([code], { type: "application/javascript" });
      const url = URL.createObjectURL(blob);
      try {
        const { default: loadedModule } = await import(url);
        return loadedModule;
      } finally {
        URL.revokeObjectURL(url);
      }
    } catch (cdnError) {
      console.error(`CDN fallback for ${packageName} also failed:`, cdnError);
      throw cdnError;
    }
  }
};

// lib/getPlatformConfig.ts
var KICAD_FOOTPRINT_CACHE_URL = "https://kicad-mod-cache.tscircuit.com";
var ngspiceEngineCache = null;
var getPlatformConfig = () => ({
  partsEngine: jlcPartsEngine,
  spiceEngineMap: {
    ngspice: {
      simulate: async (spice) => {
        if (!ngspiceEngineCache) {
          const createNgspiceSpiceEngine = await dynamicallyLoadDependencyWithCdnBackup(
            "@tscircuit/ngspice-spice-engine"
          ).catch((error) => {
            throw new Error(
              "Could not load ngspice engine from local node_modules or CDN fallback.",
              { cause: error }
            );
          });
          if (createNgspiceSpiceEngine) {
            ngspiceEngineCache = await createNgspiceSpiceEngine();
          }
        }
        if (!ngspiceEngineCache) {
          throw new Error(
            "Could not load ngspice engine from local node_modules or CDN fallback."
          );
        }
        return ngspiceEngineCache.simulate(spice);
      }
    }
  },
  footprintLibraryMap: {
    kicad: async (footprintName) => {
      const baseUrl = `${KICAD_FOOTPRINT_CACHE_URL}/${footprintName}`;
      const circuitJsonUrl = `${baseUrl}.circuit.json`;
      const res = await fetch(circuitJsonUrl);
      const raw = await res.json();
      const filtered = Array.isArray(raw) ? raw.filter(
        (el) => el?.type === "pcb_silkscreen_text" ? el?.text === "REF**" : true
      ) : raw;
      const wrlUrl = `${baseUrl}.wrl`;
      return {
        footprintCircuitJson: filtered,
        cadModel: { wrlUrl, modelUnitToMmScale: 2.54 }
      };
    }
  },
  footprintFileParserMap: {
    kicad_mod: {
      loadFromUrl: async (url) => {
        const kicadContent = await fetch(url).then((res) => res.text());
        const kicadJson = await parseKicadModToCircuitJson(kicadContent);
        return {
          footprintCircuitJson: Array.isArray(kicadJson) ? kicadJson : [kicadJson]
        };
      }
    }
  }
});

// webworker/execution-context.ts
import Debug from "debug";
var debug = Debug("tsci:eval:execution-context");
function createExecutionContext(webWorkerConfiguration, opts = {}) {
  globalThis.React = React;
  const basePlatform = opts.platform || getPlatformConfig();
  const platform = opts.projectConfig ? { ...basePlatform, ...opts.projectConfig } : basePlatform;
  if (platform.partsEngineDisabled) {
    platform.partsEngine = void 0;
  }
  const circuit = new RootCircuit({
    platform
  });
  if (opts.name) {
    circuit.name = opts.name;
  }
  if (opts.debugNamespace) {
    circuit.enableDebug(opts.debugNamespace);
  }
  const logs = [];
  return {
    fsMap: {},
    entrypoint: "",
    logger: {
      info: (message) => {
        logs.push({ msg: message });
      },
      getLogs: () => logs,
      stringifyLogs: () => logs.map((log) => log.msg).join("\n")
    },
    preSuppliedImports: {
      "@tscircuit/core": tscircuitCore,
      tscircuit: tscircuitCore,
      "@tscircuit/math-utils": tscircuitMathUtils,
      react: React,
      debug: Debug,
      // This is usually used as a type import, we can remove the shim when we
      // ignore type imports in getImportsFromCode
      "@tscircuit/props": {}
    },
    circuit,
    tsConfig: null,
    importStack: [],
    currentlyImporting: /* @__PURE__ */ new Set(),
    ...webWorkerConfiguration
  };
}

// lib/runner/normalizeFsMap.ts
function normalizeFilePath(filePath) {
  let normFilePath = filePath;
  normFilePath = normFilePath.replace(/\\/g, "/");
  normFilePath = normFilePath.trim();
  if (normFilePath.startsWith("./")) {
    normFilePath = normFilePath.slice(2);
  }
  if (normFilePath.startsWith("/")) {
    normFilePath = normFilePath.slice(1);
  }
  return normFilePath;
}
function normalizeFsMap(fsMap) {
  const normalizedFsMap = {};
  for (const [fsPath, fileContent] of Object.entries(fsMap)) {
    normalizedFsMap[normalizeFilePath(fsPath)] = fileContent;
  }
  return normalizedFsMap;
}

// lib/runner/tsconfigPaths.ts
function getTsConfig(fsMapOrAllFilePaths) {
  if (Array.isArray(fsMapOrAllFilePaths)) return null;
  const tsconfigContent = fsMapOrAllFilePaths["tsconfig.json"];
  if (!tsconfigContent) return null;
  try {
    const sanitizedContent = tsconfigContent.replace(
      /\/\*[\s\S]*?\*\/|\/\/.*/g,
      ""
    );
    const parsed = JSON.parse(sanitizedContent);
    return parsed;
  } catch (e) {
    throw new Error(`Failed to parse tsconfig.json: ${e.message}`);
  }
}
function resolveWithTsconfigPaths(opts) {
  const { importPath, normalizedFilePathMap, extensions, tsConfig } = opts;
  const paths = tsConfig?.compilerOptions?.paths;
  if (!paths) return null;
  const baseUrl = tsConfig?.compilerOptions?.baseUrl || ".";
  const tryResolveCandidate = (candidate) => {
    const normalizedCandidate = normalizeFilePath(candidate);
    if (normalizedFilePathMap.has(normalizedCandidate)) {
      return normalizedFilePathMap.get(normalizedCandidate);
    }
    for (const ext of extensions) {
      const withExt = `${normalizedCandidate}.${ext}`;
      if (normalizedFilePathMap.has(withExt)) {
        return normalizedFilePathMap.get(withExt);
      }
    }
    return null;
  };
  for (const [alias, targets] of Object.entries(paths)) {
    const hasWildcard = alias.includes("*");
    if (hasWildcard) {
      const [prefix, suffix] = alias.split("*");
      if (!importPath.startsWith(prefix) || !importPath.endsWith(suffix || "")) {
        continue;
      }
      const starMatch = importPath.slice(
        prefix.length,
        importPath.length - (suffix ? suffix.length : 0)
      );
      for (const target of targets) {
        const replaced = target.replace("*", starMatch);
        const candidate = baseUrl && !replaced.startsWith("./") && !replaced.startsWith("/") ? `${baseUrl}/${replaced}` : replaced;
        const resolved = tryResolveCandidate(candidate);
        if (resolved) return resolved;
      }
    } else {
      if (importPath !== alias) continue;
      for (const target of targets) {
        const candidate = baseUrl && !target.startsWith("./") && !target.startsWith("/") ? `${baseUrl}/${target}` : target;
        const resolved = tryResolveCandidate(candidate);
        if (resolved) return resolved;
      }
    }
  }
  const resolvedPathFromBaseUrl = resolveWithBaseUrl({
    importPath,
    normalizedFilePathMap,
    extensions,
    tsConfig
  });
  if (resolvedPathFromBaseUrl) return resolvedPathFromBaseUrl;
  return null;
}
function resolveWithBaseUrl(opts) {
  const { importPath, normalizedFilePathMap, extensions, tsConfig } = opts;
  const baseUrl = tsConfig?.compilerOptions?.baseUrl;
  if (!baseUrl) return null;
  const filePathToResolve = `${baseUrl}/${importPath}`;
  const normalizedFilePath = normalizeFilePath(filePathToResolve);
  if (normalizedFilePathMap.has(normalizedFilePath)) {
    return normalizedFilePathMap.get(normalizedFilePath);
  }
  for (const ext of extensions) {
    const withExt = `${normalizedFilePath}.${ext}`;
    if (normalizedFilePathMap.has(withExt)) {
      return normalizedFilePathMap.get(withExt);
    }
  }
  return null;
}
function matchesTsconfigPathPattern(importPath, tsConfig) {
  const paths = tsConfig?.compilerOptions?.paths;
  if (!paths) return false;
  for (const [alias] of Object.entries(paths)) {
    const hasWildcard = alias.includes("*");
    if (hasWildcard) {
      const [prefix, suffix] = alias.split("*");
      if (importPath.startsWith(prefix) && importPath.endsWith(suffix || "")) {
        return true;
      }
    } else {
      if (importPath === alias) return true;
    }
  }
  return false;
}

// lib/runner/CircuitRunner.ts
import "react";

// lib/utils/dirname.ts
function dirname(path) {
  if (!path) return ".";
  const normalizedPath = path.replace(/\\/g, "/");
  const cleanPath = normalizedPath.replace(/\/+$/, "");
  if (cleanPath.indexOf("/") === -1) return ".";
  return cleanPath.substring(0, cleanPath.lastIndexOf("/")) || "/";
}

// lib/utils/resolveRelativePath.ts
function resolveRelativePath(importPath, cwd) {
  if (importPath.startsWith("../")) {
    const parentDir = dirname(cwd);
    return resolveRelativePath(importPath.slice(3), parentDir);
  }
  if (importPath.startsWith("./")) {
    return resolveRelativePath(importPath.slice(2), cwd);
  }
  if (importPath.startsWith("/")) {
    return importPath.slice(1);
  }
  return `${cwd}/${importPath}`;
}

// lib/runner/resolveFilePath.ts
var resolveFilePath = (unknownFilePath, fsMapOrAllFilePaths, cwd, opts = {}) => {
  const resolvedPath = cwd ? resolveRelativePath(unknownFilePath, cwd) : unknownFilePath;
  const filePaths = new Set(
    Array.isArray(fsMapOrAllFilePaths) ? fsMapOrAllFilePaths : Object.keys(fsMapOrAllFilePaths)
  );
  if (filePaths.has(resolvedPath)) {
    return resolvedPath;
  }
  const normalizedFilePathMap = /* @__PURE__ */ new Map();
  for (const filePath of filePaths) {
    normalizedFilePathMap.set(normalizeFilePath(filePath), filePath);
  }
  const normalizedResolvedPath = normalizeFilePath(resolvedPath);
  if (normalizedFilePathMap.has(normalizedResolvedPath)) {
    return normalizedFilePathMap.get(normalizedResolvedPath);
  }
  const extension = ["tsx", "ts", "json", "js", "jsx", "obj", "gltf", "glb"];
  for (const ext of extension) {
    const possibleFilePath = `${normalizedResolvedPath}.${ext}`;
    if (normalizedFilePathMap.has(possibleFilePath)) {
      return normalizedFilePathMap.get(possibleFilePath);
    }
  }
  const tsConfig = opts.tsConfig ?? null;
  if (!unknownFilePath.startsWith("./") && !unknownFilePath.startsWith("../")) {
    const resolvedPathFromPaths = resolveWithTsconfigPaths({
      importPath: unknownFilePath,
      normalizedFilePathMap,
      extensions: extension,
      tsConfig
    });
    if (resolvedPathFromPaths) return resolvedPathFromPaths;
    const resolvedPathFromBaseUrl = resolveWithBaseUrl({
      importPath: unknownFilePath,
      normalizedFilePathMap,
      extensions: extension,
      tsConfig
    });
    if (resolvedPathFromBaseUrl) return resolvedPathFromBaseUrl;
  }
  if (!unknownFilePath.startsWith("./") && !unknownFilePath.startsWith("../")) {
    const normalizedUnknownFilePath = normalizeFilePath(unknownFilePath);
    if (normalizedFilePathMap.has(normalizedUnknownFilePath)) {
      return normalizedFilePathMap.get(normalizedUnknownFilePath);
    }
    for (const ext of extension) {
      const possibleFilePath = `${normalizedUnknownFilePath}.${ext}`;
      if (normalizedFilePathMap.has(possibleFilePath)) {
        return normalizedFilePathMap.get(possibleFilePath);
      }
    }
  }
  return null;
};
var resolveFilePathOrThrow = (unknownFilePath, fsMapOrAllFilePaths, cwd, opts = {}) => {
  const resolvedFilePath = resolveFilePath(
    unknownFilePath,
    fsMapOrAllFilePaths,
    cwd,
    opts
  );
  if (!resolvedFilePath) {
    throw new Error(
      `File not found "${unknownFilePath}", available paths:

${Object.keys(fsMapOrAllFilePaths).join(", ")}`
    );
  }
  return resolvedFilePath;
};

// lib/utils/get-imports-from-code.ts
var getImportsFromCode = (code) => {
  const importRegex = /^\s*import\s+(?:(?:[\w\s]+,\s*)?(?:\*\s+as\s+[\w\s]+|\{[\s\w,]+\}|\w+)\s+from\s*)?['"](.+?)['"]/gm;
  const imports = [];
  let match;
  while ((match = importRegex.exec(code)) !== null) {
    const fullMatch = match[0];
    if (/^\s*import\s+type\b/.test(fullMatch)) {
      continue;
    }
    imports.push(match[1]);
  }
  const reExportRegex = /^\s*export\s+(?:type\s+)?(?:\*\s+as\s+[\w$]+|\*|\{[^}]+\})\s+from\s*['"](.+?)['"]/gm;
  let reExportMatch;
  while ((reExportMatch = reExportRegex.exec(code)) !== null) {
    const fullMatch = reExportMatch[0];
    if (/^\s*export\s+type\b/.test(fullMatch)) {
      continue;
    }
    imports.push(reExportMatch[1]);
  }
  return imports;
};

// webworker/eval-compiled-js.ts
function evalCompiledJs(compiledCode, preSuppliedImports, cwd) {
  ;
  globalThis.__tscircuit_require = (name) => {
    const resolvedFilePath = resolveFilePath(name, preSuppliedImports, cwd);
    const hasResolvedFilePath = resolvedFilePath && preSuppliedImports[resolvedFilePath];
    if (!preSuppliedImports[name] && !hasResolvedFilePath) {
      throw new Error(`Import "${name}" not found ${cwd ? `in "${cwd}"` : ""}`);
    }
    const mod = preSuppliedImports[name] || preSuppliedImports[resolvedFilePath];
    return new Proxy(mod, {
      get(target, prop) {
        if (!(prop in target)) {
          if (prop === "default") {
            if (target.default !== void 0) {
              return target.default;
            }
            if (target.__esModule) {
              return void 0;
            }
            if (typeof target === "function" || typeof target === "object") {
              return target;
            }
            return void 0;
          }
          if (prop === "__esModule") {
            return true;
          }
          throw new Error(
            `Component "${String(prop)}" is not exported by "${name}"`
          );
        }
        return target[prop];
      }
    });
  };
  const functionBody = `
  var exports = {};
  var require = globalThis.__tscircuit_require;
  var module = { exports };
  var circuit = globalThis.__tscircuit_circuit;
  ${compiledCode};
  return module;`.trim();
  return Function(functionBody).call(globalThis);
}

// webworker/import-local-file.ts
import Debug2 from "debug";

// lib/shared/static-asset-extensions.ts
var STATIC_ASSET_EXTENSIONS = [
  ".glb",
  ".kicad_mod",
  ".gltf",
  ".obj",
  ".stl"
];
var isStaticAssetPath = (path) => STATIC_ASSET_EXTENSIONS.some((ext) => path.endsWith(ext));

// webworker/transform-with-sucrase.ts
import { transform } from "sucrase";
var TS_EXTENSIONS = /* @__PURE__ */ new Set([".ts", ".tsx", ".mts", ".cts"]);
var JSX_EXTENSIONS = /* @__PURE__ */ new Set([".tsx", ".jsx", ".ts"]);
var TYPE_STAR_EXPORT_REGEX = /^\s*export\s+type\s+\*\s+(?:as\s+[\w$]+\s+)?from\s+['"][^'"]+['"]\s*;?\s*$/gim;
var stripTypeStarExports = (code) => code.replace(TYPE_STAR_EXPORT_REGEX, "");
var stripQueryAndHash = (filePath) => {
  const queryIndex = filePath.indexOf("?");
  const hashIndex = filePath.indexOf("#");
  let endIndex = filePath.length;
  if (queryIndex !== -1 && hashIndex !== -1) {
    endIndex = Math.min(queryIndex, hashIndex);
  } else if (queryIndex !== -1) {
    endIndex = queryIndex;
  } else if (hashIndex !== -1) {
    endIndex = hashIndex;
  }
  return filePath.slice(0, endIndex);
};
var getExtension = (filePath) => {
  const normalizedPath = stripQueryAndHash(filePath);
  const lastDotIndex = normalizedPath.lastIndexOf(".");
  if (lastDotIndex === -1) {
    return "";
  }
  const lastSlashIndex = Math.max(
    normalizedPath.lastIndexOf("/"),
    normalizedPath.lastIndexOf("\\")
  );
  if (lastSlashIndex > lastDotIndex) {
    return "";
  }
  return normalizedPath.slice(lastDotIndex).toLowerCase();
};
var getTransformsForFilePath = (filePath) => {
  const extension = getExtension(filePath);
  const transforms = ["imports"];
  if (TS_EXTENSIONS.has(extension)) {
    transforms.unshift("typescript");
  }
  if (JSX_EXTENSIONS.has(extension)) {
    transforms.push("jsx");
  }
  return transforms;
};
var transformWithSucrase = (code, filePath) => {
  const transforms = getTransformsForFilePath(filePath);
  const sanitizedCode = stripTypeStarExports(code);
  const { code: transformedCode } = transform(sanitizedCode, {
    filePath,
    production: true,
    transforms
  });
  return transformedCode;
};

// webworker/import-local-file.ts
import { KicadToCircuitJsonConverter } from "kicad-to-circuit-json";
var debug2 = Debug2("tsci:eval:import-local-file");
var importLocalFile = async (importName, ctx, depth = 0) => {
  debug2("importLocalFile called with:", {
    importName
  });
  const { fsMap, preSuppliedImports, importStack, currentlyImporting } = ctx;
  const fsPath = resolveFilePathOrThrow(importName, fsMap, void 0, {
    tsConfig: ctx.tsConfig
  });
  debug2("fsPath:", fsPath);
  if (currentlyImporting.has(fsPath)) {
    const cycleStartIndex = importStack.indexOf(fsPath);
    const cyclePath = cycleStartIndex >= 0 ? importStack.slice(cycleStartIndex).concat(fsPath) : [...importStack, fsPath];
    throw new Error(
      `Circular dependency detected while importing "${fsPath}". The following import chain forms a cycle:

${cyclePath.join(
        " -> "
      )}`
    );
  }
  if (!ctx.fsMap[fsPath]) {
    debug2("fsPath not found in fsMap:", fsPath);
    throw new Error(`File "${fsPath}" not found`);
  }
  const fileContent = fsMap[fsPath];
  debug2("fileContent:", fileContent?.slice(0, 100));
  currentlyImporting.add(fsPath);
  importStack.push(fsPath);
  try {
    if (fsPath.endsWith(".json")) {
      const jsonData = JSON.parse(fileContent);
      preSuppliedImports[fsPath] = {
        __esModule: true,
        default: jsonData
      };
    } else if (fsPath.endsWith(".kicad_pcb")) {
      const converter = new KicadToCircuitJsonConverter();
      converter.addFile(fsPath, fileContent);
      converter.runUntilFinished();
      const circuitJson = converter.getOutput();
      preSuppliedImports[fsPath] = {
        __esModule: true,
        circuitJson
      };
    } else if (isStaticAssetPath(fsPath)) {
      let staticUrl;
      if (fileContent === "__STATIC_ASSET__") {
        const platformConfig = ctx.circuit.platform;
        staticUrl = `${platformConfig?.projectBaseUrl ?? ""}/${fsPath.startsWith("./") ? fsPath.slice(2) : fsPath}`;
      } else {
        const blob = new Blob([fileContent], {
          type: fsPath.endsWith(".kicad_mod") ? "text/plain" : "application/octet-stream"
        });
        staticUrl = URL.createObjectURL(blob);
      }
      preSuppliedImports[fsPath] = {
        __esModule: true,
        default: staticUrl
      };
    } else if (fsPath.endsWith(".tsx") || fsPath.endsWith(".ts")) {
      const importNames = getImportsFromCode(fileContent);
      for (const importName2 of importNames) {
        if (!preSuppliedImports[importName2]) {
          await importEvalPath(importName2, ctx, depth + 1, {
            cwd: dirname(fsPath)
          });
        }
      }
      try {
        const transformedCode = transformWithSucrase(fileContent, fsPath);
        debug2("evalCompiledJs called with:", {
          code: transformedCode.slice(0, 100),
          dirname: dirname(fsPath)
        });
        const importRunResult = evalCompiledJs(
          transformedCode,
          preSuppliedImports,
          dirname(fsPath)
        );
        debug2("importRunResult:", {
          fsPath,
          importRunResult
        });
        preSuppliedImports[fsPath] = importRunResult.exports;
      } catch (error) {
        throw new Error(
          `Eval compiled js error for "${importName}": ${error.message}`
        );
      }
    } else if (fsPath.endsWith(".js")) {
      preSuppliedImports[fsPath] = evalCompiledJs(
        transformWithSucrase(fileContent, fsPath),
        preSuppliedImports,
        dirname(fsPath)
      ).exports;
    } else {
      throw new Error(
        `Unsupported file extension "${fsPath.split(".").pop()}" for "${fsPath}"`
      );
    }
  } finally {
    importStack.pop();
    currentlyImporting.delete(fsPath);
  }
};

// webworker/import-snippet.ts
async function importSnippet(importName, ctx, depth = 0) {
  const { preSuppliedImports } = ctx;
  const fullSnippetName = importName.replace("@tsci/", "").replace(".", "/");
  const { cjs, error } = await globalThis.fetch(`${ctx.cjsRegistryUrl}/${fullSnippetName}`).then(async (res) => ({ cjs: await res.text(), error: null })).catch((e) => ({ error: e, cjs: null }));
  if (error) {
    console.error("Error fetching import", importName, error);
    return;
  }
  try {
    preSuppliedImports[importName] = evalCompiledJs(
      cjs,
      preSuppliedImports
    ).exports;
  } catch (e) {
    console.error("Error importing snippet", e);
  }
}

// lib/utils/resolve-node-module.ts
function createContext(modulePath, fsMap, basePath) {
  return {
    fsMap,
    extensions: [".js", ".jsx", ".ts", ".tsx", ".json"],
    basePath,
    modulePath
  };
}
function findPackageJson(nodeModulesPath, ctx) {
  const packageJsonPath = `${nodeModulesPath}/package.json`;
  if (!ctx.fsMap[packageJsonPath]) return null;
  try {
    return JSON.parse(ctx.fsMap[packageJsonPath]);
  } catch {
    return null;
  }
}
function tryResolveWithExtensions(path, ctx) {
  if (ctx.fsMap[path]) return path;
  for (const ext of ctx.extensions) {
    const pathWithExt = path.replace(/\.js$|\.jsx$/, "") + ext;
    if (ctx.fsMap[pathWithExt]) return pathWithExt;
  }
  return null;
}
function resolveExportPath(nodeModulesPath, exportPath, ctx) {
  const fullExportPath = `${nodeModulesPath}/${exportPath.replace(/^\.\//, "")}`;
  return tryResolveWithExtensions(fullExportPath, ctx);
}
function resolvePackageExports(nodeModulesPath, packageJson, remainingPath, ctx) {
  if (!packageJson.exports) return null;
  const defaultExport = packageJson.exports["."];
  if (remainingPath === "" && defaultExport) {
    if (typeof defaultExport === "string") {
      const resolved = resolveExportPath(nodeModulesPath, defaultExport, ctx);
      if (resolved) return resolved;
    }
  }
  const subpathExport = remainingPath ? packageJson.exports[`./${remainingPath}`] : null;
  if (subpathExport && typeof subpathExport === "string") {
    const resolved = resolveExportPath(nodeModulesPath, subpathExport, ctx);
    if (resolved) return resolved;
  }
  const importExport = packageJson.exports["import"];
  if (remainingPath === "" && importExport && typeof importExport === "string") {
    const resolved = resolveExportPath(nodeModulesPath, importExport, ctx);
    if (resolved) return resolved;
  }
  return null;
}
function resolvePackageEntryPoint(nodeModulesPath, packageJson, ctx) {
  const entryPoint = packageJson.module || packageJson.main || "index.js";
  const fullPath = `${nodeModulesPath}/${entryPoint}`;
  return tryResolveWithExtensions(fullPath, ctx);
}
function resolveRemainingPath(nodeModulesPath, remainingPath, ctx) {
  if (!remainingPath) {
    for (const ext of ctx.extensions) {
      const indexPath = `${nodeModulesPath}/index${ext}`;
      if (ctx.fsMap[indexPath]) return indexPath;
    }
    return null;
  }
  const fullPath = `${nodeModulesPath}/${remainingPath}`;
  const directMatch = tryResolveWithExtensions(fullPath, ctx);
  if (directMatch) return directMatch;
  for (const ext of ctx.extensions) {
    const indexPath = `${fullPath}/index${ext}`;
    if (ctx.fsMap[indexPath]) return indexPath;
  }
  return null;
}
function resolveNodeModuleInPath(searchPath, ctx) {
  const moduleParts = ctx.modulePath.split("/");
  const scope = moduleParts[0].startsWith("@") ? moduleParts.slice(0, 2).join("/") : moduleParts[0];
  const remainingPath = moduleParts.slice(scope.includes("/") ? 2 : 1).join("/");
  const nodeModulesPath = `${searchPath == "." ? "" : `${searchPath}/`}node_modules/${scope}`;
  const packageJson = findPackageJson(nodeModulesPath, ctx);
  if (packageJson) {
    const exportsResolution = resolvePackageExports(
      nodeModulesPath,
      packageJson,
      remainingPath,
      ctx
    );
    if (exportsResolution) return exportsResolution;
    const entryPointResolution = resolvePackageEntryPoint(
      nodeModulesPath,
      packageJson,
      ctx
    );
    if (entryPointResolution) return entryPointResolution;
  }
  const remainingPathResolution = resolveRemainingPath(
    nodeModulesPath,
    remainingPath,
    ctx
  );
  if (remainingPathResolution) return remainingPathResolution;
  const parentPath = dirname(searchPath);
  if (parentPath && parentPath !== searchPath) {
    return resolveNodeModuleInPath(parentPath, ctx);
  }
  return null;
}
function resolveNodeModule(modulePath, fsMap, basePath) {
  const ctx = createContext(modulePath, fsMap, basePath);
  return resolveNodeModuleInPath(ctx.basePath, ctx);
}

// webworker/import-node-module.ts
import Debug3 from "debug";
var debug3 = Debug3("tsci:eval:import-node-module");
var importNodeModule = async (importName, ctx, depth = 0) => {
  const { preSuppliedImports } = ctx;
  if (preSuppliedImports[importName]) {
    return;
  }
  const resolvedNodeModulePath = resolveNodeModule(importName, ctx.fsMap, "");
  if (!resolvedNodeModulePath) {
    const platform = ctx.circuit?.platform;
    if (platform?.nodeModulesResolver) {
      debug3(`Attempting to resolve "${importName}" using nodeModulesResolver`);
      try {
        const fileContent = await platform.nodeModulesResolver(importName);
        if (fileContent) {
          debug3(`Successfully resolved "${importName}" via nodeModulesResolver`);
          const syntheticPath = `node_modules/${importName}.ts`;
          ctx.fsMap[syntheticPath] = fileContent;
          await importLocalFile(syntheticPath, ctx, depth);
          preSuppliedImports[importName] = preSuppliedImports[syntheticPath];
          const unprefixedPath2 = syntheticPath.replace(/^node_modules\//, "");
          preSuppliedImports[unprefixedPath2] = preSuppliedImports[syntheticPath];
          return;
        }
        debug3(`nodeModulesResolver returned null for "${importName}"`);
      } catch (error) {
        debug3(`nodeModulesResolver failed for "${importName}":`, error);
      }
    }
    throw new Error(`Node module "${importName}" not found`);
  }
  await importLocalFile(resolvedNodeModulePath, ctx, depth);
  preSuppliedImports[importName] = preSuppliedImports[resolvedNodeModulePath];
  const unprefixedPath = resolvedNodeModulePath.replace(/^node_modules\//, "");
  preSuppliedImports[unprefixedPath] = preSuppliedImports[resolvedNodeModulePath];
  if (resolvedNodeModulePath.endsWith("/index.tsx") || resolvedNodeModulePath.endsWith("/index.ts") || resolvedNodeModulePath.endsWith("/index.js")) {
    const dirPath = resolvedNodeModulePath.replace(/\/index\.(tsx?|js)$/, "");
    const unprefixedDirPath = dirPath.replace(/^node_modules\//, "");
    preSuppliedImports[unprefixedDirPath] = preSuppliedImports[resolvedNodeModulePath];
    if (unprefixedDirPath.startsWith("@")) {
      const scopeParts = unprefixedDirPath.split("/");
      if (scopeParts.length >= 2) {
        const scopedName = `${scopeParts[0]}/${scopeParts[1]}`;
        preSuppliedImports[scopedName] = preSuppliedImports[resolvedNodeModulePath];
      }
    }
  }
};

// webworker/import-npm-package.ts
import Debug4 from "debug";
var debug4 = Debug4("tsci:eval:import-npm-package");
function extractPackagePathFromJSDelivr(url) {
  const prefix = "https://cdn.jsdelivr.net/npm/";
  if (url.startsWith(prefix)) {
    return url.substring(prefix.length).replace(/\/\+esm$/, "");
  }
  return url;
}
async function importNpmPackage(importName, ctx, depth = 0) {
  debug4(`importing npm package: ${importName}`);
  const { preSuppliedImports } = ctx;
  if (preSuppliedImports[importName]) return;
  const npmCdnUrl = `https://cdn.jsdelivr.net/npm/${importName}/+esm`;
  let finalUrl;
  const { content, error } = await globalThis.fetch(npmCdnUrl).then(async (res) => {
    finalUrl = res.url;
    if (!res.ok)
      throw new Error(
        `Could not fetch "${importName}" from jsdelivr: ${res.statusText}

${ctx.logger.stringifyLogs()}`
      );
    return { content: await res.text(), error: null };
  }).catch((e) => ({ error: e, content: null }));
  if (error) {
    console.error("Error fetching npm import", importName, error);
    throw error;
  }
  const finalImportName = extractPackagePathFromJSDelivr(finalUrl);
  const cwd = dirname(finalImportName);
  const importNames = getImportsFromCode(content);
  for (const subImportName of importNames) {
    if (!preSuppliedImports[subImportName]) {
      await importEvalPath(subImportName, ctx, depth + 1, {
        cwd
      });
    }
  }
  const transformedCode = transformWithSucrase(
    content,
    finalImportName || importName
  );
  try {
    const exports = evalCompiledJs(
      transformedCode,
      preSuppliedImports,
      cwd
    ).exports;
    preSuppliedImports[importName] = exports;
    preSuppliedImports[finalImportName] = exports;
    preSuppliedImports[finalUrl] = exports;
  } catch (e) {
    throw new Error(
      `Eval npm package error for "${importName}": ${e.message}

${ctx.logger.stringifyLogs()}`
    );
  }
}

// webworker/import-eval-path.ts
import Debug5 from "debug";
var debug5 = Debug5("tsci:eval:import-eval-path");
async function importEvalPath(importName, ctx, depth = 0, opts = {}) {
  debug5("importEvalPath called with:", {
    importName,
    depth,
    opts
  });
  ctx.logger.info(
    `importEvalPath("${importName}", {cwd: "${opts.cwd}", depth: ${depth}})`
  );
  debug5(`${"  ".repeat(depth)}\u27A1\uFE0F`, importName);
  const { preSuppliedImports } = ctx;
  if (preSuppliedImports[importName]) {
    ctx.logger.info(`Import "${importName}" in preSuppliedImports[1]`);
    return;
  }
  if (importName.startsWith("./") && preSuppliedImports[importName.slice(2)]) {
    ctx.logger.info(`Import "${importName}" in preSuppliedImports[2]`);
    return;
  }
  if (depth > 30) {
    throw new Error(
      `Max depth for imports reached (30) Import Path: ${ctx.importStack.join(" -> ")}`
    );
  }
  if (importName.startsWith("/npm/")) {
    const pkgName = importName.replace(/^\/npm\//, "").replace(/\/\+esm$/, "");
    ctx.logger.info(`importNpmPackage("${pkgName}")`);
    await importNpmPackage(pkgName, ctx, depth);
    const pkg = preSuppliedImports[pkgName];
    if (pkg) {
      preSuppliedImports[importName] = pkg;
    }
    return;
  }
  const resolvedLocalImportPath = resolveFilePath(
    importName,
    ctx.fsMap,
    opts.cwd,
    { tsConfig: ctx.tsConfig }
  );
  if (resolvedLocalImportPath) {
    ctx.logger.info(`importLocalFile("${resolvedLocalImportPath}")`);
    await importLocalFile(resolvedLocalImportPath, ctx, depth);
    if (importName !== resolvedLocalImportPath) {
      preSuppliedImports[importName] = preSuppliedImports[resolvedLocalImportPath];
    }
    return;
  }
  const tsConfig = ctx.tsConfig ?? getTsConfig(ctx.fsMap);
  if (!ctx.tsConfig && tsConfig) {
    ctx.tsConfig = tsConfig;
  }
  if (matchesTsconfigPathPattern(importName, tsConfig)) {
    throw new Error(
      `Import "${importName}" matches a tsconfig path alias but could not be resolved to an existing file${opts.cwd ? ` from directory "${opts.cwd}"` : ""}

${ctx.logger.stringifyLogs()}`
    );
  }
  const resolvedNodeModulePath = resolveNodeModule(
    importName,
    ctx.fsMap,
    opts.cwd || ""
  );
  if (resolvedNodeModulePath) {
    ctx.logger.info(`resolvedNodeModulePath="${resolvedNodeModulePath}"`);
    ctx.logger.info(`importNodeModule("${importName}")`);
    return importNodeModule(importName, ctx, depth);
  }
  if (!importName.startsWith(".") && !importName.startsWith("/") && !importName.startsWith("@tsci/")) {
    const platform = ctx.circuit?.platform;
    if (platform?.nodeModulesResolver) {
      ctx.logger.info(
        `importNodeModule("${importName}") via nodeModulesResolver`
      );
      try {
        await importNodeModule(importName, ctx, depth);
        return;
      } catch (error) {
        ctx.logger.info(
          `nodeModulesResolver failed for "${importName}", falling back to npm CDN`
        );
      }
    }
  }
  if (importName.startsWith("@tsci/")) {
    ctx.logger.info(`importSnippet("${importName}")`);
    return importSnippet(importName, ctx, depth);
  }
  if (!importName.startsWith(".") && !importName.startsWith("/")) {
    ctx.logger.info(`importNpmPackage("${importName}")`);
    return importNpmPackage(importName, ctx, depth);
  }
  throw new Error(
    `Unresolved import "${importName}" ${opts.cwd ? `from directory "${opts.cwd}"` : ""}

${ctx.logger.stringifyLogs()}`
  );
}

// lib/runner/setupDefaultEntrypointIfNeeded.ts
import Debug6 from "debug";
var debug6 = Debug6("tsci:eval:setupDefaultEntrypointIfNeeded");
var setupDefaultEntrypointIfNeeded = (opts) => {
  if (!opts.entrypoint && !opts.mainComponentPath) {
    if ("index.tsx" in opts.fsMap) {
      opts.mainComponentPath = "index.tsx";
    } else if ("index.ts" in opts.fsMap) {
      opts.mainComponentPath = "index.ts";
    } else if (Object.keys(opts.fsMap).filter((k) => k.endsWith(".tsx")).length === 1) {
      opts.mainComponentPath = Object.keys(opts.fsMap)[0];
    } else if ("tscircuit.config.json" in opts.fsMap) {
      const configContent = opts.fsMap["tscircuit.config.json"];
      try {
        const config = JSON.parse(configContent);
        if (config.mainEntrypoint) {
          opts.entrypoint = config.mainEntrypoint;
        }
      } catch (e) {
        console.warn("Failed to parse tscircuit.config.json:", e);
      }
    } else {
      throw new Error(
        "Either entrypoint or mainComponentPath must be provided (no index file, could not infer entrypoint)"
      );
    }
  }
  if (!opts.entrypoint && opts.mainComponentPath) {
    opts.entrypoint = "entrypoint.tsx";
    const mainComponentCode = opts.fsMap[resolveFilePathOrThrow(opts.mainComponentPath, opts.fsMap)];
    if (!mainComponentCode) {
      throw new Error(
        `Main component path "${opts.mainComponentPath}" not found in fsMap. Available paths: ${Object.keys(opts.fsMap).join(", ")}`
      );
    }
    opts.fsMap[opts.entrypoint] = `
     import * as UserComponents from "./${opts.mainComponentPath}";
          
      ${opts.mainComponentName ? `
        const ComponentToRender = UserComponents["${opts.mainComponentName}"]
        ` : `const ComponentToRender = UserComponents.default || 
          Object.entries(UserComponents)
          .filter(([name]) => !name.startsWith("use"))
          .map(([_, component]) => component)[0] || (() => null);`}

           ${debug6.enabled ? `
     console.log({ UserComponents })
     console.log("ComponentToRender " + ComponentToRender.toString(),  { ComponentToRender })
     ` : ""}

      circuit.add(       
          <ComponentToRender ${opts.mainComponentProps ? `{...${JSON.stringify(opts.mainComponentProps, null, 2)}}` : ""} /> 
      );
`;
  }
  if (!opts.name && opts.mainComponentName) {
    opts.name = opts.mainComponentName;
  }
};

// lib/utils/enhance-root-circuit-error.ts
var enhanceRootCircuitHasNoChildrenError = (error, entrypoint) => {
  if (error instanceof Error && entrypoint && error.message.includes("RootCircuit has no children") && !error.message.includes('"entrypoint":')) {
    const entrypointMessage = entrypoint.startsWith("./") ? entrypoint.slice(2) : entrypoint;
    error.message = `${error.message}. "entrypoint": "${entrypointMessage}" is set in the runner configuration, entrypoints must contain "circuit.add(...)", you might be looking to use mainComponentPath instead if your file exports a component.`;
  }
  return error;
};

// lib/runner/CircuitRunner.ts
import Debug7 from "debug";

// lib/shared/obj-path.ts
function getObjectPaths(obj, prefix = "") {
  const paths = [];
  for (const key in obj) {
    if (!Object.prototype.hasOwnProperty.call(obj, key)) {
      continue;
    }
    const value = obj[key];
    const path = prefix ? `${prefix}.${key}` : key;
    if (value !== null && typeof value === "object" && !Array.isArray(value)) {
      const subPaths = getObjectPaths(value, path);
      paths.push(...subPaths);
    } else {
      paths.push(path);
    }
  }
  return paths;
}
function getValueAtPath(obj, path) {
  const keys = path.split(".");
  let current = obj;
  for (const key of keys) {
    current = current[key];
  }
  return current;
}
function setValueAtPath(obj, path, value) {
  const keys = path.split(".");
  let current = obj;
  for (let i = 0; i < keys.length; i++) {
    const key = keys[i];
    if (i === keys.length - 1) {
      current[key] = value;
    } else {
      if (!Object.prototype.hasOwnProperty.call(current, key) || typeof current[key] !== "object" || current[key] === null || Array.isArray(current[key])) {
        current[key] = {};
      }
      current = current[key];
    }
  }
}

// lib/runner/CircuitRunner.ts
var debug7 = Debug7("tsci:eval:CircuitRunner");
var CircuitRunner = class {
  constructor(configuration = {}) {
    this._executionContext = null;
    this._circuitRunnerConfiguration = {
      snippetsApiBaseUrl: "https://registry-api.tscircuit.com",
      cjsRegistryUrl: "https://cjs.tscircuit.com",
      verbose: false
    };
    this._eventListeners = {};
    Object.assign(this._circuitRunnerConfiguration, configuration);
  }
  async version() {
    return "0.0.0";
  }
  async executeWithFsMap(ogOpts) {
    const opts = { ...ogOpts };
    if (this._circuitRunnerConfiguration.verbose) {
      Debug7.enable("tsci:eval:*");
    }
    debug7("executeWithFsMap called with:", {
      entrypoint: opts.entrypoint,
      fsMapKeys: Object.keys(opts.fsMap),
      name: opts.name
    });
    setupDefaultEntrypointIfNeeded(opts);
    debug7("entrypoint after setupDefaultEntrypointIfNeeded:", {
      entrypoint: opts.entrypoint
    });
    this._executionContext = createExecutionContext(
      this._circuitRunnerConfiguration,
      {
        name: opts.name,
        platform: this._circuitRunnerConfiguration.platform,
        projectConfig: this._circuitRunnerConfiguration.projectConfig,
        debugNamespace: this._debugNamespace
      }
    );
    this._bindEventListeners(this._executionContext.circuit);
    this._executionContext.entrypoint = opts.entrypoint;
    this._executionContext.fsMap = normalizeFsMap(opts.fsMap);
    this._executionContext.tsConfig = getTsConfig(this._executionContext.fsMap);
    if (!this._executionContext.fsMap[opts.entrypoint]) {
      throw new Error(`Entrypoint "${opts.entrypoint}" not found`);
    }
    ;
    globalThis.__tscircuit_circuit = this._executionContext.circuit;
    const entrypoint = opts.entrypoint.startsWith("./") ? opts.entrypoint : `./${opts.entrypoint}`;
    debug7("final entrypoint:", entrypoint);
    await importEvalPath(entrypoint, this._executionContext);
  }
  async execute(code, opts = {}) {
    if (this._circuitRunnerConfiguration.verbose) {
      console.log(
        "[CircuitRunner] execute called with code length:",
        code.length
      );
    }
    this._executionContext = createExecutionContext(
      this._circuitRunnerConfiguration,
      {
        ...opts,
        platform: this._circuitRunnerConfiguration.platform,
        projectConfig: this._circuitRunnerConfiguration.projectConfig,
        debugNamespace: this._debugNamespace
      }
    );
    this._bindEventListeners(this._executionContext.circuit);
    this._executionContext.fsMap["entrypoint.tsx"] = code;
    this._executionContext.tsConfig = getTsConfig(this._executionContext.fsMap);
    globalThis.__tscircuit_circuit = this._executionContext.circuit;
    await importEvalPath("./entrypoint.tsx", this._executionContext);
  }
  async executeComponent(component, opts = {}) {
    if (this._circuitRunnerConfiguration.verbose) {
      console.log("[CircuitRunner] executeComponent called");
    }
    this._executionContext = createExecutionContext(
      this._circuitRunnerConfiguration,
      {
        ...opts,
        platform: this._circuitRunnerConfiguration.platform,
        projectConfig: this._circuitRunnerConfiguration.projectConfig,
        debugNamespace: this._debugNamespace
      }
    );
    this._bindEventListeners(this._executionContext.circuit);
    globalThis.__tscircuit_circuit = this._executionContext.circuit;
    this._executionContext.tsConfig = null;
    const element = typeof component === "function" ? component() : component;
    this._executionContext.circuit.add(element);
  }
  on(event, callback) {
    var _a;
    (_a = this._eventListeners)[event] ?? (_a[event] = []);
    this._eventListeners[event].push(callback);
    this._executionContext?.circuit.on(event, callback);
  }
  async renderUntilSettled() {
    if (!this._executionContext) {
      throw new Error("No circuit has been created");
    }
    try {
      await this._executionContext.circuit.renderUntilSettled();
    } catch (error) {
      throw enhanceRootCircuitHasNoChildrenError(
        error,
        this._executionContext.entrypoint
      );
    }
  }
  async getCircuitJson() {
    if (!this._executionContext) {
      throw new Error("No circuit has been created");
    }
    try {
      return this._executionContext.circuit.getCircuitJson();
    } catch (error) {
      throw enhanceRootCircuitHasNoChildrenError(
        error,
        this._executionContext.entrypoint
      );
    }
  }
  clearEventListeners() {
    if (this._executionContext?.circuit) {
      for (const event in this._eventListeners) {
        for (const listener of this._eventListeners[event]) {
          const circuit = this._executionContext.circuit;
          circuit.removeListener?.(event, listener);
        }
      }
    }
    for (const event in this._eventListeners) {
      delete this._eventListeners[event];
    }
  }
  async kill() {
    this._executionContext = null;
  }
  async setSnippetsApiBaseUrl(baseUrl) {
    this._circuitRunnerConfiguration.snippetsApiBaseUrl = baseUrl;
  }
  async setPlatformConfig(platform) {
    this._circuitRunnerConfiguration.platform = platform;
  }
  async setPlatformConfigProperty(property, value) {
    if (!this._circuitRunnerConfiguration.platform) {
      this._circuitRunnerConfiguration.platform = {};
    }
    setValueAtPath(this._circuitRunnerConfiguration.platform, property, value);
  }
  async setProjectConfig(project) {
    this._circuitRunnerConfiguration.projectConfig = project;
  }
  async setProjectConfigProperty(property, value) {
    if (!this._circuitRunnerConfiguration.projectConfig) {
      this._circuitRunnerConfiguration.projectConfig = {};
    }
    setValueAtPath(
      this._circuitRunnerConfiguration.projectConfig,
      property,
      value
    );
  }
  async enableDebug(namespace) {
    this._debugNamespace = namespace;
    if (this._executionContext) {
      const circuit = this._executionContext.circuit;
      circuit.enableDebug?.(namespace);
    }
  }
  _bindEventListeners(circuit) {
    for (const event in this._eventListeners) {
      for (const listener of this._eventListeners[event]) {
        circuit.on(event, listener);
      }
    }
  }
};

// lib/runner/runTscircuitCode.ts
async function runTscircuitCode(filesystemOrCodeString, opts) {
  if (typeof filesystemOrCodeString === "string" && !filesystemOrCodeString.includes("export")) {
    throw new Error(
      `The "export" keyword wasn't found in your provided code. You need to export a component in your code, e.g.

export default () => (
  <resistor name="R1" resistance="1k" />
)`
    );
  }
  const filesystem = typeof filesystemOrCodeString === "string" ? { "user-code.tsx": filesystemOrCodeString } : filesystemOrCodeString;
  const circuitRunner = new CircuitRunner();
  await circuitRunner.executeWithFsMap({
    fsMap: filesystem,
    ...opts
  });
  await circuitRunner.renderUntilSettled();
  return await circuitRunner.getCircuitJson();
}

// lib/runner/runTscircuitModule.ts
var runTscircuitModule = async (module, opts = {}) => {
  if (!module.startsWith("@")) {
    module = `@tsci/${module.replace(/\//, ".")}`;
  }
  const circuitJson = await runTscircuitCode(
    {
      // TODO handle exports that are not the default export by scanning
      // otherExports for components
      "user-code.tsx": `
    import Module, * as otherExports from "${module}";

    let exportName = "${opts.exportName ?? ""}"

    if ((!Module || typeof Module !== "function") && !Boolean(exportName)) {
      exportName = Object.keys(otherExports).filter(key => key[0] === key[0].toUpperCase() && typeof otherExports[key] === "function")[0]
    }

    const defaultExport = exportName ? otherExports[exportName] : Module

    if (!defaultExport) {
      throw new Error(\`No export found for module "\${module}" (tried "\${exportName ?? "default"}")\`)
    }

    export default defaultExport;
    `
    },
    {
      mainComponentProps: opts.props
    }
  );
  return circuitJson;
};

// lib/worker.ts
import * as Comlink from "comlink";
var createCircuitWebWorker = async (configuration) => {
  const existingWorker = globalThis.TSCIRCUIT_GLOBAL_CIRCUIT_WORKER;
  if (existingWorker && typeof existingWorker.kill === "function") {
    if (configuration.verbose) {
      console.log("[Worker] Killing previous global worker instance...");
    }
    try {
      await existingWorker.kill();
    } catch (e) {
      if (configuration.verbose) {
        console.warn(
          "[Worker] Error killing previous global worker instance:",
          e
        );
      }
      if (globalThis.TSCIRCUIT_GLOBAL_CIRCUIT_WORKER === existingWorker) {
        globalThis.TSCIRCUIT_GLOBAL_CIRCUIT_WORKER = void 0;
      }
    }
  }
  if (configuration.verbose) {
    console.log(
      "[Worker] Creating circuit web worker with config:",
      configuration
    );
  }
  let workerBlobUrl = configuration.webWorkerBlobUrl ?? configuration.webWorkerUrl;
  if (!workerBlobUrl) {
    const cdnUrl = `https://cdn.jsdelivr.net/npm/@tscircuit/eval@${configuration.evalVersion ?? "latest"}/dist/webworker/entrypoint.js`;
    const workerBlob = await globalThis.fetch(cdnUrl).then((res) => res.blob());
    workerBlobUrl = URL.createObjectURL(workerBlob);
  }
  const rawWorker = new Worker(workerBlobUrl, { type: "module" });
  let workerInitError;
  rawWorker.addEventListener("error", (event) => {
    console.error("[Worker] Error in worker", event);
    workerInitError = event;
  });
  rawWorker.addEventListener("unhandledrejection", (event) => {
    console.error("[Worker] Unhandled rejection in worker", event);
  });
  rawWorker.addEventListener("messageerror", (event) => {
    console.error("[Worker] Message error in worker", event);
  });
  const earlyMessageHandler = (event) => {
    console.log("[Worker] Message in worker", event);
  };
  rawWorker.addEventListener("message", earlyMessageHandler);
  rawWorker.addEventListener("message", async (event) => {
    const data = event.data;
    if (data?.type !== "worker_fetch") return;
    try {
      const response = await globalThis.fetch(data.input, data.init);
      const body = await response.text();
      rawWorker.postMessage({
        type: "worker_fetch_result",
        requestId: data.requestId,
        success: true,
        response: {
          body,
          status: response.status,
          statusText: response.statusText,
          headers: (() => {
            const obj = {};
            response.headers.forEach((value, key) => {
              obj[key] = value;
            });
            return obj;
          })()
        }
      });
    } catch (err) {
      rawWorker.postMessage({
        type: "worker_fetch_result",
        requestId: data.requestId,
        success: false,
        error: {
          name: err.name,
          message: err.message,
          stack: err.stack
        }
      });
    }
  });
  if (workerInitError) {
    throw workerInitError;
  }
  const comlinkWorker = Comlink.wrap(rawWorker);
  rawWorker.removeEventListener("message", earlyMessageHandler);
  function serializeReactElement(element) {
    if (!element || typeof element !== "object") {
      return element;
    }
    if (element.type && element.props !== void 0) {
      return {
        __isSerializedReactElement: true,
        type: element.type,
        props: serializeProps(element.props),
        key: element.key
      };
    }
    return element;
  }
  function serializeProps(props) {
    if (!props || typeof props !== "object") {
      return props;
    }
    const serialized = {};
    for (const [key, value] of Object.entries(props)) {
      if (key === "children") {
        if (Array.isArray(value)) {
          serialized.children = value.map(serializeReactElement);
        } else {
          serialized.children = serializeReactElement(value);
        }
      } else {
        serialized[key] = value;
      }
    }
    return serialized;
  }
  if (configuration.enableFetchProxy) {
    rawWorker.postMessage({ type: "override_global_fetch" });
  }
  if (configuration.snippetsApiBaseUrl) {
    await comlinkWorker.setSnippetsApiBaseUrl(configuration.snippetsApiBaseUrl);
  }
  const maybeProxy = (value) => {
    if (typeof value === "function") {
      return Comlink.proxy(value);
    }
    return value;
  };
  if (configuration.platform) {
    for (const path of getObjectPaths(configuration.platform)) {
      await comlinkWorker.setPlatformConfigProperty(
        path,
        maybeProxy(getValueAtPath(configuration.platform, path))
      ).catch((e) => {
        throw new Error(
          `Error setting platform config property ${path}: ${e instanceof Error ? e.message : String(e)}`
        );
      });
    }
  }
  if (configuration.projectConfig) {
    for (const path of getObjectPaths(configuration.projectConfig)) {
      await comlinkWorker.setProjectConfigProperty(
        path,
        maybeProxy(getValueAtPath(configuration.projectConfig, path))
      ).catch((e) => {
        throw new Error(
          `Error setting project config property ${path}: ${e instanceof Error ? e.message : String(e)}`
        );
      });
    }
  }
  let isTerminated = false;
  const wrapper = {
    clearEventListeners: comlinkWorker.clearEventListeners.bind(comlinkWorker),
    enableDebug: async (...args) => {
      if (isTerminated) {
        throw new Error("CircuitWebWorker was terminated, can't enableDebug");
      }
      return comlinkWorker.enableDebug.bind(comlinkWorker)(...args);
    },
    version: comlinkWorker.version.bind(comlinkWorker),
    execute: async (...args) => {
      if (isTerminated) {
        throw new Error("CircuitWebWorker was terminated, can't execute");
      }
      return comlinkWorker.execute.bind(comlinkWorker)(...args);
    },
    executeComponent: async (component) => {
      if (isTerminated) {
        throw new Error(
          "CircuitWebWorker was terminated, can't executeComponent"
        );
      }
      if (typeof component === "function") {
        return comlinkWorker.executeComponent.bind(comlinkWorker)(component);
      }
      if (component && typeof component === "object" && component.type) {
        const serializedElement = serializeReactElement(component);
        return comlinkWorker.executeComponent.bind(comlinkWorker)(
          serializedElement
        );
      }
      return comlinkWorker.executeComponent.bind(comlinkWorker)(component);
    },
    executeWithFsMap: async (...args) => {
      if (isTerminated) {
        throw new Error(
          "CircuitWebWorker was terminated, can't executeWithFsMap"
        );
      }
      return comlinkWorker.executeWithFsMap.bind(comlinkWorker)(...args);
    },
    renderUntilSettled: comlinkWorker.renderUntilSettled.bind(comlinkWorker),
    getCircuitJson: comlinkWorker.getCircuitJson.bind(comlinkWorker),
    on: (event, callback) => {
      const proxiedCallback = Comlink.proxy(callback);
      comlinkWorker.on(event, proxiedCallback);
    },
    kill: async () => {
      comlinkWorker[Comlink.releaseProxy]();
      rawWorker.terminate();
      isTerminated = true;
      if (globalThis.TSCIRCUIT_GLOBAL_CIRCUIT_WORKER === wrapper) {
        globalThis.TSCIRCUIT_GLOBAL_CIRCUIT_WORKER = void 0;
      }
    }
  };
  wrapper.__rawWorker = rawWorker;
  globalThis.TSCIRCUIT_GLOBAL_CIRCUIT_WORKER = wrapper;
  return wrapper;
};

// lib/getPossibleEntrypointComponentPaths.ts
var getPossibleEntrypointComponentPaths = (fsMap) => {
  const normalizedFsMap = {};
  for (const [path, content] of Object.entries(fsMap)) {
    normalizedFsMap[normalizeFilePath(path)] = content;
  }
  const possible = /* @__PURE__ */ new Set();
  if ("tscircuit.config.json" in normalizedFsMap) {
    try {
      const config = JSON.parse(normalizedFsMap["tscircuit.config.json"]);
      if (typeof config.mainEntrypoint === "string") {
        possible.add(normalizeFilePath(config.mainEntrypoint));
      }
    } catch {
    }
  }
  if (normalizedFsMap["index.tsx"]) possible.add("index.tsx");
  if (normalizedFsMap["index.ts"]) possible.add("index.ts");
  const circuitFiles = Object.keys(normalizedFsMap).filter(
    (k) => k.endsWith(".circuit.tsx")
  );
  for (const file of circuitFiles) {
    possible.add(file);
  }
  const tsxFiles = Object.keys(normalizedFsMap).filter(
    (k) => k.endsWith(".tsx")
  );
  if (tsxFiles.length === 1) {
    possible.add(tsxFiles[0]);
  }
  return Array.from(possible);
};
export {
  CircuitRunner,
  STATIC_ASSET_EXTENSIONS,
  createCircuitWebWorker,
  getImportsFromCode,
  getPlatformConfig,
  getPossibleEntrypointComponentPaths,
  isStaticAssetPath,
  runTscircuitCode,
  runTscircuitModule
};
