// @tsci/Heinrich-XIAO.Key/index.cjs
'use strict';

var core = require('@tscircuit/core');

function _extends$1() { return _extends$1 = Object.assign ? Object.assign.bind() : function (n) { for (var e = 1; e < arguments.length; e++) { var t = arguments[e]; for (var r in t) ({}).hasOwnProperty.call(t, r) && (n[r] = t[r]); } return n; }, _extends$1.apply(null, arguments); }
const pinLabels$1 = {
  "1": "pin1",
  "2": "pin2"
};
const pinNames$1 = Object.values(pinLabels$1);
const KeyHotSocket = props => {
  return /*#__PURE__*/React.createElement("pushbutton", _extends$1({
    layer: "bottom"
  }, props, {
    footprint: /*#__PURE__*/React.createElement("footprint", null, /*#__PURE__*/React.createElement("hole", {
      pcbX: "3.1749999999999545mm",
      pcbY: "-1.2699999999999818mm",
      diameter: "2.9999939999999996mm"
    }), /*#__PURE__*/React.createElement("hole", {
      pcbX: "-3.1749999999999545mm",
      pcbY: "1.2700000000000955mm",
      diameter: "2.9999939999999996mm"
    }), /*#__PURE__*/React.createElement("smtpad", {
      portHints: ["2"],
      pcbX: "6.724904000000038mm",
      pcbY: "-1.2699999999999818mm",
      width: "2.8999941999999996mm",
      height: "2.4999949999999997mm",
      shape: "rect"
    }), /*#__PURE__*/React.createElement("smtpad", {
      portHints: ["1"],
      pcbX: "-6.724904000000038mm",
      pcbY: "1.2699999999999818mm",
      width: "2.8999941999999996mm",
      height: "2.4999949999999997mm",
      shape: "rect"
    }), /*#__PURE__*/React.createElement("silkscreenpath", {
      route: [{
        x: -5.461000000000013,
        y: -0.4611370000000079
      }, {
        x: -5.461000000000013,
        y: -2.9209999999999354
      }, {
        x: -5.461000000000013,
        y: -2.9209999999999354
      }]
    }), /*#__PURE__*/React.createElement("silkscreenpath", {
      route: [{
        x: 5.4500017999999955,
        y: -2.950006799999869
      }, {
        x: -5.461000000000013,
        y: -2.950006799999869
      }]
    }), /*#__PURE__*/React.createElement("silkscreenpath", {
      route: [{
        x: 0.18298159999994823,
        y: 1.0549889999999778
      }, {
        x: 5.450027199999909,
        y: 1.0549889999999778
      }]
    }), /*#__PURE__*/React.createElement("silkscreenpath", {
      route: [{
        x: 5.450027199999909,
        y: 1.0549889999999778
      }, {
        x: 5.450027199999909,
        y: 0.4611370000000079
      }]
    }), /*#__PURE__*/React.createElement("silkscreenpath", {
      route: [{
        x: -2.944114000000127,
        y: 2.9450030000000424
      }, {
        x: -5.461000000000013,
        y: 2.9450030000000424
      }]
    }), /*#__PURE__*/React.createElement("silkscreenpath", {
      route: [{
        x: -1.707006999999976,
        y: 2.9450030000000424
      }, {
        x: -2.944114000000127,
        y: 2.9450030000000424
      }]
    }), /*#__PURE__*/React.createElement("silkscreenpath", {
      route: [{
        x: -1.707006999999976,
        y: 2.9450030000000424
      }, {
        x: -1.460311886129034,
        y: 2.9288336092640748
      }, {
        x: -1.2178378661753868,
        y: 2.8806021044970294
      }, {
        x: -0.9837338087377248,
        y: 2.801133754218995
      }, {
        x: -0.762005367588813,
        y: 2.6917883071702136
      }, {
        x: -0.5564464426572613,
        y: 2.5544367262343712
      }, {
        x: -0.37057426423848483,
        y: 2.3914291751993915
      }, {
        x: -0.20756921118118044,
        y: 2.2055548061152876
      }, {
        x: -0.07022039279377168,
        y: 1.9999940353145576
      }, {
        x: 0.03912207440475868,
        y: 1.7782641246765252
      }, {
        x: 0.11858727851722506,
        y: 1.5441589992735771
      }, {
        x: 0.1668155246366041,
        y: 1.301684331151364
      }, {
        x: 0.18298159999994823,
        y: 1.0549889999999778
      }]
    })),
    supplierPartNumbers: {
      jlcpcb: ["C41430893"]
    },
    cadModel: {
      objUrl: "https://modelcdn.tscircuit.com/easyeda_models/download?uuid=c886ec2b42464573a88fc1f647577a49&pn=C5184526",
      rotationOffset: {
        x: 0,
        y: 0,
        z: 0
      },
      positionOffset: {
        x: 0,
        y: 0,
        z: 0
      }
    },
    pinLabels: pinLabels$1,
    schPortArrangement: {
      leftSide: {
        direction: "top-to-bottom",
        pins: [1]
      },
      rightSide: {
        direction: "bottom-to-top",
        pins: [2]
      }
    }
  }));
};
core.createUseComponent(KeyHotSocket, pinNames$1);

function _extends() { return _extends = Object.assign ? Object.assign.bind() : function (n) { for (var e = 1; e < arguments.length; e++) { var t = arguments[e]; for (var r in t) ({}).hasOwnProperty.call(t, r) && (n[r] = t[r]); } return n; }, _extends.apply(null, arguments); }
const pinLabels = {
  "1": "pin1",
  "2": "pin2"
};
const pinNames = Object.values(pinLabels);
const KeyShaftForHotSocket = props => {
  return /*#__PURE__*/React.createElement("chip", _extends({}, props, {
    noSchematicRepresentation: true,
    schWidth: 0.1,
    schHeight: 0.1,
    supplierPartNumbers: {
      jlcpcb: ["C400227"]
    },
    footprint: /*#__PURE__*/React.createElement("footprint", null, /*#__PURE__*/React.createElement("hole", {
      pcbX: "0.6349999999999909mm",
      pcbY: "-3.1149987999999666mm",
      diameter: "4.1999916mm"
    }), /*#__PURE__*/React.createElement("silkscreenpath", {
      route: [{
        x: -7.165009800000007,
        y: 4.685011000000031
      }, {
        x: 8.435009799999989,
        y: 4.685011000000031
      }]
    }), /*#__PURE__*/React.createElement("silkscreenpath", {
      route: [{
        x: -7.165009800000007,
        y: -10.915008600000078
      }, {
        x: 8.435009799999989,
        y: -10.915008600000078
      }]
    }), /*#__PURE__*/React.createElement("silkscreenpath", {
      route: [{
        x: -7.165009800000007,
        y: 4.685011000000031
      }, {
        x: -7.165009800000007,
        y: -10.915008600000078
      }]
    }), /*#__PURE__*/React.createElement("silkscreenpath", {
      route: [{
        x: 8.435009799999989,
        y: 4.685011000000031
      }, {
        x: 8.435009799999989,
        y: -10.915008600000078
      }]
    })),
    cadModel: {
      objUrl: "https://modelcdn.tscircuit.com/easyeda_models/download?uuid=c00f29e7afb64c29bc388e168980ded2&pn=C400227",
      rotationOffset: {
        x: 0,
        y: 0,
        z: 90
      },
      positionOffset: {
        x: 0,
        y: 0,
        z: 0
      }
    },
    pinLabels: pinLabels,
    schPinSpacing: 0.75,
    schPortArrangement: {}
  }));
};
core.createUseComponent(KeyShaftForHotSocket, pinNames);

const Key = props => {
  return /*#__PURE__*/React.createElement("group", {
    pcbX: (props.pcbX || 0) - 0.6349999999999909,
    pcbY: (props.pcbY || 0) - (-3.1149988 + -0.52),
    schX: props.schX,
    schY: props.schY
  }, /*#__PURE__*/React.createElement("hole", {
    diameter: 1.7018,
    pcbX: 0.6349999999999909 + 5.08,
    pcbY: -3.1149988 + -0.52
  }), /*#__PURE__*/React.createElement("hole", {
    diameter: 1.7018,
    pcbX: 0.6349999999999909 - 5.08,
    pcbY: -3.1149988 + -0.52
  }), /*#__PURE__*/React.createElement(KeyHotSocket, {
    name: props.name,
    connections: props.connections
  }), /*#__PURE__*/React.createElement(KeyShaftForHotSocket, {
    name: `${props.name}_shaft`,
    pcbX: 0,
    pcbY: -0.52
  }), /*#__PURE__*/React.createElement("footprint", null, /*#__PURE__*/React.createElement("silkscreentext", {
    text: props.name,
    pcbY: 5
  })));
};

exports.Key = Key;
