// connectivity-map/dist/index.js
// src/ConnectivityMap.ts
var ConnectivityMap = class {
  netMap;
  idToNetMap;
  constructor(netMap) {
    this.netMap = netMap;
    this.idToNetMap = {};
    for (const [netId, ids] of Object.entries(netMap)) {
      for (const id of ids) {
        this.idToNetMap[id] = netId;
      }
    }
  }
  addConnections(connections) {
    for (const connection of connections) {
      const existingNets = /* @__PURE__ */ new Set();
      for (const id of connection) {
        const existingNetId = this.idToNetMap[id];
        if (existingNetId) {
          existingNets.add(existingNetId);
        }
      }
      let targetNetId;
      if (existingNets.size === 0) {
        targetNetId = `connectivity_net${Object.keys(this.netMap).length}`;
        this.netMap[targetNetId] = [];
      } else if (existingNets.size === 1) {
        targetNetId = existingNets.values().next().value ?? `connectivity_net${Object.keys(this.netMap).length}`;
      } else {
        targetNetId = existingNets.values().next().value ?? `connectivity_net${Object.keys(this.netMap).length}`;
        for (const netId of existingNets) {
          if (netId !== targetNetId) {
            const targetNet = this.netMap[targetNetId];
            const sourceNet = this.netMap[netId];
            if (targetNet && sourceNet) {
              targetNet.push(...sourceNet);
              this.netMap[netId] = targetNet;
              for (const id of targetNet) {
                this.idToNetMap[id] = targetNetId;
              }
            }
          }
        }
      }
      for (const id of connection) {
        const targetNet = this.netMap[targetNetId];
        if (targetNet && !targetNet.includes(id)) {
          targetNet.push(id);
        }
        this.idToNetMap[id] = targetNetId;
      }
    }
  }
  getIdsConnectedToNet(netId) {
    return this.netMap[netId] || [];
  }
  getNetConnectedToId(id) {
    return this.idToNetMap[id];
  }
  areIdsConnected(id1, id2) {
    if (id1 === id2) return true;
    const netId1 = this.getNetConnectedToId(id1);
    if (!netId1) return false;
    const netId2 = this.getNetConnectedToId(id2);
    if (!netId2) return false;
    return netId1 === netId2 || netId2 === id1 || netId2 === id1;
  }
  areAllIdsConnected(ids) {
    if (ids.length === 0) return true;
    const netId = this.getNetConnectedToId(ids[0]);
    if (!netId) return false;
    for (const id of ids) {
      const nextNetId = this.getNetConnectedToId(id);
      if (nextNetId === void 0) {
        return false;
      }
      if (nextNetId !== netId) {
        return false;
      }
    }
    return true;
  }
};
export {
  ConnectivityMap
};
