// @tscircuit/simple-3d-svg/dist/index.js
// lib/vec3.ts
function add(a, b) {
  return { x: a.x + b.x, y: a.y + b.y, z: a.z + b.z };
}
function sub(a, b) {
  return { x: a.x - b.x, y: a.y - b.y, z: a.z - b.z };
}
function dot(a, b) {
  return a.x * b.x + a.y * b.y + a.z * b.z;
}
function cross(a, b) {
  return {
    x: a.y * b.z - a.z * b.y,
    y: a.z * b.x - a.x * b.z,
    z: a.x * b.y - a.y * b.x
  };
}
function scale(v, k) {
  return { x: v.x * k, y: v.y * k, z: v.z * k };
}
function len(v) {
  return Math.sqrt(dot(v, v));
}
function norm(v) {
  const l = len(v) || 1;
  return scale(v, 1 / l);
}
function rotLocal(p, r = { x: 0, y: 0, z: 0 }) {
  let { x, y, z } = p;
  if (r.x) {
    const c = Math.cos(r.x);
    const s = Math.sin(r.x);
    const y2 = y * c - z * s;
    z = y * s + z * c;
    y = y2;
  }
  if (r.y) {
    const c = Math.cos(r.y);
    const s = Math.sin(r.y);
    const x2 = x * c + z * s;
    z = -x * s + z * c;
    x = x2;
  }
  if (r.z) {
    const c = Math.cos(r.z);
    const s = Math.sin(r.z);
    const x2 = x * c - y * s;
    y = x * s + y * c;
    x = x2;
  }
  return { x, y, z };
}

// lib/color.ts
var NAMED_COLORS = {
  black: [0, 0, 0],
  silver: [192, 192, 192],
  gray: [128, 128, 128],
  grey: [128, 128, 128],
  white: [255, 255, 255],
  maroon: [128, 0, 0],
  red: [255, 0, 0],
  purple: [128, 0, 128],
  fuchsia: [255, 0, 255],
  green: [0, 128, 0],
  lime: [0, 255, 0],
  olive: [128, 128, 0],
  yellow: [255, 255, 0],
  navy: [0, 0, 128],
  blue: [0, 0, 255],
  teal: [0, 128, 128],
  aqua: [0, 255, 255],
  orange: [255, 165, 0]
};
function colorToCss(c) {
  if (typeof c === "string") return c;
  const [r, g, b, a] = c;
  return `rgba(${Math.round(r)},${Math.round(g)},${Math.round(b)},${a})`;
}
function colorToRGBA(c) {
  if (Array.isArray(c)) return c;
  const s = c.trim().toLowerCase();
  if (s.startsWith("#")) {
    const hex = s.slice(1);
    if (hex.length === 3) {
      const r = parseInt(hex.charAt(0) + hex.charAt(0), 16);
      const g = parseInt(hex.charAt(1) + hex.charAt(1), 16);
      const b = parseInt(hex.charAt(2) + hex.charAt(2), 16);
      return [r, g, b, 1];
    }
    if (hex.length === 6) {
      const r = parseInt(hex.slice(0, 2), 16);
      const g = parseInt(hex.slice(2, 4), 16);
      const b = parseInt(hex.slice(4, 6), 16);
      return [r, g, b, 1];
    }
  }
  const rgbm = s.match(/^rgba?\(([^)]+)\)$/);
  if (rgbm) {
    const content = rgbm[1];
    const parts = content.split(/\s*,\s*/).map(Number);
    const [r = 0, g = 0, b = 0, a = 1] = parts;
    return [r, g, b, a];
  }
  const named = NAMED_COLORS[s];
  if (named) return [named[0], named[1], named[2], 1];
  return [0, 0, 0, 1];
}
function lightenColor(c, f) {
  const [r, g, b, a] = colorToRGBA(c);
  return [r + (255 - r) * f, g + (255 - g) * f, b + (255 - b) * f, a];
}
function darkenColor(c, f) {
  const [r, g, b, a] = colorToRGBA(c);
  return [r * (1 - f), g * (1 - f), b * (1 - f), a];
}
function shadeByNormal(base, normal) {
  const n = norm(normal);
  if (n.z >= 0) {
    return colorToCss(lightenColor(base, n.z * 0.4));
  } else {
    return colorToCss(darkenColor(base, -n.z * 0.4));
  }
}

// lib/loaders/stl.ts
var stlCache = /* @__PURE__ */ new Map();
async function loadSTL(url) {
  if (stlCache.has(url)) {
    return stlCache.get(url);
  }
  const response = await fetch(url);
  const buffer = await response.arrayBuffer();
  const mesh = parseSTL(buffer);
  stlCache.set(url, mesh);
  return mesh;
}
function parseSTL(buffer) {
  const view = new DataView(buffer);
  const header = new TextDecoder().decode(buffer.slice(0, 5));
  if (header.toLowerCase() === "solid") {
    return parseASCIISTL(buffer);
  } else {
    return parseBinarySTL(view);
  }
}
function parseASCIISTL(buffer) {
  const text = new TextDecoder().decode(buffer);
  const lines = text.split("\n").map((line) => line.trim());
  const triangles = [];
  let i = 0;
  while (i < lines.length) {
    const line = lines[i];
    if (line && line.startsWith("facet normal")) {
      const normalMatch = line.match(
        /facet normal\s+([-+]?\d*\.?\d+(?:[eE][-+]?\d+)?)\s+([-+]?\d*\.?\d+(?:[eE][-+]?\d+)?)\s+([-+]?\d*\.?\d+(?:[eE][-+]?\d+)?)/
      );
      const normal = normalMatch ? {
        x: parseFloat(normalMatch[1]),
        y: parseFloat(normalMatch[2]),
        z: parseFloat(normalMatch[3])
      } : { x: 0, y: 0, z: 1 };
      i++;
      const vertices = [];
      while (i < lines.length && lines[i] && !lines[i].startsWith("endfacet")) {
        const vertexLine = lines[i];
        if (vertexLine.startsWith("vertex")) {
          const vertexMatch = vertexLine.match(
            /vertex\s+([-+]?\d*\.?\d+(?:[eE][-+]?\d+)?)\s+([-+]?\d*\.?\d+(?:[eE][-+]?\d+)?)\s+([-+]?\d*\.?\d+(?:[eE][-+]?\d+)?)/
          );
          if (vertexMatch) {
            vertices.push({
              x: parseFloat(vertexMatch[1]),
              y: parseFloat(vertexMatch[2]),
              z: parseFloat(vertexMatch[3])
            });
          }
        }
        i++;
      }
      if (vertices.length === 3) {
        triangles.push({
          vertices: [vertices[0], vertices[1], vertices[2]],
          normal
        });
      }
    }
    i++;
  }
  const rotatedTriangles = triangles.map((triangle) => ({
    ...triangle,
    vertices: triangle.vertices.map((v) => ({
      x: v.x,
      y: -v.z,
      z: v.y
    })),
    normal: {
      x: triangle.normal.x,
      y: -triangle.normal.z,
      z: triangle.normal.y
    }
  }));
  return {
    triangles: rotatedTriangles,
    boundingBox: calculateBoundingBox(rotatedTriangles)
  };
}
function parseBinarySTL(view) {
  let offset = 80;
  const numTriangles = view.getUint32(offset, true);
  offset += 4;
  const triangles = [];
  for (let i = 0; i < numTriangles; i++) {
    const normal = {
      x: view.getFloat32(offset, true),
      y: view.getFloat32(offset + 4, true),
      z: view.getFloat32(offset + 8, true)
    };
    offset += 12;
    const vertices = [
      {
        x: view.getFloat32(offset, true),
        y: view.getFloat32(offset + 4, true),
        z: view.getFloat32(offset + 8, true)
      },
      {
        x: view.getFloat32(offset + 12, true),
        y: view.getFloat32(offset + 16, true),
        z: view.getFloat32(offset + 20, true)
      },
      {
        x: view.getFloat32(offset + 24, true),
        y: view.getFloat32(offset + 28, true),
        z: view.getFloat32(offset + 32, true)
      }
    ];
    offset += 36;
    offset += 2;
    triangles.push({ vertices, normal });
  }
  const rotatedTriangles = triangles.map((triangle) => ({
    ...triangle,
    vertices: triangle.vertices.map((v) => ({
      x: v.x,
      y: -v.z,
      z: v.y
    })),
    normal: {
      x: triangle.normal.x,
      y: -triangle.normal.z,
      z: triangle.normal.y
    }
  }));
  return {
    triangles: rotatedTriangles,
    boundingBox: calculateBoundingBox(rotatedTriangles)
  };
}
function calculateBoundingBox(triangles) {
  if (triangles.length === 0) {
    return {
      min: { x: 0, y: 0, z: 0 },
      max: { x: 0, y: 0, z: 0 }
    };
  }
  let minX = Infinity, minY = Infinity, minZ = Infinity;
  let maxX = -Infinity, maxY = -Infinity, maxZ = -Infinity;
  for (const triangle of triangles) {
    for (const vertex of triangle.vertices) {
      minX = Math.min(minX, vertex.x);
      minY = Math.min(minY, vertex.y);
      minZ = Math.min(minZ, vertex.z);
      maxX = Math.max(maxX, vertex.x);
      maxY = Math.max(maxY, vertex.y);
      maxZ = Math.max(maxZ, vertex.z);
    }
  }
  return {
    min: { x: minX, y: minY, z: minZ },
    max: { x: maxX, y: maxY, z: maxZ }
  };
}

// lib/loaders/obj.ts
var objCache = /* @__PURE__ */ new Map();
async function loadOBJ(url) {
  if (objCache.has(url)) {
    return objCache.get(url);
  }
  const response = await fetch(url);
  const text = await response.text();
  const mesh = parseOBJ(text);
  objCache.set(url, mesh);
  return mesh;
}
function parseOBJ(text) {
  const lines = text.split(/\r?\n/);
  const vertices = [];
  const vertexColors = [];
  const normals = [];
  const triangles = [];
  const materialColors = {};
  let activeMaterial;
  for (const line of lines) {
    const trimmed = line.trim();
    if (trimmed.startsWith("v ")) {
      const parts = trimmed.split(/\s+/);
      const x = parts[1];
      const y = parts[2];
      const z = parts[3];
      vertices.push({ x: parseFloat(x), y: parseFloat(y), z: -parseFloat(z) });
      if (parts.length >= 7) {
        const [rStr, gStr, bStr] = parts.slice(4, 7);
        let r = Number(rStr);
        let g = Number(gStr);
        let b = Number(bStr);
        if (r <= 1 && g <= 1 && b <= 1) {
          r *= 255;
          g *= 255;
          b *= 255;
        }
        vertexColors.push([r, g, b, 1]);
      } else {
        vertexColors.push(void 0);
      }
    } else if (trimmed.startsWith("vn ")) {
      const parts = trimmed.split(/\s+/);
      const x = parts[1];
      const y = parts[2];
      const z = parts[3];
      normals.push({ x: parseFloat(x), y: parseFloat(y), z: -parseFloat(z) });
    } else if (trimmed.startsWith("newmtl ")) {
      activeMaterial = trimmed.split(/\s+/)[1];
    } else if (trimmed.startsWith("Kd ") && activeMaterial) {
      const parts = trimmed.split(/\s+/);
      const rStr = parts[1];
      const gStr = parts[2];
      const bStr = parts[3];
      let r = parseFloat(rStr);
      let g = parseFloat(gStr);
      let b = parseFloat(bStr);
      if (r <= 1 && g <= 1 && b <= 1) {
        r *= 255;
        g *= 255;
        b *= 255;
      }
      materialColors[activeMaterial] = [r, g, b, 1];
    } else if (trimmed.startsWith("usemtl ")) {
      activeMaterial = trimmed.split(/\s+/)[1];
    } else if (trimmed.startsWith("f ")) {
      const parts = trimmed.slice(2).trim().split(/\s+/);
      const idxs = parts.map((p) => {
        const [vi, , ni] = p.split("/");
        return {
          v: parseInt(vi) - 1,
          n: ni ? parseInt(ni) - 1 : void 0
        };
      });
      for (let i = 1; i < idxs.length - 1; i++) {
        const a = idxs[0];
        const b = idxs[i];
        const c = idxs[i + 1];
        const v0 = vertices[a.v];
        const v1 = vertices[b.v];
        const v2 = vertices[c.v];
        let normal;
        if (a.n !== void 0 && normals[a.n]) {
          normal = normals[a.n];
        } else if (b.n !== void 0 && normals[b.n]) {
          normal = normals[b.n];
        } else if (c.n !== void 0 && normals[c.n]) {
          normal = normals[c.n];
        } else {
          const edge1 = {
            x: v1.x - v0.x,
            y: v1.y - v0.y,
            z: v1.z - v0.z
          };
          const edge2 = {
            x: v2.x - v0.x,
            y: v2.y - v0.y,
            z: v2.z - v0.z
          };
          normal = {
            x: edge1.y * edge2.z - edge1.z * edge2.y,
            y: edge1.z * edge2.x - edge1.x * edge2.z,
            z: -(edge1.x * edge2.y - edge1.y * edge2.x)
          };
        }
        let color;
        if (activeMaterial && materialColors[activeMaterial]) {
          color = materialColors[activeMaterial];
        } else {
          color = vertexColors[a.v] ?? vertexColors[b.v] ?? vertexColors[c.v];
        }
        triangles.push({ vertices: [v0, v1, v2], normal, color });
      }
    }
  }
  return {
    triangles,
    boundingBox: calculateBoundingBox2(triangles)
  };
}
function calculateBoundingBox2(triangles) {
  if (triangles.length === 0) {
    return { min: { x: 0, y: 0, z: 0 }, max: { x: 0, y: 0, z: 0 } };
  }
  let minX = Infinity;
  let minY = Infinity;
  let minZ = Infinity;
  let maxX = -Infinity;
  let maxY = -Infinity;
  let maxZ = -Infinity;
  for (const tri of triangles) {
    for (const v of tri.vertices) {
      if (v.x < minX) minX = v.x;
      if (v.y < minY) minY = v.y;
      if (v.z < minZ) minZ = v.z;
      if (v.x > maxX) maxX = v.x;
      if (v.y > maxY) maxY = v.y;
      if (v.z > maxZ) maxZ = v.z;
    }
  }
  return {
    min: { x: minX, y: minY, z: minZ },
    max: { x: maxX, y: maxY, z: maxZ }
  };
}

// lib/loaders/threemf.ts
import { unzipSync } from "fflate";
import { XMLParser } from "fast-xml-parser";
var threeMfCache = /* @__PURE__ */ new Map();
async function load3MF(url) {
  if (threeMfCache.has(url)) {
    return threeMfCache.get(url);
  }
  const response = await fetch(url);
  const buffer = await response.arrayBuffer();
  const files = unzipSync(new Uint8Array(buffer));
  const modelFile = files["3D/3dmodel.model"];
  if (!modelFile) throw new Error("3MF model file not found");
  const modelXml = new TextDecoder().decode(modelFile);
  const mesh = parse3MF(modelXml);
  threeMfCache.set(url, mesh);
  return mesh;
}
function parse3MF(xml) {
  const parser = new XMLParser({ ignoreAttributes: false });
  const json = parser.parse(xml);
  const vertexData = json?.model?.resources?.object?.mesh?.vertices?.vertex ?? [];
  const verticesArr = Array.isArray(vertexData) ? vertexData : [vertexData];
  const vertices = verticesArr.map((v) => ({
    x: parseFloat(v["@_x"]),
    y: parseFloat(v["@_y"]),
    z: parseFloat(v["@_z"])
  }));
  const baseMaterials = json?.model?.resources?.basematerials;
  const materialColors = {};
  if (baseMaterials) {
    const mats = Array.isArray(baseMaterials) ? baseMaterials : [baseMaterials];
    for (const mat of mats) {
      const id = mat["@_id"];
      const bases = mat.base ?? [];
      const basesArr = Array.isArray(bases) ? bases : [bases];
      materialColors[id] = basesArr.map(
        (b) => parseDisplayColor(b["@_displaycolor"])
      );
    }
  }
  const triangleData = json?.model?.resources?.object?.mesh?.triangles?.triangle ?? [];
  const trianglesArr = Array.isArray(triangleData) ? triangleData : [triangleData];
  const triangles = trianglesArr.map((t) => {
    const v1 = vertices[parseInt(t["@_v1"])];
    const v2 = vertices[parseInt(t["@_v2"])];
    const v3 = vertices[parseInt(t["@_v3"])];
    const edge1 = { x: v2.x - v1.x, y: v2.y - v1.y, z: v2.z - v1.z };
    const edge2 = { x: v3.x - v1.x, y: v3.y - v1.y, z: v3.z - v1.z };
    const normal = {
      x: edge1.y * edge2.z - edge1.z * edge2.y,
      y: edge1.z * edge2.x - edge1.x * edge2.z,
      z: -(edge1.x * edge2.y - edge1.y * edge2.x)
    };
    let color;
    const pid = t["@_pid"];
    if (pid && materialColors[pid]) {
      const idx = parseInt(t["@_p1"] ?? t["@_p2"] ?? t["@_p3"] ?? "0");
      color = materialColors[pid][idx];
    }
    return color ? { vertices: [v1, v2, v3], normal, color } : { vertices: [v1, v2, v3], normal };
  });
  const rotatedTriangles = triangles.map((tri) => ({
    ...tri,
    vertices: tri.vertices.map((v) => ({
      x: v.x,
      y: -v.z,
      z: v.y
    })),
    normal: {
      x: tri.normal.x,
      y: -tri.normal.z,
      z: tri.normal.y
    }
  }));
  return {
    triangles: rotatedTriangles,
    boundingBox: calculateBoundingBox3(rotatedTriangles)
  };
}
function parseDisplayColor(str) {
  if (str?.startsWith("#")) {
    const hex = str.slice(1);
    if (hex.length === 8) {
      const r = parseInt(hex.slice(0, 2), 16);
      const g = parseInt(hex.slice(2, 4), 16);
      const b = parseInt(hex.slice(4, 6), 16);
      const a = parseInt(hex.slice(6, 8), 16) / 255;
      return [r, g, b, a];
    }
    if (hex.length === 6) {
      const r = parseInt(hex.slice(0, 2), 16);
      const g = parseInt(hex.slice(2, 4), 16);
      const b = parseInt(hex.slice(4, 6), 16);
      return [r, g, b, 1];
    }
  }
  return [0, 0, 0, 1];
}
function calculateBoundingBox3(triangles) {
  if (triangles.length === 0) {
    return { min: { x: 0, y: 0, z: 0 }, max: { x: 0, y: 0, z: 0 } };
  }
  let minX = Infinity;
  let minY = Infinity;
  let minZ = Infinity;
  let maxX = -Infinity;
  let maxY = -Infinity;
  let maxZ = -Infinity;
  for (const tri of triangles) {
    for (const v of tri.vertices) {
      if (v.x < minX) minX = v.x;
      if (v.y < minY) minY = v.y;
      if (v.z < minZ) minZ = v.z;
      if (v.x > maxX) maxX = v.x;
      if (v.y > maxY) maxY = v.y;
      if (v.z > maxZ) maxZ = v.z;
    }
  }
  return {
    min: { x: minX, y: minY, z: minZ },
    max: { x: maxX, y: maxY, z: maxZ }
  };
}

// lib/mesh.ts
function scaleAndPositionMesh(mesh, box, scaleToBox, modelType) {
  const { boundingBox } = mesh;
  const meshCenter = scale(add(boundingBox.min, boundingBox.max), 0.5);
  const centerModel = box.centerModel !== false;
  const rotatedVerts = [];
  for (const tri of mesh.triangles) {
    for (const v of tri.vertices) {
      let p = sub(v, meshCenter);
      if (modelType === "stl" && box.stlRotation)
        p = rotLocal(p, box.stlRotation);
      if (modelType === "obj" && box.objRotation)
        p = rotLocal(p, box.objRotation);
      if (modelType === "3mf" && box.threeMfRotation)
        p = rotLocal(p, box.threeMfRotation);
      if (!centerModel) p = add(p, meshCenter);
      rotatedVerts.push(p);
    }
  }
  let uniformScale = 1;
  let rotatedCenter = { x: 0, y: 0, z: 0 };
  if (scaleToBox) {
    let min = { x: Infinity, y: Infinity, z: Infinity };
    let max = { x: -Infinity, y: -Infinity, z: -Infinity };
    for (const v of rotatedVerts) {
      if (v.x < min.x) min.x = v.x;
      if (v.y < min.y) min.y = v.y;
      if (v.z < min.z) min.z = v.z;
      if (v.x > max.x) max.x = v.x;
      if (v.y > max.y) max.y = v.y;
      if (v.z > max.z) max.z = v.z;
    }
    const rotatedSize = sub(max, min);
    const boxSize = box.size;
    const scaleX = boxSize.x / rotatedSize.x;
    const scaleY = boxSize.y / rotatedSize.y;
    const scaleZ = boxSize.z / rotatedSize.z;
    uniformScale = Math.min(scaleX, scaleY, scaleZ);
    rotatedCenter = scale(add(min, max), 0.5);
  }
  const transformedVertices = [];
  for (const p of rotatedVerts) {
    let t = p;
    if (scaleToBox) {
      t = sub(t, rotatedCenter);
      t = scale(t, uniformScale);
      if (!centerModel) t = add(t, rotatedCenter);
    }
    if (box.stlPosition) t = add(t, box.stlPosition);
    if (box.objPosition) t = add(t, box.objPosition);
    if (box.threeMfPosition) t = add(t, box.threeMfPosition);
    if (box.rotation) t = rotLocal(t, box.rotation);
    t = add(t, box.center);
    transformedVertices.push(t);
  }
  return transformedVertices;
}

// lib/geometry.ts
var FACES = [
  [0, 1, 2, 3],
  [4, 7, 6, 5],
  [0, 1, 5, 4],
  [3, 2, 6, 7],
  [1, 2, 6, 5],
  [0, 3, 7, 4]
];
var EDGES = [
  [0, 1],
  [1, 2],
  [2, 3],
  [3, 0],
  [4, 5],
  [5, 6],
  [6, 7],
  [7, 4],
  [0, 4],
  [1, 5],
  [2, 6],
  [3, 7]
];
var TOP = [3, 2, 6, 7];
function verts(b) {
  const {
    size: { x: sx, y: sy, z: sz },
    center,
    rotation
  } = b;
  const offs = [
    { x: -sx / 2, y: -sy / 2, z: -sz / 2 },
    { x: sx / 2, y: -sy / 2, z: -sz / 2 },
    { x: sx / 2, y: sy / 2, z: -sz / 2 },
    { x: -sx / 2, y: sy / 2, z: -sz / 2 },
    { x: -sx / 2, y: -sy / 2, z: sz / 2 },
    { x: sx / 2, y: -sy / 2, z: sz / 2 },
    { x: sx / 2, y: sy / 2, z: sz / 2 },
    { x: -sx / 2, y: sy / 2, z: sz / 2 }
  ];
  return offs.map((o) => add(center, rotLocal(o, rotation)));
}

// lib/affine.ts
function inv3(m) {
  const a = m[0][0], d = m[0][1], g = m[0][2];
  const b = m[1][0], e = m[1][1], h = m[1][2];
  const c = m[2][0], f = m[2][1], i = m[2][2];
  const A = e * i - f * h;
  const B = -(d * i - f * g);
  const C = d * h - e * g;
  const D = -(b * i - c * h);
  const E = a * i - c * g;
  const F = -(a * h - b * g);
  const G = b * f - c * e;
  const H = -(a * f - c * d);
  const I = a * e - b * d;
  const det = a * A + d * D + g * G;
  const invDet = det ? 1 / det : 0;
  return [
    [A * invDet, B * invDet, C * invDet],
    [D * invDet, E * invDet, F * invDet],
    [G * invDet, H * invDet, I * invDet]
  ];
}
function mul3(a, b) {
  const r = [
    [0, 0, 0],
    [0, 0, 0],
    [0, 0, 0]
  ];
  for (let i = 0; i < 3; i++) {
    for (let j = 0; j < 3; j++) {
      r[i][j] = a[i][0] * b[0][j] + a[i][1] * b[1][j] + a[i][2] * b[2][j];
    }
  }
  return r;
}
function affineMatrix(src, dst) {
  const S = [
    [src[0].x, src[1].x, src[2].x],
    [src[0].y, src[1].y, src[2].y],
    [1, 1, 1]
  ];
  const D = [
    [dst[0].x, dst[1].x, dst[2].x],
    [dst[0].y, dst[1].y, dst[2].y],
    [1, 1, 1]
  ];
  const M = mul3(D, inv3(S));
  return `matrix(${M[0][0]} ${M[1][0]} ${M[0][1]} ${M[1][1]} ${M[0][2]} ${M[1][2]})`;
}

// lib/render-elements.ts
function fmtPrecise(n) {
  return (Math.round(n * 100) / 100).toString();
}
var W_DEF = 400;
var H_DEF = 400;
var FOCAL = 2;
function axes(cam) {
  const f = norm(sub(cam.lookAt, cam.position));
  const wUp = { x: 0, y: 1, z: 0 };
  let r = norm(cross(f, wUp));
  if (!len(r)) r = { x: 1, y: 0, z: 0 };
  const u = cross(r, f);
  return { r, u, f };
}
function toCam(p, cam) {
  const { r, u, f } = axes(cam);
  const d = sub(p, cam.position);
  return { x: dot(d, r), y: dot(d, u), z: dot(d, f) };
}
function proj(p, w, h, focal) {
  if (p.z <= 0) return null;
  const s = focal / p.z;
  return { x: p.x * s * w / 2, y: -p.y * s * h / 2, z: p.z };
}
async function buildRenderElements(scene, opt = {}) {
  const W = opt.width ?? W_DEF;
  const H = opt.height ?? H_DEF;
  const focal = scene.camera.focalLength ?? FOCAL;
  const faces = [];
  const images = [];
  const faceToImg = /* @__PURE__ */ new Map();
  const labels = [];
  const edges = [];
  let clipSeq = 0;
  const texId = /* @__PURE__ */ new Map();
  const stlMeshes = /* @__PURE__ */ new Map();
  const objMeshes = /* @__PURE__ */ new Map();
  const threeMfMeshes = /* @__PURE__ */ new Map();
  for (const box of scene.boxes) {
    if (box.stlUrl && !stlMeshes.has(box.stlUrl)) {
      try {
        const mesh = await loadSTL(box.stlUrl);
        stlMeshes.set(box.stlUrl, mesh);
      } catch (error) {
        console.warn(`Failed to load STL from ${box.stlUrl}:`, error);
      }
    }
    if (box.objUrl && !objMeshes.has(box.objUrl)) {
      try {
        const mesh = await loadOBJ(box.objUrl);
        objMeshes.set(box.objUrl, mesh);
      } catch (error) {
        console.warn(`Failed to load OBJ from ${box.objUrl}:`, error);
      }
    }
    if (box.threeMfUrl && !threeMfMeshes.has(box.threeMfUrl)) {
      try {
        const mesh = await load3MF(box.threeMfUrl);
        threeMfMeshes.set(box.threeMfUrl, mesh);
      } catch (error) {
        console.warn(`Failed to load 3MF from ${box.threeMfUrl}:`, error);
      }
    }
  }
  for (const box of scene.boxes) {
    const bw = verts(box);
    const bc = bw.map((v) => toCam(v, scene.camera));
    const bp = bc.map((v) => proj(v, W, H, focal));
    if (box.drawBoundingBox) {
      for (const [a, b] of EDGES) {
        const pa = bp[a];
        const pb = bp[b];
        if (pa && pb) {
          const depth = Math.max(bc[a].z, bc[b].z);
          edges.push({ pts: [pa, pb], depth, color: "rgba(0,0,0,0.5)" });
        }
      }
    }
    if (box.stlUrl && stlMeshes.has(box.stlUrl)) {
      const mesh = stlMeshes.get(box.stlUrl);
      const transformedVertices = scaleAndPositionMesh(
        mesh,
        box,
        box.scaleStlToBox ?? false,
        "stl"
      );
      for (let i = 0; i < mesh.triangles.length; i++) {
        const triangle = mesh.triangles[i];
        const vertexStart = i * 3;
        const v0w = transformedVertices[vertexStart];
        const v1w = transformedVertices[vertexStart + 1];
        const v2w = transformedVertices[vertexStart + 2];
        const v0c = toCam(v0w, scene.camera);
        const v1c = toCam(v1w, scene.camera);
        const v2c = toCam(v2w, scene.camera);
        const v0p = proj(v0c, W, H, focal);
        const v1p = proj(v1c, W, H, focal);
        const v2p = proj(v2c, W, H, focal);
        if (v0p && v1p && v2p) {
          const edge1 = sub(v1c, v0c);
          const edge2 = sub(v2c, v0c);
          const normal = cross(edge1, edge2);
          const baseColor = box.color ?? "gray";
          faces.push({
            pts: [v0p, v1p, v2p],
            cam: [v0c, v1c, v2c],
            fill: shadeByNormal(baseColor, normal),
            stroke: false
          });
        }
      }
    } else if (box.objUrl && objMeshes.has(box.objUrl)) {
      const mesh = objMeshes.get(box.objUrl);
      const transformedVertices = scaleAndPositionMesh(
        mesh,
        box,
        box.scaleObjToBox ?? false,
        "obj"
      );
      for (let i = 0; i < mesh.triangles.length; i++) {
        const vertexStart = i * 3;
        const triangle = mesh.triangles[i];
        const v0w = transformedVertices[vertexStart];
        const v1w = transformedVertices[vertexStart + 1];
        const v2w = transformedVertices[vertexStart + 2];
        const v0c = toCam(v0w, scene.camera);
        const v1c = toCam(v1w, scene.camera);
        const v2c = toCam(v2w, scene.camera);
        const v0p = proj(v0c, W, H, focal);
        const v1p = proj(v1c, W, H, focal);
        const v2p = proj(v2c, W, H, focal);
        if (v0p && v1p && v2p) {
          const edge1 = sub(v1c, v0c);
          const edge2 = sub(v2c, v0c);
          const faceNormal = cross(edge1, edge2);
          faces.push({
            pts: [v0p, v1p, v2p],
            cam: [v0c, v1c, v2c],
            fill: shadeByNormal(
              box.color ?? triangle.color ?? "gray",
              faceNormal
            ),
            stroke: false
          });
        }
      }
    } else if (box.threeMfUrl && threeMfMeshes.has(box.threeMfUrl)) {
      const mesh = threeMfMeshes.get(box.threeMfUrl);
      const transformedVertices = scaleAndPositionMesh(
        mesh,
        box,
        box.scaleThreeMfToBox ?? false,
        "3mf"
      );
      for (let i = 0; i < mesh.triangles.length; i++) {
        const vertexStart = i * 3;
        const triangle = mesh.triangles[i];
        const v0w = transformedVertices[vertexStart];
        const v1w = transformedVertices[vertexStart + 1];
        const v2w = transformedVertices[vertexStart + 2];
        const v0c = toCam(v0w, scene.camera);
        const v1c = toCam(v1w, scene.camera);
        const v2c = toCam(v2w, scene.camera);
        const v0p = proj(v0c, W, H, focal);
        const v1p = proj(v1c, W, H, focal);
        const v2p = proj(v2c, W, H, focal);
        if (v0p && v1p && v2p) {
          const edge1 = sub(v1c, v0c);
          const edge2 = sub(v2c, v0c);
          const faceNormal = cross(edge1, edge2);
          faces.push({
            pts: [v0p, v1p, v2p],
            cam: [v0c, v1c, v2c],
            fill: shadeByNormal(
              box.color ?? triangle.color ?? "gray",
              faceNormal
            ),
            stroke: false
          });
        }
      }
    } else {
      const vw = verts(box);
      const vc = vw.map((v) => toCam(v, scene.camera));
      const vp = vc.map((v) => proj(v, W, H, focal));
      for (const idx of FACES) {
        const p4 = [];
        let behind = false;
        for (const i of idx) {
          const p = vp[i];
          if (!p) {
            behind = true;
            break;
          }
          p4.push(p);
        }
        if (behind) continue;
        const cam4 = idx.map((i) => vc[i]);
        faces.push({
          pts: p4,
          cam: cam4,
          fill: colorToCss(box.color ?? "gray"),
          stroke: true
        });
      }
      if (box.faceImages?.top) {
        const pts = TOP.map((i) => vw[i]);
        if (pts.every(Boolean)) {
          const dst = pts;
          const cz = Math.max(...TOP.map((i) => vc[i].z));
          const href = box.faceImages.top;
          if (!texId.has(href)) {
            texId.set(href, `tex${texId.size}`);
          }
          const sym = texId.get(href);
          const subdivisions = box.projectionSubdivision ?? 2;
          const quadsPerSide = subdivisions;
          for (let row = 0; row < quadsPerSide; row++) {
            for (let col = 0; col < quadsPerSide; col++) {
              const u0 = col / quadsPerSide;
              const u1 = (col + 1) / quadsPerSide;
              const v0 = row / quadsPerSide;
              const v1 = (row + 1) / quadsPerSide;
              const lerp = (a, b, t) => ({
                x: a.x * (1 - t) + b.x * t,
                y: a.y * (1 - t) + b.y * t,
                z: a.z * (1 - t) + b.z * t
              });
              const c00 = toCam(
                lerp(lerp(dst[0], dst[1], u0), lerp(dst[3], dst[2], u0), v0),
                scene.camera
              );
              const c10 = toCam(
                lerp(lerp(dst[0], dst[1], u1), lerp(dst[3], dst[2], u1), v0),
                scene.camera
              );
              const c01 = toCam(
                lerp(lerp(dst[0], dst[1], u0), lerp(dst[3], dst[2], u0), v1),
                scene.camera
              );
              const c11 = toCam(
                lerp(lerp(dst[0], dst[1], u1), lerp(dst[3], dst[2], u1), v1),
                scene.camera
              );
              const p00 = proj(c00, W, H, focal);
              const p10 = proj(c10, W, H, focal);
              const p01 = proj(c01, W, H, focal);
              const p11 = proj(c11, W, H, focal);
              const tri0Mat = affineMatrix(
                [
                  { x: u0, y: v0 },
                  { x: u1, y: v0 },
                  { x: u1, y: v1 }
                ],
                [p00, p10, p11]
              );
              const id0 = `clip${clipSeq++}`;
              images.push({
                matrix: tri0Mat,
                depth: cz,
                href,
                clip: id0,
                points: `${fmtPrecise(u0)},${fmtPrecise(v0)} ${fmtPrecise(u1)},${fmtPrecise(v0)} ${fmtPrecise(u1)},${fmtPrecise(v1)}`,
                sym
              });
              const triFace0 = {
                pts: [p00, p10, p11],
                cam: [c00, c10, c11],
                fill: "none",
                stroke: false
              };
              faces.push(triFace0);
              faceToImg.set(triFace0, images[images.length - 1]);
              const tri1Mat = affineMatrix(
                [
                  { x: u0, y: v0 },
                  { x: u1, y: v1 },
                  { x: u0, y: v1 }
                ],
                [p00, p11, p01]
              );
              const id1 = `clip${clipSeq++}`;
              images.push({
                matrix: tri1Mat,
                depth: cz,
                href,
                clip: id1,
                points: `${fmtPrecise(u0)},${fmtPrecise(v0)} ${fmtPrecise(u1)},${fmtPrecise(v1)} ${fmtPrecise(u0)},${fmtPrecise(v1)}`,
                sym
              });
              const triFace1 = {
                pts: [p00, p11, p01],
                cam: [c00, c11, c01],
                fill: "none",
                stroke: false
              };
              faces.push(triFace1);
              faceToImg.set(triFace1, images[images.length - 1]);
            }
          }
        }
      }
      if (box.topLabel) {
        const pts = TOP.map((i) => vp[i]);
        if (pts.every(Boolean)) {
          const p0 = pts[0];
          const p1 = pts[1];
          const p3 = pts[3];
          const u = sub(p1, p0);
          const v = sub(p3, p0);
          const lu = len(u);
          const lv = len(v);
          if (lu && lv) {
            const uN = scale(u, 1 / lu);
            const vN = scale(v, 1 / lv);
            const cx = pts.reduce((s, p) => s + p.x, 0) / 4;
            const cy = pts.reduce((s, p) => s + p.y, 0) / 4;
            const cz = Math.max(...TOP.map((i) => vc[i].z));
            const m = `matrix(${uN.x} ${uN.y} ${vN.x} ${vN.y} ${cx} ${cy})`;
            const fillCol = box.topLabelColor ?? [0, 0, 0, 1];
            labels.push({
              matrix: m,
              depth: cz,
              text: box.topLabel,
              fill: colorToCss(fillCol)
            });
          }
        }
      }
    }
  }
  function sortFacesBSP(polys, W2, H2, focal2) {
    const EPS = 1e-6;
    function build(list) {
      if (!list.length) return null;
      const face = list[0];
      const p0 = face.cam[0];
      const p1 = face.cam[1];
      const p2 = face.cam[2];
      const normal = cross(sub(p1, p0), sub(p2, p0));
      const front = [];
      const back = [];
      for (let k = 1; k < list.length; k++) {
        const f = list[k];
        let pos = 0, neg = 0;
        const d = [];
        for (const v of f.cam) {
          const dist = dot(normal, sub(v, p0));
          d.push(dist);
          if (dist > EPS) pos++;
          else if (dist < -EPS) neg++;
        }
        if (!pos && !neg) {
          front.push(f);
        } else if (!pos) back.push(f);
        else if (!neg) front.push(f);
        else {
          const fFrontCam = [];
          const fBackCam = [];
          const fFront2D = [];
          const fBack2D = [];
          for (let i = 0; i < f.cam.length; i++) {
            const j = (i + 1) % f.cam.length;
            const aCam = f.cam[i];
            const bCam = f.cam[j];
            const a2D = f.pts[i];
            const b2D = f.pts[j];
            const da = d[i];
            const db = d[j];
            const push = (arrCam, arr2D, cCam, c2D) => {
              arrCam.push(cCam);
              arr2D.push(c2D);
            };
            if (da >= -EPS) push(fFrontCam, fFront2D, aCam, a2D);
            if (da <= EPS) push(fBackCam, fBack2D, aCam, a2D);
            if (da > 0 && db < 0 || da < 0 && db > 0) {
              const t = da / (da - db);
              const interCam = {
                x: aCam.x + (bCam.x - aCam.x) * t,
                y: aCam.y + (bCam.y - aCam.y) * t,
                z: aCam.z + (bCam.z - aCam.z) * t
              };
              const inter2D = proj(interCam, W2, H2, focal2);
              push(fFrontCam, fFront2D, interCam, inter2D);
              push(fBackCam, fBack2D, interCam, inter2D);
            }
          }
          const mk = (cam, pts) => {
            if (cam.length < 3) return null;
            const nf = { cam, pts, fill: f.fill, stroke: false };
            const img = faceToImg.get(f);
            if (img) faceToImg.set(nf, img);
            return nf;
          };
          const f1 = mk(fFrontCam, fFront2D);
          const f2 = mk(fBackCam, fBack2D);
          if (f1) front.push(f1);
          if (f2) back.push(f2);
        }
      }
      return {
        face,
        normal,
        point: p0,
        front: build(front),
        back: build(back)
      };
    }
    function traverse(node, out) {
      if (!node) return;
      const cameraSide = dot(node.normal, scale(node.point, -1));
      if (cameraSide >= 0) {
        traverse(node.back, out);
        out.push(node.face);
        traverse(node.front, out);
      } else {
        traverse(node.front, out);
        out.push(node.face);
        traverse(node.back, out);
      }
    }
    const root = build(polys);
    const ordered = [];
    traverse(root, ordered);
    return ordered;
  }
  const orderedFaces = sortFacesBSP(faces, W, H, focal);
  const elements = [];
  for (const f of orderedFaces) {
    const img = faceToImg.get(f);
    if (img) {
      elements.push({ type: "image", data: img });
    } else {
      elements.push({ type: "face", data: f });
    }
  }
  elements.push(
    ...labels.map((l) => ({ type: "label", data: l })),
    ...edges.sort((a, b) => a.depth - b.depth).map((e) => ({ type: "edge", data: e }))
  );
  return {
    width: W,
    height: H,
    backgroundColor: opt.backgroundColor,
    elements,
    images,
    texId
  };
}

// lib/render-svg.ts
function fmt(n) {
  return Math.round(n) + "";
}
async function renderScene(scene, opt = {}) {
  const {
    width: W,
    height: H,
    backgroundColor,
    elements,
    images,
    texId
  } = await buildRenderElements(scene, {
    width: opt.width,
    height: opt.height,
    backgroundColor: opt.backgroundColor
  });
  const out = [];
  out.push(
    `<svg xmlns="http://www.w3.org/2000/svg" width="${W}" height="${H}" viewBox="${-W / 2} ${-H / 2} ${W} ${H}">`
  );
  if (backgroundColor) {
    out.push(
      `  <rect x="${-W / 2}" y="${-H / 2}" width="${W}" height="${H}" fill="${colorToCss(backgroundColor)}" />
`
    );
  }
  if (images.length) {
    out.push("  <defs>\n");
    for (const [href, id] of texId) {
      out.push(
        `    <image id="${id}" href="${href}" width="1" height="1" preserveAspectRatio="none" style="image-rendering:pixelated"/>
`
      );
    }
    for (const img of images) {
      out.push(
        `    <clipPath id="${img.clip}" clipPathUnits="objectBoundingBox"><polygon points="${img.points}" /></clipPath>
`
      );
    }
    out.push("  </defs>\n");
  }
  if (opt.showGrid) {
    out.push(
      renderGrid(scene, W, H, opt.grid?.cellSize ?? 1, opt.grid?.plane ?? "xz")
    );
  }
  let inStrokeGroup = false;
  for (const element of elements) {
    if (element.type === "face" || element.type === "image") {
      if (!inStrokeGroup) {
        out.push(
          '  <g stroke="#000" stroke-width="1" stroke-linecap="round" stroke-linejoin="round">\n'
        );
        inStrokeGroup = true;
      }
      if (element.type === "face") {
        const f = element.data;
        const strokeAttr = f.stroke ? "" : ' stroke="none"';
        out.push(
          `    <polygon fill="${f.fill}"${strokeAttr} points="${f.pts.map((p) => `${fmt(p.x)},${fmt(p.y)}`).join(" ")}" />
`
        );
      } else {
        const img = element.data;
        out.push(
          `    <g transform="${img.matrix}" clip-path="url(#${img.clip})"><use href="#${img.sym}"/></g>
`
        );
      }
    } else if (element.type === "label") {
      if (inStrokeGroup) {
        out.push("  </g>\n");
        inStrokeGroup = false;
      }
      const l = element.data;
      out.push(
        `  <g font-family="sans-serif" font-size="14" text-anchor="middle" dominant-baseline="central" transform="${l.matrix}"><text x="0" y="0" fill="${l.fill}">${l.text}</text></g>
`
      );
    } else if (element.type === "edge") {
      if (inStrokeGroup) {
        out.push("  </g>\n");
        inStrokeGroup = false;
      }
      const e = element.data;
      out.push(
        `  <polyline fill="none" stroke="${e.color}" points="${e.pts.map((p) => `${p.x},${p.y}`).join(" ")}" />
`
      );
    }
  }
  if (inStrokeGroup) {
    out.push("  </g>\n");
  }
  if (opt.showOrigin) {
    out.push(renderOrigin(scene.camera, W, H));
  }
  if (opt.showAxes) {
    out.push(renderAxes(scene.camera, W, H));
  }
  out.push("</svg>");
  return out.join("");
}
function renderAxes(cam, W, H) {
  const focal = cam.focalLength ?? 2;
  const baseDist = 3;
  const margin = Math.min(W, H) * 0.08;
  const arrowDist = baseDist * 0.16 / focal;
  const baseProj = proj2({ x: 0, y: 0, z: baseDist }, W, H, focal);
  if (!baseProj) return "";
  const offsetX = -W / 2 + margin - baseProj.x;
  const offsetY = H / 2 - margin - baseProj.y;
  function t(p) {
    const pp = proj2(p, W, H, focal);
    return pp ? { x: pp.x + offsetX, y: pp.y + offsetY } : { x: 0, y: 0 };
  }
  const { r, u, f } = axes2(cam);
  const start = t({ x: 0, y: 0, z: baseDist });
  const axesData = [
    { w: { x: 1, y: 0, z: 0 }, color: "red", label: "X" },
    { w: { x: 0, y: 1, z: 0 }, color: "green", label: "Y" },
    { w: { x: 0, y: 0, z: 1 }, color: "blue", label: "Z" }
  ].map(({ w, color, label }) => ({
    dir: {
      x: w.x * r.x + w.y * r.y + w.z * r.z,
      y: w.x * u.x + w.y * u.y + w.z * u.z,
      z: w.x * f.x + w.y * f.y + w.z * f.z
    },
    color,
    label
  }));
  const parts = [];
  for (const { dir, color, label } of axesData) {
    const end = t({
      x: dir.x * arrowDist,
      y: dir.y * arrowDist,
      z: baseDist + dir.z * arrowDist
    });
    const dx = end.x - start.x;
    const dy = end.y - start.y;
    const l = Math.sqrt(dx * dx + dy * dy) || 1;
    const nx = dx / l;
    const ny = dy / l;
    const hx = end.x - nx * 8;
    const hy = end.y - ny * 8;
    const b1x = hx + -ny * 4;
    const b1y = hy + nx * 4;
    const b2x = hx - -ny * 4;
    const b2y = hy - nx * 4;
    const tx = end.x + nx * 10;
    const ty = end.y + ny * 10;
    parts.push(
      `    <line x1="${fmt(start.x)}" y1="${fmt(start.y)}" x2="${fmt(hx)}" y2="${fmt(hy)}" stroke="${color}" />`
    );
    parts.push(
      `    <polygon fill="${color}" points="${fmt(end.x)},${fmt(end.y)} ${fmt(b1x)},${fmt(b1y)} ${fmt(b2x)},${fmt(b2y)}" />`
    );
    parts.push(
      `    <text x="${fmt(tx)}" y="${fmt(ty)}" fill="${color}" font-size="12" font-family="sans-serif" text-anchor="middle" dominant-baseline="central">${label}</text>`
    );
  }
  return `  <g stroke-width="2" stroke-linecap="round" stroke-linejoin="round">
${parts.join(
    "\n"
  )}
  </g>
`;
}
function renderGrid(scene, W, H, cellSize, plane) {
  const cam = scene.camera;
  const focal = cam.focalLength ?? 2;
  const { r, u, f } = axes2(cam);
  const toCam2 = (p) => {
    const d = sub(p, cam.position);
    return { x: dot(d, r), y: dot(d, u), z: dot(d, f) };
  };
  const project = (p) => proj2(toCam2(p), W, H, focal);
  const R = cellSize * 10;
  const lines = [];
  for (let t = -R; t <= R; t += cellSize) {
    const pushLine = (a, b) => {
      const p0 = project(a);
      const p1 = project(b);
      if (p0 && p1) {
        lines.push(
          `    <line x1="${fmt(p0.x)}" y1="${fmt(p0.y)}" x2="${fmt(p1.x)}" y2="${fmt(p1.y)}" />`
        );
      }
    };
    switch (plane) {
      case "xz":
        pushLine({ x: t, y: 0, z: -R }, { x: t, y: 0, z: R });
        pushLine({ x: -R, y: 0, z: t }, { x: R, y: 0, z: t });
        break;
      case "xy":
        pushLine({ x: t, y: -R, z: 0 }, { x: t, y: R, z: 0 });
        pushLine({ x: -R, y: t, z: 0 }, { x: R, y: t, z: 0 });
        break;
      case "yz":
        pushLine({ x: 0, y: t, z: -R }, { x: 0, y: t, z: R });
        pushLine({ x: 0, y: -R, z: t }, { x: 0, y: R, z: t });
        break;
    }
  }
  return lines.length ? `  <g stroke="#ccc" stroke-width="0.5">
${lines.join("\n")}
  </g>
` : "";
}
function renderOrigin(cam, W, H) {
  const focal = cam.focalLength ?? 2;
  const { r, u, f } = axes2(cam);
  const toCam2 = (p) => {
    const d = sub(p, cam.position);
    return { x: dot(d, r), y: dot(d, u), z: dot(d, f) };
  };
  const project = (p) => proj2(p, W, H, focal);
  const axesData = [
    { dir: { x: 1, y: 0, z: 0 }, color: "red" },
    { dir: { x: 0, y: 1, z: 0 }, color: "green" },
    { dir: { x: 0, y: 0, z: 1 }, color: "blue" }
  ];
  const minLineLengthPx = Math.max(W, H) * Math.SQRT2;
  const parts = [];
  const origin = { x: 0, y: 0, z: 0 };
  const gradientDefs = [];
  axesData.forEach(({ dir, color }, i) => {
    const L = 1;
    const end = add(origin, scale(dir, L));
    const startCam = toCam2(origin);
    const endCam = toCam2(end);
    const start2d = project(startCam);
    const end2d1 = project(endCam);
    const dx = end2d1.x - start2d.x;
    const dy = end2d1.y - start2d.y;
    const len2 = Math.sqrt(dx * dx + dy * dy);
    const end2d2 = {
      x: start2d.x + dx * minLineLengthPx / len2,
      y: start2d.y + dy * minLineLengthPx / len2
    };
    if (start2d && end2d1) {
      const gradId = `axis-grad-${i}`;
      const x1 = fmt(start2d.x);
      const y1 = fmt(start2d.y);
      const x2 = fmt(end2d2.x);
      const y2 = fmt(end2d2.y);
      gradientDefs.push(
        `    <linearGradient id="${gradId}" x1="${x1}" y1="${y1}" x2="${x2}" y2="${y2}" gradientUnits="userSpaceOnUse">      <stop offset="0%" stop-color="${color}"/>      <stop offset="${Math.min(len2 / minLineLengthPx * 1e3, 100)}%" stop-color="rgba(255,255,255,0)"/>      <stop offset="100%" stop-color="rgba(255,255,255,0)"/>    </linearGradient>`
      );
      parts.push(
        `    <line x1="${x1}" y1="${y1}" x2="${x2}" y2="${y2}" stroke="url(#${gradId})" />`
      );
    }
  });
  if (gradientDefs.length) {
    parts.unshift(`  <defs>
${gradientDefs.join("\n")}
  </defs>`);
  }
  return parts.length ? `  <g stroke-width="1">
${parts.join("\n")}
  </g>
` : "";
}
function axes2(cam) {
  const f = norm(sub(cam.lookAt, cam.position));
  const wUp = { x: 0, y: 1, z: 0 };
  let r = norm(cross(f, wUp));
  if (!len(r)) r = { x: 1, y: 0, z: 0 };
  const u = cross(r, f);
  return { r, u, f };
}
function proj2(p, w, h, focal) {
  if (p.z <= 0) return null;
  const s = focal / p.z;
  return { x: p.x * s * w / 2, y: -p.y * s * h / 2 };
}
export {
  load3MF,
  loadOBJ,
  loadSTL,
  renderScene
};
