// jscad-electronics/dist/index.js
// lib/A01005.tsx
import { Cuboid } from "jscad-fiber";
import { Fragment, jsx, jsxs } from "react/jsx-runtime";
var fullLength = 0.4;
var width = 0.2;
var height = 0.13;
var terminatorWidth = 0.07;
var bodyLength = fullLength - terminatorWidth * 2;
var A01005 = ({ color = "#333" }) => {
  return /* @__PURE__ */ jsxs(Fragment, { children: [
    /* @__PURE__ */ jsx(
      Cuboid,
      {
        size: [bodyLength, width, height],
        offset: [0, 0, height / 2],
        color
      }
    ),
    /* @__PURE__ */ jsx(
      Cuboid,
      {
        size: [terminatorWidth, width, height],
        offset: [fullLength / 2 - terminatorWidth / 2, 0, height / 2],
        color: "#ccc"
      }
    ),
    /* @__PURE__ */ jsx(
      Cuboid,
      {
        size: [terminatorWidth, width, height],
        offset: [-fullLength / 2 + terminatorWidth / 2, 0, height / 2],
        color: "#ccc"
      }
    )
  ] });
};

// lib/A0201.tsx
import { Cuboid as Cuboid2 } from "jscad-fiber";
import { Fragment as Fragment2, jsx as jsx2, jsxs as jsxs2 } from "react/jsx-runtime";
var fullLength2 = 0.6;
var width2 = 0.3;
var height2 = 0.33;
var terminatorWidth2 = 0.1;
var bodyLength2 = fullLength2 - terminatorWidth2 * 2;
var A0201 = ({ color = "#333" }) => {
  return /* @__PURE__ */ jsxs2(Fragment2, { children: [
    /* @__PURE__ */ jsx2(
      Cuboid2,
      {
        size: [bodyLength2, width2, height2],
        offset: [0, 0, height2 / 2],
        color
      }
    ),
    /* @__PURE__ */ jsx2(
      Cuboid2,
      {
        size: [terminatorWidth2, width2, height2],
        offset: [fullLength2 / 2 - terminatorWidth2 / 2, 0, height2 / 2],
        color: "#ccc"
      }
    ),
    /* @__PURE__ */ jsx2(
      Cuboid2,
      {
        size: [terminatorWidth2, width2, height2],
        offset: [-fullLength2 / 2 + terminatorWidth2 / 2, 0, height2 / 2],
        color: "#ccc"
      }
    )
  ] });
};

// lib/A0402.tsx
import { Cuboid as Cuboid3 } from "jscad-fiber";
import { Fragment as Fragment3, jsx as jsx3, jsxs as jsxs3 } from "react/jsx-runtime";
var fullLength3 = 1;
var width3 = 0.5;
var height3 = 0.5;
var terminatorWidth3 = 0.2;
var bodyLength3 = fullLength3 - terminatorWidth3 * 2;
var A0402 = ({ color = "#333" }) => {
  return /* @__PURE__ */ jsxs3(Fragment3, { children: [
    /* @__PURE__ */ jsx3(
      Cuboid3,
      {
        size: [bodyLength3, width3, height3],
        offset: [0, 0, height3 / 2],
        color
      }
    ),
    /* @__PURE__ */ jsx3(
      Cuboid3,
      {
        size: [terminatorWidth3, height3, width3],
        offset: [fullLength3 / 2 - terminatorWidth3 / 2, 0, height3 / 2],
        color: "#ccc"
      }
    ),
    /* @__PURE__ */ jsx3(
      Cuboid3,
      {
        size: [terminatorWidth3, height3, width3],
        offset: [-fullLength3 / 2 + terminatorWidth3 / 2, 0, height3 / 2],
        color: "#ccc"
      }
    )
  ] });
};

// lib/A0603.tsx
import { Cuboid as Cuboid4 } from "jscad-fiber";
import { Fragment as Fragment4, jsx as jsx4, jsxs as jsxs4 } from "react/jsx-runtime";
var fullLength4 = 1.6;
var bodyLength4 = fullLength4 - 0.3 * 2;
var terminatorLength = 0.3;
var width4 = 0.85;
var height4 = 0.6;
var A0603 = ({ color = "#333" }) => {
  return /* @__PURE__ */ jsxs4(Fragment4, { children: [
    /* @__PURE__ */ jsx4(
      Cuboid4,
      {
        size: [bodyLength4, width4, height4],
        offset: [0, 0, height4 / 2],
        color
      }
    ),
    /* @__PURE__ */ jsx4(
      Cuboid4,
      {
        size: [terminatorLength, width4, height4],
        offset: [fullLength4 / 2 - terminatorLength / 2, 0, height4 / 2],
        color: "#ccc"
      }
    ),
    /* @__PURE__ */ jsx4(
      Cuboid4,
      {
        size: [terminatorLength, width4, height4],
        offset: [-fullLength4 / 2 + terminatorLength / 2, 0, height4 / 2],
        color: "#ccc"
      }
    )
  ] });
};

// lib/A0805.tsx
import { Cuboid as Cuboid5 } from "jscad-fiber";
import { Fragment as Fragment5, jsx as jsx5, jsxs as jsxs5 } from "react/jsx-runtime";
var fullLength5 = 2;
var width5 = 1.25;
var height5 = 0.55;
var terminatorWidth4 = 0.5;
var bodyLength5 = fullLength5 - terminatorWidth4 * 2;
var A0805 = ({ color = "#333" }) => {
  return /* @__PURE__ */ jsxs5(Fragment5, { children: [
    /* @__PURE__ */ jsx5(
      Cuboid5,
      {
        size: [bodyLength5, width5, height5],
        offset: [0, 0, height5 / 2],
        color
      }
    ),
    /* @__PURE__ */ jsx5(
      Cuboid5,
      {
        size: [terminatorWidth4, width5, height5],
        offset: [fullLength5 / 2 - terminatorWidth4 / 2, 0, height5 / 2],
        color: "#ccc"
      }
    ),
    /* @__PURE__ */ jsx5(
      Cuboid5,
      {
        size: [terminatorWidth4, width5, height5],
        offset: [-fullLength5 / 2 + terminatorWidth4 / 2, 0, height5 / 2],
        color: "#ccc"
      }
    )
  ] });
};

// lib/A1206.tsx
import { Cuboid as Cuboid6 } from "jscad-fiber";
import { Fragment as Fragment6, jsx as jsx6, jsxs as jsxs6 } from "react/jsx-runtime";
var fullLength6 = 3.2;
var width6 = 1.6;
var height6 = 0.9;
var terminatorWidth5 = 0.5;
var bodyLength6 = fullLength6 - terminatorWidth5 * 2;
var A1206 = ({ color = "#333" }) => {
  return /* @__PURE__ */ jsxs6(Fragment6, { children: [
    /* @__PURE__ */ jsx6(
      Cuboid6,
      {
        size: [bodyLength6, width6, height6],
        offset: [0, 0, height6 / 2],
        color
      }
    ),
    /* @__PURE__ */ jsx6(
      Cuboid6,
      {
        size: [terminatorWidth5, width6, height6],
        offset: [fullLength6 / 2 - terminatorWidth5 / 2, 0, height6 / 2],
        color: "#ccc"
      }
    ),
    /* @__PURE__ */ jsx6(
      Cuboid6,
      {
        size: [terminatorWidth5, width6, height6],
        offset: [-fullLength6 / 2 + terminatorWidth5 / 2, 0, height6 / 2],
        color: "#ccc"
      }
    )
  ] });
};

// lib/A1210.tsx
import { Cuboid as Cuboid7 } from "jscad-fiber";
import { Fragment as Fragment7, jsx as jsx7, jsxs as jsxs7 } from "react/jsx-runtime";
var fullLength7 = 3.2;
var width7 = 2.5;
var height7 = 1;
var terminatorWidth6 = 0.6;
var bodyLength7 = fullLength7 - terminatorWidth6 * 2;
var A1210 = ({ color = "#333" }) => {
  return /* @__PURE__ */ jsxs7(Fragment7, { children: [
    /* @__PURE__ */ jsx7(
      Cuboid7,
      {
        size: [bodyLength7, width7, height7],
        offset: [0, 0, height7 / 2],
        color
      }
    ),
    /* @__PURE__ */ jsx7(
      Cuboid7,
      {
        size: [terminatorWidth6, width7, height7],
        offset: [fullLength7 / 2 - terminatorWidth6 / 2, 0, height7 / 2],
        color: "#ccc"
      }
    ),
    /* @__PURE__ */ jsx7(
      Cuboid7,
      {
        size: [terminatorWidth6, width7, height7],
        offset: [-fullLength7 / 2 + terminatorWidth6 / 2, 0, height7 / 2],
        color: "#ccc"
      }
    )
  ] });
};

// lib/A2010.tsx
import { Cuboid as Cuboid8 } from "jscad-fiber";
import { Fragment as Fragment8, jsx as jsx8, jsxs as jsxs8 } from "react/jsx-runtime";
var fullLength8 = 5;
var width8 = 2.5;
var height8 = 1.4;
var terminatorWidth7 = 0.6;
var bodyLength8 = fullLength8 - terminatorWidth7 * 2;
var A2010 = ({ color = "#333" }) => {
  return /* @__PURE__ */ jsxs8(Fragment8, { children: [
    /* @__PURE__ */ jsx8(
      Cuboid8,
      {
        size: [bodyLength8, width8, height8],
        offset: [0, 0, height8 / 2],
        color
      }
    ),
    /* @__PURE__ */ jsx8(
      Cuboid8,
      {
        size: [terminatorWidth7, width8, height8],
        offset: [fullLength8 / 2 - terminatorWidth7 / 2, 0, height8 / 2],
        color: "#ccc"
      }
    ),
    /* @__PURE__ */ jsx8(
      Cuboid8,
      {
        size: [terminatorWidth7, width8, height8],
        offset: [-fullLength8 / 2 + terminatorWidth7 / 2, 0, height8 / 2],
        color: "#ccc"
      }
    )
  ] });
};

// lib/A2512.tsx
import { Cuboid as Cuboid9 } from "jscad-fiber";
import { Fragment as Fragment9, jsx as jsx9, jsxs as jsxs9 } from "react/jsx-runtime";
var fullLength9 = 6.3;
var width9 = 3.2;
var height9 = 1.8;
var terminatorWidth8 = 0.8;
var bodyLength9 = fullLength9 - terminatorWidth8 * 2;
var A2512 = ({ color = "#333" }) => {
  return /* @__PURE__ */ jsxs9(Fragment9, { children: [
    /* @__PURE__ */ jsx9(
      Cuboid9,
      {
        size: [bodyLength9, width9, height9],
        offset: [0, 0, height9 / 2],
        color
      }
    ),
    /* @__PURE__ */ jsx9(
      Cuboid9,
      {
        size: [terminatorWidth8, width9, height9],
        offset: [fullLength9 / 2 - terminatorWidth8 / 2, 0, height9 / 2],
        color: "#ccc"
      }
    ),
    /* @__PURE__ */ jsx9(
      Cuboid9,
      {
        size: [terminatorWidth8, width9, height9],
        offset: [-fullLength9 / 2 + terminatorWidth8 / 2, 0, height9 / 2],
        color: "#ccc"
      }
    )
  ] });
};

// lib/BGA.tsx
import { Cuboid as Cuboid10, Sphere, Translate as Translate2, Colorize } from "jscad-fiber";
import { fp } from "@tscircuit/footprinter";
import { useMemo } from "react";
import { Fragment as Fragment10, jsx as jsx10, jsxs as jsxs10 } from "react/jsx-runtime";
var BGA = ({
  packageWidth = 10,
  packageLength = 10,
  packageHeight = 1.2,
  standoffHeight = 0.2,
  ballPitch = 0.8,
  ballDiameter = 0.5,
  ballRows = 8,
  ballColumns = 8,
  missingBalls = [],
  footprintString
}) => {
  const bodyHeight = packageHeight - standoffHeight;
  const bodyOffset = standoffHeight + bodyHeight / 2;
  const ballOffset = -standoffHeight / 2;
  const ballsSoup = useMemo(() => {
    if (!footprintString) return null;
    const result = fp.string(footprintString);
    return result.circuitJson();
  }, [footprintString]);
  return /* @__PURE__ */ jsxs10(Fragment10, { children: [
    /* @__PURE__ */ jsx10(Translate2, { z: bodyOffset, children: /* @__PURE__ */ jsx10(Colorize, { color: "#555", children: /* @__PURE__ */ jsx10(Cuboid10, { size: [packageWidth, packageLength, bodyHeight] }) }) }),
    !footprintString && Array.from({ length: ballRows * ballColumns }).map((_, index) => {
      if (missingBalls.includes(index + 1)) return null;
      const row = Math.floor(index / ballColumns);
      const col = index % ballColumns;
      const x = (col - (ballColumns - 1) / 2) * ballPitch;
      const y = (row - (ballRows - 1) / 2) * ballPitch;
      return /* @__PURE__ */ jsx10(Translate2, { x, y, z: ballOffset, children: /* @__PURE__ */ jsx10(Sphere, { radius: ballDiameter / 2 }) }, index);
    }),
    ballsSoup && ballsSoup.map((elm, index) => {
      if (elm.type === "pcb_smtpad") {
        return /* @__PURE__ */ jsx10(
          Translate2,
          {
            x: elm.x,
            y: elm.y,
            z: ballOffset,
            children: /* @__PURE__ */ jsx10(Sphere, { radius: ballDiameter / 2 })
          },
          index
        );
      }
      return null;
    })
  ] });
};

// lib/ChipBody.tsx
import {
  Colorize as Colorize2,
  Cuboid as Cuboid11,
  Cylinder,
  Hull,
  Rotate as Rotate2,
  Subtract,
  Translate as Translate3,
  Union as Union2
} from "jscad-fiber";
import { jsx as jsx11, jsxs as jsxs11 } from "react/jsx-runtime";
var ChipBody = ({
  center,
  width: width10,
  length,
  height: height10,
  heightAboveSurface: heightAboveSurface2 = 0.15
}) => {
  const straightHeight = height10 * 0.5;
  const taperHeight = height10 - straightHeight;
  const taperInset = Math.min(width10, length) * 0.12;
  const faceWidth = Math.max(width10 - taperInset, width10 * 0.75);
  const faceLength = Math.max(length - taperInset, length * 0.75);
  const notchRadius = Math.min(width10, length) * 0.12;
  const notchCenterZ = height10 - notchRadius * 0.6;
  const notchCenterY = length / 2 - notchRadius * 0.25;
  const notchLength = 0.5;
  const notchWidth = 0.25;
  return /* @__PURE__ */ jsx11(Colorize2, { color: "#555", children: /* @__PURE__ */ jsx11(Translate3, { offset: center, children: /* @__PURE__ */ jsx11(Translate3, { offset: { x: 0, y: 0, z: heightAboveSurface2 }, children: /* @__PURE__ */ jsxs11(Subtract, { children: [
    /* @__PURE__ */ jsxs11(Union2, { children: [
      /* @__PURE__ */ jsxs11(Hull, { children: [
        /* @__PURE__ */ jsx11(Translate3, { z: 5e-3, children: /* @__PURE__ */ jsx11(Cuboid11, { size: [faceWidth, faceLength, 0.01] }) }),
        /* @__PURE__ */ jsx11(Translate3, { z: straightHeight, children: /* @__PURE__ */ jsx11(Cuboid11, { size: [width10, length, 0.01] }) })
      ] }),
      /* @__PURE__ */ jsxs11(Hull, { children: [
        /* @__PURE__ */ jsx11(Translate3, { z: straightHeight, children: /* @__PURE__ */ jsx11(Cuboid11, { size: [width10, length, 0.01] }) }),
        /* @__PURE__ */ jsx11(Translate3, { z: straightHeight + taperHeight, children: /* @__PURE__ */ jsx11(Cuboid11, { size: [faceWidth, faceLength, 0.01] }) })
      ] })
    ] }),
    /* @__PURE__ */ jsx11(Translate3, { offset: { x: 0, y: notchCenterY, z: height10 }, children: /* @__PURE__ */ jsx11(Rotate2, { rotation: [0, 0, 0], children: /* @__PURE__ */ jsx11(Cylinder, { radius: notchLength, height: notchWidth }) }) })
  ] }) }) }) });
};

// lib/ExtrudedPads.tsx
import { fp as fp2 } from "@tscircuit/footprinter";

// lib/FootprintPad.tsx
import { Colorize as Colorize3, Cuboid as Cuboid12, Translate as Translate4 } from "jscad-fiber";
import { jsx as jsx12 } from "react/jsx-runtime";
var FootprintPad = ({ pad }) => {
  if (pad.shape === "rect") {
    return /* @__PURE__ */ jsx12(Colorize3, { color: [255, 0, 0], children: /* @__PURE__ */ jsx12(Translate4, { offset: [pad.x, pad.y, 0], children: /* @__PURE__ */ jsx12(Cuboid12, { size: [pad.width, pad.height, 0.01] }) }) });
  } else {
    throw new Error("Shape not supported: " + pad.shape);
  }
};

// lib/FootprintPlatedHole.tsx
import { Colorize as Colorize4, Translate as Translate5, Cylinder as Cylinder2, Subtract as Subtract2 } from "jscad-fiber";
import { jsx as jsx13, jsxs as jsxs12 } from "react/jsx-runtime";
var FootprintPlatedHole = ({ hole }) => {
  if (hole.shape === "circle") {
    return /* @__PURE__ */ jsx13(Colorize4, { color: "#b87333", children: /* @__PURE__ */ jsx13(Translate5, { offset: [hole.x, hole.y, 0], children: /* @__PURE__ */ jsxs12(Subtract2, { children: [
      /* @__PURE__ */ jsx13(Cylinder2, { radius: hole.outer_diameter / 2, height: 0.01 }),
      /* @__PURE__ */ jsx13(Cylinder2, { radius: hole.hole_diameter / 2, height: 0.01 })
    ] }) }) });
  }
  throw new Error("Shape not supported: " + hole.shape);
};

// lib/ExtrudedPads.tsx
import { Fragment as Fragment11, jsx as jsx14, jsxs as jsxs13 } from "react/jsx-runtime";
var ExtrudedPads = ({
  circuitJson,
  footprint
}) => {
  if (!circuitJson && footprint) {
    circuitJson = fp2.string(footprint).circuitJson();
  }
  if (!circuitJson)
    throw new Error("No circuit json or footprint provided to ExtrudedPads");
  return /* @__PURE__ */ jsxs13(Fragment11, { children: [
    circuitJson.filter((s) => s.type === "pcb_smtpad").map((pad, i) => (
      // biome-ignore lint/suspicious/noArrayIndexKey:
      /* @__PURE__ */ jsx14(FootprintPad, { pad }, i)
    )),
    circuitJson.filter((s) => s.type === "pcb_plated_hole").map((hole, i) => {
      return (
        // biome-ignore lint/suspicious/noArrayIndexKey:
        /* @__PURE__ */ jsx14(FootprintPlatedHole, { hole }, i)
      );
    })
  ] });
};

// lib/Footprinter3d.tsx
import { fp as fp3 } from "@tscircuit/footprinter";

// lib/DualInlinePackage.tsx
import { Polygon, ExtrudeLinear, Rotate as Rotate3, Translate as Translate6 } from "jscad-fiber";

// lib/utils/range.ts
var range = (end) => Array.from({ length: end }, (_, i) => i);

// lib/utils/getExpandedStroke.ts
function getExpandedStroke(strokeInput, width10) {
  if (strokeInput.length < 2) {
    throw new Error("Stroke must have at least two points");
  }
  const stroke = Array.isArray(strokeInput[0]) ? strokeInput.map(([x, y]) => ({ x, y })) : strokeInput;
  const halfWidth = width10 / 2;
  const leftSide = [];
  const rightSide = [];
  function getNormal(p1, p2) {
    const dx = p2.x - p1.x;
    const dy = p2.y - p1.y;
    const length = Math.sqrt(dx * dx + dy * dy);
    return { x: -dy / length, y: dx / length };
  }
  function addPoint(point, normal, factor) {
    const newPoint = {
      x: point.x + normal.x * halfWidth * factor,
      y: point.y + normal.y * halfWidth * factor
    };
    if (factor > 0) {
      leftSide.push(newPoint);
    } else {
      rightSide.unshift(newPoint);
    }
  }
  const firstNormal = getNormal(stroke[0], stroke[1]);
  addPoint(stroke[0], firstNormal, 1);
  addPoint(stroke[0], firstNormal, -1);
  for (let i = 1; i < stroke.length - 1; i++) {
    const prev = stroke[i - 1];
    const current = stroke[i];
    const next = stroke[i + 1];
    const normalPrev = getNormal(prev, current);
    const normalNext = getNormal(current, next);
    const miterX = normalPrev.x + normalNext.x;
    const miterY = normalPrev.y + normalNext.y;
    const miterLength = Math.sqrt(miterX * miterX + miterY * miterY);
    const miterLimit = 2;
    if (miterLength / 2 > miterLimit * halfWidth) {
      addPoint(current, normalPrev, 1);
      addPoint(current, normalNext, 1);
      addPoint(current, normalPrev, -1);
      addPoint(current, normalNext, -1);
    } else {
      const scale = 1 / miterLength;
      addPoint(current, { x: miterX * scale, y: miterY * scale }, 1);
      addPoint(current, { x: miterX * scale, y: miterY * scale }, -1);
    }
  }
  const lastNormal = getNormal(
    stroke[stroke.length - 2],
    stroke[stroke.length - 1]
  );
  addPoint(stroke[stroke.length - 1], lastNormal, 1);
  addPoint(stroke[stroke.length - 1], lastNormal, -1);
  return [...leftSide, ...rightSide];
}

// lib/DualInlinePackage.tsx
import { Fragment as Fragment12, jsx as jsx15, jsxs as jsxs14 } from "react/jsx-runtime";
var normalizeOnY = (points) => {
  const minX = Math.min(...points.map((p) => p.x));
  const maxX = Math.max(...points.map((p) => p.x));
  const minY = Math.min(...points.map((p) => p.y));
  const maxY = Math.max(...points.map((p) => p.y));
  const height10 = maxY - minY;
  const factor = 5.47 / height10;
  return points.map((p) => ({
    x: (p.x - minX - (maxX - minX) / 2) * factor,
    y: (p.y - minY) * factor
  }));
};
var svgPathPoints = normalizeOnY([
  { x: 20, y: 105 },
  { x: 20, y: 109 },
  { x: 20, y: 102 },
  { x: 26, y: 102 },
  { x: 26, y: 109 },
  { x: 24, y: 111 },
  { x: 24, y: 118 },
  { x: 22, y: 118 },
  { x: 22, y: 111 },
  { x: 20, y: 109 }
]);
var DIP_PIN_HEIGHT = 5.47;
var heightAboveSurface = 0.5;
var DipPinLeg = ({ x, y, z }) => {
  const isRotated = x > 0;
  return /* @__PURE__ */ jsxs14(Fragment12, { children: [
    /* @__PURE__ */ jsx15(Translate6, { offset: { x: x + 0.25 / 2, y, z }, children: /* @__PURE__ */ jsx15(Rotate3, { rotation: ["-90deg", 0, "90deg"], children: /* @__PURE__ */ jsx15(ExtrudeLinear, { height: 0.25, children: /* @__PURE__ */ jsx15(Polygon, { points: svgPathPoints.map((p) => [p.x, p.y]) }) }) }) }),
    /* @__PURE__ */ jsx15(
      Translate6,
      {
        offset: {
          x,
          y: y + (isRotated ? 1 : -1),
          z
        },
        children: /* @__PURE__ */ jsx15(Rotate3, { rotation: ["-90deg", "90deg", isRotated ? "180deg" : "0deg"], children: /* @__PURE__ */ jsx15(ExtrudeLinear, { height: 2, children: /* @__PURE__ */ jsx15(
          Polygon,
          {
            points: getExpandedStroke(
              [
                [0, 0],
                [-1, 0],
                [-1, -1]
              ],
              0.25
            ).map((p) => [p.x, p.y])
          }
        ) }) })
      }
    )
  ] });
};
var Dip = ({
  numPins = 8,
  pitch = 2.54,
  bodyWidth = 6.4
}) => {
  const numPinsOnEachSide = Math.floor(numPins / 2);
  const crossBodyPinWidth = bodyWidth + 1;
  return /* @__PURE__ */ jsxs14(Fragment12, { children: [
    range(numPins).map((i) => {
      const yRow = i % numPinsOnEachSide;
      const xRow = (Math.floor(i / numPinsOnEachSide) - 0.5) * 2;
      return /* @__PURE__ */ jsx15(
        DipPinLeg,
        {
          x: xRow * crossBodyPinWidth / 2,
          y: yRow * pitch - (numPinsOnEachSide - 1) / 2 * pitch,
          z: DIP_PIN_HEIGHT / 2 + heightAboveSurface
        },
        i
      );
    }),
    /* @__PURE__ */ jsx15(
      ChipBody,
      {
        width: bodyWidth,
        length: numPinsOnEachSide * pitch + 0.5,
        height: DIP_PIN_HEIGHT - heightAboveSurface,
        heightAboveSurface,
        center: { x: 0, y: 0, z: heightAboveSurface }
      }
    )
  ] });
};

// lib/SmdChipLead.tsx
import {
  Polygon as Polygon2,
  ExtrudeLinear as ExtrudeLinear2,
  Translate as Translate7,
  Rotate as Rotate4,
  Colorize as Colorize5
} from "jscad-fiber";
import { jsx as jsx16 } from "react/jsx-runtime";
function calculateSCurve(x, {
  height: height10,
  padContactLength,
  curveLength,
  bodyDistance,
  steepness = 10
}) {
  if (!curveLength) curveLength = bodyDistance * 0.3;
  let flatFromBodySectionLength = bodyDistance - padContactLength - curveLength;
  if (flatFromBodySectionLength < 0) {
    curveLength += flatFromBodySectionLength;
    flatFromBodySectionLength = 0;
  }
  const curveStart = padContactLength * 0.75;
  const curveEnd = padContactLength + curveLength + (bodyDistance - padContactLength - curveLength) * 0.25;
  if (x <= curveStart) return 0;
  if (x >= curveEnd) return height10;
  const t = (x - curveStart) / (curveEnd - curveStart);
  return height10 / (1 + Math.exp(-steepness * (t - 0.5)));
}
var SmdChipLead = (props) => {
  const { thickness, width: width10, padContactLength, bodyDistance, height: height10, rotation } = props;
  const N = 15;
  const points = Array.from({ length: N }).map((_, i) => i / (N - 1) * bodyDistance).map((x) => [x, calculateSCurve(x, props)]);
  const polygon = getExpandedStroke(points, thickness);
  return /* @__PURE__ */ jsx16(Colorize5, { color: "#fff", children: /* @__PURE__ */ jsx16(Translate7, { offset: { z: 0, y: 0, x: 0, ...props.position }, children: /* @__PURE__ */ jsx16(Rotate4, { rotation: ["90deg", 0, rotation ?? 0], children: /* @__PURE__ */ jsx16(Translate7, { offset: { x: 0, y: 0, z: -width10 / 2 }, children: /* @__PURE__ */ jsx16(ExtrudeLinear2, { height: width10, children: /* @__PURE__ */ jsx16(Polygon2, { points: polygon.map((p) => [p.x, p.y]) }) }) }) }) }) });
};

// lib/Tssop.tsx
import { Fragment as Fragment13, jsx as jsx17, jsxs as jsxs15 } from "react/jsx-runtime";
var Tssop = ({
  pinCount,
  leadLength,
  leadWidth,
  pitch,
  bodyWidth
}) => {
  const sidePinCount = Math.ceil(pinCount / 2);
  const fullLength10 = pitch * pinCount / 2 + leadWidth / 2;
  const pinOffsetToCenter = (sidePinCount - 1) * pitch / 2;
  const leadThickness = 0.25;
  return /* @__PURE__ */ jsxs15(Fragment13, { children: [
    Array.from({ length: sidePinCount }).map((_, i) => /* @__PURE__ */ jsx17(
      SmdChipLead,
      {
        position: {
          x: -bodyWidth / 2 - leadLength,
          y: i * pitch - pinOffsetToCenter,
          z: leadThickness / 2
        },
        width: leadWidth,
        thickness: leadThickness,
        padContactLength: leadLength,
        bodyDistance: leadLength + 1,
        height: 0.8
      },
      i
    )),
    Array.from({ length: sidePinCount }).map((_, i) => /* @__PURE__ */ jsx17(
      SmdChipLead,
      {
        rotation: Math.PI,
        position: {
          x: bodyWidth / 2 + leadLength,
          y: i * pitch - pinOffsetToCenter,
          z: leadThickness / 2
        },
        width: leadWidth,
        thickness: leadThickness,
        padContactLength: leadLength,
        bodyDistance: leadLength + 1,
        height: 0.8
      },
      i
    )),
    /* @__PURE__ */ jsx17(
      ChipBody,
      {
        center: { x: 0, y: 0, z: leadThickness / 2 },
        width: bodyWidth - leadWidth - 1,
        length: fullLength10,
        height: 1.5
      }
    )
  ] });
};

// lib/qfp.tsx
import { Fragment as Fragment14, jsx as jsx18, jsxs as jsxs16 } from "react/jsx-runtime";
var QFP = ({
  pinCount,
  pitch,
  leadWidth,
  padContactLength,
  bodyWidth
}) => {
  const sidePinCount = pinCount / 4;
  if (!pitch) pitch = getPitch(pinCount, bodyWidth);
  if (!padContactLength) padContactLength = getPadContactLength(pinCount);
  if (!leadWidth) leadWidth = getLeadWidth(pinCount, bodyWidth);
  if (!bodyWidth) bodyWidth = pitch * (sidePinCount + 4);
  const bodyLength10 = bodyWidth;
  const pinOffsetToCenter = (sidePinCount - 1) * pitch / 2;
  const fullLength10 = bodyLength10 + 2 * padContactLength;
  const fullWidth = fullLength10;
  const leadHeight = 0.8;
  const leadThickness = 0.15;
  const bodyDistance = (fullWidth - bodyWidth) / 2 + 0.2;
  return /* @__PURE__ */ jsxs16(Fragment14, { children: [
    Array.from({ length: sidePinCount }).map((_, i) => /* @__PURE__ */ jsx18(
      SmdChipLead,
      {
        position: {
          x: -fullWidth / 2,
          y: i * pitch - pinOffsetToCenter,
          z: leadThickness / 2
        },
        width: leadWidth,
        thickness: leadThickness,
        padContactLength,
        bodyDistance,
        height: leadHeight
      },
      `left-${i}`
    )),
    Array.from({ length: sidePinCount }).map((_, i) => /* @__PURE__ */ jsx18(
      SmdChipLead,
      {
        rotation: Math.PI,
        position: {
          x: fullWidth / 2,
          y: i * pitch - pinOffsetToCenter,
          z: leadThickness / 2
        },
        width: leadWidth,
        thickness: leadThickness,
        padContactLength,
        bodyDistance,
        height: leadHeight
      },
      `right-${i}`
    )),
    Array.from({ length: sidePinCount }).map((_, i) => /* @__PURE__ */ jsx18(
      SmdChipLead,
      {
        rotation: Math.PI / 2,
        position: {
          x: i * pitch - pinOffsetToCenter,
          y: -fullLength10 / 2,
          z: leadThickness / 2
        },
        width: leadWidth,
        thickness: leadThickness,
        padContactLength,
        bodyDistance,
        height: leadHeight
      },
      `bottom-${i}`
    )),
    Array.from({ length: sidePinCount }).map((_, i) => /* @__PURE__ */ jsx18(
      SmdChipLead,
      {
        rotation: -Math.PI / 2,
        position: {
          x: i * pitch - pinOffsetToCenter,
          y: fullLength10 / 2,
          z: leadThickness / 2
        },
        width: leadWidth,
        thickness: leadThickness,
        padContactLength,
        bodyDistance,
        height: leadHeight
      },
      `top-${i}`
    )),
    /* @__PURE__ */ jsx18(
      ChipBody,
      {
        center: { x: 0, y: 0, z: leadThickness / 2 },
        width: bodyWidth,
        length: bodyLength10,
        height: 1.5
      }
    )
  ] });
};
var getPitch = (pinCount, width10) => {
  switch (pinCount) {
    case 44:
    case 64:
      return 0.8;
    case 52:
      return width10 === 14 ? 1 : 0.65;
    case 208:
      return 0.5;
    default:
      return 0.5;
  }
};
var getPadContactLength = (pinCount) => {
  switch (pinCount) {
    case 44:
    case 52:
    case 64:
      return 2.25;
    case 208:
      return 1.65;
    default:
      return 1;
  }
};
var getLeadWidth = (pinCount, width10) => {
  switch (pinCount) {
    case 44:
    case 64:
      return 0.5;
    case 52:
      return width10 === 14 ? 0.45 : 0.55;
    case 208:
      return 0.3;
    default:
      return 0.25;
  }
};

// lib/PinRow.tsx
import { Cuboid as Cuboid14, Colorize as Colorize6, Hull as Hull2 } from "jscad-fiber";
import { Fragment as Fragment15, jsx as jsx19, jsxs as jsxs17 } from "react/jsx-runtime";
var PinRow = ({
  numberOfPins,
  pitch = 2.54,
  longSidePinLength = 6
}) => {
  const pinThickness = 0.63;
  const bodyHeight = 2;
  const bodyWidth = numberOfPins * pitch;
  const shortSidePinLength = 3;
  const xoff = -((numberOfPins - 1) / 2) * pitch;
  return /* @__PURE__ */ jsxs17(Fragment15, { children: [
    /* @__PURE__ */ jsx19(
      Cuboid14,
      {
        color: "#222",
        size: [bodyWidth, pinThickness * 3, bodyHeight],
        center: [0, 0, bodyHeight / 2]
      }
    ),
    Array.from({ length: numberOfPins }, (_, i) => /* @__PURE__ */ jsxs17(Fragment15, { children: [
      /* @__PURE__ */ jsx19(Colorize6, { color: "gold", children: /* @__PURE__ */ jsxs17(Hull2, { children: [
        /* @__PURE__ */ jsx19(
          Cuboid14,
          {
            color: "gold",
            size: [pinThickness, pinThickness, shortSidePinLength * 0.9],
            center: [
              xoff + i * pitch,
              0,
              bodyHeight * 0.9 + bodyHeight / 2
            ]
          }
        ),
        /* @__PURE__ */ jsx19(
          Cuboid14,
          {
            color: "gold",
            size: [
              pinThickness / 1.8,
              pinThickness / 1.8,
              shortSidePinLength
            ],
            center: [xoff + i * pitch, 0, bodyHeight + bodyHeight / 2]
          }
        )
      ] }) }),
      /* @__PURE__ */ jsx19(Colorize6, { color: "gold", children: /* @__PURE__ */ jsxs17(Hull2, { children: [
        /* @__PURE__ */ jsx19(
          Cuboid14,
          {
            color: "gold",
            size: [pinThickness, pinThickness, longSidePinLength * 0.9],
            center: [xoff + i * pitch, 0, -longSidePinLength / 2 * 0.9]
          }
        ),
        /* @__PURE__ */ jsx19(
          Cuboid14,
          {
            color: "gold",
            size: [
              pinThickness / 1.8,
              pinThickness / 1.8,
              longSidePinLength
            ],
            center: [xoff + i * pitch, 0, -longSidePinLength / 2]
          }
        )
      ] }) })
    ] }))
  ] });
};

// lib/qfn.tsx
import { Cuboid as Cuboid15, Colorize as Colorize7 } from "jscad-fiber";

// lib/utils/getQuadCoords.ts
var getQuadCoords = (params) => {
  const SIDES_CCW = ["left", "bottom", "right", "top"];
  const { pin_count, pn, w, h, p, pl, legsoutside } = params;
  const sidePinCount = pin_count / 4;
  const side = SIDES_CCW[Math.floor((pn - 1) / sidePinCount)];
  const pos = (pn - 1) % sidePinCount;
  const ibw = p * (sidePinCount - 1);
  const ibh = p * (sidePinCount - 1);
  const pcdfe = legsoutside ? pl / 2 : -pl / 2;
  switch (side) {
    case "left":
      return { x: -w / 2 - pcdfe, y: ibh / 2 - pos * p, o: "vert" };
    case "bottom":
      return { x: -ibw / 2 + pos * p, y: -h / 2 - pcdfe, o: "horz" };
    case "right":
      return { x: w / 2 + pcdfe, y: -ibh / 2 + pos * p, o: "vert" };
    case "top":
      return { x: ibw / 2 - pos * p, y: h / 2 + pcdfe, o: "horz" };
    default:
      throw new Error("Invalid pin number");
  }
};

// lib/utils/getQuadPinMap.ts
var getQuadPinMap = ({
  num_pins,
  cw,
  ccw,
  startingpin
}) => {
  const pin_map = [];
  const pins_per_side = num_pins / 4;
  let current_position_ccw_normal = 1;
  const sfp = {};
  for (const specifier of startingpin ?? []) {
    sfp[specifier] = true;
  }
  if (!sfp.leftside && !sfp.topside && !sfp.rightside && !sfp.bottomside) {
    sfp.leftside = true;
  }
  if (!sfp.bottompin && !sfp.leftpin && !sfp.rightpin && !sfp.toppin) {
    if (sfp.leftside) {
      sfp.toppin = true;
    } else if (sfp.topside) {
      sfp.rightpin = true;
    } else if (sfp.rightside) {
      sfp.bottompin = true;
    } else if (sfp.bottomside) {
      sfp.leftpin = true;
    }
  }
  if (sfp.leftside && sfp.toppin) {
    current_position_ccw_normal = 1;
  } else if (sfp.leftside && sfp.bottompin) {
    current_position_ccw_normal = pins_per_side;
  } else if (sfp.bottomside && sfp.leftpin) {
    current_position_ccw_normal = pins_per_side + 1;
  } else if (sfp.bottomside && sfp.rightpin) {
    current_position_ccw_normal = pins_per_side * 2;
  } else if (sfp.rightside && sfp.bottompin) {
    current_position_ccw_normal = pins_per_side * 2 + 1;
  } else if (sfp.rightside && sfp.toppin) {
    current_position_ccw_normal = pins_per_side * 3;
  } else if (sfp.topside && sfp.rightpin) {
    current_position_ccw_normal = pins_per_side * 3 + 1;
  } else if (sfp.topside && sfp.leftpin) {
    current_position_ccw_normal = pins_per_side * 4;
  }
  pin_map.push(-1);
  for (let i = 0; i < num_pins; i++) {
    pin_map[current_position_ccw_normal] = i + 1;
    if (ccw || !cw) {
      current_position_ccw_normal++;
      if (current_position_ccw_normal > num_pins) {
        current_position_ccw_normal = 1;
      }
    } else {
      current_position_ccw_normal--;
      if (current_position_ccw_normal < 1) {
        current_position_ccw_normal = num_pins;
      }
    }
  }
  return pin_map;
};

// lib/qfn.tsx
import { Fragment as Fragment16, jsx as jsx20, jsxs as jsxs18 } from "react/jsx-runtime";
var QFN = ({
  num_pins = 16,
  bodyWidth = 9,
  bodyLength: bodyLength10 = 9,
  bodyThickness = 0.8,
  thermalPadSize,
  padWidth = 0.25,
  padLength = 0.25,
  pitch = 0.5,
  thermalPadThickness = 0.05
}) => {
  const pin_map = getQuadPinMap({
    num_pins,
    cw: true,
    ccw: true
  });
  const pinPositions = [];
  const spc = num_pins / 4;
  for (let i = 0; i < num_pins; i++) {
    const {
      x,
      y,
      o: orientation
    } = getQuadCoords({
      pin_count: num_pins,
      pn: i + 1,
      w: bodyWidth,
      h: bodyLength10,
      p: pitch,
      pl: padLength,
      legsoutside: false
    });
    let pw = padWidth;
    let pl = padLength;
    if (orientation === "vert") {
      ;
      [pw, pl] = [pl, pw];
    }
    const pn = pin_map[i + 1];
    pinPositions.push({ pn, x, y, pw, pl });
  }
  return /* @__PURE__ */ jsxs18(Fragment16, { children: [
    /* @__PURE__ */ jsx20(Colorize7, { color: "grey", children: /* @__PURE__ */ jsx20(
      Cuboid15,
      {
        center: { x: 0, y: 0, z: bodyThickness / 2 },
        size: [bodyWidth, bodyLength10, bodyThickness]
      }
    ) }),
    pinPositions.map((p, i) => /* @__PURE__ */ jsx20(
      Cuboid15,
      {
        center: { x: p.x, y: p.y, z: thermalPadThickness / 2 },
        size: [p.pw, p.pl, thermalPadThickness]
      },
      i
    )),
    thermalPadSize?.length !== void 0 && thermalPadSize?.width !== void 0 && /* @__PURE__ */ jsx20(
      Cuboid15,
      {
        center: { x: 0, y: 0, z: thermalPadThickness / 2 },
        size: [
          thermalPadSize.width,
          thermalPadSize.length,
          thermalPadThickness
        ]
      }
    )
  ] });
};
var qfn_default = QFN;

// lib/SOT-235.tsx
import { Fragment as Fragment17, jsx as jsx21, jsxs as jsxs19 } from "react/jsx-runtime";
var SOT235 = () => {
  const fullWidth = 2.8;
  const bodyWidth = 1.6;
  const bodyLength10 = 2.9;
  const bodyHeight = 1.2;
  const leadWidth = 0.4;
  const leadThickness = 0.15;
  const leadHeight = 0.95;
  const padContactLength = 0.5;
  const padPitch = 0.95;
  const extendedBodyDistance = fullWidth - bodyWidth;
  return /* @__PURE__ */ jsxs19(Fragment17, { children: [
    /* @__PURE__ */ jsx21(
      SmdChipLead,
      {
        rotation: Math.PI,
        position: {
          x: fullWidth / 2 + extendedBodyDistance / 4,
          y: -1,
          z: leadThickness / 2
        },
        width: leadWidth,
        thickness: leadThickness,
        padContactLength,
        bodyDistance: extendedBodyDistance,
        height: leadHeight
      },
      1
    ),
    /* @__PURE__ */ jsx21(
      SmdChipLead,
      {
        rotation: Math.PI,
        position: {
          x: fullWidth / 2 + extendedBodyDistance / 4,
          y: 1,
          z: leadThickness / 2
        },
        width: leadWidth,
        thickness: leadThickness,
        padContactLength,
        bodyDistance: extendedBodyDistance,
        height: leadHeight
      },
      2
    ),
    /* @__PURE__ */ jsx21(
      SmdChipLead,
      {
        position: {
          x: -fullWidth / 2 - extendedBodyDistance / 4,
          y: 0,
          z: leadThickness / 2
        },
        width: leadWidth,
        thickness: leadThickness,
        padContactLength,
        bodyDistance: extendedBodyDistance,
        height: leadHeight
      },
      3
    ),
    /* @__PURE__ */ jsx21(
      SmdChipLead,
      {
        position: {
          x: -fullWidth / 2 - extendedBodyDistance / 4,
          y: -1,
          z: leadThickness / 2
        },
        width: leadWidth,
        thickness: leadThickness,
        padContactLength,
        bodyDistance: extendedBodyDistance,
        height: leadHeight
      },
      1
    ),
    /* @__PURE__ */ jsx21(
      SmdChipLead,
      {
        position: {
          x: -fullWidth / 2 - extendedBodyDistance / 4,
          y: 1,
          z: leadThickness / 2
        },
        width: leadWidth,
        thickness: leadThickness,
        padContactLength,
        bodyDistance: extendedBodyDistance,
        height: leadHeight
      },
      2
    ),
    /* @__PURE__ */ jsx21(
      ChipBody,
      {
        center: { x: 0, y: 0, z: 0 },
        width: bodyWidth,
        length: bodyLength10,
        height: bodyHeight
      }
    )
  ] });
};
var SOT_235_default = SOT235;

// lib/FemaleHeader.tsx
import { Cuboid as Cuboid16, Colorize as Colorize8, Hull as Hull3, Subtract as Subtract3, Cylinder as Cylinder4 } from "jscad-fiber";
import { Fragment as Fragment18, jsx as jsx22, jsxs as jsxs20 } from "react/jsx-runtime";
var FemaleHeader = ({
  numberOfPins,
  pitch = 2.54,
  legsLength = 3,
  outerDiameter = 0.945,
  innerDiameter = 0.945
}) => {
  const pinThickness = innerDiameter / 1.5;
  const bodyDepth = pinThickness * 2 + outerDiameter;
  const bodyHeight = 5;
  const bodyWidth = (numberOfPins - 1) * pitch + outerDiameter + pitch / 2;
  const gapWidth = pinThickness * 1.6;
  const xoff = -((numberOfPins - 1) / 2) * pitch;
  const Body = /* @__PURE__ */ jsx22(Colorize8, { color: "#1a1a1a", children: /* @__PURE__ */ jsxs20(Subtract3, { children: [
    /* @__PURE__ */ jsx22(
      Cuboid16,
      {
        color: "#000",
        size: [bodyWidth, bodyDepth, bodyHeight],
        center: [0, 0, bodyHeight / 2]
      }
    ),
    Array.from(
      { length: numberOfPins },
      (_, i) => innerDiameter ? /* @__PURE__ */ jsx22(
        Cylinder4,
        {
          height: bodyHeight + 0.1,
          radius: innerDiameter / 2,
          center: [xoff + i * pitch, 0, bodyHeight / 2],
          color: "#222"
        },
        i
      ) : /* @__PURE__ */ jsx22(
        Cuboid16,
        {
          size: [gapWidth, gapWidth, bodyHeight],
          center: [xoff + i * pitch, 0, bodyHeight / 2]
        },
        i
      )
    )
  ] }) });
  return /* @__PURE__ */ jsxs20(Fragment18, { children: [
    Body,
    Array.from({ length: numberOfPins }, (_, i) => /* @__PURE__ */ jsxs20(Colorize8, { color: "silver", children: [
      /* @__PURE__ */ jsxs20(Hull3, { children: [
        /* @__PURE__ */ jsx22(
          Cuboid16,
          {
            color: "silver",
            size: [pinThickness, pinThickness, legsLength * 0.9],
            center: [xoff + i * pitch, 0, -legsLength / 2 * 0.9]
          }
        ),
        /* @__PURE__ */ jsx22(
          Cuboid16,
          {
            color: "silver",
            size: [pinThickness / 1.8, pinThickness / 1.8, legsLength],
            center: [xoff + i * pitch, 0, -legsLength / 2]
          }
        )
      ] }),
      /* @__PURE__ */ jsx22(
        Cuboid16,
        {
          color: "silver",
          size: [gapWidth, gapWidth, gapWidth * 0.5],
          center: [xoff + i * pitch, 0, gapWidth / 2 * 0.5]
        }
      )
    ] }, i))
  ] });
};

// lib/PushButton.tsx
import {
  Colorize as Colorize9,
  Cylinder as Cylinder5,
  ExtrudeLinear as ExtrudeLinear3,
  Polygon as Polygon3,
  Rotate as Rotate5,
  RoundedCuboid,
  Translate as Translate8
} from "jscad-fiber";
import { Fragment as Fragment19, jsx as jsx23, jsxs as jsxs21 } from "react/jsx-runtime";
var PushButton = ({
  width: width10,
  length,
  innerDiameter = 1
}) => {
  const bodyWidth = width10;
  const bodyLength10 = length;
  const bodyHeight = width10 * 0.7;
  const legWidth = innerDiameter / 2.5;
  return /* @__PURE__ */ jsxs21(Fragment19, { children: [
    /* @__PURE__ */ jsx23(
      RoundedCuboid,
      {
        color: "#1a1a1f",
        center: [0, 0, bodyHeight / 2],
        size: [bodyLength10, bodyLength10, bodyHeight],
        roundRadius: 0.3
      }
    ),
    /* @__PURE__ */ jsx23(
      RoundedCuboid,
      {
        color: "#f2f2f2",
        center: [0, 0, bodyHeight + bodyHeight * 0.1 / 2.5],
        size: [bodyLength10, bodyLength10, bodyHeight * 0.1],
        roundRadius: 0.14
      }
    ),
    /* @__PURE__ */ jsx23(
      Cylinder5,
      {
        color: "#1a1a1f",
        height: bodyHeight * 0.8,
        radius: bodyWidth / 3,
        center: [0, 0, bodyHeight + bodyHeight * 0.8 / 2]
      }
    ),
    /* @__PURE__ */ jsx23(
      Cylinder5,
      {
        color: "#1a1a1f",
        height: bodyHeight * 0.2,
        radius: innerDiameter / 2,
        center: [
          bodyLength10 / 3,
          bodyLength10 / 3,
          bodyHeight + bodyHeight * 0.1 / 2
        ]
      }
    ),
    /* @__PURE__ */ jsx23(
      Cylinder5,
      {
        color: "#1a1a1f",
        height: bodyHeight * 0.2,
        radius: innerDiameter / 2,
        center: [
          -bodyLength10 / 3,
          -bodyLength10 / 3,
          bodyHeight + bodyHeight * 0.1 / 2
        ]
      }
    ),
    /* @__PURE__ */ jsx23(
      Cylinder5,
      {
        color: "#1a1a1f",
        height: bodyHeight * 0.2,
        radius: innerDiameter / 2,
        center: [
          -bodyLength10 / 3,
          bodyLength10 / 3,
          bodyHeight + bodyHeight * 0.1 / 2
        ]
      }
    ),
    /* @__PURE__ */ jsx23(
      Cylinder5,
      {
        color: "#1a1a1f",
        height: bodyHeight * 0.2,
        radius: innerDiameter / 2,
        center: [
          bodyLength10 / 3,
          -bodyLength10 / 3,
          bodyHeight + bodyHeight * 0.1 / 2
        ]
      }
    ),
    /* @__PURE__ */ jsx23(
      PushButtonLeg,
      {
        thickness: innerDiameter / 3,
        width: legWidth,
        horizontalLength: bodyLength10 * 0.8,
        verticalLength: bodyHeight / 2,
        position: {
          x: -bodyWidth / 2,
          y: -bodyLength10 / 2,
          z: -bodyHeight * 1.2
        }
      }
    ),
    /* @__PURE__ */ jsx23(
      PushButtonLeg,
      {
        thickness: innerDiameter / 3,
        width: legWidth,
        horizontalLength: bodyLength10 * 0.8,
        verticalLength: bodyHeight / 2,
        position: {
          x: -bodyWidth / 2,
          y: bodyLength10 / 2,
          z: -bodyHeight * 1.2
        },
        rotation: Math.PI
      }
    ),
    /* @__PURE__ */ jsx23(
      PushButtonLeg,
      {
        thickness: innerDiameter / 3,
        width: legWidth,
        horizontalLength: bodyLength10 * 0.8,
        verticalLength: bodyHeight / 2,
        position: { x: bodyWidth / 2, y: bodyLength10 / 2, z: -bodyHeight * 1.2 },
        rotation: Math.PI
      }
    ),
    /* @__PURE__ */ jsx23(
      PushButtonLeg,
      {
        thickness: innerDiameter / 3,
        width: legWidth,
        horizontalLength: bodyLength10 * 0.8,
        verticalLength: bodyHeight / 2,
        position: {
          x: bodyWidth / 2 + innerDiameter / 3.6,
          y: -bodyLength10 / 2,
          z: -bodyHeight * 1.2
        }
      }
    )
  ] });
};
var PushButtonLeg = (props) => {
  const {
    thickness,
    width: width10,
    horizontalLength,
    verticalLength,
    rotation = 0,
    position
  } = props;
  const points = [
    [0, horizontalLength],
    [-verticalLength / 3, horizontalLength / 3],
    [-verticalLength / 5, horizontalLength / 4],
    [0, 0]
  ];
  const polygon = getExpandedStroke(points, thickness);
  return /* @__PURE__ */ jsx23(Colorize9, { color: "#f2f2f2", children: /* @__PURE__ */ jsx23(
    Translate8,
    {
      offset: {
        x: position?.x || 0,
        y: position?.y || 0,
        z: position?.z || 0
      },
      children: /* @__PURE__ */ jsx23(Rotate5, { rotation: [0, 55, rotation], children: /* @__PURE__ */ jsx23(ExtrudeLinear3, { height: width10, children: /* @__PURE__ */ jsx23(Polygon3, { points: polygon.map((p) => [p.y, p.x]) }) }) })
    }
  ) });
};

// lib/SOIC.tsx
import { Fragment as Fragment20, jsx as jsx24, jsxs as jsxs22 } from "react/jsx-runtime";
var SOIC = ({
  pinCount,
  leadLength,
  leadWidth,
  pitch,
  bodyWidth
}) => {
  const sidePinCount = Math.ceil(pinCount / 2);
  const pinOffsetToCenter = (sidePinCount - 1) * pitch / 2;
  const leadThickness = 0.25;
  const bodyHeight = 1;
  const leadHeight = 0.8;
  const leadBodyOffset = leadLength * 0;
  const fullLength10 = pitch * (sidePinCount - 1) + leadWidth + 0.2;
  const bodyWidthAdjusted = bodyWidth * 0.55;
  return /* @__PURE__ */ jsxs22(Fragment20, { children: [
    Array.from({ length: sidePinCount }).map((_, i) => /* @__PURE__ */ jsx24(
      SmdChipLead,
      {
        position: {
          x: -bodyWidth / 2 - leadBodyOffset,
          y: i * pitch - pinOffsetToCenter,
          z: leadThickness / 2
        },
        width: leadWidth,
        thickness: leadThickness,
        padContactLength: leadLength / 2,
        bodyDistance: leadLength + 0.3,
        height: leadHeight
      },
      i
    )),
    Array.from({ length: sidePinCount }).map((_, i) => /* @__PURE__ */ jsx24(
      SmdChipLead,
      {
        rotation: Math.PI,
        position: {
          x: bodyWidth / 2 + leadBodyOffset,
          y: i * pitch - pinOffsetToCenter,
          z: leadThickness / 2
        },
        width: leadWidth,
        thickness: leadThickness,
        padContactLength: leadLength / 2,
        bodyDistance: leadLength + 0.3,
        height: leadHeight
      },
      i
    )),
    /* @__PURE__ */ jsx24(
      ChipBody,
      {
        center: { x: 0, y: 0, z: leadThickness / 2 },
        width: bodyWidthAdjusted,
        length: fullLength10,
        height: bodyHeight
      }
    )
  ] });
};

// lib/VSSOP.tsx
import { Fragment as Fragment21, jsx as jsx25, jsxs as jsxs23 } from "react/jsx-runtime";
var VSSOP = ({
  pinCount,
  pitch,
  leadWidth,
  leadLength,
  bodyWidth,
  bodyLength: bodyLength10
}) => {
  const defaults = pinCount === 8 ? {
    pitch: 0.65,
    leadWidth: 0.3,
    leadLength: 1.6
  } : {
    pitch: 0.5,
    leadWidth: 0.225,
    leadLength: 1.45
  };
  const _pitch = !isNaN(parseFloat(pitch)) ? parseFloat(pitch) : defaults.pitch;
  const _leadWidth = !isNaN(parseFloat(leadWidth)) ? parseFloat(leadWidth) * 0.8 : defaults.leadWidth;
  const _leadLength = !isNaN(parseFloat(leadLength)) ? parseFloat(leadLength) : defaults.leadLength;
  const _bodyWidth = !isNaN(parseFloat(bodyWidth)) ? parseFloat(bodyWidth) * 0.8 : 2.6;
  const _bodyLength = !isNaN(parseFloat(bodyLength10)) ? parseFloat(bodyLength10) : 2 + Number(_pitch) * 1.4;
  const sidePinCount = pinCount / 2;
  const pinOffsetToCenter = (sidePinCount - 1) * _pitch / 2;
  const leadThickness = 0.15;
  const leadHeight = 0.8;
  const componentFullWidth = 4.5;
  const leadBodyDistance = (componentFullWidth - _bodyWidth) / 2;
  const padContactLength = leadBodyDistance * 0.5;
  return /* @__PURE__ */ jsxs23(Fragment21, { children: [
    Array.from({ length: sidePinCount }).map((_, i) => /* @__PURE__ */ jsx25(
      SmdChipLead,
      {
        position: {
          x: -componentFullWidth / 2,
          y: pinOffsetToCenter - i * _pitch,
          z: leadThickness / 2
        },
        width: _leadWidth,
        thickness: leadThickness,
        padContactLength: padContactLength + 0.05,
        bodyDistance: leadBodyDistance + 0.05,
        height: leadHeight
      },
      `left-${i}`
    )),
    Array.from({ length: sidePinCount }).map((_, i) => /* @__PURE__ */ jsx25(
      SmdChipLead,
      {
        rotation: Math.PI,
        position: {
          x: componentFullWidth / 2,
          y: pinOffsetToCenter - i * _pitch,
          z: leadThickness / 2
        },
        width: _leadWidth,
        thickness: leadThickness,
        padContactLength: padContactLength + 0.05,
        bodyDistance: leadBodyDistance + 0.05,
        height: leadHeight
      },
      `right-${i}`
    )),
    /* @__PURE__ */ jsx25(
      ChipBody,
      {
        center: { x: 0, y: 0, z: leadThickness / 2 },
        width: _bodyWidth,
        length: _bodyLength,
        height: 1
      }
    )
  ] });
};

// lib/SOD523.tsx
import { Colorize as Colorize10, Cuboid as Cuboid18, Hull as Hull4, Translate as Translate9, Union as Union5 } from "jscad-fiber";
import { Fragment as Fragment22, jsx as jsx26, jsxs as jsxs24 } from "react/jsx-runtime";
var SOD523 = () => {
  const fullWidth = 2.15;
  const bodyLength10 = 0.8;
  const bodyHeight = 0.6;
  const padWidth = 0.6;
  const padLength = 0.5;
  const padThickness = 0.12;
  const bodyWidth = fullWidth - padLength;
  const leftPadCenterX = -bodyWidth / 2 + padLength / 2 - 0.15;
  const rightPadCenterX = bodyWidth / 2 - padLength / 2 + 0.15;
  const taperOffset = 0.2;
  const straightHeight = bodyHeight * 0.5;
  const Body = /* @__PURE__ */ jsx26(Colorize10, { color: "#222", children: /* @__PURE__ */ jsxs24(Union5, { children: [
    /* @__PURE__ */ jsx26(Translate9, { z: straightHeight / 2, children: /* @__PURE__ */ jsx26(Cuboid18, { size: [bodyWidth, bodyLength10, straightHeight] }) }),
    /* @__PURE__ */ jsxs24(Hull4, { children: [
      /* @__PURE__ */ jsx26(Translate9, { z: straightHeight, children: /* @__PURE__ */ jsx26(Cuboid18, { size: [bodyWidth, bodyLength10, 0.01] }) }),
      /* @__PURE__ */ jsx26(Translate9, { z: bodyHeight, children: /* @__PURE__ */ jsx26(
        Cuboid18,
        {
          size: [bodyWidth - taperOffset, bodyLength10 - taperOffset, 0.01]
        }
      ) })
    ] })
  ] }) });
  return /* @__PURE__ */ jsxs24(Fragment22, { children: [
    /* @__PURE__ */ jsx26(
      Cuboid18,
      {
        color: "#ccc",
        size: [padLength, padWidth, padThickness],
        center: [leftPadCenterX, 0, padThickness / 2]
      }
    ),
    /* @__PURE__ */ jsx26(
      Cuboid18,
      {
        color: "#ccc",
        size: [padLength, padWidth, padThickness],
        center: [rightPadCenterX, 0, padThickness / 2]
      }
    ),
    Body
  ] });
};

// lib/Footprinter3d.tsx
import { jsx as jsx27 } from "react/jsx-runtime";
var Footprinter3d = ({ footprint }) => {
  const fpJson = fp3.string(footprint).json();
  switch (fpJson.fn) {
    case "dip":
      return /* @__PURE__ */ jsx27(Dip, { numPins: fpJson.num_pins, pitch: fpJson.p, bodyWidth: fpJson.w });
    case "tssop":
      return /* @__PURE__ */ jsx27(
        Tssop,
        {
          pinCount: fpJson.num_pins,
          leadLength: fpJson.pl,
          leadWidth: fpJson.pw,
          pitch: fpJson.p,
          bodyWidth: fpJson.w
        }
      );
    case "vssop":
      return /* @__PURE__ */ jsx27(
        VSSOP,
        {
          pinCount: fpJson.num_pins,
          leadLength: fpJson.pl,
          leadWidth: fpJson.pw,
          pitch: fpJson.p,
          bodyWidth: fpJson.w,
          bodyLength: fpJson.h
        }
      );
    case "qfp":
      return /* @__PURE__ */ jsx27(
        QFP,
        {
          pinCount: fpJson.num_pins,
          pitch: fpJson.p,
          leadWidth: fpJson.pw,
          padContactLength: fpJson.pl,
          bodyWidth: fpJson.w
        }
      );
    case "qfn":
      return /* @__PURE__ */ jsx27(
        qfn_default,
        {
          num_pins: fpJson.num_pins,
          bodyWidth: fpJson.w,
          bodyLength: fpJson.h,
          pitch: fpJson.p,
          padLength: fpJson.pl,
          padWidth: fpJson.pw,
          thermalPadSize: {
            width: fpJson.thermalpad.x,
            length: fpJson.thermalpad.y
          }
        }
      );
    case "pinrow":
      if (fpJson.male)
        return /* @__PURE__ */ jsx27(PinRow, { numberOfPins: fpJson.num_pins, pitch: fpJson.p });
      if (fpJson.female)
        return /* @__PURE__ */ jsx27(FemaleHeader, { numberOfPins: fpJson.num_pins, pitch: fpJson.p });
    case "cap": {
      switch (fpJson.imperial) {
        case "0402":
          return /* @__PURE__ */ jsx27(A0402, { color: "#856c4d" });
        case "0603":
          return /* @__PURE__ */ jsx27(A0603, { color: "#856c4d" });
        case "0805":
          return /* @__PURE__ */ jsx27(A0805, { color: "#856c4d" });
        case "0201":
          return /* @__PURE__ */ jsx27(A0201, { color: "#856c4d" });
        case "01005":
          return /* @__PURE__ */ jsx27(A01005, { color: "#856c4d" });
        case "1206":
          return /* @__PURE__ */ jsx27(A1206, { color: "#856c4d" });
        case "1210":
          return /* @__PURE__ */ jsx27(A1210, { color: "#856c4d" });
        case "2010":
          return /* @__PURE__ */ jsx27(A2010, { color: "#856c4d" });
        case "2512":
          return /* @__PURE__ */ jsx27(A2512, { color: "#856c4d" });
      }
    }
    case "sot235":
      return /* @__PURE__ */ jsx27(SOT_235_default, {});
    case "pushbutton":
      return /* @__PURE__ */ jsx27(
        PushButton,
        {
          width: fpJson.w,
          length: fpJson.h,
          innerDiameter: fpJson.id
        }
      );
    case "soic":
      return /* @__PURE__ */ jsx27(
        SOIC,
        {
          pinCount: fpJson.num_pins,
          leadLength: fpJson.pl,
          leadWidth: fpJson.pw,
          pitch: fpJson.p,
          bodyWidth: fpJson.w
        }
      );
    case "sod523":
      return /* @__PURE__ */ jsx27(SOD523, {});
  }
  const colorMatch = footprint.match(/_color\(([^)]+)\)/);
  const color = colorMatch ? colorMatch[1] : void 0;
  switch (fpJson.imperial) {
    case "0402":
      return /* @__PURE__ */ jsx27(A0402, { color });
    case "0603":
      return /* @__PURE__ */ jsx27(A0603, { color });
    case "0805":
      return /* @__PURE__ */ jsx27(A0805, { color });
    case "0201":
      return /* @__PURE__ */ jsx27(A0201, { color });
    case "01005":
      return /* @__PURE__ */ jsx27(A01005, { color });
    case "1206":
      return /* @__PURE__ */ jsx27(A1206, { color });
    case "1210":
      return /* @__PURE__ */ jsx27(A1210, { color });
    case "2010":
      return /* @__PURE__ */ jsx27(A2010, { color });
    case "2512":
      return /* @__PURE__ */ jsx27(A2512, { color });
  }
  return null;
};

// lib/SOT-23-3P.tsx
import { Fragment as Fragment23, jsx as jsx28, jsxs as jsxs25 } from "react/jsx-runtime";
var SOT233P = ({ fullWidth = 2.9, fullLength: fullLength10 = 2.8 }) => {
  const bodyWidth = 1.3;
  const bodyLength10 = 2.9;
  const bodyHeight = 1.1;
  const leadWidth = 0.4;
  const leadThickness = 0.15;
  const leadHeight = 0.95;
  const padContactLength = 0.4;
  const padThickness = leadThickness / 2;
  const extendedBodyDistance = (fullWidth - bodyWidth) / 2 + 0.3;
  return /* @__PURE__ */ jsxs25(Fragment23, { children: [
    /* @__PURE__ */ jsx28(
      SmdChipLead,
      {
        rotation: Math.PI,
        position: {
          x: fullWidth / 2,
          y: -1,
          z: padThickness
        },
        width: leadWidth,
        thickness: leadThickness,
        padContactLength,
        bodyDistance: extendedBodyDistance,
        height: leadHeight
      },
      1
    ),
    /* @__PURE__ */ jsx28(
      SmdChipLead,
      {
        rotation: Math.PI,
        position: {
          x: fullWidth / 2,
          y: 1,
          z: padThickness
        },
        width: leadWidth,
        thickness: leadThickness,
        padContactLength,
        bodyDistance: extendedBodyDistance,
        height: leadHeight
      },
      2
    ),
    /* @__PURE__ */ jsx28(
      SmdChipLead,
      {
        position: {
          x: -fullWidth / 2,
          y: 0,
          z: padThickness
        },
        width: leadWidth,
        thickness: leadThickness,
        padContactLength,
        bodyDistance: extendedBodyDistance,
        height: leadHeight
      },
      3
    ),
    /* @__PURE__ */ jsx28(
      ChipBody,
      {
        center: { x: 0, y: 0, z: 0 },
        width: bodyWidth,
        length: bodyLength10,
        height: bodyHeight
      }
    )
  ] });
};

// lib/SOT-563.tsx
import { Cuboid as Cuboid19, Translate as Translate10, Rotate as Rotate6, Colorize as Colorize11 } from "jscad-fiber";
import { Fragment as Fragment24, jsx as jsx29, jsxs as jsxs26 } from "react/jsx-runtime";
var SOT563 = ({ fullWidth = 1.94, fullLength: fullLength10 = 1.6 }) => {
  const bodyWidth = 1.2;
  const bodyLength10 = 1.6;
  const bodyHeight = 0.55;
  const leadWidth = 0.3;
  const leadLength = 0.67;
  const leadHeight = 0.13;
  const leadSpacing = 0.5;
  const bodyZOffset = -0.4;
  return /* @__PURE__ */ jsxs26(Fragment24, { children: [
    /* @__PURE__ */ jsx29(Rotate6, { rotation: [45 * Math.PI, 0, 0], children: /* @__PURE__ */ jsx29(Translate10, { center: [0, 0, bodyZOffset], children: /* @__PURE__ */ jsx29(Colorize11, { color: "grey", children: /* @__PURE__ */ jsx29(Cuboid19, { size: [bodyWidth, bodyLength10, bodyHeight] }) }) }) }),
    [-1, 0, 1].flatMap((yOffset, index) => [
      // Left lead
      /* @__PURE__ */ jsx29(
        Translate10,
        {
          center: [
            -bodyWidth / 2 - 0.03,
            yOffset * leadSpacing,
            leadHeight / 2
          ],
          children: /* @__PURE__ */ jsx29(Cuboid19, { size: [leadLength, leadWidth, leadHeight] })
        },
        `left-${index}`
      ),
      // Right lead
      /* @__PURE__ */ jsx29(
        Translate10,
        {
          center: [bodyWidth / 2 + 0.03, yOffset * leadSpacing, leadHeight / 2],
          children: /* @__PURE__ */ jsx29(Cuboid19, { size: [leadLength, leadWidth, leadHeight] })
        },
        `right-${index}`
      )
    ])
  ] });
};

// lib/SOT-723.tsx
import { Cuboid as Cuboid20, Translate as Translate11, Rotate as Rotate7, Colorize as Colorize12 } from "jscad-fiber";
import { Fragment as Fragment25, jsx as jsx30, jsxs as jsxs27 } from "react/jsx-runtime";
var getCcwSot723Coords = (pn) => {
  if (pn === 1) {
    return { x: 0, y: 0 };
  } else if (pn === 2) {
    return { x: 1, y: -0.4 };
  } else {
    return { x: 1, y: 0.4 };
  }
};
var SOT723 = () => {
  const bodyWidth = 0.8;
  const bodyLength10 = 1.2;
  const bodyHeight = 0.5;
  const leadWidth = 0.32;
  const leadLength = 0.3;
  const leadHeight = 0.1;
  const centerLeadWidth = 0.42;
  return /* @__PURE__ */ jsxs27(Fragment25, { children: [
    /* @__PURE__ */ jsx30(Rotate7, { rotation: [45 * Math.PI, 0, 0], children: /* @__PURE__ */ jsx30(Translate11, { center: [0.475, leadHeight / 2, -0.25], children: /* @__PURE__ */ jsx30(Colorize12, { color: "grey", children: /* @__PURE__ */ jsx30(Cuboid20, { size: [bodyWidth, bodyLength10, bodyHeight] }) }) }) }),
    [1, 2, 3].map((pn) => {
      const { x, y } = getCcwSot723Coords(pn);
      return /* @__PURE__ */ jsx30(Translate11, { center: [x, y, 0], children: /* @__PURE__ */ jsx30(
        Cuboid20,
        {
          size: [
            leadLength,
            pn === 1 ? centerLeadWidth : leadWidth,
            leadHeight
          ]
        }
      ) }, `lead-${pn}`);
    })
  ] });
};

// lib/sod-123.tsx
import { Fragment as Fragment26, jsx as jsx31, jsxs as jsxs28 } from "react/jsx-runtime";
var SOD123 = ({ fullWidth = 3.8, fullLength: fullLength10 = 1.6 }) => {
  const bodyWidth = 2.9;
  const bodyLength10 = 1.3;
  const bodyHeight = 1.1;
  const leadWidth = 0.8;
  const leadThickness = 0.15;
  const leadHeight = 0.5;
  const padContactLength = 0.4;
  const padThickness = leadThickness / 2;
  const bodyDistance = (fullWidth - bodyWidth) / 2;
  return /* @__PURE__ */ jsxs28(Fragment26, { children: [
    /* @__PURE__ */ jsx31(
      SmdChipLead,
      {
        position: {
          x: -fullWidth / 2 + leadWidth / 2,
          y: 0,
          z: padThickness
        },
        width: leadWidth,
        thickness: leadThickness,
        padContactLength,
        bodyDistance,
        height: leadHeight
      },
      1
    ),
    /* @__PURE__ */ jsx31(
      SmdChipLead,
      {
        rotation: Math.PI,
        position: {
          x: fullWidth / 2 - leadWidth / 2,
          y: 0,
          z: padThickness
        },
        width: leadWidth,
        thickness: leadThickness,
        padContactLength,
        bodyDistance,
        height: leadHeight
      },
      2
    ),
    /* @__PURE__ */ jsx31(
      ChipBody,
      {
        center: { x: 0, y: 0, z: 0 },
        width: bodyWidth,
        length: bodyLength10,
        height: bodyHeight
      }
    )
  ] });
};
export {
  A01005,
  A0201,
  A0402,
  A0603,
  A0805,
  A1206,
  A1210,
  A2010,
  A2512,
  BGA,
  ChipBody,
  ExtrudedPads,
  FootprintPad,
  FootprintPlatedHole,
  Footprinter3d,
  PinRow,
  QFN,
  QFP,
  SOD123,
  SOD523,
  SOT233P,
  SOT563,
  SOT723,
  SmdChipLead,
  Tssop,
  VSSOP
};
