// @tscircuit/common/dist/index.js
var __create = Object.create;
var __defProp = Object.defineProperty;
var __defProps = Object.defineProperties;
var __getOwnPropDesc = Object.getOwnPropertyDescriptor;
var __getOwnPropDescs = Object.getOwnPropertyDescriptors;
var __getOwnPropNames = Object.getOwnPropertyNames;
var __getOwnPropSymbols = Object.getOwnPropertySymbols;
var __getProtoOf = Object.getPrototypeOf;
var __hasOwnProp = Object.prototype.hasOwnProperty;
var __propIsEnum = Object.prototype.propertyIsEnumerable;
var __pow = Math.pow;
var __defNormalProp = (obj, key, value) => key in obj ? __defProp(obj, key, { enumerable: true, configurable: true, writable: true, value }) : obj[key] = value;
var __spreadValues = (a, b) => {
  for (var prop in b || (b = {}))
    if (__hasOwnProp.call(b, prop))
      __defNormalProp(a, prop, b[prop]);
  if (__getOwnPropSymbols)
    for (var prop of __getOwnPropSymbols(b)) {
      if (__propIsEnum.call(b, prop))
        __defNormalProp(a, prop, b[prop]);
    }
  return a;
};
var __spreadProps = (a, b) => __defProps(a, __getOwnPropDescs(b));
var __objRest = (source, exclude) => {
  var target = {};
  for (var prop in source)
    if (__hasOwnProp.call(source, prop) && exclude.indexOf(prop) < 0)
      target[prop] = source[prop];
  if (source != null && __getOwnPropSymbols)
    for (var prop of __getOwnPropSymbols(source)) {
      if (exclude.indexOf(prop) < 0 && __propIsEnum.call(source, prop))
        target[prop] = source[prop];
    }
  return target;
};
var __commonJS = (cb, mod) => function __require() {
  return mod || (0, cb[__getOwnPropNames(cb)[0]])((mod = { exports: {} }).exports, mod), mod.exports;
};
var __copyProps = (to, from, except, desc) => {
  if (from && typeof from === "object" || typeof from === "function") {
    for (let key of __getOwnPropNames(from))
      if (!__hasOwnProp.call(to, key) && key !== except)
        __defProp(to, key, { get: () => from[key], enumerable: !(desc = __getOwnPropDesc(from, key)) || desc.enumerable });
  }
  return to;
};
var __toESM = (mod, isNodeMode, target) => (target = mod != null ? __create(__getProtoOf(mod)) : {}, __copyProps(
  // If the importer is in node compatibility mode or this is not an ESM
  // file that has been converted to a CommonJS file using a Babel-
  // compatible transform (i.e. "__esModule" has not been set), then set
  // "default" to the CommonJS "module.exports" for node compatibility.
  isNodeMode || !mod || !mod.__esModule ? __defProp(target, "default", { value: mod, enumerable: true }) : target,
  mod
));

// node_modules/react/cjs/react-jsx-runtime.production.js
var require_react_jsx_runtime_production = __commonJS({
  "node_modules/react/cjs/react-jsx-runtime.production.js"(exports) {
    "use strict";
    var REACT_ELEMENT_TYPE = Symbol.for("react.transitional.element");
    var REACT_FRAGMENT_TYPE = Symbol.for("react.fragment");
    function jsxProd(type, config, maybeKey) {
      var key = null;
      void 0 !== maybeKey && (key = "" + maybeKey);
      void 0 !== config.key && (key = "" + config.key);
      if ("key" in config) {
        maybeKey = {};
        for (var propName in config)
          "key" !== propName && (maybeKey[propName] = config[propName]);
      } else maybeKey = config;
      config = maybeKey.ref;
      return {
        $$typeof: REACT_ELEMENT_TYPE,
        type,
        key,
        ref: void 0 !== config ? config : null,
        props: maybeKey
      };
    }
    exports.Fragment = REACT_FRAGMENT_TYPE;
    exports.jsx = jsxProd;
    exports.jsxs = jsxProd;
  }
});

// node_modules/react/cjs/react.production.js
var require_react_production = __commonJS({
  "node_modules/react/cjs/react.production.js"(exports) {
    "use strict";
    var REACT_ELEMENT_TYPE = Symbol.for("react.transitional.element");
    var REACT_PORTAL_TYPE = Symbol.for("react.portal");
    var REACT_FRAGMENT_TYPE = Symbol.for("react.fragment");
    var REACT_STRICT_MODE_TYPE = Symbol.for("react.strict_mode");
    var REACT_PROFILER_TYPE = Symbol.for("react.profiler");
    var REACT_CONSUMER_TYPE = Symbol.for("react.consumer");
    var REACT_CONTEXT_TYPE = Symbol.for("react.context");
    var REACT_FORWARD_REF_TYPE = Symbol.for("react.forward_ref");
    var REACT_SUSPENSE_TYPE = Symbol.for("react.suspense");
    var REACT_MEMO_TYPE = Symbol.for("react.memo");
    var REACT_LAZY_TYPE = Symbol.for("react.lazy");
    var REACT_ACTIVITY_TYPE = Symbol.for("react.activity");
    var MAYBE_ITERATOR_SYMBOL = Symbol.iterator;
    function getIteratorFn(maybeIterable) {
      if (null === maybeIterable || "object" !== typeof maybeIterable) return null;
      maybeIterable = MAYBE_ITERATOR_SYMBOL && maybeIterable[MAYBE_ITERATOR_SYMBOL] || maybeIterable["@@iterator"];
      return "function" === typeof maybeIterable ? maybeIterable : null;
    }
    var ReactNoopUpdateQueue = {
      isMounted: function() {
        return false;
      },
      enqueueForceUpdate: function() {
      },
      enqueueReplaceState: function() {
      },
      enqueueSetState: function() {
      }
    };
    var assign = Object.assign;
    var emptyObject = {};
    function Component(props, context, updater) {
      this.props = props;
      this.context = context;
      this.refs = emptyObject;
      this.updater = updater || ReactNoopUpdateQueue;
    }
    Component.prototype.isReactComponent = {};
    Component.prototype.setState = function(partialState, callback) {
      if ("object" !== typeof partialState && "function" !== typeof partialState && null != partialState)
        throw Error(
          "takes an object of state variables to update or a function which returns an object of state variables."
        );
      this.updater.enqueueSetState(this, partialState, callback, "setState");
    };
    Component.prototype.forceUpdate = function(callback) {
      this.updater.enqueueForceUpdate(this, callback, "forceUpdate");
    };
    function ComponentDummy() {
    }
    ComponentDummy.prototype = Component.prototype;
    function PureComponent(props, context, updater) {
      this.props = props;
      this.context = context;
      this.refs = emptyObject;
      this.updater = updater || ReactNoopUpdateQueue;
    }
    var pureComponentPrototype = PureComponent.prototype = new ComponentDummy();
    pureComponentPrototype.constructor = PureComponent;
    assign(pureComponentPrototype, Component.prototype);
    pureComponentPrototype.isPureReactComponent = true;
    var isArrayImpl = Array.isArray;
    function noop() {
    }
    var ReactSharedInternals = { H: null, A: null, T: null, S: null };
    var hasOwnProperty = Object.prototype.hasOwnProperty;
    function ReactElement4(type, key, props) {
      var refProp = props.ref;
      return {
        $$typeof: REACT_ELEMENT_TYPE,
        type,
        key,
        ref: void 0 !== refProp ? refProp : null,
        props
      };
    }
    function cloneAndReplaceKey(oldElement, newKey) {
      return ReactElement4(oldElement.type, newKey, oldElement.props);
    }
    function isValidElement(object) {
      return "object" === typeof object && null !== object && object.$$typeof === REACT_ELEMENT_TYPE;
    }
    function escape(key) {
      var escaperLookup = { "=": "=0", ":": "=2" };
      return "$" + key.replace(/[=:]/g, function(match) {
        return escaperLookup[match];
      });
    }
    var userProvidedKeyEscapeRegex = /\/+/g;
    function getElementKey(element, index) {
      return "object" === typeof element && null !== element && null != element.key ? escape("" + element.key) : index.toString(36);
    }
    function resolveThenable(thenable) {
      switch (thenable.status) {
        case "fulfilled":
          return thenable.value;
        case "rejected":
          throw thenable.reason;
        default:
          switch ("string" === typeof thenable.status ? thenable.then(noop, noop) : (thenable.status = "pending", thenable.then(
            function(fulfilledValue) {
              "pending" === thenable.status && (thenable.status = "fulfilled", thenable.value = fulfilledValue);
            },
            function(error) {
              "pending" === thenable.status && (thenable.status = "rejected", thenable.reason = error);
            }
          )), thenable.status) {
            case "fulfilled":
              return thenable.value;
            case "rejected":
              throw thenable.reason;
          }
      }
      throw thenable;
    }
    function mapIntoArray(children, array, escapedPrefix, nameSoFar, callback) {
      var type = typeof children;
      if ("undefined" === type || "boolean" === type) children = null;
      var invokeCallback = false;
      if (null === children) invokeCallback = true;
      else
        switch (type) {
          case "bigint":
          case "string":
          case "number":
            invokeCallback = true;
            break;
          case "object":
            switch (children.$$typeof) {
              case REACT_ELEMENT_TYPE:
              case REACT_PORTAL_TYPE:
                invokeCallback = true;
                break;
              case REACT_LAZY_TYPE:
                return invokeCallback = children._init, mapIntoArray(
                  invokeCallback(children._payload),
                  array,
                  escapedPrefix,
                  nameSoFar,
                  callback
                );
            }
        }
      if (invokeCallback)
        return callback = callback(children), invokeCallback = "" === nameSoFar ? "." + getElementKey(children, 0) : nameSoFar, isArrayImpl(callback) ? (escapedPrefix = "", null != invokeCallback && (escapedPrefix = invokeCallback.replace(userProvidedKeyEscapeRegex, "$&/") + "/"), mapIntoArray(callback, array, escapedPrefix, "", function(c) {
          return c;
        })) : null != callback && (isValidElement(callback) && (callback = cloneAndReplaceKey(
          callback,
          escapedPrefix + (null == callback.key || children && children.key === callback.key ? "" : ("" + callback.key).replace(
            userProvidedKeyEscapeRegex,
            "$&/"
          ) + "/") + invokeCallback
        )), array.push(callback)), 1;
      invokeCallback = 0;
      var nextNamePrefix = "" === nameSoFar ? "." : nameSoFar + ":";
      if (isArrayImpl(children))
        for (var i = 0; i < children.length; i++)
          nameSoFar = children[i], type = nextNamePrefix + getElementKey(nameSoFar, i), invokeCallback += mapIntoArray(
            nameSoFar,
            array,
            escapedPrefix,
            type,
            callback
          );
      else if (i = getIteratorFn(children), "function" === typeof i)
        for (children = i.call(children), i = 0; !(nameSoFar = children.next()).done; )
          nameSoFar = nameSoFar.value, type = nextNamePrefix + getElementKey(nameSoFar, i++), invokeCallback += mapIntoArray(
            nameSoFar,
            array,
            escapedPrefix,
            type,
            callback
          );
      else if ("object" === type) {
        if ("function" === typeof children.then)
          return mapIntoArray(
            resolveThenable(children),
            array,
            escapedPrefix,
            nameSoFar,
            callback
          );
        array = String(children);
        throw Error(
          "Objects are not valid as a React child (found: " + ("[object Object]" === array ? "object with keys {" + Object.keys(children).join(", ") + "}" : array) + "). If you meant to render a collection of children, use an array instead."
        );
      }
      return invokeCallback;
    }
    function mapChildren(children, func, context) {
      if (null == children) return children;
      var result = [], count = 0;
      mapIntoArray(children, result, "", "", function(child) {
        return func.call(context, child, count++);
      });
      return result;
    }
    function lazyInitializer(payload) {
      if (-1 === payload._status) {
        var ctor = payload._result;
        ctor = ctor();
        ctor.then(
          function(moduleObject) {
            if (0 === payload._status || -1 === payload._status)
              payload._status = 1, payload._result = moduleObject;
          },
          function(error) {
            if (0 === payload._status || -1 === payload._status)
              payload._status = 2, payload._result = error;
          }
        );
        -1 === payload._status && (payload._status = 0, payload._result = ctor);
      }
      if (1 === payload._status) return payload._result.default;
      throw payload._result;
    }
    var reportGlobalError = "function" === typeof reportError ? reportError : function(error) {
      if ("object" === typeof window && "function" === typeof window.ErrorEvent) {
        var event = new window.ErrorEvent("error", {
          bubbles: true,
          cancelable: true,
          message: "object" === typeof error && null !== error && "string" === typeof error.message ? String(error.message) : String(error),
          error
        });
        if (!window.dispatchEvent(event)) return;
      } else if ("object" === typeof process && "function" === typeof process.emit) {
        process.emit("uncaughtException", error);
        return;
      }
      console.error(error);
    };
    var Children = {
      map: mapChildren,
      forEach: function(children, forEachFunc, forEachContext) {
        mapChildren(
          children,
          function() {
            forEachFunc.apply(this, arguments);
          },
          forEachContext
        );
      },
      count: function(children) {
        var n = 0;
        mapChildren(children, function() {
          n++;
        });
        return n;
      },
      toArray: function(children) {
        return mapChildren(children, function(child) {
          return child;
        }) || [];
      },
      only: function(children) {
        if (!isValidElement(children))
          throw Error(
            "React.Children.only expected to receive a single React element child."
          );
        return children;
      }
    };
    exports.Activity = REACT_ACTIVITY_TYPE;
    exports.Children = Children;
    exports.Component = Component;
    exports.Fragment = REACT_FRAGMENT_TYPE;
    exports.Profiler = REACT_PROFILER_TYPE;
    exports.PureComponent = PureComponent;
    exports.StrictMode = REACT_STRICT_MODE_TYPE;
    exports.Suspense = REACT_SUSPENSE_TYPE;
    exports.__CLIENT_INTERNALS_DO_NOT_USE_OR_WARN_USERS_THEY_CANNOT_UPGRADE = ReactSharedInternals;
    exports.__COMPILER_RUNTIME = {
      __proto__: null,
      c: function(size) {
        return ReactSharedInternals.H.useMemoCache(size);
      }
    };
    exports.cache = function(fn) {
      return function() {
        return fn.apply(null, arguments);
      };
    };
    exports.cacheSignal = function() {
      return null;
    };
    exports.cloneElement = function(element, config, children) {
      if (null === element || void 0 === element)
        throw Error(
          "The argument must be a React element, but you passed " + element + "."
        );
      var props = assign({}, element.props), key = element.key;
      if (null != config)
        for (propName in void 0 !== config.key && (key = "" + config.key), config)
          !hasOwnProperty.call(config, propName) || "key" === propName || "__self" === propName || "__source" === propName || "ref" === propName && void 0 === config.ref || (props[propName] = config[propName]);
      var propName = arguments.length - 2;
      if (1 === propName) props.children = children;
      else if (1 < propName) {
        for (var childArray = Array(propName), i = 0; i < propName; i++)
          childArray[i] = arguments[i + 2];
        props.children = childArray;
      }
      return ReactElement4(element.type, key, props);
    };
    exports.createContext = function(defaultValue) {
      defaultValue = {
        $$typeof: REACT_CONTEXT_TYPE,
        _currentValue: defaultValue,
        _currentValue2: defaultValue,
        _threadCount: 0,
        Provider: null,
        Consumer: null
      };
      defaultValue.Provider = defaultValue;
      defaultValue.Consumer = {
        $$typeof: REACT_CONSUMER_TYPE,
        _context: defaultValue
      };
      return defaultValue;
    };
    exports.createElement = function(type, config, children) {
      var propName, props = {}, key = null;
      if (null != config)
        for (propName in void 0 !== config.key && (key = "" + config.key), config)
          hasOwnProperty.call(config, propName) && "key" !== propName && "__self" !== propName && "__source" !== propName && (props[propName] = config[propName]);
      var childrenLength = arguments.length - 2;
      if (1 === childrenLength) props.children = children;
      else if (1 < childrenLength) {
        for (var childArray = Array(childrenLength), i = 0; i < childrenLength; i++)
          childArray[i] = arguments[i + 2];
        props.children = childArray;
      }
      if (type && type.defaultProps)
        for (propName in childrenLength = type.defaultProps, childrenLength)
          void 0 === props[propName] && (props[propName] = childrenLength[propName]);
      return ReactElement4(type, key, props);
    };
    exports.createRef = function() {
      return { current: null };
    };
    exports.forwardRef = function(render) {
      return { $$typeof: REACT_FORWARD_REF_TYPE, render };
    };
    exports.isValidElement = isValidElement;
    exports.lazy = function(ctor) {
      return {
        $$typeof: REACT_LAZY_TYPE,
        _payload: { _status: -1, _result: ctor },
        _init: lazyInitializer
      };
    };
    exports.memo = function(type, compare) {
      return {
        $$typeof: REACT_MEMO_TYPE,
        type,
        compare: void 0 === compare ? null : compare
      };
    };
    exports.startTransition = function(scope) {
      var prevTransition = ReactSharedInternals.T, currentTransition = {};
      ReactSharedInternals.T = currentTransition;
      try {
        var returnValue = scope(), onStartTransitionFinish = ReactSharedInternals.S;
        null !== onStartTransitionFinish && onStartTransitionFinish(currentTransition, returnValue);
        "object" === typeof returnValue && null !== returnValue && "function" === typeof returnValue.then && returnValue.then(noop, reportGlobalError);
      } catch (error) {
        reportGlobalError(error);
      } finally {
        null !== prevTransition && null !== currentTransition.types && (prevTransition.types = currentTransition.types), ReactSharedInternals.T = prevTransition;
      }
    };
    exports.unstable_useCacheRefresh = function() {
      return ReactSharedInternals.H.useCacheRefresh();
    };
    exports.use = function(usable) {
      return ReactSharedInternals.H.use(usable);
    };
    exports.useActionState = function(action, initialState, permalink) {
      return ReactSharedInternals.H.useActionState(action, initialState, permalink);
    };
    exports.useCallback = function(callback, deps) {
      return ReactSharedInternals.H.useCallback(callback, deps);
    };
    exports.useContext = function(Context) {
      return ReactSharedInternals.H.useContext(Context);
    };
    exports.useDebugValue = function() {
    };
    exports.useDeferredValue = function(value, initialValue) {
      return ReactSharedInternals.H.useDeferredValue(value, initialValue);
    };
    exports.useEffect = function(create, deps) {
      return ReactSharedInternals.H.useEffect(create, deps);
    };
    exports.useEffectEvent = function(callback) {
      return ReactSharedInternals.H.useEffectEvent(callback);
    };
    exports.useId = function() {
      return ReactSharedInternals.H.useId();
    };
    exports.useImperativeHandle = function(ref, create, deps) {
      return ReactSharedInternals.H.useImperativeHandle(ref, create, deps);
    };
    exports.useInsertionEffect = function(create, deps) {
      return ReactSharedInternals.H.useInsertionEffect(create, deps);
    };
    exports.useLayoutEffect = function(create, deps) {
      return ReactSharedInternals.H.useLayoutEffect(create, deps);
    };
    exports.useMemo = function(create, deps) {
      return ReactSharedInternals.H.useMemo(create, deps);
    };
    exports.useOptimistic = function(passthrough, reducer) {
      return ReactSharedInternals.H.useOptimistic(passthrough, reducer);
    };
    exports.useReducer = function(reducer, initialArg, init) {
      return ReactSharedInternals.H.useReducer(reducer, initialArg, init);
    };
    exports.useRef = function(initialValue) {
      return ReactSharedInternals.H.useRef(initialValue);
    };
    exports.useState = function(initialState) {
      return ReactSharedInternals.H.useState(initialState);
    };
    exports.useSyncExternalStore = function(subscribe, getSnapshot, getServerSnapshot) {
      return ReactSharedInternals.H.useSyncExternalStore(
        subscribe,
        getSnapshot,
        getServerSnapshot
      );
    };
    exports.useTransition = function() {
      return ReactSharedInternals.H.useTransition();
    };
    exports.version = "19.2.0";
  }
});

// node_modules/react/cjs/react.development.js
var require_react_development = __commonJS({
  "node_modules/react/cjs/react.development.js"(exports, module) {
    "use strict";
    "production" !== process.env.NODE_ENV && (function() {
      function defineDeprecationWarning(methodName, info) {
        Object.defineProperty(Component.prototype, methodName, {
          get: function() {
            console.warn(
              "%s(...) is deprecated in plain JavaScript React classes. %s",
              info[0],
              info[1]
            );
          }
        });
      }
      function getIteratorFn(maybeIterable) {
        if (null === maybeIterable || "object" !== typeof maybeIterable)
          return null;
        maybeIterable = MAYBE_ITERATOR_SYMBOL && maybeIterable[MAYBE_ITERATOR_SYMBOL] || maybeIterable["@@iterator"];
        return "function" === typeof maybeIterable ? maybeIterable : null;
      }
      function warnNoop(publicInstance, callerName) {
        publicInstance = (publicInstance = publicInstance.constructor) && (publicInstance.displayName || publicInstance.name) || "ReactClass";
        var warningKey = publicInstance + "." + callerName;
        didWarnStateUpdateForUnmountedComponent[warningKey] || (console.error(
          "Can't call %s on a component that is not yet mounted. This is a no-op, but it might indicate a bug in your application. Instead, assign to `this.state` directly or define a `state = {};` class property with the desired state in the %s component.",
          callerName,
          publicInstance
        ), didWarnStateUpdateForUnmountedComponent[warningKey] = true);
      }
      function Component(props, context, updater) {
        this.props = props;
        this.context = context;
        this.refs = emptyObject;
        this.updater = updater || ReactNoopUpdateQueue;
      }
      function ComponentDummy() {
      }
      function PureComponent(props, context, updater) {
        this.props = props;
        this.context = context;
        this.refs = emptyObject;
        this.updater = updater || ReactNoopUpdateQueue;
      }
      function noop() {
      }
      function testStringCoercion(value) {
        return "" + value;
      }
      function checkKeyStringCoercion(value) {
        try {
          testStringCoercion(value);
          var JSCompiler_inline_result = false;
        } catch (e) {
          JSCompiler_inline_result = true;
        }
        if (JSCompiler_inline_result) {
          JSCompiler_inline_result = console;
          var JSCompiler_temp_const = JSCompiler_inline_result.error;
          var JSCompiler_inline_result$jscomp$0 = "function" === typeof Symbol && Symbol.toStringTag && value[Symbol.toStringTag] || value.constructor.name || "Object";
          JSCompiler_temp_const.call(
            JSCompiler_inline_result,
            "The provided key is an unsupported type %s. This value must be coerced to a string before using it here.",
            JSCompiler_inline_result$jscomp$0
          );
          return testStringCoercion(value);
        }
      }
      function getComponentNameFromType(type) {
        if (null == type) return null;
        if ("function" === typeof type)
          return type.$$typeof === REACT_CLIENT_REFERENCE ? null : type.displayName || type.name || null;
        if ("string" === typeof type) return type;
        switch (type) {
          case REACT_FRAGMENT_TYPE:
            return "Fragment";
          case REACT_PROFILER_TYPE:
            return "Profiler";
          case REACT_STRICT_MODE_TYPE:
            return "StrictMode";
          case REACT_SUSPENSE_TYPE:
            return "Suspense";
          case REACT_SUSPENSE_LIST_TYPE:
            return "SuspenseList";
          case REACT_ACTIVITY_TYPE:
            return "Activity";
        }
        if ("object" === typeof type)
          switch ("number" === typeof type.tag && console.error(
            "Received an unexpected object in getComponentNameFromType(). This is likely a bug in React. Please file an issue."
          ), type.$$typeof) {
            case REACT_PORTAL_TYPE:
              return "Portal";
            case REACT_CONTEXT_TYPE:
              return type.displayName || "Context";
            case REACT_CONSUMER_TYPE:
              return (type._context.displayName || "Context") + ".Consumer";
            case REACT_FORWARD_REF_TYPE:
              var innerType = type.render;
              type = type.displayName;
              type || (type = innerType.displayName || innerType.name || "", type = "" !== type ? "ForwardRef(" + type + ")" : "ForwardRef");
              return type;
            case REACT_MEMO_TYPE:
              return innerType = type.displayName || null, null !== innerType ? innerType : getComponentNameFromType(type.type) || "Memo";
            case REACT_LAZY_TYPE:
              innerType = type._payload;
              type = type._init;
              try {
                return getComponentNameFromType(type(innerType));
              } catch (x) {
              }
          }
        return null;
      }
      function getTaskName(type) {
        if (type === REACT_FRAGMENT_TYPE) return "<>";
        if ("object" === typeof type && null !== type && type.$$typeof === REACT_LAZY_TYPE)
          return "<...>";
        try {
          var name = getComponentNameFromType(type);
          return name ? "<" + name + ">" : "<...>";
        } catch (x) {
          return "<...>";
        }
      }
      function getOwner() {
        var dispatcher = ReactSharedInternals.A;
        return null === dispatcher ? null : dispatcher.getOwner();
      }
      function UnknownOwner() {
        return Error("react-stack-top-frame");
      }
      function hasValidKey(config) {
        if (hasOwnProperty.call(config, "key")) {
          var getter = Object.getOwnPropertyDescriptor(config, "key").get;
          if (getter && getter.isReactWarning) return false;
        }
        return void 0 !== config.key;
      }
      function defineKeyPropWarningGetter(props, displayName) {
        function warnAboutAccessingKey() {
          specialPropKeyWarningShown || (specialPropKeyWarningShown = true, console.error(
            "%s: `key` is not a prop. Trying to access it will result in `undefined` being returned. If you need to access the same value within the child component, you should pass it as a different prop. (https://react.dev/link/special-props)",
            displayName
          ));
        }
        warnAboutAccessingKey.isReactWarning = true;
        Object.defineProperty(props, "key", {
          get: warnAboutAccessingKey,
          configurable: true
        });
      }
      function elementRefGetterWithDeprecationWarning() {
        var componentName = getComponentNameFromType(this.type);
        didWarnAboutElementRef[componentName] || (didWarnAboutElementRef[componentName] = true, console.error(
          "Accessing element.ref was removed in React 19. ref is now a regular prop. It will be removed from the JSX Element type in a future release."
        ));
        componentName = this.props.ref;
        return void 0 !== componentName ? componentName : null;
      }
      function ReactElement4(type, key, props, owner, debugStack, debugTask) {
        var refProp = props.ref;
        type = {
          $$typeof: REACT_ELEMENT_TYPE,
          type,
          key,
          props,
          _owner: owner
        };
        null !== (void 0 !== refProp ? refProp : null) ? Object.defineProperty(type, "ref", {
          enumerable: false,
          get: elementRefGetterWithDeprecationWarning
        }) : Object.defineProperty(type, "ref", { enumerable: false, value: null });
        type._store = {};
        Object.defineProperty(type._store, "validated", {
          configurable: false,
          enumerable: false,
          writable: true,
          value: 0
        });
        Object.defineProperty(type, "_debugInfo", {
          configurable: false,
          enumerable: false,
          writable: true,
          value: null
        });
        Object.defineProperty(type, "_debugStack", {
          configurable: false,
          enumerable: false,
          writable: true,
          value: debugStack
        });
        Object.defineProperty(type, "_debugTask", {
          configurable: false,
          enumerable: false,
          writable: true,
          value: debugTask
        });
        Object.freeze && (Object.freeze(type.props), Object.freeze(type));
        return type;
      }
      function cloneAndReplaceKey(oldElement, newKey) {
        newKey = ReactElement4(
          oldElement.type,
          newKey,
          oldElement.props,
          oldElement._owner,
          oldElement._debugStack,
          oldElement._debugTask
        );
        oldElement._store && (newKey._store.validated = oldElement._store.validated);
        return newKey;
      }
      function validateChildKeys(node) {
        isValidElement(node) ? node._store && (node._store.validated = 1) : "object" === typeof node && null !== node && node.$$typeof === REACT_LAZY_TYPE && ("fulfilled" === node._payload.status ? isValidElement(node._payload.value) && node._payload.value._store && (node._payload.value._store.validated = 1) : node._store && (node._store.validated = 1));
      }
      function isValidElement(object) {
        return "object" === typeof object && null !== object && object.$$typeof === REACT_ELEMENT_TYPE;
      }
      function escape(key) {
        var escaperLookup = { "=": "=0", ":": "=2" };
        return "$" + key.replace(/[=:]/g, function(match) {
          return escaperLookup[match];
        });
      }
      function getElementKey(element, index) {
        return "object" === typeof element && null !== element && null != element.key ? (checkKeyStringCoercion(element.key), escape("" + element.key)) : index.toString(36);
      }
      function resolveThenable(thenable) {
        switch (thenable.status) {
          case "fulfilled":
            return thenable.value;
          case "rejected":
            throw thenable.reason;
          default:
            switch ("string" === typeof thenable.status ? thenable.then(noop, noop) : (thenable.status = "pending", thenable.then(
              function(fulfilledValue) {
                "pending" === thenable.status && (thenable.status = "fulfilled", thenable.value = fulfilledValue);
              },
              function(error) {
                "pending" === thenable.status && (thenable.status = "rejected", thenable.reason = error);
              }
            )), thenable.status) {
              case "fulfilled":
                return thenable.value;
              case "rejected":
                throw thenable.reason;
            }
        }
        throw thenable;
      }
      function mapIntoArray(children, array, escapedPrefix, nameSoFar, callback) {
        var type = typeof children;
        if ("undefined" === type || "boolean" === type) children = null;
        var invokeCallback = false;
        if (null === children) invokeCallback = true;
        else
          switch (type) {
            case "bigint":
            case "string":
            case "number":
              invokeCallback = true;
              break;
            case "object":
              switch (children.$$typeof) {
                case REACT_ELEMENT_TYPE:
                case REACT_PORTAL_TYPE:
                  invokeCallback = true;
                  break;
                case REACT_LAZY_TYPE:
                  return invokeCallback = children._init, mapIntoArray(
                    invokeCallback(children._payload),
                    array,
                    escapedPrefix,
                    nameSoFar,
                    callback
                  );
              }
          }
        if (invokeCallback) {
          invokeCallback = children;
          callback = callback(invokeCallback);
          var childKey = "" === nameSoFar ? "." + getElementKey(invokeCallback, 0) : nameSoFar;
          isArrayImpl(callback) ? (escapedPrefix = "", null != childKey && (escapedPrefix = childKey.replace(userProvidedKeyEscapeRegex, "$&/") + "/"), mapIntoArray(callback, array, escapedPrefix, "", function(c) {
            return c;
          })) : null != callback && (isValidElement(callback) && (null != callback.key && (invokeCallback && invokeCallback.key === callback.key || checkKeyStringCoercion(callback.key)), escapedPrefix = cloneAndReplaceKey(
            callback,
            escapedPrefix + (null == callback.key || invokeCallback && invokeCallback.key === callback.key ? "" : ("" + callback.key).replace(
              userProvidedKeyEscapeRegex,
              "$&/"
            ) + "/") + childKey
          ), "" !== nameSoFar && null != invokeCallback && isValidElement(invokeCallback) && null == invokeCallback.key && invokeCallback._store && !invokeCallback._store.validated && (escapedPrefix._store.validated = 2), callback = escapedPrefix), array.push(callback));
          return 1;
        }
        invokeCallback = 0;
        childKey = "" === nameSoFar ? "." : nameSoFar + ":";
        if (isArrayImpl(children))
          for (var i = 0; i < children.length; i++)
            nameSoFar = children[i], type = childKey + getElementKey(nameSoFar, i), invokeCallback += mapIntoArray(
              nameSoFar,
              array,
              escapedPrefix,
              type,
              callback
            );
        else if (i = getIteratorFn(children), "function" === typeof i)
          for (i === children.entries && (didWarnAboutMaps || console.warn(
            "Using Maps as children is not supported. Use an array of keyed ReactElements instead."
          ), didWarnAboutMaps = true), children = i.call(children), i = 0; !(nameSoFar = children.next()).done; )
            nameSoFar = nameSoFar.value, type = childKey + getElementKey(nameSoFar, i++), invokeCallback += mapIntoArray(
              nameSoFar,
              array,
              escapedPrefix,
              type,
              callback
            );
        else if ("object" === type) {
          if ("function" === typeof children.then)
            return mapIntoArray(
              resolveThenable(children),
              array,
              escapedPrefix,
              nameSoFar,
              callback
            );
          array = String(children);
          throw Error(
            "Objects are not valid as a React child (found: " + ("[object Object]" === array ? "object with keys {" + Object.keys(children).join(", ") + "}" : array) + "). If you meant to render a collection of children, use an array instead."
          );
        }
        return invokeCallback;
      }
      function mapChildren(children, func, context) {
        if (null == children) return children;
        var result = [], count = 0;
        mapIntoArray(children, result, "", "", function(child) {
          return func.call(context, child, count++);
        });
        return result;
      }
      function lazyInitializer(payload) {
        if (-1 === payload._status) {
          var ioInfo = payload._ioInfo;
          null != ioInfo && (ioInfo.start = ioInfo.end = performance.now());
          ioInfo = payload._result;
          var thenable = ioInfo();
          thenable.then(
            function(moduleObject) {
              if (0 === payload._status || -1 === payload._status) {
                payload._status = 1;
                payload._result = moduleObject;
                var _ioInfo = payload._ioInfo;
                null != _ioInfo && (_ioInfo.end = performance.now());
                void 0 === thenable.status && (thenable.status = "fulfilled", thenable.value = moduleObject);
              }
            },
            function(error) {
              if (0 === payload._status || -1 === payload._status) {
                payload._status = 2;
                payload._result = error;
                var _ioInfo2 = payload._ioInfo;
                null != _ioInfo2 && (_ioInfo2.end = performance.now());
                void 0 === thenable.status && (thenable.status = "rejected", thenable.reason = error);
              }
            }
          );
          ioInfo = payload._ioInfo;
          if (null != ioInfo) {
            ioInfo.value = thenable;
            var displayName = thenable.displayName;
            "string" === typeof displayName && (ioInfo.name = displayName);
          }
          -1 === payload._status && (payload._status = 0, payload._result = thenable);
        }
        if (1 === payload._status)
          return ioInfo = payload._result, void 0 === ioInfo && console.error(
            "lazy: Expected the result of a dynamic import() call. Instead received: %s\n\nYour code should look like: \n  const MyComponent = lazy(() => import('./MyComponent'))\n\nDid you accidentally put curly braces around the import?",
            ioInfo
          ), "default" in ioInfo || console.error(
            "lazy: Expected the result of a dynamic import() call. Instead received: %s\n\nYour code should look like: \n  const MyComponent = lazy(() => import('./MyComponent'))",
            ioInfo
          ), ioInfo.default;
        throw payload._result;
      }
      function resolveDispatcher() {
        var dispatcher = ReactSharedInternals.H;
        null === dispatcher && console.error(
          "Invalid hook call. Hooks can only be called inside of the body of a function component. This could happen for one of the following reasons:\n1. You might have mismatching versions of React and the renderer (such as React DOM)\n2. You might be breaking the Rules of Hooks\n3. You might have more than one copy of React in the same app\nSee https://react.dev/link/invalid-hook-call for tips about how to debug and fix this problem."
        );
        return dispatcher;
      }
      function releaseAsyncTransition() {
        ReactSharedInternals.asyncTransitions--;
      }
      function enqueueTask(task) {
        if (null === enqueueTaskImpl)
          try {
            var requireString = ("require" + Math.random()).slice(0, 7);
            enqueueTaskImpl = (module && module[requireString]).call(
              module,
              "timers"
            ).setImmediate;
          } catch (_err) {
            enqueueTaskImpl = function(callback) {
              false === didWarnAboutMessageChannel && (didWarnAboutMessageChannel = true, "undefined" === typeof MessageChannel && console.error(
                "This browser does not have a MessageChannel implementation, so enqueuing tasks via await act(async () => ...) will fail. Please file an issue at https://github.com/facebook/react/issues if you encounter this warning."
              ));
              var channel = new MessageChannel();
              channel.port1.onmessage = callback;
              channel.port2.postMessage(void 0);
            };
          }
        return enqueueTaskImpl(task);
      }
      function aggregateErrors(errors) {
        return 1 < errors.length && "function" === typeof AggregateError ? new AggregateError(errors) : errors[0];
      }
      function popActScope(prevActQueue, prevActScopeDepth) {
        prevActScopeDepth !== actScopeDepth - 1 && console.error(
          "You seem to have overlapping act() calls, this is not supported. Be sure to await previous act() calls before making a new one. "
        );
        actScopeDepth = prevActScopeDepth;
      }
      function recursivelyFlushAsyncActWork(returnValue, resolve, reject) {
        var queue = ReactSharedInternals.actQueue;
        if (null !== queue)
          if (0 !== queue.length)
            try {
              flushActQueue(queue);
              enqueueTask(function() {
                return recursivelyFlushAsyncActWork(returnValue, resolve, reject);
              });
              return;
            } catch (error) {
              ReactSharedInternals.thrownErrors.push(error);
            }
          else ReactSharedInternals.actQueue = null;
        0 < ReactSharedInternals.thrownErrors.length ? (queue = aggregateErrors(ReactSharedInternals.thrownErrors), ReactSharedInternals.thrownErrors.length = 0, reject(queue)) : resolve(returnValue);
      }
      function flushActQueue(queue) {
        if (!isFlushing) {
          isFlushing = true;
          var i = 0;
          try {
            for (; i < queue.length; i++) {
              var callback = queue[i];
              do {
                ReactSharedInternals.didUsePromise = false;
                var continuation = callback(false);
                if (null !== continuation) {
                  if (ReactSharedInternals.didUsePromise) {
                    queue[i] = callback;
                    queue.splice(0, i);
                    return;
                  }
                  callback = continuation;
                } else break;
              } while (1);
            }
            queue.length = 0;
          } catch (error) {
            queue.splice(0, i + 1), ReactSharedInternals.thrownErrors.push(error);
          } finally {
            isFlushing = false;
          }
        }
      }
      "undefined" !== typeof __REACT_DEVTOOLS_GLOBAL_HOOK__ && "function" === typeof __REACT_DEVTOOLS_GLOBAL_HOOK__.registerInternalModuleStart && __REACT_DEVTOOLS_GLOBAL_HOOK__.registerInternalModuleStart(Error());
      var REACT_ELEMENT_TYPE = Symbol.for("react.transitional.element"), REACT_PORTAL_TYPE = Symbol.for("react.portal"), REACT_FRAGMENT_TYPE = Symbol.for("react.fragment"), REACT_STRICT_MODE_TYPE = Symbol.for("react.strict_mode"), REACT_PROFILER_TYPE = Symbol.for("react.profiler"), REACT_CONSUMER_TYPE = Symbol.for("react.consumer"), REACT_CONTEXT_TYPE = Symbol.for("react.context"), REACT_FORWARD_REF_TYPE = Symbol.for("react.forward_ref"), REACT_SUSPENSE_TYPE = Symbol.for("react.suspense"), REACT_SUSPENSE_LIST_TYPE = Symbol.for("react.suspense_list"), REACT_MEMO_TYPE = Symbol.for("react.memo"), REACT_LAZY_TYPE = Symbol.for("react.lazy"), REACT_ACTIVITY_TYPE = Symbol.for("react.activity"), MAYBE_ITERATOR_SYMBOL = Symbol.iterator, didWarnStateUpdateForUnmountedComponent = {}, ReactNoopUpdateQueue = {
        isMounted: function() {
          return false;
        },
        enqueueForceUpdate: function(publicInstance) {
          warnNoop(publicInstance, "forceUpdate");
        },
        enqueueReplaceState: function(publicInstance) {
          warnNoop(publicInstance, "replaceState");
        },
        enqueueSetState: function(publicInstance) {
          warnNoop(publicInstance, "setState");
        }
      }, assign = Object.assign, emptyObject = {};
      Object.freeze(emptyObject);
      Component.prototype.isReactComponent = {};
      Component.prototype.setState = function(partialState, callback) {
        if ("object" !== typeof partialState && "function" !== typeof partialState && null != partialState)
          throw Error(
            "takes an object of state variables to update or a function which returns an object of state variables."
          );
        this.updater.enqueueSetState(this, partialState, callback, "setState");
      };
      Component.prototype.forceUpdate = function(callback) {
        this.updater.enqueueForceUpdate(this, callback, "forceUpdate");
      };
      var deprecatedAPIs = {
        isMounted: [
          "isMounted",
          "Instead, make sure to clean up subscriptions and pending requests in componentWillUnmount to prevent memory leaks."
        ],
        replaceState: [
          "replaceState",
          "Refactor your code to use setState instead (see https://github.com/facebook/react/issues/3236)."
        ]
      };
      for (fnName in deprecatedAPIs)
        deprecatedAPIs.hasOwnProperty(fnName) && defineDeprecationWarning(fnName, deprecatedAPIs[fnName]);
      ComponentDummy.prototype = Component.prototype;
      deprecatedAPIs = PureComponent.prototype = new ComponentDummy();
      deprecatedAPIs.constructor = PureComponent;
      assign(deprecatedAPIs, Component.prototype);
      deprecatedAPIs.isPureReactComponent = true;
      var isArrayImpl = Array.isArray, REACT_CLIENT_REFERENCE = Symbol.for("react.client.reference"), ReactSharedInternals = {
        H: null,
        A: null,
        T: null,
        S: null,
        actQueue: null,
        asyncTransitions: 0,
        isBatchingLegacy: false,
        didScheduleLegacyUpdate: false,
        didUsePromise: false,
        thrownErrors: [],
        getCurrentStack: null,
        recentlyCreatedOwnerStacks: 0
      }, hasOwnProperty = Object.prototype.hasOwnProperty, createTask = console.createTask ? console.createTask : function() {
        return null;
      };
      deprecatedAPIs = {
        react_stack_bottom_frame: function(callStackForError) {
          return callStackForError();
        }
      };
      var specialPropKeyWarningShown, didWarnAboutOldJSXRuntime;
      var didWarnAboutElementRef = {};
      var unknownOwnerDebugStack = deprecatedAPIs.react_stack_bottom_frame.bind(
        deprecatedAPIs,
        UnknownOwner
      )();
      var unknownOwnerDebugTask = createTask(getTaskName(UnknownOwner));
      var didWarnAboutMaps = false, userProvidedKeyEscapeRegex = /\/+/g, reportGlobalError = "function" === typeof reportError ? reportError : function(error) {
        if ("object" === typeof window && "function" === typeof window.ErrorEvent) {
          var event = new window.ErrorEvent("error", {
            bubbles: true,
            cancelable: true,
            message: "object" === typeof error && null !== error && "string" === typeof error.message ? String(error.message) : String(error),
            error
          });
          if (!window.dispatchEvent(event)) return;
        } else if ("object" === typeof process && "function" === typeof process.emit) {
          process.emit("uncaughtException", error);
          return;
        }
        console.error(error);
      }, didWarnAboutMessageChannel = false, enqueueTaskImpl = null, actScopeDepth = 0, didWarnNoAwaitAct = false, isFlushing = false, queueSeveralMicrotasks = "function" === typeof queueMicrotask ? function(callback) {
        queueMicrotask(function() {
          return queueMicrotask(callback);
        });
      } : enqueueTask;
      deprecatedAPIs = Object.freeze({
        __proto__: null,
        c: function(size) {
          return resolveDispatcher().useMemoCache(size);
        }
      });
      var fnName = {
        map: mapChildren,
        forEach: function(children, forEachFunc, forEachContext) {
          mapChildren(
            children,
            function() {
              forEachFunc.apply(this, arguments);
            },
            forEachContext
          );
        },
        count: function(children) {
          var n = 0;
          mapChildren(children, function() {
            n++;
          });
          return n;
        },
        toArray: function(children) {
          return mapChildren(children, function(child) {
            return child;
          }) || [];
        },
        only: function(children) {
          if (!isValidElement(children))
            throw Error(
              "React.Children.only expected to receive a single React element child."
            );
          return children;
        }
      };
      exports.Activity = REACT_ACTIVITY_TYPE;
      exports.Children = fnName;
      exports.Component = Component;
      exports.Fragment = REACT_FRAGMENT_TYPE;
      exports.Profiler = REACT_PROFILER_TYPE;
      exports.PureComponent = PureComponent;
      exports.StrictMode = REACT_STRICT_MODE_TYPE;
      exports.Suspense = REACT_SUSPENSE_TYPE;
      exports.__CLIENT_INTERNALS_DO_NOT_USE_OR_WARN_USERS_THEY_CANNOT_UPGRADE = ReactSharedInternals;
      exports.__COMPILER_RUNTIME = deprecatedAPIs;
      exports.act = function(callback) {
        var prevActQueue = ReactSharedInternals.actQueue, prevActScopeDepth = actScopeDepth;
        actScopeDepth++;
        var queue = ReactSharedInternals.actQueue = null !== prevActQueue ? prevActQueue : [], didAwaitActCall = false;
        try {
          var result = callback();
        } catch (error) {
          ReactSharedInternals.thrownErrors.push(error);
        }
        if (0 < ReactSharedInternals.thrownErrors.length)
          throw popActScope(prevActQueue, prevActScopeDepth), callback = aggregateErrors(ReactSharedInternals.thrownErrors), ReactSharedInternals.thrownErrors.length = 0, callback;
        if (null !== result && "object" === typeof result && "function" === typeof result.then) {
          var thenable = result;
          queueSeveralMicrotasks(function() {
            didAwaitActCall || didWarnNoAwaitAct || (didWarnNoAwaitAct = true, console.error(
              "You called act(async () => ...) without await. This could lead to unexpected testing behaviour, interleaving multiple act calls and mixing their scopes. You should - await act(async () => ...);"
            ));
          });
          return {
            then: function(resolve, reject) {
              didAwaitActCall = true;
              thenable.then(
                function(returnValue) {
                  popActScope(prevActQueue, prevActScopeDepth);
                  if (0 === prevActScopeDepth) {
                    try {
                      flushActQueue(queue), enqueueTask(function() {
                        return recursivelyFlushAsyncActWork(
                          returnValue,
                          resolve,
                          reject
                        );
                      });
                    } catch (error$0) {
                      ReactSharedInternals.thrownErrors.push(error$0);
                    }
                    if (0 < ReactSharedInternals.thrownErrors.length) {
                      var _thrownError = aggregateErrors(
                        ReactSharedInternals.thrownErrors
                      );
                      ReactSharedInternals.thrownErrors.length = 0;
                      reject(_thrownError);
                    }
                  } else resolve(returnValue);
                },
                function(error) {
                  popActScope(prevActQueue, prevActScopeDepth);
                  0 < ReactSharedInternals.thrownErrors.length ? (error = aggregateErrors(
                    ReactSharedInternals.thrownErrors
                  ), ReactSharedInternals.thrownErrors.length = 0, reject(error)) : reject(error);
                }
              );
            }
          };
        }
        var returnValue$jscomp$0 = result;
        popActScope(prevActQueue, prevActScopeDepth);
        0 === prevActScopeDepth && (flushActQueue(queue), 0 !== queue.length && queueSeveralMicrotasks(function() {
          didAwaitActCall || didWarnNoAwaitAct || (didWarnNoAwaitAct = true, console.error(
            "A component suspended inside an `act` scope, but the `act` call was not awaited. When testing React components that depend on asynchronous data, you must await the result:\n\nawait act(() => ...)"
          ));
        }), ReactSharedInternals.actQueue = null);
        if (0 < ReactSharedInternals.thrownErrors.length)
          throw callback = aggregateErrors(ReactSharedInternals.thrownErrors), ReactSharedInternals.thrownErrors.length = 0, callback;
        return {
          then: function(resolve, reject) {
            didAwaitActCall = true;
            0 === prevActScopeDepth ? (ReactSharedInternals.actQueue = queue, enqueueTask(function() {
              return recursivelyFlushAsyncActWork(
                returnValue$jscomp$0,
                resolve,
                reject
              );
            })) : resolve(returnValue$jscomp$0);
          }
        };
      };
      exports.cache = function(fn) {
        return function() {
          return fn.apply(null, arguments);
        };
      };
      exports.cacheSignal = function() {
        return null;
      };
      exports.captureOwnerStack = function() {
        var getCurrentStack = ReactSharedInternals.getCurrentStack;
        return null === getCurrentStack ? null : getCurrentStack();
      };
      exports.cloneElement = function(element, config, children) {
        if (null === element || void 0 === element)
          throw Error(
            "The argument must be a React element, but you passed " + element + "."
          );
        var props = assign({}, element.props), key = element.key, owner = element._owner;
        if (null != config) {
          var JSCompiler_inline_result;
          a: {
            if (hasOwnProperty.call(config, "ref") && (JSCompiler_inline_result = Object.getOwnPropertyDescriptor(
              config,
              "ref"
            ).get) && JSCompiler_inline_result.isReactWarning) {
              JSCompiler_inline_result = false;
              break a;
            }
            JSCompiler_inline_result = void 0 !== config.ref;
          }
          JSCompiler_inline_result && (owner = getOwner());
          hasValidKey(config) && (checkKeyStringCoercion(config.key), key = "" + config.key);
          for (propName in config)
            !hasOwnProperty.call(config, propName) || "key" === propName || "__self" === propName || "__source" === propName || "ref" === propName && void 0 === config.ref || (props[propName] = config[propName]);
        }
        var propName = arguments.length - 2;
        if (1 === propName) props.children = children;
        else if (1 < propName) {
          JSCompiler_inline_result = Array(propName);
          for (var i = 0; i < propName; i++)
            JSCompiler_inline_result[i] = arguments[i + 2];
          props.children = JSCompiler_inline_result;
        }
        props = ReactElement4(
          element.type,
          key,
          props,
          owner,
          element._debugStack,
          element._debugTask
        );
        for (key = 2; key < arguments.length; key++)
          validateChildKeys(arguments[key]);
        return props;
      };
      exports.createContext = function(defaultValue) {
        defaultValue = {
          $$typeof: REACT_CONTEXT_TYPE,
          _currentValue: defaultValue,
          _currentValue2: defaultValue,
          _threadCount: 0,
          Provider: null,
          Consumer: null
        };
        defaultValue.Provider = defaultValue;
        defaultValue.Consumer = {
          $$typeof: REACT_CONSUMER_TYPE,
          _context: defaultValue
        };
        defaultValue._currentRenderer = null;
        defaultValue._currentRenderer2 = null;
        return defaultValue;
      };
      exports.createElement = function(type, config, children) {
        for (var i = 2; i < arguments.length; i++)
          validateChildKeys(arguments[i]);
        i = {};
        var key = null;
        if (null != config)
          for (propName in didWarnAboutOldJSXRuntime || !("__self" in config) || "key" in config || (didWarnAboutOldJSXRuntime = true, console.warn(
            "Your app (or one of its dependencies) is using an outdated JSX transform. Update to the modern JSX transform for faster performance: https://react.dev/link/new-jsx-transform"
          )), hasValidKey(config) && (checkKeyStringCoercion(config.key), key = "" + config.key), config)
            hasOwnProperty.call(config, propName) && "key" !== propName && "__self" !== propName && "__source" !== propName && (i[propName] = config[propName]);
        var childrenLength = arguments.length - 2;
        if (1 === childrenLength) i.children = children;
        else if (1 < childrenLength) {
          for (var childArray = Array(childrenLength), _i = 0; _i < childrenLength; _i++)
            childArray[_i] = arguments[_i + 2];
          Object.freeze && Object.freeze(childArray);
          i.children = childArray;
        }
        if (type && type.defaultProps)
          for (propName in childrenLength = type.defaultProps, childrenLength)
            void 0 === i[propName] && (i[propName] = childrenLength[propName]);
        key && defineKeyPropWarningGetter(
          i,
          "function" === typeof type ? type.displayName || type.name || "Unknown" : type
        );
        var propName = 1e4 > ReactSharedInternals.recentlyCreatedOwnerStacks++;
        return ReactElement4(
          type,
          key,
          i,
          getOwner(),
          propName ? Error("react-stack-top-frame") : unknownOwnerDebugStack,
          propName ? createTask(getTaskName(type)) : unknownOwnerDebugTask
        );
      };
      exports.createRef = function() {
        var refObject = { current: null };
        Object.seal(refObject);
        return refObject;
      };
      exports.forwardRef = function(render) {
        null != render && render.$$typeof === REACT_MEMO_TYPE ? console.error(
          "forwardRef requires a render function but received a `memo` component. Instead of forwardRef(memo(...)), use memo(forwardRef(...))."
        ) : "function" !== typeof render ? console.error(
          "forwardRef requires a render function but was given %s.",
          null === render ? "null" : typeof render
        ) : 0 !== render.length && 2 !== render.length && console.error(
          "forwardRef render functions accept exactly two parameters: props and ref. %s",
          1 === render.length ? "Did you forget to use the ref parameter?" : "Any additional parameter will be undefined."
        );
        null != render && null != render.defaultProps && console.error(
          "forwardRef render functions do not support defaultProps. Did you accidentally pass a React component?"
        );
        var elementType = { $$typeof: REACT_FORWARD_REF_TYPE, render }, ownName;
        Object.defineProperty(elementType, "displayName", {
          enumerable: false,
          configurable: true,
          get: function() {
            return ownName;
          },
          set: function(name) {
            ownName = name;
            render.name || render.displayName || (Object.defineProperty(render, "name", { value: name }), render.displayName = name);
          }
        });
        return elementType;
      };
      exports.isValidElement = isValidElement;
      exports.lazy = function(ctor) {
        ctor = { _status: -1, _result: ctor };
        var lazyType = {
          $$typeof: REACT_LAZY_TYPE,
          _payload: ctor,
          _init: lazyInitializer
        }, ioInfo = {
          name: "lazy",
          start: -1,
          end: -1,
          value: null,
          owner: null,
          debugStack: Error("react-stack-top-frame"),
          debugTask: console.createTask ? console.createTask("lazy()") : null
        };
        ctor._ioInfo = ioInfo;
        lazyType._debugInfo = [{ awaited: ioInfo }];
        return lazyType;
      };
      exports.memo = function(type, compare) {
        null == type && console.error(
          "memo: The first argument must be a component. Instead received: %s",
          null === type ? "null" : typeof type
        );
        compare = {
          $$typeof: REACT_MEMO_TYPE,
          type,
          compare: void 0 === compare ? null : compare
        };
        var ownName;
        Object.defineProperty(compare, "displayName", {
          enumerable: false,
          configurable: true,
          get: function() {
            return ownName;
          },
          set: function(name) {
            ownName = name;
            type.name || type.displayName || (Object.defineProperty(type, "name", { value: name }), type.displayName = name);
          }
        });
        return compare;
      };
      exports.startTransition = function(scope) {
        var prevTransition = ReactSharedInternals.T, currentTransition = {};
        currentTransition._updatedFibers = /* @__PURE__ */ new Set();
        ReactSharedInternals.T = currentTransition;
        try {
          var returnValue = scope(), onStartTransitionFinish = ReactSharedInternals.S;
          null !== onStartTransitionFinish && onStartTransitionFinish(currentTransition, returnValue);
          "object" === typeof returnValue && null !== returnValue && "function" === typeof returnValue.then && (ReactSharedInternals.asyncTransitions++, returnValue.then(releaseAsyncTransition, releaseAsyncTransition), returnValue.then(noop, reportGlobalError));
        } catch (error) {
          reportGlobalError(error);
        } finally {
          null === prevTransition && currentTransition._updatedFibers && (scope = currentTransition._updatedFibers.size, currentTransition._updatedFibers.clear(), 10 < scope && console.warn(
            "Detected a large number of updates inside startTransition. If this is due to a subscription please re-write it to use React provided hooks. Otherwise concurrent mode guarantees are off the table."
          )), null !== prevTransition && null !== currentTransition.types && (null !== prevTransition.types && prevTransition.types !== currentTransition.types && console.error(
            "We expected inner Transitions to have transferred the outer types set and that you cannot add to the outer Transition while inside the inner.This is a bug in React."
          ), prevTransition.types = currentTransition.types), ReactSharedInternals.T = prevTransition;
        }
      };
      exports.unstable_useCacheRefresh = function() {
        return resolveDispatcher().useCacheRefresh();
      };
      exports.use = function(usable) {
        return resolveDispatcher().use(usable);
      };
      exports.useActionState = function(action, initialState, permalink) {
        return resolveDispatcher().useActionState(
          action,
          initialState,
          permalink
        );
      };
      exports.useCallback = function(callback, deps) {
        return resolveDispatcher().useCallback(callback, deps);
      };
      exports.useContext = function(Context) {
        var dispatcher = resolveDispatcher();
        Context.$$typeof === REACT_CONSUMER_TYPE && console.error(
          "Calling useContext(Context.Consumer) is not supported and will cause bugs. Did you mean to call useContext(Context) instead?"
        );
        return dispatcher.useContext(Context);
      };
      exports.useDebugValue = function(value, formatterFn) {
        return resolveDispatcher().useDebugValue(value, formatterFn);
      };
      exports.useDeferredValue = function(value, initialValue) {
        return resolveDispatcher().useDeferredValue(value, initialValue);
      };
      exports.useEffect = function(create, deps) {
        null == create && console.warn(
          "React Hook useEffect requires an effect callback. Did you forget to pass a callback to the hook?"
        );
        return resolveDispatcher().useEffect(create, deps);
      };
      exports.useEffectEvent = function(callback) {
        return resolveDispatcher().useEffectEvent(callback);
      };
      exports.useId = function() {
        return resolveDispatcher().useId();
      };
      exports.useImperativeHandle = function(ref, create, deps) {
        return resolveDispatcher().useImperativeHandle(ref, create, deps);
      };
      exports.useInsertionEffect = function(create, deps) {
        null == create && console.warn(
          "React Hook useInsertionEffect requires an effect callback. Did you forget to pass a callback to the hook?"
        );
        return resolveDispatcher().useInsertionEffect(create, deps);
      };
      exports.useLayoutEffect = function(create, deps) {
        null == create && console.warn(
          "React Hook useLayoutEffect requires an effect callback. Did you forget to pass a callback to the hook?"
        );
        return resolveDispatcher().useLayoutEffect(create, deps);
      };
      exports.useMemo = function(create, deps) {
        return resolveDispatcher().useMemo(create, deps);
      };
      exports.useOptimistic = function(passthrough, reducer) {
        return resolveDispatcher().useOptimistic(passthrough, reducer);
      };
      exports.useReducer = function(reducer, initialArg, init) {
        return resolveDispatcher().useReducer(reducer, initialArg, init);
      };
      exports.useRef = function(initialValue) {
        return resolveDispatcher().useRef(initialValue);
      };
      exports.useState = function(initialState) {
        return resolveDispatcher().useState(initialState);
      };
      exports.useSyncExternalStore = function(subscribe, getSnapshot, getServerSnapshot) {
        return resolveDispatcher().useSyncExternalStore(
          subscribe,
          getSnapshot,
          getServerSnapshot
        );
      };
      exports.useTransition = function() {
        return resolveDispatcher().useTransition();
      };
      exports.version = "19.2.0";
      "undefined" !== typeof __REACT_DEVTOOLS_GLOBAL_HOOK__ && "function" === typeof __REACT_DEVTOOLS_GLOBAL_HOOK__.registerInternalModuleStop && __REACT_DEVTOOLS_GLOBAL_HOOK__.registerInternalModuleStop(Error());
    })();
  }
});

// node_modules/react/index.js
var require_react = __commonJS({
  "node_modules/react/index.js"(exports, module) {
    "use strict";
    if (process.env.NODE_ENV === "production") {
      module.exports = require_react_production();
    } else {
      module.exports = require_react_development();
    }
  }
});

// node_modules/react/cjs/react-jsx-runtime.development.js
var require_react_jsx_runtime_development = __commonJS({
  "node_modules/react/cjs/react-jsx-runtime.development.js"(exports) {
    "use strict";
    "production" !== process.env.NODE_ENV && (function() {
      function getComponentNameFromType(type) {
        if (null == type) return null;
        if ("function" === typeof type)
          return type.$$typeof === REACT_CLIENT_REFERENCE ? null : type.displayName || type.name || null;
        if ("string" === typeof type) return type;
        switch (type) {
          case REACT_FRAGMENT_TYPE:
            return "Fragment";
          case REACT_PROFILER_TYPE:
            return "Profiler";
          case REACT_STRICT_MODE_TYPE:
            return "StrictMode";
          case REACT_SUSPENSE_TYPE:
            return "Suspense";
          case REACT_SUSPENSE_LIST_TYPE:
            return "SuspenseList";
          case REACT_ACTIVITY_TYPE:
            return "Activity";
        }
        if ("object" === typeof type)
          switch ("number" === typeof type.tag && console.error(
            "Received an unexpected object in getComponentNameFromType(). This is likely a bug in React. Please file an issue."
          ), type.$$typeof) {
            case REACT_PORTAL_TYPE:
              return "Portal";
            case REACT_CONTEXT_TYPE:
              return type.displayName || "Context";
            case REACT_CONSUMER_TYPE:
              return (type._context.displayName || "Context") + ".Consumer";
            case REACT_FORWARD_REF_TYPE:
              var innerType = type.render;
              type = type.displayName;
              type || (type = innerType.displayName || innerType.name || "", type = "" !== type ? "ForwardRef(" + type + ")" : "ForwardRef");
              return type;
            case REACT_MEMO_TYPE:
              return innerType = type.displayName || null, null !== innerType ? innerType : getComponentNameFromType(type.type) || "Memo";
            case REACT_LAZY_TYPE:
              innerType = type._payload;
              type = type._init;
              try {
                return getComponentNameFromType(type(innerType));
              } catch (x) {
              }
          }
        return null;
      }
      function testStringCoercion(value) {
        return "" + value;
      }
      function checkKeyStringCoercion(value) {
        try {
          testStringCoercion(value);
          var JSCompiler_inline_result = false;
        } catch (e) {
          JSCompiler_inline_result = true;
        }
        if (JSCompiler_inline_result) {
          JSCompiler_inline_result = console;
          var JSCompiler_temp_const = JSCompiler_inline_result.error;
          var JSCompiler_inline_result$jscomp$0 = "function" === typeof Symbol && Symbol.toStringTag && value[Symbol.toStringTag] || value.constructor.name || "Object";
          JSCompiler_temp_const.call(
            JSCompiler_inline_result,
            "The provided key is an unsupported type %s. This value must be coerced to a string before using it here.",
            JSCompiler_inline_result$jscomp$0
          );
          return testStringCoercion(value);
        }
      }
      function getTaskName(type) {
        if (type === REACT_FRAGMENT_TYPE) return "<>";
        if ("object" === typeof type && null !== type && type.$$typeof === REACT_LAZY_TYPE)
          return "<...>";
        try {
          var name = getComponentNameFromType(type);
          return name ? "<" + name + ">" : "<...>";
        } catch (x) {
          return "<...>";
        }
      }
      function getOwner() {
        var dispatcher = ReactSharedInternals.A;
        return null === dispatcher ? null : dispatcher.getOwner();
      }
      function UnknownOwner() {
        return Error("react-stack-top-frame");
      }
      function hasValidKey(config) {
        if (hasOwnProperty.call(config, "key")) {
          var getter = Object.getOwnPropertyDescriptor(config, "key").get;
          if (getter && getter.isReactWarning) return false;
        }
        return void 0 !== config.key;
      }
      function defineKeyPropWarningGetter(props, displayName) {
        function warnAboutAccessingKey() {
          specialPropKeyWarningShown || (specialPropKeyWarningShown = true, console.error(
            "%s: `key` is not a prop. Trying to access it will result in `undefined` being returned. If you need to access the same value within the child component, you should pass it as a different prop. (https://react.dev/link/special-props)",
            displayName
          ));
        }
        warnAboutAccessingKey.isReactWarning = true;
        Object.defineProperty(props, "key", {
          get: warnAboutAccessingKey,
          configurable: true
        });
      }
      function elementRefGetterWithDeprecationWarning() {
        var componentName = getComponentNameFromType(this.type);
        didWarnAboutElementRef[componentName] || (didWarnAboutElementRef[componentName] = true, console.error(
          "Accessing element.ref was removed in React 19. ref is now a regular prop. It will be removed from the JSX Element type in a future release."
        ));
        componentName = this.props.ref;
        return void 0 !== componentName ? componentName : null;
      }
      function ReactElement4(type, key, props, owner, debugStack, debugTask) {
        var refProp = props.ref;
        type = {
          $$typeof: REACT_ELEMENT_TYPE,
          type,
          key,
          props,
          _owner: owner
        };
        null !== (void 0 !== refProp ? refProp : null) ? Object.defineProperty(type, "ref", {
          enumerable: false,
          get: elementRefGetterWithDeprecationWarning
        }) : Object.defineProperty(type, "ref", { enumerable: false, value: null });
        type._store = {};
        Object.defineProperty(type._store, "validated", {
          configurable: false,
          enumerable: false,
          writable: true,
          value: 0
        });
        Object.defineProperty(type, "_debugInfo", {
          configurable: false,
          enumerable: false,
          writable: true,
          value: null
        });
        Object.defineProperty(type, "_debugStack", {
          configurable: false,
          enumerable: false,
          writable: true,
          value: debugStack
        });
        Object.defineProperty(type, "_debugTask", {
          configurable: false,
          enumerable: false,
          writable: true,
          value: debugTask
        });
        Object.freeze && (Object.freeze(type.props), Object.freeze(type));
        return type;
      }
      function jsxDEVImpl(type, config, maybeKey, isStaticChildren, debugStack, debugTask) {
        var children = config.children;
        if (void 0 !== children)
          if (isStaticChildren)
            if (isArrayImpl(children)) {
              for (isStaticChildren = 0; isStaticChildren < children.length; isStaticChildren++)
                validateChildKeys(children[isStaticChildren]);
              Object.freeze && Object.freeze(children);
            } else
              console.error(
                "React.jsx: Static children should always be an array. You are likely explicitly calling React.jsxs or React.jsxDEV. Use the Babel transform instead."
              );
          else validateChildKeys(children);
        if (hasOwnProperty.call(config, "key")) {
          children = getComponentNameFromType(type);
          var keys = Object.keys(config).filter(function(k) {
            return "key" !== k;
          });
          isStaticChildren = 0 < keys.length ? "{key: someKey, " + keys.join(": ..., ") + ": ...}" : "{key: someKey}";
          didWarnAboutKeySpread[children + isStaticChildren] || (keys = 0 < keys.length ? "{" + keys.join(": ..., ") + ": ...}" : "{}", console.error(
            'A props object containing a "key" prop is being spread into JSX:\n  let props = %s;\n  <%s {...props} />\nReact keys must be passed directly to JSX without using spread:\n  let props = %s;\n  <%s key={someKey} {...props} />',
            isStaticChildren,
            children,
            keys,
            children
          ), didWarnAboutKeySpread[children + isStaticChildren] = true);
        }
        children = null;
        void 0 !== maybeKey && (checkKeyStringCoercion(maybeKey), children = "" + maybeKey);
        hasValidKey(config) && (checkKeyStringCoercion(config.key), children = "" + config.key);
        if ("key" in config) {
          maybeKey = {};
          for (var propName in config)
            "key" !== propName && (maybeKey[propName] = config[propName]);
        } else maybeKey = config;
        children && defineKeyPropWarningGetter(
          maybeKey,
          "function" === typeof type ? type.displayName || type.name || "Unknown" : type
        );
        return ReactElement4(
          type,
          children,
          maybeKey,
          getOwner(),
          debugStack,
          debugTask
        );
      }
      function validateChildKeys(node) {
        isValidElement(node) ? node._store && (node._store.validated = 1) : "object" === typeof node && null !== node && node.$$typeof === REACT_LAZY_TYPE && ("fulfilled" === node._payload.status ? isValidElement(node._payload.value) && node._payload.value._store && (node._payload.value._store.validated = 1) : node._store && (node._store.validated = 1));
      }
      function isValidElement(object) {
        return "object" === typeof object && null !== object && object.$$typeof === REACT_ELEMENT_TYPE;
      }
      var React4 = require_react(), REACT_ELEMENT_TYPE = Symbol.for("react.transitional.element"), REACT_PORTAL_TYPE = Symbol.for("react.portal"), REACT_FRAGMENT_TYPE = Symbol.for("react.fragment"), REACT_STRICT_MODE_TYPE = Symbol.for("react.strict_mode"), REACT_PROFILER_TYPE = Symbol.for("react.profiler"), REACT_CONSUMER_TYPE = Symbol.for("react.consumer"), REACT_CONTEXT_TYPE = Symbol.for("react.context"), REACT_FORWARD_REF_TYPE = Symbol.for("react.forward_ref"), REACT_SUSPENSE_TYPE = Symbol.for("react.suspense"), REACT_SUSPENSE_LIST_TYPE = Symbol.for("react.suspense_list"), REACT_MEMO_TYPE = Symbol.for("react.memo"), REACT_LAZY_TYPE = Symbol.for("react.lazy"), REACT_ACTIVITY_TYPE = Symbol.for("react.activity"), REACT_CLIENT_REFERENCE = Symbol.for("react.client.reference"), ReactSharedInternals = React4.__CLIENT_INTERNALS_DO_NOT_USE_OR_WARN_USERS_THEY_CANNOT_UPGRADE, hasOwnProperty = Object.prototype.hasOwnProperty, isArrayImpl = Array.isArray, createTask = console.createTask ? console.createTask : function() {
        return null;
      };
      React4 = {
        react_stack_bottom_frame: function(callStackForError) {
          return callStackForError();
        }
      };
      var specialPropKeyWarningShown;
      var didWarnAboutElementRef = {};
      var unknownOwnerDebugStack = React4.react_stack_bottom_frame.bind(
        React4,
        UnknownOwner
      )();
      var unknownOwnerDebugTask = createTask(getTaskName(UnknownOwner));
      var didWarnAboutKeySpread = {};
      exports.Fragment = REACT_FRAGMENT_TYPE;
      exports.jsx = function(type, config, maybeKey) {
        var trackActualOwner = 1e4 > ReactSharedInternals.recentlyCreatedOwnerStacks++;
        return jsxDEVImpl(
          type,
          config,
          maybeKey,
          false,
          trackActualOwner ? Error("react-stack-top-frame") : unknownOwnerDebugStack,
          trackActualOwner ? createTask(getTaskName(type)) : unknownOwnerDebugTask
        );
      };
      exports.jsxs = function(type, config, maybeKey) {
        var trackActualOwner = 1e4 > ReactSharedInternals.recentlyCreatedOwnerStacks++;
        return jsxDEVImpl(
          type,
          config,
          maybeKey,
          true,
          trackActualOwner ? Error("react-stack-top-frame") : unknownOwnerDebugStack,
          trackActualOwner ? createTask(getTaskName(type)) : unknownOwnerDebugTask
        );
      };
    })();
  }
});

// node_modules/react/jsx-runtime.js
var require_jsx_runtime = __commonJS({
  "node_modules/react/jsx-runtime.js"(exports, module) {
    "use strict";
    if (process.env.NODE_ENV === "production") {
      module.exports = require_react_jsx_runtime_production();
    } else {
      module.exports = require_react_jsx_runtime_development();
    }
  }
});

// lib/ArduinoShield/ArduinoShieldFootprint.tsx
var import_jsx_runtime = __toESM(require_jsx_runtime(), 1);
var ArduinoShieldFootprint = () => /* @__PURE__ */ (0, import_jsx_runtime.jsxs)("footprint", { children: [
  /* @__PURE__ */ (0, import_jsx_runtime.jsx)(
    "platedhole",
    {
      holeDiameter: "1.016mm",
      portHints: ["pin1"],
      name: "A0",
      pcbX: 31.75,
      pcbY: -23.99,
      outerDiameter: "1.88mm",
      shape: "circle"
    }
  ),
  /* @__PURE__ */ (0, import_jsx_runtime.jsx)(
    "platedhole",
    {
      holeDiameter: "1.016mm",
      portHints: ["pin2"],
      name: "A1",
      pcbX: 29.19,
      pcbY: -23.99,
      outerDiameter: "1.88mm",
      shape: "circle"
    }
  ),
  /* @__PURE__ */ (0, import_jsx_runtime.jsx)(
    "platedhole",
    {
      holeDiameter: "1.016mm",
      portHints: ["pin3"],
      name: "A2",
      pcbX: 26.63,
      pcbY: -23.99,
      outerDiameter: "1.88mm",
      shape: "circle"
    }
  ),
  /* @__PURE__ */ (0, import_jsx_runtime.jsx)(
    "platedhole",
    {
      holeDiameter: "1.016mm",
      portHints: ["pin4"],
      name: "A3",
      pcbX: 24.07,
      pcbY: -23.99,
      outerDiameter: "1.88mm",
      shape: "circle"
    }
  ),
  /* @__PURE__ */ (0, import_jsx_runtime.jsx)(
    "platedhole",
    {
      holeDiameter: "1.016mm",
      portHints: ["pin5"],
      name: "A4",
      pcbX: 21.51,
      pcbY: -23.99,
      outerDiameter: "1.88mm",
      shape: "circle"
    }
  ),
  /* @__PURE__ */ (0, import_jsx_runtime.jsx)(
    "platedhole",
    {
      holeDiameter: "1.016mm",
      portHints: ["pin6"],
      name: "A5",
      pcbX: 18.95,
      pcbY: -23.99,
      outerDiameter: "1.88mm",
      shape: "circle"
    }
  ),
  /* @__PURE__ */ (0, import_jsx_runtime.jsx)("silkscreentext", { text: "A0", fontSize: 0.8, pcbX: 31.75, pcbY: -21.99 }),
  /* @__PURE__ */ (0, import_jsx_runtime.jsx)("silkscreentext", { text: "A1", fontSize: 0.8, pcbX: 29.19, pcbY: -21.99 }),
  /* @__PURE__ */ (0, import_jsx_runtime.jsx)("silkscreentext", { text: "A2", fontSize: 0.8, pcbX: 26.63, pcbY: -21.99 }),
  /* @__PURE__ */ (0, import_jsx_runtime.jsx)("silkscreentext", { text: "A3", fontSize: 0.8, pcbX: 24.07, pcbY: -21.99 }),
  /* @__PURE__ */ (0, import_jsx_runtime.jsx)("silkscreentext", { text: "A4", fontSize: 0.8, pcbX: 21.51, pcbY: -21.99 }),
  /* @__PURE__ */ (0, import_jsx_runtime.jsx)("silkscreentext", { text: "A5", fontSize: 0.8, pcbX: 18.95, pcbY: -21.99 }),
  /* @__PURE__ */ (0, import_jsx_runtime.jsx)(
    "platedhole",
    {
      holeDiameter: "1.016mm",
      portHints: ["pin7"],
      name: "VIN",
      pcbX: 13.91,
      pcbY: -23.99,
      outerDiameter: "1.88mm",
      shape: "circle"
    }
  ),
  /* @__PURE__ */ (0, import_jsx_runtime.jsx)(
    "platedhole",
    {
      holeDiameter: "1.016mm",
      portHints: ["pin8"],
      name: "GND1",
      pcbX: 11.35,
      pcbY: -23.99,
      outerDiameter: "1.88mm",
      shape: "circle"
    }
  ),
  /* @__PURE__ */ (0, import_jsx_runtime.jsx)(
    "platedhole",
    {
      holeDiameter: "1.016mm",
      portHints: ["pin9"],
      name: "GND0",
      pcbX: 8.79,
      pcbY: -23.99,
      outerDiameter: "1.88mm",
      shape: "circle"
    }
  ),
  /* @__PURE__ */ (0, import_jsx_runtime.jsx)(
    "platedhole",
    {
      holeDiameter: "1.016mm",
      portHints: ["pin10"],
      name: "V5",
      pcbX: 6.23,
      pcbY: -23.99,
      outerDiameter: "1.88mm",
      shape: "circle"
    }
  ),
  /* @__PURE__ */ (0, import_jsx_runtime.jsx)(
    "platedhole",
    {
      holeDiameter: "1.016mm",
      portHints: ["pin11"],
      name: "V3_3",
      pcbX: 3.67,
      pcbY: -23.99,
      outerDiameter: "1.88mm",
      shape: "circle"
    }
  ),
  /* @__PURE__ */ (0, import_jsx_runtime.jsx)(
    "platedhole",
    {
      holeDiameter: "1.016mm",
      portHints: ["pin12"],
      name: "RES",
      pcbX: 1.11,
      pcbY: -23.99,
      outerDiameter: "1.88mm",
      shape: "circle"
    }
  ),
  /* @__PURE__ */ (0, import_jsx_runtime.jsx)(
    "platedhole",
    {
      holeDiameter: "1.016mm",
      portHints: ["pin13"],
      name: "IOREF",
      pcbX: -1.11,
      pcbY: -23.99,
      outerDiameter: "1.88mm",
      shape: "circle"
    }
  ),
  /* @__PURE__ */ (0, import_jsx_runtime.jsx)(
    "platedhole",
    {
      holeDiameter: "1.016mm",
      portHints: ["pin14"],
      name: "NC",
      pcbX: -3.67,
      pcbY: -23.99,
      outerDiameter: "1.88mm",
      shape: "circle"
    }
  ),
  /* @__PURE__ */ (0, import_jsx_runtime.jsx)("silkscreentext", { text: "VIN", fontSize: 0.8, pcbX: 13.91, pcbY: -21.99 }),
  /* @__PURE__ */ (0, import_jsx_runtime.jsx)("silkscreentext", { text: "GND1", fontSize: 0.8, pcbX: 11.35, pcbY: -21.99 }),
  /* @__PURE__ */ (0, import_jsx_runtime.jsx)("silkscreentext", { text: "GND0", fontSize: 0.8, pcbX: 8.79, pcbY: -21.99 }),
  /* @__PURE__ */ (0, import_jsx_runtime.jsx)("silkscreentext", { text: "V5", fontSize: 0.8, pcbX: 6.23, pcbY: -21.99 }),
  /* @__PURE__ */ (0, import_jsx_runtime.jsx)("silkscreentext", { text: "V3_3", fontSize: 0.8, pcbX: 3.67, pcbY: -21.99 }),
  /* @__PURE__ */ (0, import_jsx_runtime.jsx)("silkscreentext", { text: "RES", fontSize: 0.8, pcbX: 1.11, pcbY: -21.99 }),
  /* @__PURE__ */ (0, import_jsx_runtime.jsx)("silkscreentext", { text: "IOREF", fontSize: 0.8, pcbX: -1.11, pcbY: -21.99 }),
  /* @__PURE__ */ (0, import_jsx_runtime.jsx)("silkscreentext", { text: "NC", fontSize: 0.8, pcbX: -3.67, pcbY: -21.99 }),
  /* @__PURE__ */ (0, import_jsx_runtime.jsx)(
    "platedhole",
    {
      holeDiameter: "1.016mm",
      portHints: ["pin15"],
      name: "RX",
      pcbX: 31.75,
      pcbY: 23.99,
      outerDiameter: "1.88mm",
      shape: "circle"
    }
  ),
  /* @__PURE__ */ (0, import_jsx_runtime.jsx)(
    "platedhole",
    {
      holeDiameter: "1.016mm",
      portHints: ["pin16"],
      name: "TX",
      pcbX: 29.19,
      pcbY: 23.99,
      outerDiameter: "1.88mm",
      shape: "circle"
    }
  ),
  /* @__PURE__ */ (0, import_jsx_runtime.jsx)(
    "platedhole",
    {
      holeDiameter: "1.016mm",
      portHints: ["pin17"],
      name: "D2",
      pcbX: 26.63,
      pcbY: 23.99,
      outerDiameter: "1.88mm",
      shape: "circle"
    }
  ),
  /* @__PURE__ */ (0, import_jsx_runtime.jsx)(
    "platedhole",
    {
      holeDiameter: "1.016mm",
      portHints: ["pin18"],
      name: "D3",
      pcbX: 24.07,
      pcbY: 23.99,
      outerDiameter: "1.88mm",
      shape: "circle"
    }
  ),
  /* @__PURE__ */ (0, import_jsx_runtime.jsx)(
    "platedhole",
    {
      holeDiameter: "1.016mm",
      portHints: ["pin19"],
      name: "D4",
      pcbX: 21.51,
      pcbY: 23.99,
      outerDiameter: "1.88mm",
      shape: "circle"
    }
  ),
  /* @__PURE__ */ (0, import_jsx_runtime.jsx)(
    "platedhole",
    {
      holeDiameter: "1.016mm",
      portHints: ["pin20"],
      name: "D5",
      pcbX: 18.95,
      pcbY: 23.99,
      outerDiameter: "1.88mm",
      shape: "circle"
    }
  ),
  /* @__PURE__ */ (0, import_jsx_runtime.jsx)(
    "platedhole",
    {
      holeDiameter: "1.016mm",
      portHints: ["pin21"],
      name: "D6",
      pcbX: 16.39,
      pcbY: 23.99,
      outerDiameter: "1.88mm",
      shape: "circle"
    }
  ),
  /* @__PURE__ */ (0, import_jsx_runtime.jsx)(
    "platedhole",
    {
      holeDiameter: "1.016mm",
      portHints: ["pin22"],
      name: "D7",
      pcbX: 13.83,
      pcbY: 23.99,
      outerDiameter: "1.88mm",
      shape: "circle"
    }
  ),
  /* @__PURE__ */ (0, import_jsx_runtime.jsx)("silkscreentext", { text: "D7", fontSize: 0.8, pcbX: 13.83, pcbY: 22.2 }),
  /* @__PURE__ */ (0, import_jsx_runtime.jsx)("silkscreentext", { text: "D6", fontSize: 0.8, pcbX: 16.39, pcbY: 22.2 }),
  /* @__PURE__ */ (0, import_jsx_runtime.jsx)("silkscreentext", { text: "D5", fontSize: 0.8, pcbX: 18.95, pcbY: 22.2 }),
  /* @__PURE__ */ (0, import_jsx_runtime.jsx)("silkscreentext", { text: "D4", fontSize: 0.8, pcbX: 21.51, pcbY: 22.2 }),
  /* @__PURE__ */ (0, import_jsx_runtime.jsx)("silkscreentext", { text: "D3", fontSize: 0.8, pcbX: 24.07, pcbY: 22.2 }),
  /* @__PURE__ */ (0, import_jsx_runtime.jsx)("silkscreentext", { text: "D2", fontSize: 0.8, pcbX: 26.63, pcbY: 22.2 }),
  /* @__PURE__ */ (0, import_jsx_runtime.jsx)("silkscreentext", { text: "TX", fontSize: 0.8, pcbX: 29.19, pcbY: 22.2 }),
  /* @__PURE__ */ (0, import_jsx_runtime.jsx)("silkscreentext", { text: "RX", fontSize: 0.8, pcbX: 31.75, pcbY: 22.2 }),
  /* @__PURE__ */ (0, import_jsx_runtime.jsx)(
    "platedhole",
    {
      holeDiameter: "1.016mm",
      portHints: ["pin23"],
      name: "D8",
      pcbX: 9.96,
      pcbY: 23.99,
      outerDiameter: "1.88mm",
      shape: "circle"
    }
  ),
  /* @__PURE__ */ (0, import_jsx_runtime.jsx)(
    "platedhole",
    {
      holeDiameter: "1.016mm",
      portHints: ["pin24"],
      name: "D9",
      pcbX: 7.4,
      pcbY: 23.99,
      outerDiameter: "1.88mm",
      shape: "circle"
    }
  ),
  /* @__PURE__ */ (0, import_jsx_runtime.jsx)(
    "platedhole",
    {
      holeDiameter: "1.016mm",
      portHints: ["pin25"],
      name: "D10",
      pcbX: 4.84,
      pcbY: 23.99,
      outerDiameter: "1.88mm",
      shape: "circle"
    }
  ),
  /* @__PURE__ */ (0, import_jsx_runtime.jsx)(
    "platedhole",
    {
      holeDiameter: "1.016mm",
      portHints: ["pin26"],
      name: "D11",
      pcbX: 2.28,
      pcbY: 23.99,
      outerDiameter: "1.88mm",
      shape: "circle"
    }
  ),
  /* @__PURE__ */ (0, import_jsx_runtime.jsx)(
    "platedhole",
    {
      holeDiameter: "1.016mm",
      portHints: ["pin27"],
      name: "D12",
      pcbX: -0.28,
      pcbY: 23.99,
      outerDiameter: "1.88mm",
      shape: "circle"
    }
  ),
  /* @__PURE__ */ (0, import_jsx_runtime.jsx)(
    "platedhole",
    {
      holeDiameter: "1.016mm",
      portHints: ["pin28"],
      name: "D13",
      pcbX: -2.82,
      pcbY: 23.99,
      outerDiameter: "1.88mm",
      shape: "circle"
    }
  ),
  /* @__PURE__ */ (0, import_jsx_runtime.jsx)(
    "platedhole",
    {
      holeDiameter: "1.016mm",
      portHints: ["pin29"],
      name: "GND2",
      pcbX: -5.36,
      pcbY: 23.99,
      outerDiameter: "1.88mm",
      shape: "circle"
    }
  ),
  /* @__PURE__ */ (0, import_jsx_runtime.jsx)(
    "platedhole",
    {
      holeDiameter: "1.016mm",
      portHints: ["pin30"],
      name: "AREF",
      pcbX: -7.9,
      pcbY: 23.99,
      outerDiameter: "1.88mm",
      shape: "circle"
    }
  ),
  /* @__PURE__ */ (0, import_jsx_runtime.jsx)(
    "platedhole",
    {
      holeDiameter: "1.016mm",
      portHints: ["pin31"],
      name: "SDA",
      pcbX: -10.46,
      pcbY: 23.99,
      outerDiameter: "1.88mm",
      shape: "circle"
    }
  ),
  /* @__PURE__ */ (0, import_jsx_runtime.jsx)(
    "platedhole",
    {
      holeDiameter: "1.016mm",
      portHints: ["pin32"],
      name: "SCL",
      pcbX: -12.9,
      pcbY: 23.99,
      outerDiameter: "1.88mm",
      shape: "circle"
    }
  ),
  /* @__PURE__ */ (0, import_jsx_runtime.jsx)("silkscreentext", { text: "D8", fontSize: 0.8, pcbX: 9.96, pcbY: 22.2 }),
  /* @__PURE__ */ (0, import_jsx_runtime.jsx)("silkscreentext", { text: "D9", fontSize: 0.8, pcbX: 7.4, pcbY: 22.2 }),
  /* @__PURE__ */ (0, import_jsx_runtime.jsx)("silkscreentext", { text: "D10", fontSize: 0.8, pcbX: 4.84, pcbY: 22.2 }),
  /* @__PURE__ */ (0, import_jsx_runtime.jsx)("silkscreentext", { text: "D11", fontSize: 0.8, pcbX: 2.28, pcbY: 22.2 }),
  /* @__PURE__ */ (0, import_jsx_runtime.jsx)("silkscreentext", { text: "D12", fontSize: 0.8, pcbX: -0.28, pcbY: 22.2 }),
  /* @__PURE__ */ (0, import_jsx_runtime.jsx)("silkscreentext", { text: "D13", fontSize: 0.8, pcbX: -2.82, pcbY: 22.2 }),
  /* @__PURE__ */ (0, import_jsx_runtime.jsx)("silkscreentext", { text: "GND2", fontSize: 0.8, pcbX: -5.36, pcbY: 22.2 }),
  /* @__PURE__ */ (0, import_jsx_runtime.jsx)("silkscreentext", { text: "AREF", fontSize: 0.8, pcbX: -7.9, pcbY: 22.2 }),
  /* @__PURE__ */ (0, import_jsx_runtime.jsx)("silkscreentext", { text: "SDA", fontSize: 0.8, pcbX: -10.46, pcbY: 22.2 }),
  /* @__PURE__ */ (0, import_jsx_runtime.jsx)("silkscreentext", { text: "SCL", fontSize: 0.8, pcbX: -12.9, pcbY: 22.2 })
] });

// util/splitBoardAndChipProps.ts
var splitBoardAndChipProps = (props = {}) => {
  const boardOnlyProps = [
    "autorouter",
    // Autorouter configuration
    "boardAnchorAlignment",
    // Board anchor alignment
    "boardAnchorPosition",
    // Board anchor position
    "pcbGap"
    // PCB gap
  ];
  const boardProps = {};
  const chipProps = {};
  for (const prop of boardOnlyProps) {
    if (prop in props) {
      boardProps[prop] = props[prop];
    }
  }
  Object.entries(props).forEach(([key, value]) => {
    if (!boardOnlyProps.includes(key)) {
      chipProps[key] = value;
    }
  });
  return { boardProps, chipProps };
};

// lib/ArduinoShield/ArduinoShield.circuit.tsx
var import_jsx_runtime2 = __toESM(require_jsx_runtime(), 1);
var ArduinoShield = (_a) => {
  var _b = _a, { children } = _b, rest = __objRest(_b, ["children"]);
  const { boardProps, chipProps = {} } = splitBoardAndChipProps(__spreadValues({}, rest));
  const resolvedName = `${chipProps.name}_chip`;
  const _a2 = chipProps, { name: _ } = _a2, chipRest = __objRest(_a2, ["name"]);
  return /* @__PURE__ */ (0, import_jsx_runtime2.jsxs)(
    "board",
    __spreadProps(__spreadValues({}, boardProps), {
      outline: [
        { x: -34.29, y: 26.67 },
        // top-left corner
        { x: 32.29, y: 26.67 },
        // top-right (sharp)
        { x: 34.29, y: 24.67 },
        // top-right (sharp)
        { x: 34.29, y: 13.89 },
        // start top slanted transition
        { x: 36.83, y: 11.35 },
        // outward notch top
        { x: 36.83, y: -21.35 },
        // outward notch bottom
        { x: 34.29, y: -23.89 },
        // end bottom slanted transition
        { x: 34.29, y: -26.67 },
        // bottom-right corner
        { x: -34.29, y: -26.67 }
        // bottom-left corner
      ],
      children: [
        /* @__PURE__ */ (0, import_jsx_runtime2.jsx)(
          "chip",
          __spreadProps(__spreadValues({}, chipRest), {
            obstructsWithinBounds: false,
            doNotPlace: true,
            name: resolvedName,
            pcbX: 0,
            pcbY: 0,
            pinLabels: {
              pin1: "A0",
              pin2: "A1",
              pin3: "A2",
              pin4: "A3",
              pin5: "A4",
              pin6: "A5",
              pin7: "VIN",
              pin8: "NC",
              pin9: "IOREF",
              pin10: "RES",
              pin11: "V3_3",
              pin12: "V5",
              pin13: "GND0",
              pin14: "GND1",
              pin15: "RX",
              pin16: "TX",
              pin17: "D2",
              pin18: "D3",
              pin19: "D4",
              pin20: "D5",
              pin21: "D6",
              pin22: "D7",
              pin23: "D8",
              pin24: "D9",
              pin25: "D10",
              pin26: "D11",
              pin27: "D12",
              pin28: "D13",
              pin29: "GND2",
              pin30: "AREF",
              pin31: "SDA",
              pin32: "SCL"
            },
            schWidth: 1.5,
            schPinArrangement: {
              leftSide: {
                direction: "top-to-bottom",
                pins: [
                  "A0",
                  "A1",
                  "A2",
                  "A3",
                  "A4",
                  "A5",
                  "IOREF",
                  "RES",
                  "VIN",
                  "V5",
                  "V3_3",
                  "AREF",
                  "GND0",
                  "GND1",
                  "GND2"
                ]
              },
              rightSide: {
                direction: "top-to-bottom",
                pins: [
                  "RX",
                  "TX",
                  "D2",
                  "D3",
                  "D4",
                  "D5",
                  "D6",
                  "D7",
                  "D8",
                  "D9",
                  "D10",
                  "D11",
                  "D12",
                  "D13",
                  "SDA",
                  "SCL"
                ]
              }
            },
            schPinStyle: {
              pin6: {
                marginBottom: 0.5
              },
              pin16: {
                marginBottom: 0.3
              }
            },
            footprint: /* @__PURE__ */ (0, import_jsx_runtime2.jsx)(ArduinoShieldFootprint, {})
          })
        ),
        children
      ]
    })
  );
};

// lib/MicroModBoard/MicroModBoardFootprint.tsx
var import_react = __toESM(require_react(), 1);
var import_jsx_runtime3 = __toESM(require_jsx_runtime(), 1);
var createSmtPad = (props) => import_react.default.createElement("smtpad", props);
var createPlatedHole = (props) => import_react.default.createElement("platedhole", props);
var MicroModBoardFootprint = ({
  variant = "processor"
}) => {
  const padWidth = 0.35;
  const topPadHeight = 1.45;
  const bottomPadHeight = 1.95;
  const pitch = 0.5;
  const pcbYValue = variant === "processor" ? -9.725 : -31.225;
  const generatePads = () => {
    const pads = [];
    let x = 9.25;
    for (let pn = 1; pn <= 75; pn += 2) {
      if (pn >= 24 && pn <= 31) {
        x -= pitch;
        continue;
      }
      pads.push(
        createSmtPad({
          portHints: [`pin${pn}`],
          width: padWidth,
          height: topPadHeight,
          pcbX: x,
          pcbY: pcbYValue,
          layer: "top",
          shape: "rect"
        })
      );
      x -= pitch;
    }
    x = 9;
    for (let pn = 2; pn <= 74; pn += 2) {
      if (pn >= 24 && pn <= 31) {
        x -= pitch;
        continue;
      }
      pads.push(
        createSmtPad({
          portHints: [`pin${pn}`],
          width: padWidth,
          height: bottomPadHeight,
          pcbX: x,
          pcbY: pcbYValue + 0.25,
          layer: "bottom",
          shape: "rect"
        })
      );
      x -= pitch;
    }
    return pads;
  };
  const mountingHoles = variant === "function" ? [
    createPlatedHole({
      portHints: ["HOLE_PAD_1"],
      holeDiameter: 3.82,
      outerDiameter: 5.62,
      pcbX: -19,
      pcbY: 11.5,
      shape: "circle"
    }),
    createPlatedHole({
      portHints: ["HOLE_PAD_2"],
      holeDiameter: 3.82,
      outerDiameter: 5.62,
      pcbX: 19,
      pcbY: 11.5,
      shape: "circle"
    })
  ] : createPlatedHole({
    portHints: ["HOLE_PAD_1"],
    holeDiameter: 3.82,
    outerDiameter: 5.62,
    pcbX: 4.2,
    pcbY: 11,
    shape: "circle"
  });
  return /* @__PURE__ */ (0, import_jsx_runtime3.jsxs)("footprint", { children: [
    generatePads(),
    Array.isArray(mountingHoles) ? mountingHoles : [mountingHoles]
  ] });
};

// util/outlineBuilder.ts
var OutlineBuilder = class {
  constructor(startX, startY) {
    this.points = [];
    this.lastDirection = null;
    this.currentX = startX;
    this.currentY = startY;
    this.points.push({ x: startX, y: startY });
  }
  updateDirection(x, y) {
    if (x > this.currentX) this.lastDirection = "right";
    else if (x < this.currentX) this.lastDirection = "left";
    else if (y > this.currentY) this.lastDirection = "down";
    else if (y < this.currentY) this.lastDirection = "up";
  }
  lineTo(x, y) {
    if (x !== this.currentX || y !== this.currentY) {
      this.updateDirection(x, y);
      this.currentX = x;
      this.currentY = y;
      this.points.push({ x, y });
    }
    return this;
  }
  // helper: convert cardinal direction to unit vector
  dirToUnit(dir) {
    switch (dir) {
      case "right":
        return { x: 1, y: 0 };
      case "left":
        return { x: -1, y: 0 };
      case "down":
        return { x: 0, y: 1 };
      case "up":
        return { x: 0, y: -1 };
    }
  }
  // helper: compute outgoing unit & direction given lastDirection and turn
  outgoingFor(last, turn) {
    switch (last) {
      case "right":
        return turn === "cw" ? { unit: { x: 0, y: 1 }, dir: "down" } : { unit: { x: 0, y: -1 }, dir: "up" };
      case "left":
        return turn === "cw" ? { unit: { x: 0, y: -1 }, dir: "up" } : { unit: { x: 0, y: 1 }, dir: "down" };
      case "down":
        return turn === "cw" ? { unit: { x: -1, y: 0 }, dir: "left" } : { unit: { x: 1, y: 0 }, dir: "right" };
      case "up":
        return turn === "cw" ? { unit: { x: 1, y: 0 }, dir: "right" } : { unit: { x: -1, y: 0 }, dir: "left" };
    }
  }
  corner({ radius, turn }) {
    if (this.points.length < 2) {
      throw new Error("Need at least two points to create a corner");
    }
    if (!this.lastDirection) {
      throw new Error(
        "Cannot create a corner without a known incoming direction"
      );
    }
    const ix = this.currentX;
    const iy = this.currentY;
    const prevIndex = this.points.length - 2;
    const prevPoint = this.points[prevIndex];
    this.points.pop();
    const incomingUnit = this.dirToUnit(this.lastDirection);
    const { unit: outgoingUnit, dir: outgoingDir } = this.outgoingFor(
      this.lastDirection,
      turn
    );
    const dxPrev = ix - prevPoint.x;
    const dyPrev = iy - prevPoint.y;
    const distPrev = Math.hypot(dxPrev, dyPrev);
    const r = Math.min(radius, distPrev);
    const startX = ix - incomingUnit.x * r;
    const startY = iy - incomingUnit.y * r;
    const endX = ix + outgoingUnit.x * r;
    const endY = iy + outgoingUnit.y * r;
    const inVec = { x: -incomingUnit.x, y: -incomingUnit.y };
    const outVec = { x: outgoingUnit.x, y: outgoingUnit.y };
    const bisector = { x: inVec.x + outVec.x, y: inVec.y + outVec.y };
    const bisLen = Math.hypot(bisector.x, bisector.y);
    let dot = inVec.x * outVec.x + inVec.y * outVec.y;
    dot = Math.max(-1, Math.min(1, dot));
    const angle = Math.acos(dot);
    if (bisLen < 1e-9 || Math.abs(Math.sin(angle / 2)) < 1e-9) {
      this.points.push({ x: ix, y: iy });
      this.currentX = ix;
      this.currentY = iy;
      return this;
    }
    const offset = r / Math.sin(angle / 2);
    const bisUnit = { x: bisector.x / bisLen, y: bisector.y / bisLen };
    const centerX = ix + bisUnit.x * offset;
    const centerY = iy + bisUnit.y * offset;
    const startAngle = Math.atan2(startY - centerY, startX - centerX);
    const endAngle = Math.atan2(endY - centerY, endX - centerX);
    let angleDelta = endAngle - startAngle;
    while (angleDelta <= -Math.PI) angleDelta += 2 * Math.PI;
    while (angleDelta > Math.PI) angleDelta -= 2 * Math.PI;
    const lastAfterPop = this.points[this.points.length - 1];
    const almostEqual = (a, b) => Math.abs(a - b) < 1e-9;
    if (!lastAfterPop || !almostEqual(lastAfterPop.x, startX) || !almostEqual(lastAfterPop.y, startY)) {
      this.points.push({ x: startX, y: startY });
    }
    const segments = 8;
    for (let i = 1; i <= segments; i++) {
      const t = i / segments;
      const angle2 = startAngle + angleDelta * t;
      const x = centerX + r * Math.cos(angle2);
      const y = centerY + r * Math.sin(angle2);
      this.points.push({ x, y });
    }
    const last = this.points[this.points.length - 1];
    this.currentX = last.x;
    this.currentY = last.y;
    this.lastDirection = outgoingDir;
    return this;
  }
  arcTo(x, y, { radius, sweep = true, segments = 16 }) {
    const start = { x: this.currentX, y: this.currentY };
    const end = { x, y };
    const mx = (start.x + end.x) / 2;
    const my = (start.y + end.y) / 2;
    const dx = end.x - start.x;
    const dy = end.y - start.y;
    const dist = Math.sqrt(dx * dx + dy * dy);
    if (dist === 0) return this;
    const h = Math.sqrt(Math.max(0, radius * radius - __pow(dist / 2, 2)));
    const perpX = -dy / dist;
    const perpY = dx / dist;
    const cx = mx + (sweep ? 1 : -1) * perpX * h;
    const cy = my + (sweep ? 1 : -1) * perpY * h;
    const startAngle = Math.atan2(start.y - cy, start.x - cx);
    const endAngle = Math.atan2(end.y - cy, end.x - cx);
    let sweepAngle = endAngle - startAngle;
    if (sweep && sweepAngle < 0) sweepAngle += 2 * Math.PI;
    if (!sweep && sweepAngle > 0) sweepAngle -= 2 * Math.PI;
    for (let i = 1; i <= segments; i++) {
      const t = i / segments;
      const angle = startAngle + sweepAngle * t;
      const px = cx + radius * Math.cos(angle);
      const py = cy + radius * Math.sin(angle);
      this.points.push({ x: px, y: py });
    }
    this.currentX = end.x;
    this.currentY = end.y;
    return this;
  }
  toArray() {
    return [...this.points];
  }
};
var outlineBuilder = (x, y) => {
  return new OutlineBuilder(x, y);
};

// lib/MicroModBoard/outlines/boardOutlines.ts
var processorOutline = outlineBuilder(-9, 11).lineTo(11, 11).corner({ radius: 1, turn: "ccw" }).lineTo(11, -7).corner({ radius: 0.5, turn: "ccw" }).lineTo(10, -7).corner({ radius: 0.5, turn: "cw" }).lineTo(10, -11).lineTo(3.28, -11).lineTo(3.28, -8).arcTo(2.08, -8, { radius: 0.6, sweep: true }).lineTo(2.08, -11).lineTo(-10, -11).lineTo(-10, -7).corner({ radius: 0.5, turn: "cw" }).lineTo(-11, -7).corner({ radius: 0.5, turn: "ccw" }).lineTo(-11, 11).corner({ radius: 1, turn: "ccw" }).toArray();
var functionOutline = outlineBuilder(0, 32.5).lineTo(19, 32.5).corner({ radius: 1, turn: "ccw" }).lineTo(19, -28.5).corner({ radius: 0.5, turn: "ccw" }).lineTo(10, -28.5).corner({ radius: 0.5, turn: "cw" }).lineTo(10, -32.5).lineTo(3.28, -32.5).lineTo(3.28, -29.5).arcTo(2.08, -29.5, { radius: 0.6, sweep: true }).lineTo(2.08, -32.5).lineTo(-10, -32.5).lineTo(-10, -28.5).corner({ radius: 0.5, turn: "cw" }).lineTo(-19, -28.5).corner({ radius: 0.5, turn: "ccw" }).lineTo(-19, 32.5).corner({ radius: 1, turn: "ccw" }).toArray();

// lib/MicroModBoard/MicroModBoard.tsx
var import_jsx_runtime4 = __toESM(require_jsx_runtime(), 1);
var MicroModBoard = (_a) => {
  var _b = _a, {
    variant = "processor",
    children
  } = _b, rest = __objRest(_b, [
    "variant",
    "children"
  ]);
  const { boardProps, chipProps = {} } = splitBoardAndChipProps(__spreadProps(__spreadValues({}, rest), {
    variant
  }));
  const resolvedName = `${chipProps.name}_chip`;
  const _a2 = chipProps, { name: _ } = _a2, chipRest = __objRest(_a2, ["name"]);
  let outline;
  const pinLabels = __spreadValues({
    pin2: ["V3_3_1"],
    pin74: ["V3_3_2"],
    pin6: ["N_RESET"],
    pin11: ["N_BOOT"],
    pin4: ["V3_3_EN"],
    pin72: ["RTC_V3"],
    pin9: ["USB_VIN"],
    pin5: ["USB_D_N"],
    pin3: ["USB_D_P"],
    pin37: ["USBHOST_D_N"],
    pin35: ["USBHOST_D_P"],
    pin43: ["CAN_TX"],
    pin41: ["CAN_RX"],
    pin23: ["SWDIO"],
    pin21: ["SWDCK"],
    pin58: ["AUD_MCLK"],
    pin56: ["AUD_OUT"],
    pin54: ["AUD_IN"],
    pin52: ["AUD_LRCLK"],
    pin50: ["AUD_BCLK"],
    pin14: ["I2C_SCL"],
    pin12: ["I2C_SDA"],
    pin16: ["I2C_N_INT"],
    pin53: ["I2C_SCL1"],
    pin51: ["I2C_SDA1"],
    pin49: ["BATT_VIN_3"],
    pin1: ["GND1"],
    pin7: ["GND2"],
    pin33: ["GND3"],
    pin36: ["GND4"],
    pin39: ["GND5"],
    pin45: ["GND6"],
    pin75: ["GND7"],
    pin60: ["SPI_SCK1"],
    pin62: ["SPI_SDO1"],
    pin64: ["SPI_SDI1"],
    pin66: ["SDIO_DATA1"],
    pin68: ["SDIO_DATA2"],
    pin70: ["SPI_N_CS1"],
    pin57: ["SPI_SCK"],
    pin59: ["SPI_SDO"],
    pin61: ["SPI_SDI"],
    pin55: ["SPI_N_CS"],
    pin34: ["A0"],
    pin38: ["A1"],
    pin32: ["PWM0"],
    pin47: ["PWM1"],
    pin10: ["D0"],
    pin18: ["D1"],
    pin17: ["TX1"],
    pin19: ["RX1"],
    pin13: ["RTS1"],
    pin15: ["CTS1"],
    pin22: ["TX2"],
    pin20: ["RX2"],
    pin40: ["G0"],
    pin42: ["G1"],
    pin44: ["G2"],
    pin46: ["G3"],
    pin48: ["G4"],
    pin73: ["G5"],
    pin71: ["G6", "PWR_EN"],
    pin69: ["G7"],
    pin67: ["G8"],
    pin65: ["G9"],
    pin63: ["G10"],
    pin8: ["G11"],
    pin76: ["HOLE_PAD_1"]
  }, variant === "function" ? { pin77: ["HOLE_PAD_2"] } : {});
  outline = variant === "processor" ? processorOutline : functionOutline;
  return /* @__PURE__ */ (0, import_jsx_runtime4.jsxs)("board", __spreadProps(__spreadValues({}, boardProps), { outline, children: [
    /* @__PURE__ */ (0, import_jsx_runtime4.jsx)(
      "chip",
      __spreadProps(__spreadValues({}, chipRest), {
        name: resolvedName,
        footprint: /* @__PURE__ */ (0, import_jsx_runtime4.jsx)(MicroModBoardFootprint, { variant }),
        schWidth: 2.8,
        pcbX: 0,
        pcbY: 0,
        pinLabels,
        schPinArrangement: {
          leftSide: {
            direction: "top-to-bottom",
            pins: [
              "V3_3_1",
              "V3_3_2",
              "N_RESET",
              "N_BOOT",
              "V3_3_EN",
              "RTC_V3",
              "USB_VIN",
              "USB_D_N",
              "USB_D_P",
              "USBHOST_D_N",
              "USBHOST_D_P",
              "CAN_TX",
              "CAN_RX",
              "SWDIO",
              "SWDCK",
              "AUD_MCLK",
              "AUD_OUT",
              "AUD_IN",
              "AUD_LRCLK",
              "AUD_BCLK",
              "I2C_SCL",
              "I2C_SDA",
              "I2C_N_INT",
              "I2C_SCL1",
              "I2C_SDA1",
              "BATT_VIN_3",
              "GND1",
              "GND2",
              "GND3",
              "GND4",
              "GND5",
              "GND6",
              "GND7"
            ]
          },
          rightSide: {
            direction: "top-to-bottom",
            pins: [
              "SPI_SCK1",
              "SPI_SDO1",
              "SPI_SDI1",
              "SDIO_DATA1",
              "SDIO_DATA2",
              "SPI_N_CS1",
              "SPI_SCK",
              "SPI_SDO",
              "SPI_SDI",
              "SPI_N_CS",
              "A0",
              "A1",
              "PWM0",
              "PWM1",
              "D0",
              "D1",
              "TX1",
              "RX1",
              "RTS1",
              "CTS1",
              "TX2",
              "RX2",
              "G0",
              "G1",
              "G2",
              "G3",
              "G4",
              "G5",
              "G6",
              "G7",
              "G8",
              "G9",
              "G10",
              "G11",
              "HOLE_PAD_1",
              ...variant === "function" ? ["HOLE_PAD_2"] : []
            ]
          }
        },
        schPinStyle: {
          pin11: {
            marginBottom: 0.3
          },
          pin72: {
            marginBottom: 0.3
          },
          pin3: {
            marginBottom: 0.3
          },
          pin35: {
            marginBottom: 0.3
          },
          pin41: {
            marginBottom: 0.3
          },
          pin21: {
            marginBottom: 0.3
          },
          pin50: {
            marginBottom: 0.3
          },
          pin51: {
            marginBottom: 0.3
          },
          pin49: {
            marginBottom: 0.4
          },
          pin70: {
            marginBottom: 0.3
          },
          pin55: {
            marginBottom: 0.3
          },
          pin18: {
            marginBottom: 0.6
          },
          pin20: {
            marginBottom: 1
          },
          pin8: {
            marginBottom: 0.3
          }
        }
      })
    ),
    children
  ] }));
};

// lib/RaspberryPiHatBoard/RaspberryPiHatBoard.circuit.tsx
var import_jsx_runtime5 = __toESM(require_jsx_runtime(), 1);
var RaspberryPiHatBoard = (_a) => {
  var _b = _a, {
    children
  } = _b, rest = __objRest(_b, [
    "children"
  ]);
  const { boardProps, chipProps = {} } = splitBoardAndChipProps(__spreadValues({}, rest));
  const resolvedChipName = `${chipProps.name}_chip`;
  const _a2 = chipProps, { name: _ } = _a2, chipRest = __objRest(_a2, ["name"]);
  const outline = new OutlineBuilder(0, 28).lineTo(32.5, 28).corner({ radius: 1, turn: "ccw" }).lineTo(32.5, -28).corner({ radius: 1, turn: "ccw" }).lineTo(13.5, -28).corner({ radius: 1, turn: "ccw" }).lineTo(13.5, -9).arcTo(11.5, -9, { radius: 1, sweep: true }).lineTo(11.5, -28).corner({ radius: 1, turn: "ccw" }).lineTo(-32.5, -28).corner({ radius: 1, turn: "ccw" }).lineTo(-32.5, -8.5).corner({ radius: 1, turn: "ccw" }).lineTo(-27.5, -8.5).corner({ radius: 1, turn: "cw" }).lineTo(-27.5, 8.5).corner({ radius: 1, turn: "cw" }).lineTo(-32.5, 8.5).corner({ radius: 1, turn: "ccw" }).lineTo(-32.5, 28).corner({ radius: 1, turn: "ccw" }).toArray();
  const pinLabels = {
    pin1: ["V3_3_1"],
    pin2: ["V5_1"],
    pin3: ["V5_2"],
    pin4: ["GND_1"],
    pin5: ["GPIO_14", "TX"],
    pin6: ["GPIO_15", "RX"],
    pin7: ["GPIO_18", "PCM_CLK"],
    pin8: ["GND_2"],
    pin9: ["GPIO_23"],
    pin10: ["GPIO_24"],
    pin11: ["GND_3"],
    pin12: ["GPIO_25"],
    pin13: ["GPIO_8", "CE0"],
    pin14: ["GPIO_7", "CE1"],
    pin15: ["GPIO_1", "ID_SC"],
    pin16: ["GND_4"],
    pin17: ["GPIO_12", "PWM0"],
    pin18: ["GND_5"],
    pin19: ["GPIO_16"],
    pin20: ["GPIO_20", "PCM_DIN"],
    pin21: ["GPIO_21", "PCM_DOUT"],
    pin22: ["GND_6"],
    pin23: ["GPIO_26"],
    pin24: ["GPIO_19", "PCM_FS"],
    pin25: ["GPIO_13", "PWM1"],
    pin26: ["GPIO_6"],
    pin27: ["GPIO_5"],
    pin28: ["GPIO_0", "ID_SD"],
    pin29: ["GND_7"],
    pin30: ["GPIO_11", "SCLK"],
    pin31: ["GPIO_9", "MISO"],
    pin32: ["GPIO_10", "MOSI"],
    pin33: ["V3_3_2"],
    pin34: ["GPIO_22"],
    pin35: ["GPIO_27"],
    pin36: ["GPIO_17"],
    pin37: ["GND_8"],
    pin38: ["GPIO_4", "GPCLK0"],
    pin39: ["GPIO_3", "SCL"],
    pin40: ["GPIO_2", "SDA"]
  };
  return /* @__PURE__ */ (0, import_jsx_runtime5.jsxs)("board", __spreadProps(__spreadValues({}, boardProps), { outline, children: [
    /* @__PURE__ */ (0, import_jsx_runtime5.jsx)(
      "chip",
      __spreadProps(__spreadValues({}, chipRest), {
        name: resolvedChipName,
        schWidth: 3.5,
        pinLabels,
        layer: "bottom",
        footprint: "pinrow40_rows2_nopinlabels_p2.54_id1.016_od1.524",
        pcbX: 0,
        pcbY: 23.23,
        pcbRotation: 180,
        showPinAliases: true,
        schPinArrangement: {
          leftSide: {
            direction: "top-to-bottom",
            pins: [
              "V3_3_1",
              "V3_3_2",
              "V5_1",
              "V5_2",
              "GPIO_14",
              "GPIO_15",
              "GPIO_3",
              "GPIO_2",
              "GPIO_20",
              "GPIO_21",
              "GPIO_9",
              "GPIO_10",
              "GND_1",
              "GND_2",
              "GND_3",
              "GND_4",
              "GND_5",
              "GND_6",
              "GND_7",
              "GND_8"
            ]
          },
          rightSide: {
            direction: "top-to-bottom",
            pins: [
              "GPIO_18",
              "GPIO_11",
              "GPIO_24",
              "GPIO_25",
              "GPIO_8",
              "GPIO_7",
              "GPIO_12",
              "GPIO_13",
              "GPIO_26",
              "GPIO_19",
              "GPIO_16",
              "GPIO_6",
              "GPIO_5",
              "GPIO_0",
              "GPIO_22",
              "GPIO_27",
              "GPIO_17",
              "GPIO_4",
              "GPIO_1"
            ]
          }
        },
        schPinStyle: {
          pin3: {
            marginBottom: 0.3
          },
          pin6: {
            marginBottom: 0.3
          },
          pin40: {
            marginBottom: 0.3
          },
          pin21: {
            marginBottom: 0.3
          },
          pin32: {
            marginBottom: 0.3
          },
          pin30: {
            marginBottom: 0.3
          },
          pin12: {
            marginBottom: 0.3
          },
          pin14: {
            marginBottom: 0.3
          },
          pin25: {
            marginBottom: 0.3
          }
        }
      })
    ),
    /* @__PURE__ */ (0, import_jsx_runtime5.jsx)("hole", { diameter: 2.8, pcbX: -29, pcbY: 24.5 }),
    /* @__PURE__ */ (0, import_jsx_runtime5.jsx)("hole", { diameter: 2.8, pcbX: 29, pcbY: -24.5 }),
    /* @__PURE__ */ (0, import_jsx_runtime5.jsx)("hole", { diameter: 2.8, pcbX: -29, pcbY: -24.5 }),
    /* @__PURE__ */ (0, import_jsx_runtime5.jsx)("hole", { diameter: 2.8, pcbX: 29, pcbY: 24.5 }),
    children
  ] }));
};

// lib/XiaoBoard/XiaoBoardFootprint.tsx
var import_react2 = __toESM(require_react(), 1);
var import_jsx_runtime6 = __toESM(require_jsx_runtime(), 1);
var createSmtPad2 = (props) => import_react2.default.createElement("smtpad", props);
var createPlatedHole2 = (props) => import_react2.default.createElement("platedhole", props);
var createHole = (props) => import_react2.default.createElement("hole", props);
var XiaoBoardFootprint = ({
  top = 0,
  bottom = 0,
  left = 7,
  right = 7,
  pitch = 2.54,
  padLength = 2,
  padWidth = 3,
  edgeClearance = 1.2,
  componentWidth = 17,
  bottomPadWidth = 1.016,
  bottomPadHeight = 2.032,
  bottomPitch = 2.54,
  variant,
  withPlatedHoles = false
}) => {
  const isRP2040 = variant === "RP2040";
  if (isRP2040) {
    top = 4;
    bottom = 2;
    left = 7;
    right = 7;
    pitch = 2.54;
    padLength = 2;
    padWidth = 3;
    edgeClearance = 1.2;
    componentWidth = 17;
    bottomPadWidth = 1.016;
    bottomPadHeight = 2.032;
    bottomPitch = 2.54;
  }
  const leftRowX = -componentWidth / 2 + 0.25;
  const rightRowX = componentWidth / 2 - 0.25;
  const generatePads = () => {
    const pads = [];
    let pinNumber = 1;
    if (top > 0) {
      pads.push(
        createSmtPad2({
          portHints: [`pin${pinNumber++}`],
          radius: 0.5715,
          pcbX: -1.396,
          pcbY: 8.442,
          layer: "bottom",
          shape: "circle"
        }),
        createSmtPad2({
          portHints: [`pin${pinNumber++}`],
          radius: 0.5715,
          pcbX: 1.144,
          pcbY: 8.442,
          layer: "bottom",
          shape: "circle"
        })
      );
      pads.push(
        createSmtPad2({
          portHints: [`pin${pinNumber++}`],
          radius: 0.5715,
          pcbX: -1.396,
          pcbY: 5.902,
          layer: "bottom",
          shape: "circle"
        }),
        createSmtPad2({
          portHints: [`pin${pinNumber++}`],
          radius: 0.5715,
          pcbX: 1.144,
          pcbY: 5.902,
          layer: "bottom",
          shape: "circle"
        })
      );
    }
    if (bottom > 0) {
      const actualBottomPitch = bottomPitch != null ? bottomPitch : pitch;
      const xOffset = (bottom - 1) / 2 * actualBottomPitch;
      const bottomPadW = bottomPadWidth != null ? bottomPadWidth : padWidth;
      const bottomPadH = bottomPadHeight != null ? bottomPadHeight : padLength;
      for (let i = 0; i < bottom; i++) {
        pads.push(
          createSmtPad2({
            portHints: [`pin${pinNumber++}`],
            width: bottomPadW,
            height: bottomPadH,
            pcbX: xOffset - i * actualBottomPitch,
            pcbY: -9.7 + edgeClearance,
            layer: "bottom",
            shape: "rect"
          })
        );
      }
    }
    if (left > 0) {
      const yOffset = (left - 1) / 2 * pitch;
      for (let i = 0; i < left; i++) {
        const halfPad = padWidth / 2;
        if (variant !== "Receiver") {
          pads.push(
            createHole({
              diameter: 0.7,
              pcbX: leftRowX - halfPad + 0.86,
              pcbY: yOffset - i * pitch
            })
          );
        }
        if (withPlatedHoles) {
          pads.push(
            createPlatedHole2({
              portHints: [`pin${pinNumber}`],
              holeDiameter: 0.85,
              rectPadWidth: padWidth,
              rectPadHeight: padLength,
              pcbX: leftRowX,
              pcbY: yOffset - i * pitch,
              shape: "circular_hole_with_rect_pad",
              holeOffsetX: 0.63
            })
          );
          pinNumber++;
        } else {
          pads.push(
            createSmtPad2({
              portHints: [`pin${pinNumber}`],
              width: padWidth,
              height: padLength,
              pcbX: leftRowX,
              pcbY: yOffset - i * pitch,
              layer: "top",
              shape: "rect"
            })
          );
          pinNumber++;
        }
      }
    }
    if (right > 0) {
      const yOffset = (right - 1) / 2 * pitch;
      for (let i = 0; i < right; i++) {
        const halfPad = padWidth / 2;
        if (variant !== "Receiver") {
          pads.push(
            createHole({
              diameter: 0.7,
              pcbX: rightRowX + halfPad - 0.86,
              pcbY: yOffset - i * pitch
            })
          );
        }
        if (withPlatedHoles) {
          pads.push(
            createPlatedHole2({
              portHints: [`pin${pinNumber}`],
              holeDiameter: 0.85,
              rectPadWidth: padWidth,
              rectPadHeight: padLength,
              pcbX: rightRowX,
              pcbY: yOffset - i * pitch,
              shape: "circular_hole_with_rect_pad",
              holeOffsetX: -0.63
            })
          );
          pinNumber++;
        } else {
          pads.push(
            createSmtPad2({
              portHints: [`pin${pinNumber}`],
              width: padWidth,
              height: padLength,
              pcbX: rightRowX,
              pcbY: yOffset - i * pitch,
              layer: "top",
              shape: "rect"
            })
          );
          pinNumber++;
        }
      }
    }
    return pads;
  };
  return /* @__PURE__ */ (0, import_jsx_runtime6.jsx)("footprint", { children: generatePads() });
};

// lib/XiaoBoard/XiaoBoard.circuit.tsx
var import_jsx_runtime7 = __toESM(require_jsx_runtime(), 1);
var XiaoBoard = (_a) => {
  var _b = _a, {
    variant,
    withPlatedHoles = false,
    children
  } = _b, rest = __objRest(_b, [
    "variant",
    "withPlatedHoles",
    "children"
  ]);
  const { boardProps, chipProps = {} } = splitBoardAndChipProps(__spreadProps(__spreadValues({}, rest), {
    variant,
    withPlatedHoles
  }));
  const resolvedName = chipProps.name;
  const _a2 = chipProps, { name: _ } = _a2, chipRest = __objRest(_a2, ["name"]);
  const DefaultPinLabels = {
    pin1: "A0",
    pin2: "A1",
    pin3: "A2",
    pin4: "A3",
    pin5: "SDA",
    pin6: "SCL",
    pin7: "TX",
    pin8: "VBUS",
    pin9: "GND1",
    pin10: "V3_3",
    pin11: "MOSI",
    pin12: "MISO",
    pin13: "SCK",
    pin14: "RX"
  };
  const RP2040PinLabels = {
    pin1: "SWDIO",
    pin2: "SWCLK",
    pin3: "RUN",
    pin4: "GND1",
    pin5: "GND2",
    pin6: "VIN",
    pin7: "A0",
    pin8: "A1",
    pin9: "A2",
    pin10: "A3",
    pin11: "SDA",
    pin12: "SCL",
    pin13: "TX",
    pin14: "VBUS",
    pin15: "GND3",
    pin16: "V3_3",
    pin17: "MOSI",
    pin18: "MISO",
    pin19: "SCK",
    pin20: "RX"
  };
  const defaultPinArrangement = {
    leftSide: {
      direction: "top-to-bottom",
      pins: ["A0", "A1", "A2", "A3", "SDA", "SCL", "TX"]
    },
    rightSide: {
      direction: "top-to-bottom",
      pins: ["RX", "SCK", "MISO", "MOSI", "V3_3", "GND1", "VBUS"]
    }
  };
  const rp2040PinArrangement = {
    leftSide: {
      direction: "top-to-bottom",
      pins: ["SWDIO", "RUN", "A0", "A1", "A2", "A3", "SDA", "SCL", "TX", "VIN"]
    },
    rightSide: {
      direction: "top-to-bottom",
      pins: [
        "SWCLK",
        "RX",
        "SCK",
        "MISO",
        "MOSI",
        "V3_3",
        "GND1",
        "GND2",
        "GND3",
        "VBUS"
      ]
    }
  };
  const outline = outlineBuilder(0, 10.55).lineTo(8.89, 10.55).corner({ radius: 1.8, turn: "ccw" }).lineTo(8.89, -10.55).corner({ radius: 1.8, turn: "ccw" }).lineTo(-8.89, -10.55).corner({ radius: 1.8, turn: "ccw" }).lineTo(-8.89, 10.55).corner({ radius: 1.8, turn: "ccw" }).lineTo(0, 10.55).toArray();
  return /* @__PURE__ */ (0, import_jsx_runtime7.jsxs)("board", __spreadProps(__spreadValues({}, boardProps), { outline, children: [
    /* @__PURE__ */ (0, import_jsx_runtime7.jsx)(
      "chip",
      __spreadValues(__spreadProps(__spreadValues({}, chipRest), {
        name: resolvedName,
        footprint: /* @__PURE__ */ (0, import_jsx_runtime7.jsx)(
          XiaoBoardFootprint,
          {
            variant,
            withPlatedHoles
          }
        ),
        doNotPlace: true,
        pcbX: 0,
        pcbY: 0,
        pinLabels: variant === "RP2040" ? RP2040PinLabels : DefaultPinLabels,
        schWidth: 1.5,
        schPinArrangement: variant === "RP2040" ? rp2040PinArrangement : defaultPinArrangement
      }), variant === "RP2040" && {
        schPinStyle: {
          pin2: {
            marginBottom: 0.2
          },
          pin3: {
            marginBottom: 0.3
          },
          pin15: {
            marginBottom: 0.2
          },
          pin16: {
            marginBottom: 0.2
          },
          pin13: {
            marginBottom: 0.3
          }
        }
      })
    ),
    children
  ] }));
};

// lib/XiaoBoard/XiaoReceiver.circuit.tsx
var import_jsx_runtime8 = __toESM(require_jsx_runtime(), 1);
var PIN_LABELS = {
  pin1: "A0",
  pin2: "A1",
  pin3: "A2",
  pin4: "A3",
  pin5: "SDA",
  pin6: "SCL",
  pin7: "TX",
  pin8: "VBUS",
  pin9: "GND1",
  pin10: "V3_3",
  pin11: "MOSI",
  pin12: "MISO",
  pin13: "SCK",
  pin14: "RX"
};
var PIN_ARRANGEMENT = {
  leftSide: {
    direction: "top-to-bottom",
    pins: ["A0", "A1", "A2", "A3", "SDA", "SCL", "TX"]
  },
  rightSide: {
    direction: "top-to-bottom",
    pins: ["RX", "SCK", "MISO", "MOSI", "V3_3", "GND1", "VBUS"]
  }
};
var XiaoReceiver = (props) => {
  const _a = props, { children } = _a, rest = __objRest(_a, ["children"]);
  return /* @__PURE__ */ (0, import_jsx_runtime8.jsxs)(
    "chip",
    __spreadProps(__spreadValues({}, rest), {
      footprint: /* @__PURE__ */ (0, import_jsx_runtime8.jsx)(XiaoBoardFootprint, { variant: "Receiver" }),
      pinLabels: PIN_LABELS,
      schPinArrangement: PIN_ARRANGEMENT,
      schWidth: 1.5,
      doNotPlace: true,
      children: [
        /* @__PURE__ */ (0, import_jsx_runtime8.jsx)(
          "silkscreenpath",
          {
            route: [
              { x: -9.1166, y: -8.5345 },
              { x: -9.1166, y: 8.6105 }
            ]
          }
        ),
        /* @__PURE__ */ (0, import_jsx_runtime8.jsx)(
          "silkscreenpath",
          {
            route: [
              { x: -7.2116, y: -10.4395 },
              { x: 6.7584, y: -10.4395 }
            ]
          }
        ),
        /* @__PURE__ */ (0, import_jsx_runtime8.jsx)(
          "silkscreenpath",
          {
            route: [
              { x: -4.7216, y: 10.5155 },
              { x: -4.717872, y: 11.525772 }
            ]
          }
        ),
        /* @__PURE__ */ (0, import_jsx_runtime8.jsx)(
          "silkscreenpath",
          {
            route: [
              { x: -4.217872, y: 12.0255 },
              { x: 3.7774, y: 12.0255 }
            ]
          }
        ),
        /* @__PURE__ */ (0, import_jsx_runtime8.jsx)(
          "silkscreenpath",
          {
            route: [
              { x: 4.2774, y: 11.5255 },
              { x: 4.2774, y: 10.5155 }
            ]
          }
        ),
        /* @__PURE__ */ (0, import_jsx_runtime8.jsx)(
          "silkscreenpath",
          {
            route: [
              { x: 6.7584, y: 10.5155 },
              { x: -7.2116, y: 10.5155 }
            ]
          }
        ),
        /* @__PURE__ */ (0, import_jsx_runtime8.jsx)(
          "silkscreenpath",
          {
            route: [
              { x: 8.6634, y: -8.5345 },
              { x: 8.6634, y: 8.6105 }
            ]
          }
        ),
        /* @__PURE__ */ (0, import_jsx_runtime8.jsx)(
          "silkscreenpath",
          {
            route: [
              { x: -9.1166, y: 8.6145 },
              { x: -8.86137796498721, y: 9.566999698293817 },
              { x: -8.164099698293818, y: 10.26427796498721 },
              { x: -7.2116, y: 10.5195 }
            ]
          }
        ),
        /* @__PURE__ */ (0, import_jsx_runtime8.jsx)(
          "silkscreenpath",
          {
            route: [
              { x: -7.2116, y: -10.4395 },
              { x: -8.164099698293818, y: -10.18427796498721 },
              { x: -8.861377964987211, y: -9.486999698293818 },
              { x: -9.116600000000002, y: -8.5345 }
            ]
          }
        ),
        /* @__PURE__ */ (0, import_jsx_runtime8.jsx)(
          "silkscreenpath",
          {
            route: [
              { x: -4.717872, y: 11.525772 },
              { x: -4.217872, y: 12.025499999999997 }
            ]
          }
        ),
        /* @__PURE__ */ (0, import_jsx_runtime8.jsx)(
          "silkscreenpath",
          {
            route: [
              { x: 3.7774, y: 12.0255 },
              { x: 4.277400000000001, y: 11.5255 }
            ]
          }
        ),
        /* @__PURE__ */ (0, import_jsx_runtime8.jsx)(
          "silkscreenpath",
          {
            route: [
              { x: 6.7584, y: 10.5155 },
              { x: 7.710899698293817, y: 10.26027796498721 },
              { x: 8.408177964987209, y: 9.562999698293817 },
              { x: 8.663399999999998, y: 8.6105 }
            ]
          }
        ),
        /* @__PURE__ */ (0, import_jsx_runtime8.jsx)(
          "silkscreenpath",
          {
            route: [
              { x: 8.6634, y: -8.5345 },
              { x: 8.40817796498721, y: -9.486999698293818 },
              { x: 7.710899698293818, y: -10.184277964987212 },
              { x: 6.7584, y: -10.4395 }
            ]
          }
        ),
        children
      ]
    })
  );
};

// lib/ProMicroBoard/ProMicroBoardFootprint.tsx
var import_react3 = __toESM(require_react(), 1);
var import_jsx_runtime9 = __toESM(require_jsx_runtime(), 1);
var createPlatedHole3 = (props) => import_react3.default.createElement("platedhole", props);
var ProMicroBoardFootprint = ({
  left = 12,
  right = 12,
  pitch = 2.54
}) => {
  const leftRowX = -7.62;
  const rightRowX = 7.62;
  const boardHeight = 33.02;
  const topMargin = 3.81;
  const boardTop = boardHeight / 2;
  const firstHoleY = boardTop - topMargin;
  const pads = [];
  let pinNumber = 1;
  for (let i = 0; i < right; i++) {
    pads.push(
      createPlatedHole3({
        portHints: [`${pinNumber}`],
        holeDiameter: 1.01,
        outerDiameter: 1.87,
        pcbX: rightRowX,
        pcbY: firstHoleY - i * pitch,
        shape: "circle"
      })
    );
    pinNumber++;
  }
  for (let i = 0; i < left; i++) {
    pads.push(
      createPlatedHole3({
        portHints: [`${pinNumber}`],
        holeDiameter: 1.01,
        outerDiameter: 1.87,
        pcbX: leftRowX,
        pcbY: firstHoleY - i * pitch,
        shape: "circle"
      })
    );
    pinNumber++;
  }
  return /* @__PURE__ */ (0, import_jsx_runtime9.jsx)("footprint", { children: pads });
};

// lib/ProMicroBoard/ProMicroBoard.circuit.tsx
var import_jsx_runtime10 = __toESM(require_jsx_runtime(), 1);
var ProMicroBoard = (_a) => {
  var _b = _a, { children } = _b, rest = __objRest(_b, ["children"]);
  const { boardProps, chipProps = {} } = splitBoardAndChipProps(__spreadValues({}, rest));
  const resolvedName = chipProps.name;
  const _a2 = chipProps, { name: _ } = _a2, chipRest = __objRest(_a2, ["name"]);
  const DefaultPinLabels = {
    pin1: "RAW",
    pin2: "GND1",
    pin3: "RST",
    pin4: "VCC",
    pin5: "A3",
    pin6: "A2",
    pin7: "A1",
    pin8: "A0",
    pin9: "SCK",
    pin10: "MISO",
    pin11: "MOSI",
    pin12: "D10",
    pin13: "TXO",
    pin14: "RX1",
    pin15: "GND2",
    pin16: "GND3",
    pin17: "D2",
    pin18: "D3",
    pin19: "D4",
    pin20: "D5",
    pin21: "D6",
    pin22: "D7",
    pin23: "D8",
    pin24: "D9"
  };
  const defaultPinArrangement = {
    leftSide: {
      direction: "top-to-bottom",
      pins: [
        "TXO",
        "RX1",
        "GND2",
        "GND3",
        "D2",
        "D3",
        "D4",
        "D5",
        "D6",
        "D7",
        "D8",
        "D9"
      ]
    },
    rightSide: {
      direction: "top-to-bottom",
      pins: [
        "RAW",
        "GND1",
        "RST",
        "VCC",
        "A3",
        "A2",
        "A1",
        "A0",
        "SCK",
        "MISO",
        "MOSI",
        "D10"
      ]
    }
  };
  return /* @__PURE__ */ (0, import_jsx_runtime10.jsxs)(
    "board",
    __spreadProps(__spreadValues({}, boardProps), {
      outline: [
        { x: -8.89, y: 16.51 },
        { x: 8.89, y: 16.51 },
        { x: 8.89, y: -16.51 },
        { x: -8.89, y: -16.51 },
        { x: -8.89, y: 16.51 }
      ],
      children: [
        /* @__PURE__ */ (0, import_jsx_runtime10.jsx)(
          "chip",
          __spreadProps(__spreadValues({}, chipRest), {
            name: resolvedName,
            footprint: /* @__PURE__ */ (0, import_jsx_runtime10.jsx)(ProMicroBoardFootprint, {}),
            doNotPlace: true,
            pcbX: 0,
            pcbY: 0,
            pinLabels: DefaultPinLabels,
            schWidth: 1.8,
            schPinArrangement: defaultPinArrangement
          })
        ),
        children
      ]
    })
  );
};

// node_modules/@tscircuit/math-utils/dist/chunk-U45EKA3R.js
function grid({
  rows,
  cols,
  xSpacing,
  ySpacing,
  width,
  height,
  offsetX = 0,
  offsetY = 0,
  yDirection = "cartesian",
  centered = true
}) {
  const effectiveXSpacing = xSpacing != null ? xSpacing : 1;
  const effectiveYSpacing = ySpacing != null ? ySpacing : 1;
  const totalWidth = width != null ? width : cols * effectiveXSpacing;
  const totalHeight = height != null ? height : rows * effectiveYSpacing;
  const centeringOffsetX = centered ? -totalWidth / 2 : 0;
  const centeringOffsetY = centered ? -totalHeight / 2 : 0;
  const cellWidth = width ? width / cols : effectiveXSpacing;
  const cellHeight = height ? height / rows : effectiveYSpacing;
  const cells = [];
  for (let row = 0; row < rows; row++) {
    for (let col = 0; col < cols; col++) {
      const index = row * cols + col;
      const centerX = offsetX + centeringOffsetX + col * cellWidth + cellWidth / 2;
      const rawCenterY = offsetY + row * cellHeight + cellHeight / 2;
      const centerY = yDirection === "cartesian" ? offsetY + centeringOffsetY + (rows - 1 - row) * cellHeight + cellHeight / 2 : offsetY + centeringOffsetY + row * cellHeight + cellHeight / 2;
      cells.push({
        index,
        center: { x: centerX, y: centerY },
        topLeft: {
          x: centerX - cellWidth / 2,
          y: centerY + cellHeight / 2
        },
        bottomRight: {
          x: centerX + cellWidth / 2,
          y: centerY - cellHeight / 2
        },
        row,
        col
      });
    }
  }
  return cells;
}

// lib/ViaGridBoard/viaGridElements.tsx
var import_jsx_runtime11 = __toESM(require_jsx_runtime(), 1);
var ViaGridVia = (props) => {
  const _a = props, { viaIndex } = _a, restProps = __objRest(_a, ["viaIndex"]);
  return /* @__PURE__ */ (0, import_jsx_runtime11.jsx)(
    "via",
    __spreadValues({
      name: `via_${viaIndex}`,
      fromLayer: "top",
      toLayer: "bottom",
      outerDiameter: "1.5mm",
      holeDiameter: "0.3mm"
    }, restProps)
  );
};
var ViaGridPlus = (props) => {
  const _a = props, { startIndex } = _a, restProps = __objRest(_a, ["startIndex"]);
  return /* @__PURE__ */ (0, import_jsx_runtime11.jsxs)("group", __spreadProps(__spreadValues({}, restProps), { children: [
    /* @__PURE__ */ (0, import_jsx_runtime11.jsx)(ViaGridVia, { pcbX: 0, pcbY: 0, viaIndex: startIndex }),
    /* @__PURE__ */ (0, import_jsx_runtime11.jsx)(ViaGridVia, { pcbX: 2.5, pcbY: 0, viaIndex: startIndex + 1 }),
    /* @__PURE__ */ (0, import_jsx_runtime11.jsx)(ViaGridVia, { pcbX: 0, pcbY: 2.5, viaIndex: startIndex + 2 }),
    /* @__PURE__ */ (0, import_jsx_runtime11.jsx)(ViaGridVia, { pcbX: -2.5, pcbY: 0, viaIndex: startIndex + 3 }),
    /* @__PURE__ */ (0, import_jsx_runtime11.jsx)(ViaGridVia, { pcbX: 0, pcbY: -2.5, viaIndex: startIndex + 4 })
  ] }));
};
var pacmanPolygonOutline = new OutlineBuilder(-0.2, -0.2).lineTo(-0.2, 2.5).arcTo(-2.5, 0, { radius: 2.5, sweep: true }).arcTo(0, -2.5, { radius: 2.5, sweep: true }).arcTo(2.5, -0.2, { radius: 2.5, sweep: true }).lineTo(-0.2, -0.2).toArray();

// lib/ViaGridBoard/ViaGridBoard.circuit.tsx
var import_jsx_runtime12 = __toESM(require_jsx_runtime(), 1);
var ViaGridBoard = (_a) => {
  var _b = _a, { children } = _b, rest = __objRest(_b, ["children"]);
  const { boardProps = {} } = splitBoardAndChipProps(__spreadValues({}, rest));
  return /* @__PURE__ */ (0, import_jsx_runtime12.jsxs)("board", __spreadProps(__spreadValues({}, boardProps), { width: "100mm", height: "65mm", children: [
    /* @__PURE__ */ (0, import_jsx_runtime12.jsx)(
      "pcbnoterect",
      {
        pcbX: 0,
        pcbY: 0,
        width: 90,
        height: 55,
        strokeWidth: 0.3,
        color: "blue"
      }
    ),
    ["BL", "TL", "TR", "BR"].map((cornerPositionName, index) => {
      const x = (cornerPositionName.includes("R") ? 90 : 0) + 5 - 50;
      const y = (cornerPositionName.includes("T") ? 55 : 0) + 5 - 32.5;
      const rotation = index * 90;
      return /* @__PURE__ */ (0, import_jsx_runtime12.jsx)(
        "chip",
        {
          name: cornerPositionName,
          pcbX: x,
          pcbY: y,
          noSchematicRepresentation: true,
          pcbRotation: -rotation,
          footprint: /* @__PURE__ */ (0, import_jsx_runtime12.jsx)("footprint", { children: /* @__PURE__ */ (0, import_jsx_runtime12.jsx)(
            "smtpad",
            {
              pcbX: "0mm",
              pcbY: "0mm",
              layer: "top",
              shape: "polygon",
              portHints: ["pin1"],
              points: pacmanPolygonOutline
            }
          ) })
        },
        "pacman_" + cornerPositionName
      );
    }),
    /* @__PURE__ */ (0, import_jsx_runtime12.jsx)(ViaGridPlus, { pcbX: -20, pcbY: -7.5, startIndex: 0 }),
    /* @__PURE__ */ (0, import_jsx_runtime12.jsx)(ViaGridPlus, { pcbX: 20, pcbY: -7.5, startIndex: 5 }),
    /* @__PURE__ */ (0, import_jsx_runtime12.jsx)(ViaGridPlus, { pcbX: -20, pcbY: 7.5, startIndex: 10 }),
    /* @__PURE__ */ (0, import_jsx_runtime12.jsx)(ViaGridPlus, { pcbX: 20, pcbY: 7.5, startIndex: 15 }),
    horizontalEdgeViaGridCells.map((cell) => /* @__PURE__ */ (0, import_jsx_runtime12.jsx)(
      ViaGridVia,
      {
        pcbX: cell.center.x - 50,
        pcbY: cell.center.y - 32.5,
        viaIndex: cell.index + 20
      },
      cell.index
    )),
    verticalEdgeViaGridCells.map((cell) => /* @__PURE__ */ (0, import_jsx_runtime12.jsx)(
      ViaGridVia,
      {
        pcbX: cell.center.x - 50,
        pcbY: cell.center.y - 32.5,
        viaIndex: cell.index + 54
      },
      cell.index
    )),
    /* @__PURE__ */ (0, import_jsx_runtime12.jsx)(
      "chip",
      {
        name: "TOP_RECT",
        pcbX: 0,
        pcbY: 30,
        noSchematicRepresentation: true,
        footprint: /* @__PURE__ */ (0, import_jsx_runtime12.jsx)("footprint", { children: /* @__PURE__ */ (0, import_jsx_runtime12.jsx)(
          "smtpad",
          {
            layer: "top",
            shape: "rect",
            width: "40mm",
            height: "3mm",
            portHints: ["pin1"],
            cornerRadius: 0.5
          }
        ) })
      }
    ),
    /* @__PURE__ */ (0, import_jsx_runtime12.jsx)("silkscreentext", { text: "VIAGRID TOP", fontSize: "1.5mm", pcbX: 0, pcbY: -30 }),
    /* @__PURE__ */ (0, import_jsx_runtime12.jsx)("net", { name: "GND" }),
    children
  ] }));
};
var horizontalEdgeViaGridCells = grid({
  rows: 2,
  cols: 17,
  // width: 80,
  // height: 45,
  xSpacing: 5,
  // if you want to provide the spacing instead of width
  ySpacing: 45,
  // if you want to provide the spacing instead of height
  offsetX: 50,
  // optional
  offsetY: 32.5,
  // optional
  yDirection: "up-is-negative"
  // optional, default: "cartesian"
  // centered: true  // optional, default: true
});
var verticalEdgeViaGridCells = grid({
  rows: 8,
  cols: 2,
  // width: 80,
  // height: 45,
  xSpacing: 80,
  // if you want to provide the spacing instead of width
  ySpacing: 5,
  // if you want to provide the spacing instead of height
  offsetX: 50,
  // optional
  offsetY: 32.5,
  // optional
  yDirection: "up-is-negative"
  // optional, default: "cartesian"
  // centered: true  // optional, default: true
});
export {
  ArduinoShield,
  MicroModBoard,
  ProMicroBoard,
  RaspberryPiHatBoard,
  ViaGridBoard,
  XiaoBoard,
  XiaoReceiver
};
/*! Bundled license information:

react/cjs/react-jsx-runtime.production.js:
  (**
   * @license React
   * react-jsx-runtime.production.js
   *
   * Copyright (c) Meta Platforms, Inc. and affiliates.
   *
   * This source code is licensed under the MIT license found in the
   * LICENSE file in the root directory of this source tree.
   *)

react/cjs/react.production.js:
  (**
   * @license React
   * react.production.js
   *
   * Copyright (c) Meta Platforms, Inc. and affiliates.
   *
   * This source code is licensed under the MIT license found in the
   * LICENSE file in the root directory of this source tree.
   *)

react/cjs/react.development.js:
  (**
   * @license React
   * react.development.js
   *
   * Copyright (c) Meta Platforms, Inc. and affiliates.
   *
   * This source code is licensed under the MIT license found in the
   * LICENSE file in the root directory of this source tree.
   *)

react/cjs/react-jsx-runtime.development.js:
  (**
   * @license React
   * react-jsx-runtime.development.js
   *
   * Copyright (c) Meta Platforms, Inc. and affiliates.
   *
   * This source code is licensed under the MIT license found in the
   * LICENSE file in the root directory of this source tree.
   *)
*/
