// @tscircuit/eval/dist/worker.js
// lib/worker.ts
import * as Comlink from "comlink";

// lib/utils/get-imports-from-code.ts
var getImportsFromCode = (code) => {
  const importRegex = /^\s*import\s+(?:(?:[\w\s]+,\s*)?(?:\*\s+as\s+[\w\s]+|\{[\s\w,]+\}|\w+)\s+from\s*)?['"](.+?)['"]/gm;
  const imports = [];
  let match;
  while ((match = importRegex.exec(code)) !== null) {
    const fullMatch = match[0];
    if (/^\s*import\s+type\b/.test(fullMatch)) {
      continue;
    }
    imports.push(match[1]);
  }
  const reExportRegex = /^\s*export\s+(?:type\s+)?(?:\*\s+as\s+[\w$]+|\*|\{[^}]+\})\s+from\s*['"](.+?)['"]/gm;
  let reExportMatch;
  while ((reExportMatch = reExportRegex.exec(code)) !== null) {
    const fullMatch = reExportMatch[0];
    if (/^\s*export\s+type\b/.test(fullMatch)) {
      continue;
    }
    imports.push(reExportMatch[1]);
  }
  return imports;
};

// lib/shared/obj-path.ts
function getObjectPaths(obj, prefix = "") {
  const paths = [];
  for (const key in obj) {
    if (!Object.prototype.hasOwnProperty.call(obj, key)) {
      continue;
    }
    const value = obj[key];
    const path = prefix ? `${prefix}.${key}` : key;
    if (value !== null && typeof value === "object" && !Array.isArray(value)) {
      const subPaths = getObjectPaths(value, path);
      paths.push(...subPaths);
    } else {
      paths.push(path);
    }
  }
  return paths;
}
function getValueAtPath(obj, path) {
  const keys = path.split(".");
  let current = obj;
  for (const key of keys) {
    current = current[key];
  }
  return current;
}

// lib/worker.ts
var createCircuitWebWorker = async (configuration) => {
  const existingWorker = globalThis.TSCIRCUIT_GLOBAL_CIRCUIT_WORKER;
  if (existingWorker && typeof existingWorker.kill === "function") {
    if (configuration.verbose) {
      console.log("[Worker] Killing previous global worker instance...");
    }
    try {
      await existingWorker.kill();
    } catch (e) {
      if (configuration.verbose) {
        console.warn(
          "[Worker] Error killing previous global worker instance:",
          e
        );
      }
      if (globalThis.TSCIRCUIT_GLOBAL_CIRCUIT_WORKER === existingWorker) {
        globalThis.TSCIRCUIT_GLOBAL_CIRCUIT_WORKER = void 0;
      }
    }
  }
  if (configuration.verbose) {
    console.log(
      "[Worker] Creating circuit web worker with config:",
      configuration
    );
  }
  let workerBlobUrl = configuration.webWorkerBlobUrl ?? configuration.webWorkerUrl;
  if (!workerBlobUrl) {
    const cdnUrl = `https://cdn.jsdelivr.net/npm/@tscircuit/eval@${configuration.evalVersion ?? "latest"}/dist/webworker/entrypoint.js`;
    const workerBlob = await globalThis.fetch(cdnUrl).then((res) => res.blob());
    workerBlobUrl = URL.createObjectURL(workerBlob);
  }
  const rawWorker = new Worker(workerBlobUrl, { type: "module" });
  let workerInitError;
  rawWorker.addEventListener("error", (event) => {
    console.error("[Worker] Error in worker", event);
    workerInitError = event;
  });
  rawWorker.addEventListener("unhandledrejection", (event) => {
    console.error("[Worker] Unhandled rejection in worker", event);
  });
  rawWorker.addEventListener("messageerror", (event) => {
    console.error("[Worker] Message error in worker", event);
  });
  const earlyMessageHandler = (event) => {
    console.log("[Worker] Message in worker", event);
  };
  rawWorker.addEventListener("message", earlyMessageHandler);
  rawWorker.addEventListener("message", async (event) => {
    const data = event.data;
    if (data?.type !== "worker_fetch") return;
    try {
      const response = await globalThis.fetch(data.input, data.init);
      const body = await response.text();
      rawWorker.postMessage({
        type: "worker_fetch_result",
        requestId: data.requestId,
        success: true,
        response: {
          body,
          status: response.status,
          statusText: response.statusText,
          headers: (() => {
            const obj = {};
            response.headers.forEach((value, key) => {
              obj[key] = value;
            });
            return obj;
          })()
        }
      });
    } catch (err) {
      rawWorker.postMessage({
        type: "worker_fetch_result",
        requestId: data.requestId,
        success: false,
        error: {
          name: err.name,
          message: err.message,
          stack: err.stack
        }
      });
    }
  });
  if (workerInitError) {
    throw workerInitError;
  }
  const comlinkWorker = Comlink.wrap(rawWorker);
  rawWorker.removeEventListener("message", earlyMessageHandler);
  function serializeReactElement(element) {
    if (!element || typeof element !== "object") {
      return element;
    }
    if (element.type && element.props !== void 0) {
      return {
        __isSerializedReactElement: true,
        type: element.type,
        props: serializeProps(element.props),
        key: element.key
      };
    }
    return element;
  }
  function serializeProps(props) {
    if (!props || typeof props !== "object") {
      return props;
    }
    const serialized = {};
    for (const [key, value] of Object.entries(props)) {
      if (key === "children") {
        if (Array.isArray(value)) {
          serialized.children = value.map(serializeReactElement);
        } else {
          serialized.children = serializeReactElement(value);
        }
      } else {
        serialized[key] = value;
      }
    }
    return serialized;
  }
  if (configuration.enableFetchProxy) {
    rawWorker.postMessage({ type: "override_global_fetch" });
  }
  if (configuration.snippetsApiBaseUrl) {
    await comlinkWorker.setSnippetsApiBaseUrl(configuration.snippetsApiBaseUrl);
  }
  const maybeProxy = (value) => {
    if (typeof value === "function") {
      return Comlink.proxy(value);
    }
    return value;
  };
  if (configuration.platform) {
    for (const path of getObjectPaths(configuration.platform)) {
      await comlinkWorker.setPlatformConfigProperty(
        path,
        maybeProxy(getValueAtPath(configuration.platform, path))
      ).catch((e) => {
        throw new Error(
          `Error setting platform config property ${path}: ${e instanceof Error ? e.message : String(e)}`
        );
      });
    }
  }
  if (configuration.projectConfig) {
    for (const path of getObjectPaths(configuration.projectConfig)) {
      await comlinkWorker.setProjectConfigProperty(
        path,
        maybeProxy(getValueAtPath(configuration.projectConfig, path))
      ).catch((e) => {
        throw new Error(
          `Error setting project config property ${path}: ${e instanceof Error ? e.message : String(e)}`
        );
      });
    }
  }
  let isTerminated = false;
  const wrapper = {
    clearEventListeners: comlinkWorker.clearEventListeners.bind(comlinkWorker),
    enableDebug: async (...args) => {
      if (isTerminated) {
        throw new Error("CircuitWebWorker was terminated, can't enableDebug");
      }
      return comlinkWorker.enableDebug.bind(comlinkWorker)(...args);
    },
    version: comlinkWorker.version.bind(comlinkWorker),
    execute: async (...args) => {
      if (isTerminated) {
        throw new Error("CircuitWebWorker was terminated, can't execute");
      }
      return comlinkWorker.execute.bind(comlinkWorker)(...args);
    },
    executeComponent: async (component) => {
      if (isTerminated) {
        throw new Error(
          "CircuitWebWorker was terminated, can't executeComponent"
        );
      }
      if (typeof component === "function") {
        return comlinkWorker.executeComponent.bind(comlinkWorker)(component);
      }
      if (component && typeof component === "object" && component.type) {
        const serializedElement = serializeReactElement(component);
        return comlinkWorker.executeComponent.bind(comlinkWorker)(
          serializedElement
        );
      }
      return comlinkWorker.executeComponent.bind(comlinkWorker)(component);
    },
    executeWithFsMap: async (...args) => {
      if (isTerminated) {
        throw new Error(
          "CircuitWebWorker was terminated, can't executeWithFsMap"
        );
      }
      return comlinkWorker.executeWithFsMap.bind(comlinkWorker)(...args);
    },
    renderUntilSettled: comlinkWorker.renderUntilSettled.bind(comlinkWorker),
    getCircuitJson: comlinkWorker.getCircuitJson.bind(comlinkWorker),
    on: (event, callback) => {
      const proxiedCallback = Comlink.proxy(callback);
      comlinkWorker.on(event, proxiedCallback);
    },
    kill: async () => {
      comlinkWorker[Comlink.releaseProxy]();
      rawWorker.terminate();
      isTerminated = true;
      if (globalThis.TSCIRCUIT_GLOBAL_CIRCUIT_WORKER === wrapper) {
        globalThis.TSCIRCUIT_GLOBAL_CIRCUIT_WORKER = void 0;
      }
    }
  };
  wrapper.__rawWorker = rawWorker;
  globalThis.TSCIRCUIT_GLOBAL_CIRCUIT_WORKER = wrapper;
  return wrapper;
};
export {
  createCircuitWebWorker,
  getImportsFromCode
};
